// @solar-republic/cosmos-grpc/build/dist/proto/juno/feeshare/v1/genesis.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeJunoFeeshareFeeShare } from './feeshare.js';
/**
 * Encodes a `GenesisState` protobuf message: defines the module's genesis state.
 * @param g_params - `params`: the feeshare module parameters
 * @param a_fee_shares - `fee_share`: FeeShare is a slice of active registered contracts for fee distribution
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoFeeshareGenesisState = (atu8_params, a_fee_shares) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .B(a_fee_shares) // FeeShare[] fee_share = 2
    .o;
/**
 * Encodes a `Params` protobuf message: defines the feeshare module params
 * @param b_enable_fee_share - `enable_fee_share`: enable_feeshare defines a parameter to enable the feeshare module
 * @param s_developer_shares - `developer_shares`: defines the proportion of the transaction fees to be distributed to the registered contract owner
 * @param a_allowed_denoms - `allowed_denoms`: defines the list of denoms that are allowed to be paid to the contract withdraw addresses. If said denom is not in the list, the fees will ONLY be sent to the community pool. If this list is empty, all denoms are allowed.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoFeeshareParams = (b_enable_fee_share, s_developer_shares, a_allowed_denoms) => Protobuf() // ...
    .v(b_enable_fee_share) // boolean enable_fee_share = 1
    .s(s_developer_shares) // string developer_shares = 2
    .S(a_allowed_denoms) // string[] allowed_denoms = 3
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoFeeshareGenesisState}
 */
export const decodeJunoFeeshareGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0], [decodeJunoFeeshareParams, decodeJunoFeeshareFeeShare]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoFeeshareParams}
 */
export const decodeJunoFeeshareParams = (atu8_payload) => decode_protobuf(atu8_payload, [1, 5, 4]);
/**
 * Destructures the fields of a {@link JunoFeeshareParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: enable_fee_share - enable_feeshare defines a parameter to enable the feeshare module
 *   - 1: developer_shares - defines the proportion of the transaction fees to be distributed to the registered contract owner
 *   - 2: allowed_denoms - defines the list of denoms that are allowed to be paid to the contract withdraw addresses. If said denom is not in the list, the fees will ONLY be sent to the community pool. If this list is empty, all denoms are allowed.
 */
export const destructJunoFeeshareParams = (g_struct) => [g_struct.enable_fee_share, g_struct.developer_shares, g_struct.allowed_denoms];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_FEESHARE_PARAMS = '/juno.feeshare.v1.Params';
/**
 * Destructures the fields of a {@link JunoFeeshareGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - the feeshare module parameters
 *   - 1: fee_share - FeeShare is a slice of active registered contracts for fee distribution
 */
export const destructJunoFeeshareGenesisState = (g_struct) => [g_struct.params, g_struct.fee_share];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_FEESHARE_GENESIS_STATE = '/juno.feeshare.v1.GenesisState';
