// @solar-republic/cosmos-grpc/build/dist/proto/ibc/applications/interchain_accounts/v1/packet.js
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../../api/util.js';
import { decodeGoogleProtobufAny } from '../../../../google/protobuf/any.js';
/**
 * Encodes a `InterchainAccountPacketData` protobuf message: comprised of a raw transaction, type of transaction and optional memo field.
 * @param xc_type - `type`:
 * @param atu8_data - `data`:
 * @param s_memo - `memo`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsInterchainAccountPacketData = (xc_type, atu8_data, s_memo) => Protobuf() // ...
    .v(xc_type) // Type type = 1
    .b(atu8_data) // bytes data = 2
    .s(s_memo) // string memo = 3
    .o;
/**
 * Encodes a `CosmosTx` protobuf message: contains a list of sdk.Msg's. It should be used when sending transactions to an SDK host chain.
 * @param a_messages - `messages`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsCosmosTx = (a_messages) => Protobuf() // ...
    .B(a_messages) // Any[] messages = 1
    .o;
/**
 * Decodes a protobuf InterchainAccountPacketData message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsInterchainAccountPacketData}
 */
export const decodeIbcApplicationsInterchainAccountsInterchainAccountPacketData = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 5]);
/**
 * Decodes a protobuf CosmosTx message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsCosmosTx}
 */
export const decodeIbcApplicationsInterchainAccountsCosmosTx = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeGoogleProtobufAny]);
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsInterchainAccountPacketData} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: type -
 *   - 1: data -
 *   - 2: memo -
 */
export const destructIbcApplicationsInterchainAccountsInterchainAccountPacketData = (g_struct) => [g_struct.type, safe_base64_to_bytes(g_struct.data), g_struct.memo];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_INTERCHAIN_ACCOUNT_PACKET_DATA = '/ibc.applications.interchain_accounts.v1.InterchainAccountPacketData';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsCosmosTx} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: messages -
 */
export const destructIbcApplicationsInterchainAccountsCosmosTx = (g_struct) => [g_struct.messages];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_COSMOS_TX = '/ibc.applications.interchain_accounts.v1.CosmosTx';
/**
 * Protobuf enum value for `ibc.applications.interchain_accounts.v1.Type`.
 *
 * **TYPE_UNSPECIFIED** - Default zero value enumeration
 *
 * Belongs to enum type {@link ProtoEnumIbcApplicationsInterchainAccountsType}
 */
export const XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED = 0;
/**
 * JSON enum value for `ibc.applications.interchain_accounts.v1.Type`.
 *
 * **TYPE_UNSPECIFIED** - Default zero value enumeration
 *
 * Belongs to enum type {@link JsonEnumIbcApplicationsInterchainAccountsType}
 */
export const SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED = 'UNSPECIFIED';
/**
 * Protobuf enum value for `ibc.applications.interchain_accounts.v1.Type`.
 *
 * **TYPE_EXECUTE_TX** - Execute a transaction on an interchain accounts host chain
 *
 * Belongs to enum type {@link ProtoEnumIbcApplicationsInterchainAccountsType}
 */
export const XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX = 1;
/**
 * JSON enum value for `ibc.applications.interchain_accounts.v1.Type`.
 *
 * **TYPE_EXECUTE_TX** - Execute a transaction on an interchain accounts host chain
 *
 * Belongs to enum type {@link JsonEnumIbcApplicationsInterchainAccountsType}
 */
export const SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX = 'EXECUTE_TX';
/**
 * Maps a protobuf enum int value for `ibc.applications.interchain_accounts.v1.Type` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumIbcApplicationsInterchainAccountsType = {
    [XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED]: SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED,
    [XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX]: SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX,
};
/**
 * Maps a JSON enum string value for `ibc.applications.interchain_accounts.v1.Type` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumIbcApplicationsInterchainAccountsType = {
    [SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED]: XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_UNSPECIFIED,
    [SI_JSON_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX]: XC_PROTO_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_TYPE_EXECUTE_TX,
};
