// @solar-republic/cosmos-grpc/build/dist/proto/akash/base/v1beta2/resource.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeAkashBaseAttribute } from './attribute.js';
import { decodeAkashBaseResourceValue } from './resourcevalue.js';
/**
 * Encodes a `CPU` protobuf message: stores resource units and cpu config attributes
 * @param g_units - `units`:
 * @param a_attributes - `attributes`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseCPU = (atu8_units, a_attributes) => Protobuf() // ...
    .b(atu8_units) // ResourceValue units = 1
    .B(a_attributes) // Attribute[] attributes = 2
    .o;
/**
 * Encodes a `Memory` protobuf message: stores resource quantity and memory attributes
 * @param g_quantity - `quantity`:
 * @param a_attributes - `attributes`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseMemory = (atu8_quantity, a_attributes) => Protobuf() // ...
    .b(atu8_quantity) // ResourceValue quantity = 1
    .B(a_attributes) // Attribute[] attributes = 2
    .o;
/**
 * Encodes a `Storage` protobuf message: stores resource quantity and storage attributes
 * @param s_name - `name`:
 * @param g_quantity - `quantity`:
 * @param a_attributes - `attributes`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseStorage = (s_name, atu8_quantity, a_attributes) => Protobuf() // ...
    .s(s_name) // string name = 1
    .b(atu8_quantity) // ResourceValue quantity = 2
    .B(a_attributes) // Attribute[] attributes = 3
    .o;
/**
 * Decodes a protobuf CPU message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseCPU}
 */
export const decodeAkashBaseCPU = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0], [decodeAkashBaseResourceValue, decodeAkashBaseAttribute]);
/**
 * Decodes a protobuf Memory message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseMemory}
 */
export const decodeAkashBaseMemory = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0], [decodeAkashBaseResourceValue, decodeAkashBaseAttribute]);
/**
 * Decodes a protobuf Storage message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseStorage}
 */
export const decodeAkashBaseStorage = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 0], [0, decodeAkashBaseResourceValue, decodeAkashBaseAttribute]);
/**
 * Destructures the fields of a {@link AkashBaseCPU} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: units -
 *   - 1: attributes -
 */
export const destructAkashBaseCPU = (g_struct) => [g_struct.units, g_struct.attributes];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_C_P_U = '/akash.base.v1beta2.CPU';
/**
 * Destructures the fields of a {@link AkashBaseMemory} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: quantity -
 *   - 1: attributes -
 */
export const destructAkashBaseMemory = (g_struct) => [g_struct.quantity, g_struct.attributes];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_MEMORY = '/akash.base.v1beta2.Memory';
/**
 * Destructures the fields of a {@link AkashBaseStorage} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: quantity -
 *   - 2: attributes -
 */
export const destructAkashBaseStorage = (g_struct) => [g_struct.name, g_struct.quantity, g_struct.attributes];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_STORAGE = '/akash.base.v1beta2.Storage';
