// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/feegrant/v1beta1/feegrant.js
import { parse_duration, parse_timestamp } from '../../../../api/json.js';
import { decode_protobuf, decode_coin, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, temporal, any, coins } from '../../../../api/protobuf-writer.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
/**
 * Encodes a `BasicAllowance` protobuf message wrapped in the `Any` container: implements Allowance with a one-time grant of coins that optionally expires. The grantee can use up to SpendLimit to cover fees.
 * @param a_spend_limits - `spend_limit`: specifies the maximum amount of coins that can be spent by this allowance and will be updated as coins are spent. If it is empty, there is no spend limit and any amount of coins can be spent.
 * @param xt_expiration - `expiration`: specifies an optional time when this allowance expires
 * @returns a strongly subtyped Uint8Array representing an `Any` protobuf message
 */
export const anyBasicAllowance = (a_spend_limits, xt_expiration) => any('/cosmos.feegrant.v1beta1.BasicAllowance', Protobuf() // ...
    .B(coins(a_spend_limits)) // Coin[] spend_limit = 1
    .b(temporal(xt_expiration)) // Timestamp expiration = 2
    .o);
/**
 * Encodes a `PeriodicAllowance` protobuf message wrapped in the `Any` container: extends Allowance to allow for both a maximum cap, as well as a limit per time period.
 * @param g_basic - `basic`: specifies a struct of `BasicAllowance`
 * @param xt_period - `period`: specifies the time duration in which period_spend_limit coins can be spent before that allowance is reset
 * @param a_period_spend_limits - `period_spend_limit`: specifies the maximum number of coins that can be spent in the period
 * @param a_period_can_spends - `period_can_spend`: the number of coins left to be spent before the period_reset time
 * @param xt_period_reset - `period_reset`: the time at which this period resets and a new one begins, it is calculated from the start time of the first transaction after the last period ended
 * @returns a strongly subtyped Uint8Array representing an `Any` protobuf message
 */
export const anyPeriodicAllowance = (atu8_basic, xt_period, a_period_spend_limits, a_period_can_spends, xt_period_reset) => any('/cosmos.feegrant.v1beta1.PeriodicAllowance', Protobuf() // ...
    .b(atu8_basic) // BasicAllowance basic = 1
    .b(temporal(xt_period)) // Duration period = 2
    .B(coins(a_period_spend_limits)) // Coin[] period_spend_limit = 3
    .B(coins(a_period_can_spends)) // Coin[] period_can_spend = 4
    .b(temporal(xt_period_reset)) // Timestamp period_reset = 5
    .o);
/**
 * Encodes a `AllowedMsgAllowance` protobuf message wrapped in the `Any` container: creates allowance only for specified message types.
 * @param atu8_allowance - `allowance`: can be any of basic and periodic fee allowance.
 * @param a_allowed_messages - `allowed_messages`: the messages for which the grantee has the access.
 * @returns a strongly subtyped Uint8Array representing an `Any` protobuf message
 */
export const anyAllowedMsgAllowance = (atu8_allowance, a_allowed_messages) => any('/cosmos.feegrant.v1beta1.AllowedMsgAllowance', Protobuf() // ...
    .b(atu8_allowance) // Any allowance = 1
    .S(a_allowed_messages) // string[] allowed_messages = 2
    .o);
/**
 * Encodes a `BasicAllowance` protobuf message: implements Allowance with a one-time grant of coins that optionally expires. The grantee can use up to SpendLimit to cover fees.
 * @param a_spend_limits - `spend_limit`: specifies the maximum amount of coins that can be spent by this allowance and will be updated as coins are spent. If it is empty, there is no spend limit and any amount of coins can be spent.
 * @param xt_expiration - `expiration`: specifies an optional time when this allowance expires
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosFeegrantBasicAllowance = (a_spend_limits, xt_expiration) => Protobuf() // ...
    .B(coins(a_spend_limits)) // Coin[] spend_limit = 1
    .b(temporal(xt_expiration)) // Timestamp expiration = 2
    .o;
/**
 * Encodes a `PeriodicAllowance` protobuf message: extends Allowance to allow for both a maximum cap, as well as a limit per time period.
 * @param g_basic - `basic`: specifies a struct of `BasicAllowance`
 * @param xt_period - `period`: specifies the time duration in which period_spend_limit coins can be spent before that allowance is reset
 * @param a_period_spend_limits - `period_spend_limit`: specifies the maximum number of coins that can be spent in the period
 * @param a_period_can_spends - `period_can_spend`: the number of coins left to be spent before the period_reset time
 * @param xt_period_reset - `period_reset`: the time at which this period resets and a new one begins, it is calculated from the start time of the first transaction after the last period ended
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosFeegrantPeriodicAllowance = (atu8_basic, xt_period, a_period_spend_limits, a_period_can_spends, xt_period_reset) => Protobuf() // ...
    .b(atu8_basic) // BasicAllowance basic = 1
    .b(temporal(xt_period)) // Duration period = 2
    .B(coins(a_period_spend_limits)) // Coin[] period_spend_limit = 3
    .B(coins(a_period_can_spends)) // Coin[] period_can_spend = 4
    .b(temporal(xt_period_reset)) // Timestamp period_reset = 5
    .o;
/**
 * Encodes a `AllowedMsgAllowance` protobuf message: creates allowance only for specified message types.
 * @param atu8_allowance - `allowance`: can be any of basic and periodic fee allowance.
 * @param a_allowed_messages - `allowed_messages`: the messages for which the grantee has the access.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosFeegrantAllowedMsgAllowance = (atu8_allowance, a_allowed_messages) => Protobuf() // ...
    .b(atu8_allowance) // Any allowance = 1
    .S(a_allowed_messages) // string[] allowed_messages = 2
    .o;
/**
 * Encodes a `Grant` protobuf message: stored in the KVStore to record a grant with full context
 * @param sa_granter - `granter`: the address of the user granting an allowance of their funds.
 * @param sa_grantee - `grantee`: the address of the user being granted an allowance of another user's funds.
 * @param atu8_allowance - `allowance`: can be any of basic, periodic, allowed fee allowance.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosFeegrantGrant = (sa_granter, sa_grantee, atu8_allowance) => Protobuf() // ...
    .s(sa_granter) // string granter = 1
    .s(sa_grantee) // string grantee = 2
    .b(atu8_allowance) // Any allowance = 3
    .o;
/**
 * Decodes a protobuf BasicAllowance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosFeegrantBasicAllowance}
 */
export const decodeCosmosFeegrantBasicAllowance = (atu8_payload) => decode_protobuf(atu8_payload, [0, [3, 1]], [decode_coin, decode_temporal]);
/**
 * Decodes a protobuf PeriodicAllowance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosFeegrantPeriodicAllowance}
 */
export const decodeCosmosFeegrantPeriodicAllowance = (atu8_payload) => decode_protobuf(atu8_payload, [1, [3, 1], 0, 0, [3, 1]], [decodeCosmosFeegrantBasicAllowance, decode_temporal, decode_coin, decode_coin, decode_temporal]);
/**
 * Decodes a protobuf AllowedMsgAllowance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosFeegrantAllowedMsgAllowance}
 */
export const decodeCosmosFeegrantAllowedMsgAllowance = (atu8_payload) => decode_protobuf(atu8_payload, [1, 4], [decodeGoogleProtobufAny]);
/**
 * Decodes a protobuf Grant message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosFeegrantGrant}
 */
export const decodeCosmosFeegrantGrant = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 1], [0, 0, decodeGoogleProtobufAny]);
/**
 * Destructures the fields of a {@link CosmosFeegrantGrant} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: granter - the address of the user granting an allowance of their funds.
 *   - 1: grantee - the address of the user being granted an allowance of another user's funds.
 *   - 2: allowance - can be any of basic, periodic, allowed fee allowance.
 */
export const destructCosmosFeegrantGrant = (g_struct) => [g_struct.granter, g_struct.grantee, g_struct.allowance];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_FEEGRANT_GRANT = '/cosmos.feegrant.v1beta1.Grant';
/**
 * Destructures the fields of a {@link CosmosFeegrantBasicAllowance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: spend_limit - specifies the maximum amount of coins that can be spent by this allowance and will be updated as coins are spent. If it is empty, there is no spend limit and any amount of coins can be spent.
 *   - 1: expiration - specifies an optional time when this allowance expires
 */
export const destructCosmosFeegrantBasicAllowance = (g_struct) => [g_struct.spend_limit, parse_timestamp(g_struct.expiration)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_FEEGRANT_BASIC_ALLOWANCE = '/cosmos.feegrant.v1beta1.BasicAllowance';
/**
 * Destructures the fields of a {@link CosmosFeegrantPeriodicAllowance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: basic - specifies a struct of `BasicAllowance`
 *   - 1: period - specifies the time duration in which period_spend_limit coins can be spent before that allowance is reset
 *   - 2: period_spend_limit - specifies the maximum number of coins that can be spent in the period
 *   - 3: period_can_spend - the number of coins left to be spent before the period_reset time
 *   - 4: period_reset - the time at which this period resets and a new one begins, it is calculated from the start time of the first transaction after the last period ended
 */
export const destructCosmosFeegrantPeriodicAllowance = (g_struct) => [g_struct.basic, parse_duration(g_struct.period), g_struct.period_spend_limit, g_struct.period_can_spend, parse_timestamp(g_struct.period_reset)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_FEEGRANT_PERIODIC_ALLOWANCE = '/cosmos.feegrant.v1beta1.PeriodicAllowance';
/**
 * Destructures the fields of a {@link CosmosFeegrantAllowedMsgAllowance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: allowance - can be any of basic and periodic fee allowance.
 *   - 1: allowed_messages - the messages for which the grantee has the access.
 */
export const destructCosmosFeegrantAllowedMsgAllowance = (g_struct) => [g_struct.allowance, g_struct.allowed_messages];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_FEEGRANT_ALLOWED_MSG_ALLOWANCE = '/cosmos.feegrant.v1beta1.AllowedMsgAllowance';
