// @solar-republic/cosmos-grpc/build/dist/proto/ibc/core/client/v1/genesis.js
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../../api/util.js';
import { decodeIbcCoreClientIdentifiedClientState, decodeIbcCoreClientClientConsensusStates, decodeIbcCoreClientParams } from './client.js';
/**
 * Encodes a `GenesisState` protobuf message: defines the ibc client submodule's genesis state.
 * @param a_clients - `clients`: client states with their corresponding identifiers
 * @param a_clients_consensus - `clients_consensus`: consensus states from each client
 * @param a_clients_metadatas - `clients_metadata`: metadata from each client
 * @param g_params - `params`:
 * @param b_create_localhost - `create_localhost`: Deprecated: create_localhost has been deprecated. The localhost client is automatically created at genesis.
 * @param sg_next_client_sequence - `next_client_sequence`: the sequence for the next generated client identifier
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcCoreClientGenesisState = (a_clients, a_clients_consensus, a_clients_metadatas, atu8_params, b_create_localhost, sg_next_client_sequence) => Protobuf() // ...
    .B(a_clients) // IdentifiedClientState[] clients = 1
    .B(a_clients_consensus) // ClientConsensusStates[] clients_consensus = 2
    .B(a_clients_metadatas) // IdentifiedGenesisMetadata[] clients_metadata = 3
    .b(atu8_params) // Params params = 4
    .v(b_create_localhost) // boolean create_localhost = 5
    .g(sg_next_client_sequence) // uint64 next_client_sequence = 6
    .o;
/**
 * Encodes a `GenesisMetadata` protobuf message: defines the genesis type for metadata that clients may return with ExportMetadata
 * @param atu8_key - `key`: store key of metadata without clientID-prefix
 * @param atu8_value - `value`: metadata value
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcCoreClientGenesisMetadata = (atu8_key, atu8_value) => Protobuf() // ...
    .b(atu8_key) // bytes key = 1
    .b(atu8_value) // bytes value = 2
    .o;
/**
 * Encodes a `IdentifiedGenesisMetadata` protobuf message: has the client metadata with the corresponding client id.
 * @param si_client - `client_id`:
 * @param a_client_metadatas - `client_metadata`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcCoreClientIdentifiedGenesisMetadata = (si_client, a_client_metadatas) => Protobuf() // ...
    .s(si_client) // string client_id = 1
    .B(a_client_metadatas) // GenesisMetadata[] client_metadata = 2
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcCoreClientGenesisState}
 */
export const decodeIbcCoreClientGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [0, 0, 0, 1, 1, 3], [decodeIbcCoreClientIdentifiedClientState, decodeIbcCoreClientClientConsensusStates, decodeIbcCoreClientIdentifiedGenesisMetadata, decodeIbcCoreClientParams]);
/**
 * Decodes a protobuf GenesisMetadata message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcCoreClientGenesisMetadata}
 */
export const decodeIbcCoreClientGenesisMetadata = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Decodes a protobuf IdentifiedGenesisMetadata message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcCoreClientIdentifiedGenesisMetadata}
 */
export const decodeIbcCoreClientIdentifiedGenesisMetadata = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0], [0, decodeIbcCoreClientGenesisMetadata]);
/**
 * Destructures the fields of a {@link IbcCoreClientGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: clients - client states with their corresponding identifiers
 *   - 1: clients_consensus - consensus states from each client
 *   - 2: clients_metadata - metadata from each client
 *   - 3: params -
 *   - 4: create_localhost - Deprecated: create_localhost has been deprecated. The localhost client is automatically created at genesis.
 *   - 5: next_client_sequence - the sequence for the next generated client identifier
 */
export const destructIbcCoreClientGenesisState = (g_struct) => [g_struct.clients, g_struct.clients_consensus, g_struct.clients_metadata, g_struct.params, g_struct.create_localhost, g_struct.next_client_sequence];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_CORE_CLIENT_GENESIS_STATE = '/ibc.core.client.v1.GenesisState';
/**
 * Destructures the fields of a {@link IbcCoreClientIdentifiedGenesisMetadata} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: client_id -
 *   - 1: client_metadata -
 */
export const destructIbcCoreClientIdentifiedGenesisMetadata = (g_struct) => [g_struct.client_id, g_struct.client_metadata];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_CORE_CLIENT_IDENTIFIED_GENESIS_METADATA = '/ibc.core.client.v1.IdentifiedGenesisMetadata';
/**
 * Destructures the fields of a {@link IbcCoreClientGenesisMetadata} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key - store key of metadata without clientID-prefix
 *   - 1: value - metadata value
 */
export const destructIbcCoreClientGenesisMetadata = (g_struct) => [safe_base64_to_bytes(g_struct.key), safe_base64_to_bytes(g_struct.value)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_CORE_CLIENT_GENESIS_METADATA = '/ibc.core.client.v1.GenesisMetadata';
