// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/store/snapshots/v1/snapshot.js
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf, map } from '../../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../../api/util.js';
/**
 * Encodes a `Snapshot` protobuf message: contains Tendermint state sync snapshot info.
 * @param sg_height - `height`:
 * @param n_format - `format`:
 * @param n_chunks - `chunks`:
 * @param atu8_hash - `hash`:
 * @param g_metadata - `metadata`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshot = (sg_height, n_format, n_chunks, atu8_hash, atu8_metadata) => Protobuf() // ...
    .g(sg_height) // uint64 height = 1
    .v(n_format) // uint32 format = 2
    .v(n_chunks) // uint32 chunks = 3
    .b(atu8_hash) // bytes hash = 4
    .b(atu8_metadata) // Metadata metadata = 5
    .o;
/**
 * Encodes a `Metadata` protobuf message: contains SDK-specific snapshot metadata.
 * @param a_chunk_hashes - `chunk_hashes`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsMetadata = (a_chunk_hashes) => Protobuf() // ...
    .B(a_chunk_hashes) // bytes[] chunk_hashes = 1
    .o;
/**
 * Encodes a `SnapshotItem` protobuf message: an item contained in a rootmulti.Store snapshot. Since: cosmos-sdk 0.46
 * @param g_store - `store`:
 * @param g_iavl - `iavl`:
 * @param g_extension - `extension`:
 * @param g_extension_payload - `extension_payload`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshotItem = (atu8_store, atu8_iavl, atu8_extension, atu8_extension_payload) => Protobuf() // ...
    .b(atu8_store) // SnapshotStoreItem store = 1
    .b(atu8_iavl) // SnapshotIAVLItem iavl = 2
    .b(atu8_extension) // SnapshotExtensionMeta extension = 3
    .b(atu8_extension_payload) // SnapshotExtensionPayload extension_payload = 4
    .o;
/**
 * Encodes a `SnapshotStoreItem` protobuf message: contains metadata about a snapshotted store. Since: cosmos-sdk 0.46
 * @param s_name - `name`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshotStoreItem = (s_name) => Protobuf() // ...
    .s(s_name) // string name = 1
    .o;
/**
 * Encodes a `SnapshotIAVLItem` protobuf message: an exported IAVL node. Since: cosmos-sdk 0.46
 * @param atu8_key - `key`:
 * @param atu8_value - `value`:
 * @param sg_version - `version`: block height
 * @param n_height - `height`: depth of the tree.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshotIAVLItem = (atu8_key, atu8_value, sg_version, n_height) => Protobuf() // ...
    .b(atu8_key) // bytes key = 1
    .b(atu8_value) // bytes value = 2
    .g(sg_version) // int32 version = 3
    .v(n_height) // int32 height = 4
    .o;
/**
 * Encodes a `SnapshotExtensionMeta` protobuf message: contains metadata about an external snapshotter. Since: cosmos-sdk 0.46
 * @param s_name - `name`:
 * @param n_format - `format`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshotExtensionMeta = (s_name, n_format) => Protobuf() // ...
    .s(s_name) // string name = 1
    .v(n_format) // uint32 format = 2
    .o;
/**
 * Encodes a `SnapshotExtensionPayload` protobuf message: contains payloads of an external snapshotter. Since: cosmos-sdk 0.46
 * @param atu8_payload - `payload`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStoreSnapshotsSnapshotExtensionPayload = (atu8_payload) => Protobuf() // ...
    .b(atu8_payload) // bytes payload = 1
    .o;
/**
 * Decodes a protobuf Snapshot message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshot}
 */
export const decodeCosmosStoreSnapshotsSnapshot = (atu8_payload) => decode_protobuf(atu8_payload, [3, 1, 1, 1, 1], [0, 0, 0, 0, decodeCosmosStoreSnapshotsMetadata]);
/**
 * Decodes a protobuf Metadata message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsMetadata}
 */
export const decodeCosmosStoreSnapshotsMetadata = (atu8_payload) => decode_protobuf(atu8_payload, [0]);
/**
 * Decodes a protobuf SnapshotItem message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshotItem}
 */
export const decodeCosmosStoreSnapshotsSnapshotItem = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1, 1], [decodeCosmosStoreSnapshotsSnapshotStoreItem, decodeCosmosStoreSnapshotsSnapshotIAVLItem, decodeCosmosStoreSnapshotsSnapshotExtensionMeta, decodeCosmosStoreSnapshotsSnapshotExtensionPayload]);
/**
 * Decodes a protobuf SnapshotStoreItem message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshotStoreItem}
 */
export const decodeCosmosStoreSnapshotsSnapshotStoreItem = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf SnapshotIAVLItem message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshotIAVLItem}
 */
export const decodeCosmosStoreSnapshotsSnapshotIAVLItem = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 3, 1]);
/**
 * Decodes a protobuf SnapshotExtensionMeta message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshotExtensionMeta}
 */
export const decodeCosmosStoreSnapshotsSnapshotExtensionMeta = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1]);
/**
 * Decodes a protobuf SnapshotExtensionPayload message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStoreSnapshotsSnapshotExtensionPayload}
 */
export const decodeCosmosStoreSnapshotsSnapshotExtensionPayload = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshot} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: height -
 *   - 1: format -
 *   - 2: chunks -
 *   - 3: hash -
 *   - 4: metadata -
 */
export const destructCosmosStoreSnapshotsSnapshot = (g_struct) => [g_struct.height, g_struct.format, g_struct.chunks, safe_base64_to_bytes(g_struct.hash), g_struct.metadata];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT = '/cosmos.store.snapshots.v1.Snapshot';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsMetadata} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chunk_hashes -
 */
export const destructCosmosStoreSnapshotsMetadata = (g_struct) => [map(g_struct.chunk_hashes, safe_base64_to_bytes)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_METADATA = '/cosmos.store.snapshots.v1.Metadata';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshotItem} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: store -
 *   - 1: iavl -
 *   - 2: extension -
 *   - 3: extension_payload -
 */
export const destructCosmosStoreSnapshotsSnapshotItem = (g_struct) => [g_struct.store, g_struct.iavl, g_struct.extension, g_struct.extension_payload];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT_ITEM = '/cosmos.store.snapshots.v1.SnapshotItem';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshotStoreItem} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 */
export const destructCosmosStoreSnapshotsSnapshotStoreItem = (g_struct) => [g_struct.name];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT_STORE_ITEM = '/cosmos.store.snapshots.v1.SnapshotStoreItem';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshotIAVLItem} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 *   - 2: version - block height
 *   - 3: height - depth of the tree.
 */
export const destructCosmosStoreSnapshotsSnapshotIAVLItem = (g_struct) => [safe_base64_to_bytes(g_struct.key), safe_base64_to_bytes(g_struct.value), g_struct.version, g_struct.height];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT_I_A_V_L_ITEM = '/cosmos.store.snapshots.v1.SnapshotIAVLItem';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshotExtensionMeta} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: format -
 */
export const destructCosmosStoreSnapshotsSnapshotExtensionMeta = (g_struct) => [g_struct.name, g_struct.format];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT_EXTENSION_META = '/cosmos.store.snapshots.v1.SnapshotExtensionMeta';
/**
 * Destructures the fields of a {@link CosmosStoreSnapshotsSnapshotExtensionPayload} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: payload -
 */
export const destructCosmosStoreSnapshotsSnapshotExtensionPayload = (g_struct) => [safe_base64_to_bytes(g_struct.payload)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STORE_SNAPSHOTS_SNAPSHOT_EXTENSION_PAYLOAD = '/cosmos.store.snapshots.v1.SnapshotExtensionPayload';
