// @solar-republic/cosmos-grpc/build/dist/api/json.js
import { __UNDEFINED } from '@blake.regalia/belt';
export const parse_duration = (s_duration) => s_duration
    ? +/^([\d.]+)s$/.exec(s_duration)[0] * 1e3
    : __UNDEFINED;
export const parse_timestamp = (s_timestamp) => s_timestamp
    ? Date.parse(s_timestamp)
    : __UNDEFINED;
export const duration_to_json = (xn_milliseconds) => xn_milliseconds
    ? (xn_milliseconds / 1e3) + ''
    : __UNDEFINED;
export const timestamp_to_json = (xt_timestamp) => xt_timestamp
    ? new Date(xt_timestamp).toISOString()
    : __UNDEFINED;
export const expand_duration = (a_temporal) => (a_temporal ? a_temporal[0] + '.' + ('' + (a_temporal[1] | 0)).padStart(9, '0') : __UNDEFINED);
export const expand_timestamp = (a_temporal) => {
    // @ts-expect-error implicit void
    if (!a_temporal)
        return;
    const [sg_seconds, xn_nanos] = a_temporal;
    const xn_millis = xn_nanos / 1e6;
    const s_basis = new Date((+sg_seconds * 1e3) + (xn_millis | 0)).toISOString();
    const xn_remainder = xn_millis % 1;
    if (xn_remainder) {
        return s_basis.replace(/Z$/, xn_remainder.toFixed(6) + 'Z');
    }
    return s_basis;
};
export const expand_coin = (a_coin) => (a_coin ? {
    denom: a_coin[0],
    amount: a_coin[1],
} : __UNDEFINED);
