// @solar-republic/cosmos-grpc/build/dist/proto/akash/base/v1beta3/attribute.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `Attribute` protobuf message: represents key value pair
 * @param s_key - `key`:
 * @param s_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseAttribute = (s_key, s_value) => Protobuf() // ...
    .s(s_key) // string key = 1
    .s(s_value) // string value = 2
    .o;
/**
 * Encodes a `SignedBy` protobuf message: represents validation accounts that tenant expects signatures for provider attributes AllOf has precedence i.e. if there is at least one entry AnyOf is ignored regardless to how many entries there this behaviour to be discussed
 * @param a_all_ofs - `all_of`: all keys in this list must have signed attributes
 * @param a_any_ofs - `any_of`: at least of of the keys from the list must have signed attributes
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseSignedBy = (a_all_ofs, a_any_ofs) => Protobuf() // ...
    .S(a_all_ofs) // string[] all_of = 1
    .S(a_any_ofs) // string[] any_of = 2
    .o;
/**
 * Encodes a `PlacementRequirements` protobuf message:
 * @param g_signed_by - `signed_by`: SignedBy list of keys that tenants expect to have signatures from
 * @param a_attributes - `attributes`: Attribute list of attributes tenant expects from the provider
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBasePlacementRequirements = (atu8_signed_by, a_attributes) => Protobuf() // ...
    .b(atu8_signed_by) // SignedBy signed_by = 1
    .B(a_attributes) // Attribute[] attributes = 2
    .o;
/**
 * Decodes a protobuf Attribute message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseAttribute}
 */
export const decodeAkashBaseAttribute = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf SignedBy message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseSignedBy}
 */
export const decodeAkashBaseSignedBy = (atu8_payload) => decode_protobuf(atu8_payload, [4, 4]);
/**
 * Decodes a protobuf PlacementRequirements message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBasePlacementRequirements}
 */
export const decodeAkashBasePlacementRequirements = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0], [decodeAkashBaseSignedBy, decodeAkashBaseAttribute]);
/**
 * Destructures the fields of a {@link AkashBaseAttribute} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 */
export const destructAkashBaseAttribute = (g_struct) => [g_struct.key, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_ATTRIBUTE = '/akash.base.v1beta3.Attribute';
/**
 * Destructures the fields of a {@link AkashBasePlacementRequirements} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: signed_by - SignedBy list of keys that tenants expect to have signatures from
 *   - 1: attributes - Attribute list of attributes tenant expects from the provider
 */
export const destructAkashBasePlacementRequirements = (g_struct) => [g_struct.signed_by, g_struct.attributes];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_PLACEMENT_REQUIREMENTS = '/akash.base.v1beta3.PlacementRequirements';
/**
 * Destructures the fields of a {@link AkashBaseSignedBy} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: all_of - all keys in this list must have signed attributes
 *   - 1: any_of - at least of of the keys from the list must have signed attributes
 */
export const destructAkashBaseSignedBy = (g_struct) => [g_struct.all_of, g_struct.any_of];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_SIGNED_BY = '/akash.base.v1beta3.SignedBy';
