// @solar-republic/cosmos-grpc/build/dist/proto/ibc/lightclients/tendermint/v1/tendermint.js
import { parse_duration, parse_timestamp } from '../../../../../api/json.js';
import { decode_protobuf, decode_temporal } from '../../../../../api/protobuf-reader.js';
import { Protobuf, temporal } from '../../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../../api/util.js';
import { decodeCosmosIcs23ProofSpec } from '../../../../cosmos/ics23/v1/proofs.js';
import { decodeIbcCoreClientHeight } from '../../../core/client/v1/client.js';
import { decodeIbcCoreCommitmentMerkleRoot } from '../../../core/commitment/v1/commitment.js';
import { decodeTendermintTypesSignedHeader } from '../../../../tendermint/types/types.js';
import { decodeTendermintTypesValidatorSet } from '../../../../tendermint/types/validator.js';
/**
 * Encodes a `ClientState` protobuf message: from Tendermint tracks the current validator set, latest height, and a possible frozen height.
 * @param si_chain - `chain_id`:
 * @param g_trust_level - `trust_level`:
 * @param xt_trusting_period - `trusting_period`: duration of the period since the LatestTimestamp during which the submitted headers are valid for upgrade
 * @param xt_unbonding_period - `unbonding_period`: duration of the staking unbonding period
 * @param xt_max_clock_drift - `max_clock_drift`: defines how much new (untrusted) header's Time can drift into the future.
 * @param g_frozen_height - `frozen_height`: Block height when the client was frozen due to a misbehaviour
 * @param g_latest_height - `latest_height`: Latest height the client was updated to
 * @param a_proof_specs - `proof_specs`: Proof specifications used in verifying counterparty state
 * @param a_upgrade_paths - `upgrade_path`: Path at which next upgraded client will be committed. Each element corresponds to the key for a single CommitmentProof in the chained proof. NOTE: ClientState must stored under `{upgradePath}/{upgradeHeight}/clientState` ConsensusState must be stored under `{upgradepath}/{upgradeHeight}/consensusState` For SDK chains using the default upgrade module, upgrade_path should be []string{"upgrade", "upgradedIBCState"}`
 * @param b_allow_update_after_expiry - `allow_update_after_expiry`: deprecated
 * @param b_allow_update_after_misbehaviour - `allow_update_after_misbehaviour`: deprecated
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcLightclientsTendermintClientState = (si_chain, atu8_trust_level, xt_trusting_period, xt_unbonding_period, xt_max_clock_drift, atu8_frozen_height, atu8_latest_height, a_proof_specs, a_upgrade_paths, b_allow_update_after_expiry, b_allow_update_after_misbehaviour) => Protobuf() // ...
    .s(si_chain) // string chain_id = 1
    .b(atu8_trust_level) // Fraction trust_level = 2
    .b(temporal(xt_trusting_period)) // Duration trusting_period = 3
    .b(temporal(xt_unbonding_period)) // Duration unbonding_period = 4
    .b(temporal(xt_max_clock_drift)) // Duration max_clock_drift = 5
    .b(atu8_frozen_height) // Height frozen_height = 6
    .b(atu8_latest_height) // Height latest_height = 7
    .B(a_proof_specs) // ProofSpec[] proof_specs = 8
    .S(a_upgrade_paths) // string[] upgrade_path = 9
    .v(b_allow_update_after_expiry) // boolean allow_update_after_expiry = 10
    .v(b_allow_update_after_misbehaviour) // boolean allow_update_after_misbehaviour = 11
    .o;
/**
 * Encodes a `ConsensusState` protobuf message: defines the consensus state from Tendermint.
 * @param xt_timestamp - `timestamp`: that corresponds to the block height in which the ConsensusState was stored.
 * @param g_root - `root`: commitment root (i.e app hash)
 * @param atu8_next_validators_hash - `next_validators_hash`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcLightclientsTendermintConsensusState = (xt_timestamp, atu8_root, atu8_next_validators_hash) => Protobuf() // ...
    .b(temporal(xt_timestamp)) // Timestamp timestamp = 1
    .b(atu8_root) // MerkleRoot root = 2
    .b(atu8_next_validators_hash) // bytes next_validators_hash = 3
    .o;
/**
 * Encodes a `Misbehaviour` protobuf message: a wrapper over two conflicting Headers that implements Misbehaviour interface expected by ICS-02
 * @param si_client - `client_id`: ClientID is deprecated
 * @param g_header_1 - `header_1`:
 * @param g_header_2 - `header_2`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcLightclientsTendermintMisbehaviour = (si_client, atu8_header_1, atu8_header_2) => Protobuf() // ...
    .s(si_client) // string client_id = 1
    .b(atu8_header_1) // Header header_1 = 2
    .b(atu8_header_2) // Header header_2 = 3
    .o;
/**
 * Encodes a `Header` protobuf message: defines the Tendermint client consensus Header. It encapsulates all the information necessary to update from a trusted Tendermint ConsensusState. The inclusion of TrustedHeight and TrustedValidators allows this update to process correctly, so long as the ConsensusState for the TrustedHeight exists, this removes race conditions among relayers The SignedHeader and ValidatorSet are the new untrusted update fields for the client. The TrustedHeight is the height of a stored ConsensusState on the client that will be used to verify the new untrusted header. The Trusted ConsensusState must be within the unbonding period of current time in order to correctly verify, and the TrustedValidators must hash to TrustedConsensusState.NextValidatorsHash since that is the last trusted validator set at the TrustedHeight.
 * @param g_signed_header - `signed_header`:
 * @param g_validator_set - `validator_set`:
 * @param g_trusted_height - `trusted_height`:
 * @param g_trusted_validators - `trusted_validators`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcLightclientsTendermintHeader = (atu8_signed_header, atu8_validator_set, atu8_trusted_height, atu8_trusted_validators) => Protobuf() // ...
    .b(atu8_signed_header) // SignedHeader signed_header = 1
    .b(atu8_validator_set) // ValidatorSet validator_set = 2
    .b(atu8_trusted_height) // Height trusted_height = 3
    .b(atu8_trusted_validators) // ValidatorSet trusted_validators = 4
    .o;
/**
 * Encodes a `Fraction` protobuf message: defines the protobuf message type for tmmath.Fraction that only supports positive values.
 * @param sg_numerator - `numerator`:
 * @param sg_denominator - `denominator`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcLightclientsTendermintFraction = (sg_numerator, sg_denominator) => Protobuf() // ...
    .g(sg_numerator) // uint64 numerator = 1
    .g(sg_denominator) // uint64 denominator = 2
    .o;
/**
 * Decodes a protobuf ClientState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcLightclientsTendermintClientState}
 */
export const decodeIbcLightclientsTendermintClientState = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, [3, 1], [3, 1], [3, 1], 1, 1, 0, 4, 1, 1], [0, decodeIbcLightclientsTendermintFraction, decode_temporal, decode_temporal, decode_temporal, decodeIbcCoreClientHeight, decodeIbcCoreClientHeight, decodeCosmosIcs23ProofSpec]);
/**
 * Decodes a protobuf ConsensusState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcLightclientsTendermintConsensusState}
 */
export const decodeIbcLightclientsTendermintConsensusState = (atu8_payload) => decode_protobuf(atu8_payload, [[3, 1], 1, 1], [decode_temporal, decodeIbcCoreCommitmentMerkleRoot]);
/**
 * Decodes a protobuf Misbehaviour message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcLightclientsTendermintMisbehaviour}
 */
export const decodeIbcLightclientsTendermintMisbehaviour = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 1], [0, decodeIbcLightclientsTendermintHeader, decodeIbcLightclientsTendermintHeader]);
/**
 * Decodes a protobuf Header message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcLightclientsTendermintHeader}
 */
export const decodeIbcLightclientsTendermintHeader = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1, 1], [decodeTendermintTypesSignedHeader, decodeTendermintTypesValidatorSet, decodeIbcCoreClientHeight, decodeTendermintTypesValidatorSet]);
/**
 * Decodes a protobuf Fraction message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcLightclientsTendermintFraction}
 */
export const decodeIbcLightclientsTendermintFraction = (atu8_payload) => decode_protobuf(atu8_payload, [3, 3]);
/**
 * Destructures the fields of a {@link IbcLightclientsTendermintClientState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain_id -
 *   - 1: trust_level -
 *   - 2: trusting_period - duration of the period since the LatestTimestamp during which the submitted headers are valid for upgrade
 *   - 3: unbonding_period - duration of the staking unbonding period
 *   - 4: max_clock_drift - defines how much new (untrusted) header's Time can drift into the future.
 *   - 5: frozen_height - Block height when the client was frozen due to a misbehaviour
 *   - 6: latest_height - Latest height the client was updated to
 *   - 7: proof_specs - Proof specifications used in verifying counterparty state
 *   - 8: upgrade_path - Path at which next upgraded client will be committed. Each element corresponds to the key for a single CommitmentProof in the chained proof. NOTE: ClientState must stored under `{upgradePath}/{upgradeHeight}/clientState` ConsensusState must be stored under `{upgradepath}/{upgradeHeight}/consensusState` For SDK chains using the default upgrade module, upgrade_path should be []string{"upgrade", "upgradedIBCState"}`
 *   - 9: allow_update_after_expiry - deprecated
 *   - 10: allow_update_after_misbehaviour - deprecated
 */
export const destructIbcLightclientsTendermintClientState = (g_struct) => [g_struct.chain_id, g_struct.trust_level, parse_duration(g_struct.trusting_period), parse_duration(g_struct.unbonding_period), parse_duration(g_struct.max_clock_drift), g_struct.frozen_height, g_struct.latest_height, g_struct.proof_specs, g_struct.upgrade_path, g_struct.allow_update_after_expiry, g_struct.allow_update_after_misbehaviour];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_LIGHTCLIENTS_TENDERMINT_CLIENT_STATE = '/ibc.lightclients.tendermint.v1.ClientState';
/**
 * Destructures the fields of a {@link IbcLightclientsTendermintFraction} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: numerator -
 *   - 1: denominator -
 */
export const destructIbcLightclientsTendermintFraction = (g_struct) => [g_struct.numerator, g_struct.denominator];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_LIGHTCLIENTS_TENDERMINT_FRACTION = '/ibc.lightclients.tendermint.v1.Fraction';
/**
 * Destructures the fields of a {@link IbcLightclientsTendermintConsensusState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: timestamp - that corresponds to the block height in which the ConsensusState was stored.
 *   - 1: root - commitment root (i.e app hash)
 *   - 2: next_validators_hash -
 */
export const destructIbcLightclientsTendermintConsensusState = (g_struct) => [parse_timestamp(g_struct.timestamp), g_struct.root, safe_base64_to_bytes(g_struct.next_validators_hash)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_LIGHTCLIENTS_TENDERMINT_CONSENSUS_STATE = '/ibc.lightclients.tendermint.v1.ConsensusState';
/**
 * Destructures the fields of a {@link IbcLightclientsTendermintMisbehaviour} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: client_id - ClientID is deprecated
 *   - 1: header_1 -
 *   - 2: header_2 -
 */
export const destructIbcLightclientsTendermintMisbehaviour = (g_struct) => [g_struct.client_id, g_struct.header_1, g_struct.header_2];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_LIGHTCLIENTS_TENDERMINT_MISBEHAVIOUR = '/ibc.lightclients.tendermint.v1.Misbehaviour';
/**
 * Destructures the fields of a {@link IbcLightclientsTendermintHeader} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: signed_header -
 *   - 1: validator_set -
 *   - 2: trusted_height -
 *   - 3: trusted_validators -
 */
export const destructIbcLightclientsTendermintHeader = (g_struct) => [g_struct.signed_header, g_struct.validator_set, g_struct.trusted_height, g_struct.trusted_validators];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_LIGHTCLIENTS_TENDERMINT_HEADER = '/ibc.lightclients.tendermint.v1.Header';
