// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/pool-models/balancer/tx/tx.js
import { decode_protobuf } from '../../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../../api/protobuf-writer.js';
import { decodeOsmosisGammPoolParams, decodeOsmosisGammPoolAsset } from '../balancerPool.js';
/**
 * Encodes a `MsgCreateBalancerPool` protobuf message: ===================== MsgCreatePool
 * @param sa_sender - `sender`:
 * @param g_pool_params - `pool_params`:
 * @param a_pool_assets - `pool_assets`:
 * @param s_future_pool_governor - `future_pool_governor`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsBalancerMsgCreateBalancerPool = (sa_sender, atu8_pool_params, a_pool_assets, s_future_pool_governor) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .b(atu8_pool_params) // PoolParams pool_params = 2
    .B(a_pool_assets) // PoolAsset[] pool_assets = 3
    .s(s_future_pool_governor) // string future_pool_governor = 4
    .o;
/**
 * Encodes a `MsgCreateBalancerPoolResponse` protobuf message: Returns the poolID
 * @param sg_pool_id - `pool_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsBalancerMsgCreateBalancerPoolResponse = (sg_pool_id) => Protobuf() // ...
    .g(sg_pool_id) // uint64 pool_id = 1
    .o;
/**
 * Decodes a protobuf CreateBalancerPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsBalancerMsgCreateBalancerPool}
 */
export const decodeOsmosisGammPoolmodelsBalancerMsgCreateBalancerPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 0, 5], [0, decodeOsmosisGammPoolParams, decodeOsmosisGammPoolAsset]);
/**
 * Decodes a protobuf CreateBalancerPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsBalancerMsgCreateBalancerPoolResponse}
 */
export const decodeOsmosisGammPoolmodelsBalancerMsgCreateBalancerPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsBalancerMsgCreateBalancerPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_params -
 *   - 2: pool_assets -
 *   - 3: future_pool_governor -
 */
export const destructOsmosisGammPoolmodelsBalancerMsgCreateBalancerPool = (g_struct) => [g_struct.sender, g_struct.pool_params, g_struct.pool_assets, g_struct.future_pool_governor];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_BALANCER_MSG_CREATE_BALANCER_POOL = '/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPool';
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsBalancerMsgCreateBalancerPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_id -
 */
export const destructOsmosisGammPoolmodelsBalancerMsgCreateBalancerPoolResponse = (g_struct) => [g_struct.pool_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_BALANCER_MSG_CREATE_BALANCER_POOL_RESPONSE = '/osmosis.gamm.poolmodels.balancer.v1beta1.MsgCreateBalancerPoolResponse';
