// @solar-republic/cosmos-grpc/build/dist/proto/axelar/vote/v1beta1/types.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
/**
 * Encodes a `TalliedVote` protobuf message: represents a vote for a poll with the accumulated stake of all validators voting for the same data
 * @param atu8_tally - `tally`:
 * @param atu8_data - `data`:
 * @param sg_poll_id - `poll_id`:
 * @param a_is_voter_lates - `is_voter_late`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarVoteTalliedVote = (atu8_tally, atu8_data, sg_poll_id, a_is_voter_lates) => Protobuf() // ...
    .b(atu8_tally) // bytes tally = 1
    .b(atu8_data, 3) // Any data = 3
    .g(sg_poll_id, 4) // uint64 poll_id = 4
    .B(a_is_voter_lates, 5) // IsVoterLateEntry[] is_voter_late = 5
    .o;
/**
 * Encodes a `IsVoterLateEntry` protobuf message:
 * @param s_key - `key`:
 * @param b_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarVoteTalliedVoteIsVoterLateEntry = (s_key, b_value) => Protobuf() // ...
    .s(s_key) // string key = 1
    .v(b_value) // boolean value = 2
    .o;
/**
 * Decodes a protobuf TalliedVote message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedAxelarVoteTalliedVote}
 */
export const decodeAxelarVoteTalliedVote = ((atu8_payload, [atu8_tally, , a_data, sg_poll_id, a_is_voter_late] = decode_protobuf(atu8_payload, [1, 0, 1, 3, 0], [0, 0, decodeGoogleProtobufAny, 0, decodeAxelarVoteTalliedVoteIsVoterLateEntry])) => [atu8_tally, __UNDEFINED, a_data, sg_poll_id, a_is_voter_late]);
/**
 * Decodes a protobuf IsVoterLateEntry message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarVoteTalliedVoteIsVoterLateEntry}
 */
export const decodeAxelarVoteTalliedVoteIsVoterLateEntry = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1]);
/**
 * Destructures the fields of a {@link AxelarVoteTalliedVote} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: tally -
 *   - 2: data -
 *   - 3: poll_id -
 *   - 4: is_voter_late -
 */
export const destructAxelarVoteTalliedVote = (g_struct) => [safe_base64_to_bytes(g_struct.tally), __UNDEFINED, g_struct.data, g_struct.poll_id, g_struct.is_voter_late];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_VOTE_TALLIED_VOTE = '/axelar.vote.v1beta1.TalliedVote';
/**
 * Destructures the fields of a {@link AxelarVoteTalliedVoteIsVoterLateEntry} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 */
export const destructAxelarVoteTalliedVoteIsVoterLateEntry = (g_struct) => [g_struct.key, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_VOTE_TALLIED_VOTE_IS_VOTER_LATE_ENTRY = '/axelar.vote.v1beta1.TalliedVote.IsVoterLateEntry';
