// @solar-republic/cosmos-grpc/build/dist/proto/axelar/axelarnet/v1beta1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeAxelarAxelarnetParams } from './params.js';
/**
 * Encodes a `PendingIBCTransferCountRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetPendingIBCTransferCountRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `PendingIBCTransferCountResponse` protobuf message:
 * @param a_transfers_by_chains - `transfers_by_chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetPendingIBCTransferCountResponse = (a_transfers_by_chains) => Protobuf() // ...
    .B(a_transfers_by_chains) // TransfersByChainEntry[] transfers_by_chain = 1
    .o;
/**
 * Encodes a `ParamsRequest` protobuf message: represents a message that queries the params
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `ParamsResponse` protobuf message:
 * @param g_params - `params`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Encodes a `TransfersByChainEntry` protobuf message:
 * @param s_key - `key`:
 * @param n_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry = (s_key, n_value) => Protobuf() // ...
    .s(s_key) // string key = 1
    .v(n_value) // uint32 value = 2
    .o;
/**
 * Decodes a protobuf PendingIBCTransferCountRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetPendingIBCTransferCountRequest}
 */
export const decodeAxelarAxelarnetPendingIBCTransferCountRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf PendingIBCTransferCount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetPendingIBCTransferCountResponse}
 */
export const decodeAxelarAxelarnetPendingIBCTransferCountResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeAxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry]);
/**
 * Decodes a protobuf ParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetParamsRequest}
 */
export const decodeAxelarAxelarnetParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetParamsResponse}
 */
export const decodeAxelarAxelarnetParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarAxelarnetParams]);
/**
 * Decodes a protobuf TransfersByChainEntry message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry}
 */
export const decodeAxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1]);
/**
 * Destructures the fields of a {@link AxelarAxelarnetPendingIBCTransferCountRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAxelarAxelarnetPendingIBCTransferCountRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PENDING_I_B_C_TRANSFER_COUNT_REQUEST = '/axelar.axelarnet.v1beta1.PendingIBCTransferCountRequest';
/**
 * Destructures the fields of a {@link AxelarAxelarnetPendingIBCTransferCountResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: transfers_by_chain -
 */
export const destructAxelarAxelarnetPendingIBCTransferCountResponse = (g_struct) => [g_struct.transfers_by_chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PENDING_I_B_C_TRANSFER_COUNT_RESPONSE = '/axelar.axelarnet.v1beta1.PendingIBCTransferCountResponse';
/**
 * Destructures the fields of a {@link AxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 */
export const destructAxelarAxelarnetPendingIBCTransferCountResponseTransfersByChainEntry = (g_struct) => [g_struct.key, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PENDING_I_B_C_TRANSFER_COUNT_RESPONSE_TRANSFERS_BY_CHAIN_ENTRY = '/axelar.axelarnet.v1beta1.PendingIBCTransferCountResponse.TransfersByChainEntry';
/**
 * Destructures the fields of a {@link AxelarAxelarnetParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAxelarAxelarnetParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PARAMS_REQUEST = '/axelar.axelarnet.v1beta1.ParamsRequest';
/**
 * Destructures the fields of a {@link AxelarAxelarnetParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params -
 */
export const destructAxelarAxelarnetParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PARAMS_RESPONSE = '/axelar.axelarnet.v1beta1.ParamsResponse';
