// @solar-republic/neutrino/dist/cjs/secret-app.js
import { __UNDEFINED, odv, safe_json } from '@blake.regalia/belt';
import { exec_secret_contract, query_secret_contract_infer } from './app-layer.js';
export const exec_fees = (xg_limit, x_gas_price, s_denom = 'uscrt') => [[
        '' + Math.ceil(Number(xg_limit) * x_gas_price), s_denom
    ],
];
export const SecretApp = (k_wallet, k_contract, xn_gas_price, sa_granter) => ({
    price: (xn_price) => xn_gas_price = xn_price,
    granter: (sa_granter_new) => sa_granter = sa_granter_new,
    query: (si_method, h_args = __UNDEFINED, z_auth = __UNDEFINED) => query_secret_contract_infer(k_contract, si_method, h_args || {}, z_auth),
    exec: async (si_method, h_args, xg_limit, a_funds, s_memo) => {
        const [xc_code, s_res, g_tx, si_txn] = await exec_secret_contract(k_contract, k_wallet, { [si_method]: h_args }, exec_fees(xg_limit, xn_gas_price), xg_limit + '', sa_granter, a_funds, s_memo);
        const g_ans = safe_json(s_res) || {};
        return [xc_code ? __UNDEFINED : g_ans ? odv(g_ans)[0] : g_ans, xc_code, s_res, g_tx, si_txn];
    },
});
