// @solar-republic/cosmos-grpc/build/dist/proto/cosmwasm/wasm/v1/types.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
/**
 * Encodes a `AccessTypeParam` protobuf message:
 * @param xc_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmAccessTypeParam = (xc_value) => Protobuf() // ...
    .v(xc_value) // AccessType value = 1
    .o;
/**
 * Encodes a `AccessConfig` protobuf message: access control type.
 * @param xc_permission - `permission`:
 * @param a_addresses - `addresses`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmAccessConfig = (xc_permission, a_addresses) => Protobuf() // ...
    .v(xc_permission) // AccessType permission = 1
    .S(a_addresses, 3) // string[] addresses = 3
    .o;
/**
 * Encodes a `Params` protobuf message: defines the set of wasm parameters.
 * @param g_code_upload_access - `code_upload_access`:
 * @param xc_instantiate_default_permission - `instantiate_default_permission`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmParams = (atu8_code_upload_access, xc_instantiate_default_permission) => Protobuf() // ...
    .b(atu8_code_upload_access) // AccessConfig code_upload_access = 1
    .v(xc_instantiate_default_permission) // AccessType instantiate_default_permission = 2
    .o;
/**
 * Encodes a `CodeInfo` protobuf message: data for the uploaded contract WASM code
 * @param atu8_code_hash - `code_hash`: CodeHash is the unique identifier created by wasmvm
 * @param sa_creator - `creator`: Creator address who initially stored the code
 * @param g_instantiate_config - `instantiate_config`: InstantiateConfig access control to apply on contract creation, optional
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmCodeInfo = (atu8_code_hash, sa_creator, atu8_instantiate_config) => Protobuf() // ...
    .b(atu8_code_hash) // bytes code_hash = 1
    .s(sa_creator) // string creator = 2
    .b(atu8_instantiate_config, 5) // AccessConfig instantiate_config = 5
    .o;
/**
 * Encodes a `ContractInfo` protobuf message: stores a WASM contract instance
 * @param sg_code_id - `code_id`: CodeID is the reference to the stored Wasm code
 * @param sa_creator - `creator`: Creator address who initially instantiated the contract
 * @param s_admin - `admin`: Admin is an optional address that can execute migrations
 * @param s_label - `label`: Label is optional metadata to be stored with a contract instance.
 * @param g_created - `created`: Created Tx position when the contract was instantiated.
 * @param si_ibc_port - `ibc_port_id`:
 * @param atu8_extension - `extension`: Extension is an extension point to store custom metadata within the persistence model.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmContractInfo = (sg_code_id, sa_creator, s_admin, s_label, atu8_created, si_ibc_port, atu8_extension) => Protobuf() // ...
    .g(sg_code_id) // uint64 code_id = 1
    .s(sa_creator) // string creator = 2
    .s(s_admin) // string admin = 3
    .s(s_label) // string label = 4
    .b(atu8_created) // AbsoluteTxPosition created = 5
    .s(si_ibc_port) // string ibc_port_id = 6
    .b(atu8_extension) // Any extension = 7
    .o;
/**
 * Encodes a `ContractCodeHistoryEntry` protobuf message: metadata to a contract.
 * @param xc_operation - `operation`:
 * @param sg_code_id - `code_id`: CodeID is the reference to the stored WASM code
 * @param g_updated - `updated`: Updated Tx position when the operation was executed.
 * @param atu8_msg - `msg`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmContractCodeHistoryEntry = (xc_operation, sg_code_id, atu8_updated, atu8_msg) => Protobuf() // ...
    .v(xc_operation) // ContractCodeHistoryOperationType operation = 1
    .g(sg_code_id) // uint64 code_id = 2
    .b(atu8_updated) // AbsoluteTxPosition updated = 3
    .b(atu8_msg) // bytes msg = 4
    .o;
/**
 * Encodes a `AbsoluteTxPosition` protobuf message: a unique transaction position that allows for global ordering of transactions.
 * @param sg_block_height - `block_height`: BlockHeight is the block the contract was created at
 * @param sg_tx_index - `tx_index`: TxIndex is a monotonic counter within the block (actual transaction index, or gas consumed)
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmAbsoluteTxPosition = (sg_block_height, sg_tx_index) => Protobuf() // ...
    .g(sg_block_height) // uint64 block_height = 1
    .g(sg_tx_index) // uint64 tx_index = 2
    .o;
/**
 * Encodes a `Model` protobuf message: a struct that holds a KV pair
 * @param atu8_key - `key`: hex-encode key to read it better (this is often ascii)
 * @param atu8_value - `value`: base64-encode raw value
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmwasmWasmModel = (atu8_key, atu8_value) => Protobuf() // ...
    .b(atu8_key) // bytes key = 1
    .b(atu8_value) // bytes value = 2
    .o;
/**
 * Decodes a protobuf AccessTypeParam message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmAccessTypeParam}
 */
export const decodeCosmwasmWasmAccessTypeParam = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Decodes a protobuf AccessConfig message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedCosmwasmWasmAccessConfig}
 */
export const decodeCosmwasmWasmAccessConfig = ((atu8_payload, [xc_permission, , a_addresses] = decode_protobuf(atu8_payload, [1, 0, 4])) => [xc_permission, __UNDEFINED, a_addresses]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmParams}
 */
export const decodeCosmwasmWasmParams = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1], [decodeCosmwasmWasmAccessConfig]);
/**
 * Decodes a protobuf CodeInfo message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedCosmwasmWasmCodeInfo}
 */
export const decodeCosmwasmWasmCodeInfo = ((atu8_payload, [atu8_code_hash, sa_creator, , , a_instantiate_config] = decode_protobuf(atu8_payload, [1, 5, 0, 0, 1], [0, 0, 0, 0, decodeCosmwasmWasmAccessConfig])) => [atu8_code_hash, sa_creator, __UNDEFINED, __UNDEFINED, a_instantiate_config]);
/**
 * Decodes a protobuf ContractInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmContractInfo}
 */
export const decodeCosmwasmWasmContractInfo = (atu8_payload) => decode_protobuf(atu8_payload, [3, 5, 5, 5, 1, 5, 1], [0, 0, 0, 0, decodeCosmwasmWasmAbsoluteTxPosition, 0, decodeGoogleProtobufAny]);
/**
 * Decodes a protobuf ContractCodeHistoryEntry message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmContractCodeHistoryEntry}
 */
export const decodeCosmwasmWasmContractCodeHistoryEntry = (atu8_payload) => decode_protobuf(atu8_payload, [1, 3, 1, 1], [0, 0, decodeCosmwasmWasmAbsoluteTxPosition]);
/**
 * Decodes a protobuf AbsoluteTxPosition message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmAbsoluteTxPosition}
 */
export const decodeCosmwasmWasmAbsoluteTxPosition = (atu8_payload) => decode_protobuf(atu8_payload, [3, 3]);
/**
 * Decodes a protobuf Model message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmwasmWasmModel}
 */
export const decodeCosmwasmWasmModel = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Destructures the fields of a {@link CosmwasmWasmAccessConfig} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: permission -
 *   - 2: addresses -
 */
export const destructCosmwasmWasmAccessConfig = (g_struct) => [g_struct.permission, __UNDEFINED, g_struct.addresses];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_ACCESS_CONFIG = '/cosmwasm.wasm.v1.AccessConfig';
/**
 * Destructures the fields of a {@link CosmwasmWasmParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: code_upload_access -
 *   - 1: instantiate_default_permission -
 */
export const destructCosmwasmWasmParams = (g_struct) => [g_struct.code_upload_access, g_struct.instantiate_default_permission];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_PARAMS = '/cosmwasm.wasm.v1.Params';
/**
 * Destructures the fields of a {@link CosmwasmWasmAccessTypeParam} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: value -
 */
export const destructCosmwasmWasmAccessTypeParam = (g_struct) => [g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_ACCESS_TYPE_PARAM = '/cosmwasm.wasm.v1.AccessTypeParam';
/**
 * Destructures the fields of a {@link CosmwasmWasmCodeInfo} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: code_hash - CodeHash is the unique identifier created by wasmvm
 *   - 1: creator - Creator address who initially stored the code
 *   - 4: instantiate_config - InstantiateConfig access control to apply on contract creation, optional
 */
export const destructCosmwasmWasmCodeInfo = (g_struct) => [safe_base64_to_bytes(g_struct.code_hash), g_struct.creator, __UNDEFINED, __UNDEFINED, g_struct.instantiate_config];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_CODE_INFO = '/cosmwasm.wasm.v1.CodeInfo';
/**
 * Destructures the fields of a {@link CosmwasmWasmContractInfo} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: code_id - CodeID is the reference to the stored Wasm code
 *   - 1: creator - Creator address who initially instantiated the contract
 *   - 2: admin - Admin is an optional address that can execute migrations
 *   - 3: label - Label is optional metadata to be stored with a contract instance.
 *   - 4: created - Created Tx position when the contract was instantiated.
 *   - 5: ibc_port_id -
 *   - 6: extension - Extension is an extension point to store custom metadata within the persistence model.
 */
export const destructCosmwasmWasmContractInfo = (g_struct) => [g_struct.code_id, g_struct.creator, g_struct.admin, g_struct.label, g_struct.created, g_struct.ibc_port_id, g_struct.extension];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_CONTRACT_INFO = '/cosmwasm.wasm.v1.ContractInfo';
/**
 * Destructures the fields of a {@link CosmwasmWasmAbsoluteTxPosition} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: block_height - BlockHeight is the block the contract was created at
 *   - 1: tx_index - TxIndex is a monotonic counter within the block (actual transaction index, or gas consumed)
 */
export const destructCosmwasmWasmAbsoluteTxPosition = (g_struct) => [g_struct.block_height, g_struct.tx_index];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_ABSOLUTE_TX_POSITION = '/cosmwasm.wasm.v1.AbsoluteTxPosition';
/**
 * Destructures the fields of a {@link CosmwasmWasmContractCodeHistoryEntry} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: operation -
 *   - 1: code_id - CodeID is the reference to the stored WASM code
 *   - 2: updated - Updated Tx position when the operation was executed.
 *   - 3: msg -
 */
export const destructCosmwasmWasmContractCodeHistoryEntry = (g_struct) => [g_struct.operation, g_struct.code_id, g_struct.updated, safe_base64_to_bytes(g_struct.msg)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_CONTRACT_CODE_HISTORY_ENTRY = '/cosmwasm.wasm.v1.ContractCodeHistoryEntry';
/**
 * Destructures the fields of a {@link CosmwasmWasmModel} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key - hex-encode key to read it better (this is often ascii)
 *   - 1: value - base64-encode raw value
 */
export const destructCosmwasmWasmModel = (g_struct) => [safe_base64_to_bytes(g_struct.key), safe_base64_to_bytes(g_struct.value)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMWASM_WASM_MODEL = '/cosmwasm.wasm.v1.Model';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_UNSPECIFIED** - AccessTypeUnspecified placeholder for empty value
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmAccessType}
 */
export const XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED = 0;
/**
 * JSON enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_UNSPECIFIED** - AccessTypeUnspecified placeholder for empty value
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmAccessType}
 */
export const SI_JSON_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED = 'AccessTypeUnspecified';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_NOBODY** - AccessTypeNobody forbidden
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmAccessType}
 */
export const XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_NOBODY = 1;
/**
 * JSON enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_NOBODY** - AccessTypeNobody forbidden
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmAccessType}
 */
export const SI_JSON_COSMWASM_WASM_ACCESS_TYPE_NOBODY = 'AccessTypeNobody';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_EVERYBODY** - AccessTypeEverybody unrestricted
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmAccessType}
 */
export const XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY = 3;
/**
 * JSON enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_EVERYBODY** - AccessTypeEverybody unrestricted
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmAccessType}
 */
export const SI_JSON_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY = 'AccessTypeEverybody';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_ANY_OF_ADDRESSES** - AccessTypeAnyOfAddresses allow any of the addresses
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmAccessType}
 */
export const XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES = 4;
/**
 * JSON enum value for `cosmwasm.wasm.v1.AccessType`.
 *
 * **ACCESS_TYPE_ANY_OF_ADDRESSES** - AccessTypeAnyOfAddresses allow any of the addresses
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmAccessType}
 */
export const SI_JSON_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES = 'AccessTypeAnyOfAddresses';
/**
 * Maps a protobuf enum int value for `cosmwasm.wasm.v1.AccessType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumCosmwasmWasmAccessType = {
    [XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED]: SI_JSON_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED,
    [XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_NOBODY]: SI_JSON_COSMWASM_WASM_ACCESS_TYPE_NOBODY,
    [XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY]: SI_JSON_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY,
    [XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES]: SI_JSON_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES,
};
/**
 * Maps a JSON enum string value for `cosmwasm.wasm.v1.AccessType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumCosmwasmWasmAccessType = {
    [SI_JSON_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED]: XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_UNSPECIFIED,
    [SI_JSON_COSMWASM_WASM_ACCESS_TYPE_NOBODY]: XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_NOBODY,
    [SI_JSON_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY]: XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_EVERYBODY,
    [SI_JSON_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES]: XC_PROTO_COSMWASM_WASM_ACCESS_TYPE_ANY_OF_ADDRESSES,
};
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED** - ContractCodeHistoryOperationTypeUnspecified placeholder for empty value
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED = 0;
/**
 * JSON enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED** - ContractCodeHistoryOperationTypeUnspecified placeholder for empty value
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED = 'ContractCodeHistoryOperationTypeUnspecified';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT** - ContractCodeHistoryOperationTypeInit on chain contract instantiation
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT = 1;
/**
 * JSON enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT** - ContractCodeHistoryOperationTypeInit on chain contract instantiation
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT = 'ContractCodeHistoryOperationTypeInit';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE** - ContractCodeHistoryOperationTypeMigrate code migration
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE = 2;
/**
 * JSON enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE** - ContractCodeHistoryOperationTypeMigrate code migration
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE = 'ContractCodeHistoryOperationTypeMigrate';
/**
 * Protobuf enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS** - ContractCodeHistoryOperationTypeGenesis based on genesis data
 *
 * Belongs to enum type {@link ProtoEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS = 3;
/**
 * JSON enum value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType`.
 *
 * **CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS** - ContractCodeHistoryOperationTypeGenesis based on genesis data
 *
 * Belongs to enum type {@link JsonEnumCosmwasmWasmContractCodeHistoryOperationType}
 */
export const SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS = 'ContractCodeHistoryOperationTypeGenesis';
/**
 * Maps a protobuf enum int value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumCosmwasmWasmContractCodeHistoryOperationType = {
    [XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED]: SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED,
    [XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT]: SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT,
    [XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE]: SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE,
    [XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS]: SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS,
};
/**
 * Maps a JSON enum string value for `cosmwasm.wasm.v1.ContractCodeHistoryOperationType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumCosmwasmWasmContractCodeHistoryOperationType = {
    [SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED]: XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_UNSPECIFIED,
    [SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT]: XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_INIT,
    [SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE]: XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_MIGRATE,
    [SI_JSON_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS]: XC_PROTO_COSMWASM_WASM_CONTRACT_CODE_HISTORY_OPERATION_TYPE_GENESIS,
};
