// @solar-republic/cosmos-grpc/build/dist/proto/ibc/applications/interchain_accounts/host/v1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../../../api/transport.js';
import { decodeIbcApplicationsInterchainAccountsHostParams } from './host.js';
const SR_LCD_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_QUERY = '/ibc/apps/interchain_accounts/host/v1/params';
/**
 * Queries all parameters of the ICA host submodule.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const queryIbcApplicationsInterchainAccountsHostParams = restful_grpc(() => [SR_LCD_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_QUERY]);
/**
 * Encodes a `QueryParamsRequest` protobuf message: the request type for the Query/Params RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsHostQueryParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryParamsResponse` protobuf message: the response type for the Query/Params RPC method.
 * @param g_params - `params`: defines the parameters of the module.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsHostQueryParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Decodes a protobuf QueryParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsHostQueryParamsRequest}
 */
export const decodeIbcApplicationsInterchainAccountsHostQueryParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsHostQueryParamsResponse}
 */
export const decodeIbcApplicationsInterchainAccountsHostQueryParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeIbcApplicationsInterchainAccountsHostParams]);
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsHostQueryParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructIbcApplicationsInterchainAccountsHostQueryParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_QUERY_PARAMS_REQUEST = '/ibc.applications.interchain_accounts.host.v1.QueryParamsRequest';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsHostQueryParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const destructIbcApplicationsInterchainAccountsHostQueryParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_QUERY_PARAMS_RESPONSE = '/ibc.applications.interchain_accounts.host.v1.QueryParamsResponse';
