// @solar-republic/cosmos-grpc/build/dist/proto/ibc/applications/interchain_accounts/host/v1/tx.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../../api/protobuf-writer.js';
import { decodeIbcApplicationsInterchainAccountsHostParams } from './host.js';
/**
 * Encodes a `MsgUpdateParams` protobuf message: defines the payload for Msg/UpdateParams
 * @param s_signer - `signer`: address
 * @param g_params - `params`: defines the 27-interchain-accounts/host parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsHostMsgUpdateParams = (s_signer, atu8_params) => Protobuf() // ...
    .s(s_signer) // string signer = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response for Msg/UpdateParams
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsHostMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsHostMsgUpdateParams}
 */
export const decodeIbcApplicationsInterchainAccountsHostMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeIbcApplicationsInterchainAccountsHostParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsHostMsgUpdateParamsResponse}
 */
export const decodeIbcApplicationsInterchainAccountsHostMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsHostMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: signer - address
 *   - 1: params - defines the 27-interchain-accounts/host parameters to update. NOTE: All parameters must be supplied.
 */
export const destructIbcApplicationsInterchainAccountsHostMsgUpdateParams = (g_struct) => [g_struct.signer, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_MSG_UPDATE_PARAMS = '/ibc.applications.interchain_accounts.host.v1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsHostMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructIbcApplicationsInterchainAccountsHostMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_HOST_MSG_UPDATE_PARAMS_RESPONSE = '/ibc.applications.interchain_accounts.host.v1.MsgUpdateParamsResponse';
