// @solar-republic/cosmos-grpc/build/dist/proto/google/protobuf/descriptor.js
import { __UNDEFINED, oda } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../api/protobuf-reader.js';
import { Protobuf } from '../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../api/util.js';
/**
 * Encodes a `FileDescriptorProto` protobuf message: Describes a complete .proto file.
 * @param s_name - `name`:
 * @param s_package - `package`:
 * @param a_dependencys - `dependency`: Names of files imported by this file.
 * @param a_message_types - `message_type`: Indexes of the public imported files in the dependency list above.
 * @param a_enum_types - `enum_type`: Indexes of the weak imported files in the dependency list. For Google-internal migration only. Do not use.
 * @param a_services - `service`: All top-level definitions in this file.
 * @param a_extensions - `extension`:
 * @param g_options - `options`:
 * @param g_source_code_info - `source_code_info`:
 * @param a_public_dependencys - `public_dependency`:
 * @param a_weak_dependencys - `weak_dependency`: This field contains optional information about the original source code. You may safely remove this entire field without harming runtime functionality of the descriptors -- the information is needed only by development tools.
 * @param s_syntax - `syntax`: The syntax of the proto file. The supported values are "proto2", "proto3", and "editions". If `edition` is present, this value must be "editions".
 * @param xc_edition - `edition`: The edition of the proto file.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFileDescriptorProto = (s_name, s_package, a_dependencys, a_message_types, a_enum_types, a_services, a_extensions, atu8_options, atu8_source_code_info, a_public_dependencys, a_weak_dependencys, s_syntax, xc_edition) => Protobuf() // ...
    .s(s_name) // string name = 1
    .s(s_package) // string package = 2
    .S(a_dependencys) // string[] dependency = 3
    .B(a_message_types) // DescriptorProto[] message_type = 4
    .B(a_enum_types) // EnumDescriptorProto[] enum_type = 5
    .B(a_services) // ServiceDescriptorProto[] service = 6
    .B(a_extensions) // FieldDescriptorProto[] extension = 7
    .b(atu8_options) // FileOptions options = 8
    .b(atu8_source_code_info) // SourceCodeInfo source_code_info = 9
    .V(a_public_dependencys) // int32[] public_dependency = 10
    .V(a_weak_dependencys) // int32[] weak_dependency = 11
    .s(s_syntax) // string syntax = 12
    .v(xc_edition, 14) // Edition edition = 14
    .o;
/**
 * Encodes a `DescriptorProto` protobuf message: Describes a message type.
 * @param s_name - `name`:
 * @param a_fields - `field`:
 * @param a_nested_types - `nested_type`:
 * @param a_enum_types - `enum_type`:
 * @param a_extension_ranges - `extension_range`:
 * @param a_extensions - `extension`:
 * @param g_options - `options`:
 * @param a_oneof_decls - `oneof_decl`:
 * @param a_reserved_ranges - `reserved_range`:
 * @param a_reserved_names - `reserved_name`: Reserved field names, which may not be used by fields in the same message. A given name may only be reserved once.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufDescriptorProto = (s_name, a_fields, a_nested_types, a_enum_types, a_extension_ranges, a_extensions, atu8_options, a_oneof_decls, a_reserved_ranges, a_reserved_names) => Protobuf() // ...
    .s(s_name) // string name = 1
    .B(a_fields) // FieldDescriptorProto[] field = 2
    .B(a_nested_types) // DescriptorProto[] nested_type = 3
    .B(a_enum_types) // EnumDescriptorProto[] enum_type = 4
    .B(a_extension_ranges) // ExtensionRange[] extension_range = 5
    .B(a_extensions) // FieldDescriptorProto[] extension = 6
    .b(atu8_options) // MessageOptions options = 7
    .B(a_oneof_decls) // OneofDescriptorProto[] oneof_decl = 8
    .B(a_reserved_ranges) // ReservedRange[] reserved_range = 9
    .S(a_reserved_names) // string[] reserved_name = 10
    .o;
/**
 * Encodes a `FieldDescriptorProto` protobuf message: Describes a field within a message.
 * @param s_name - `name`:
 * @param s_extendee - `extendee`:
 * @param n_number - `number`:
 * @param xc_label - `label`: If type_name is set, this need not be set.  If both this and type_name are set, this must be one of TYPE_ENUM, TYPE_MESSAGE or TYPE_GROUP.
 * @param xc_type - `type`: For message and enum types, this is the name of the type.  If the name starts with a '.', it is fully-qualified.  Otherwise, C++-like scoping rules are used to find the type (i.e. first the nested types within this message are searched, then within the parent, on up to the root namespace).
 * @param s_type_name - `type_name`: For extensions, this is the name of the type being extended.  It is resolved in the same manner as type_name.
 * @param s_default_value - `default_value`: For numeric types, contains the original text representation of the value. For booleans, "true" or "false". For strings, contains the default text contents (not escaped in any way). For bytes, contains the C escaped value.  All bytes >= 128 are escaped.
 * @param g_options - `options`: If set, gives the index of a oneof in the containing type's oneof_decl list.  This field is a member of that oneof.
 * @param n_oneof_index - `oneof_index`: JSON name of this field. The value is set by protocol compiler. If the user has set a "json_name" option on this field, that option's value will be used. Otherwise, it's deduced from the field's name by converting it to camelCase.
 * @param s_json_name - `json_name`:
 * @param b_proto3_optional - `proto3_optional`: If true, this is a proto3 "optional". When a proto3 field is optional, it tracks presence regardless of field type. When proto3_optional is true, this field must be belong to a oneof to signal to old proto3 clients that presence is tracked for this field. This oneof is known as a "synthetic" oneof, and this field must be its sole member (each proto3 optional field gets its own synthetic oneof). Synthetic oneofs exist in the descriptor only, and do not generate any API. Synthetic oneofs must be ordered after all "real" oneofs. For message fields, proto3_optional doesn't create any semantic change, since non-repeated message fields always track presence. However it still indicates the semantic detail of whether the user wrote "optional" or not. This can be useful for round-tripping the .proto file. For consistency we give message fields a synthetic oneof also, even though it is not required to track presence. This is especially important because the parser can't tell if a field is a message or an enum, so it must always create a synthetic oneof. Proto2 optional fields do not set this flag, because they already indicate optional with `LABEL_OPTIONAL`.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFieldDescriptorProto = (s_name, s_extendee, n_number, xc_label, xc_type, s_type_name, s_default_value, atu8_options, n_oneof_index, s_json_name, b_proto3_optional) => Protobuf() // ...
    .s(s_name) // string name = 1
    .s(s_extendee) // string extendee = 2
    .v(n_number) // int32 number = 3
    .v(xc_label) // Label label = 4
    .v(xc_type) // Type type = 5
    .s(s_type_name) // string type_name = 6
    .s(s_default_value) // string default_value = 7
    .b(atu8_options) // FieldOptions options = 8
    .v(n_oneof_index) // int32 oneof_index = 9
    .s(s_json_name) // string json_name = 10
    .v(b_proto3_optional, 17) // boolean proto3_optional = 17
    .o;
/**
 * Encodes a `FieldOptions` protobuf message:
 * @param xc_ctype - `ctype`: The ctype option instructs the C++ code generator to use a different representation of the field than it normally would.  See the specific options below.  This option is only implemented to support use of [ctype=CORD] and [ctype=STRING] (the default) on non-repeated fields of type "bytes" in the open source release -- sorry, we'll try to include other types in a future version!
 * @param b_packed - `packed`: The packed option can be enabled for repeated primitive fields to enable a more efficient representation on the wire. Rather than repeatedly writing the tag and type for each element, the entire array is encoded as a single length-delimited blob. In proto3, only explicit setting it to false will avoid using packed encoding.  This option is prohibited in Editions, but the `repeated_field_encoding` feature can be used to control the behavior.
 * @param b_deprecated - `deprecated`: The jstype option determines the JavaScript type used for values of the field.  The option is permitted only for 64 bit integral and fixed types (int64, uint64, sint64, fixed64, sfixed64).  A field with jstype JS_STRING is represented as JavaScript string, which avoids loss of precision that can happen when a large value is converted to a floating point JavaScript. Specifying JS_NUMBER for the jstype causes the generated JavaScript code to use the JavaScript "number" type.  The behavior of the default option JS_NORMAL is implementation dependent. This option is an enum to permit additional types to be added, e.g. goog.math.Integer.
 * @param b_lazy - `lazy`: Should this field be parsed lazily?  Lazy applies only to message-type fields.  It means that when the outer message is initially parsed, the inner message's contents will not be parsed but instead stored in encoded form.  The inner message will actually be parsed when it is first accessed. This is only a hint.  Implementations are free to choose whether to use eager or lazy parsing regardless of the value of this option.  However, setting this option true suggests that the protocol author believes that using lazy parsing on this field is worth the additional bookkeeping overhead typically needed to implement it. This option does not affect the public interface of any generated code; all method signatures remain the same.  Furthermore, thread-safety of the interface is not affected by this option; const methods remain safe to call from multiple threads concurrently, while non-const methods continue to require exclusive access. Note that implementations may choose not to check required fields within a lazy sub-message.  That is, calling IsInitialized() on the outer message may return true even if the inner message has missing required fields. This is necessary because otherwise the inner message would have to be parsed in order to perform the check, defeating the purpose of lazy parsing.  An implementation which chooses not to check required fields must be consistent about it.  That is, for any particular sub-message, the implementation must either *always* check its required fields, or *never* check its required fields, regardless of whether or not the message has been parsed. As of May 2022, lazy verifies the contents of the byte stream during parsing.  An invalid byte stream will cause the overall parsing to fail.
 * @param xc_jstype - `jstype`: unverified_lazy does no correctness checks on the byte stream. This should only be used where lazy with verification is prohibitive for performance reasons.
 * @param b_weak - `weak`: Is this field deprecated? Depending on the target platform, this can emit Deprecated annotations for accessors, or it will be completely ignored; in the very least, this is a formalization for deprecating fields.
 * @param b_unverified_lazy - `unverified_lazy`: For Google-internal migration only. Do not use.
 * @param b_debug_redact - `debug_redact`: Indicate that the field value should not be printed out when using debug formats, e.g. when the field contains sensitive credentials.
 * @param xc_retention - `retention`:
 * @param a_targets - `targets`:
 * @param a_edition_defaults - `edition_defaults`:
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFieldOptions = (xc_ctype, b_packed, b_deprecated, b_lazy, xc_jstype, b_weak, b_unverified_lazy, b_debug_redact, xc_retention, a_targets, a_edition_defaults, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .v(xc_ctype) // CType ctype = 1
    .v(b_packed) // boolean packed = 2
    .v(b_deprecated) // boolean deprecated = 3
    .v(b_lazy, 5) // boolean lazy = 5
    .v(xc_jstype, 6) // JSType jstype = 6
    .v(b_weak, 10) // boolean weak = 10
    .v(b_unverified_lazy, 15) // boolean unverified_lazy = 15
    .v(b_debug_redact, 16) // boolean debug_redact = 16
    .v(xc_retention, 17) // OptionRetention retention = 17
    .V(a_targets, 19) // OptionTargetType[] targets = 19
    .B(a_edition_defaults, 20) // EditionDefault[] edition_defaults = 20
    .b(atu8_features, 21) // FeatureSet features = 21
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `EditionDefault` protobuf message:
 * @param s_value - `value`: Default mode.
 * @param xc_edition - `edition`: The option [ctype=CORD] may be applied to a non-repeated field of type "bytes". It indicates that in C++, the data should be stored in a Cord instead of a string.  For very large strings, this may reduce memory fragmentation. It may also allow better performance when parsing from a Cord, or when parsing with aliasing enabled, as the parsed Cord may then alias the original buffer.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFieldOptionsEditionDefault = (s_value, xc_edition) => Protobuf() // ...
    .s(s_value, 2) // string value = 2
    .v(xc_edition, 3) // Edition edition = 3
    .o;
/**
 * Encodes a `FeatureSet` protobuf message: TODO Enums in C++ gencode (and potentially other languages) are not well scoped.  This means that each of the feature enums below can clash with each other.  The short names we've chosen maximize call-site readability, but leave us very open to this scenario.  A future feature will be designed and implemented to handle this, hopefully before we ever hit a conflict here.
 * @param xc_field_presence - `field_presence`:
 * @param xc_enum_type - `enum_type`:
 * @param xc_repeated_field_encoding - `repeated_field_encoding`:
 * @param xc_utf8_validation - `utf8_validation`:
 * @param xc_message_encoding - `message_encoding`:
 * @param xc_json_format - `json_format`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFeatureSet = (xc_field_presence, xc_enum_type, xc_repeated_field_encoding, xc_utf8_validation, xc_message_encoding, xc_json_format) => Protobuf() // ...
    .v(xc_field_presence) // FieldPresence field_presence = 1
    .v(xc_enum_type) // EnumType enum_type = 2
    .v(xc_repeated_field_encoding) // RepeatedFieldEncoding repeated_field_encoding = 3
    .v(xc_utf8_validation) // Utf8Validation utf8_validation = 4
    .v(xc_message_encoding) // MessageEncoding message_encoding = 5
    .v(xc_json_format) // JsonFormat json_format = 6
    .o;
/**
 * Encodes a `UninterpretedOption` protobuf message: A message representing a option the parser does not recognize. This only appears in options protos created by the compiler::Parser class. DescriptorPool resolves these when building Descriptor objects. Therefore, options protos in descriptor objects (e.g. returned by Descriptor::options(), or produced by Descriptor::CopyTo()) will never have UninterpretedOptions in them.
 * @param a_names - `name`:
 * @param s_identifier_value - `identifier_value`: The value of the uninterpreted option, in whatever type the tokenizer identified it as during parsing. Exactly one of these should be set.
 * @param sg_positive_int_value - `positive_int_value`:
 * @param sg_negative_int_value - `negative_int_value`:
 * @param x_double_value - `double_value`:
 * @param atu8_string_value - `string_value`:
 * @param s_aggregate_value - `aggregate_value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufUninterpretedOption = (a_names, s_identifier_value, sg_positive_int_value, sg_negative_int_value, x_double_value, atu8_string_value, s_aggregate_value) => Protobuf() // ...
    .B(a_names, 2) // NamePart[] name = 2
    .s(s_identifier_value, 3) // string identifier_value = 3
    .g(sg_positive_int_value, 4) // uint64 positive_int_value = 4
    .g(sg_negative_int_value, 5) // int32 negative_int_value = 5
    .i(x_double_value, 6) // unknown double_value = 6
    .b(atu8_string_value, 7) // bytes string_value = 7
    .s(s_aggregate_value, 8) // string aggregate_value = 8
    .o;
/**
 * Encodes a `NamePart` protobuf message:
 * @param s_name_part - `name_part`:
 * @param b_is_extension - `is_extension`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufUninterpretedOptionNamePart = (s_name_part, b_is_extension) => Protobuf() // ...
    .s(s_name_part) // string name_part = 1
    .v(b_is_extension) // boolean is_extension = 2
    .o;
/**
 * Encodes a `EnumDescriptorProto` protobuf message: Describes an enum type.
 * @param s_name - `name`:
 * @param a_values - `value`:
 * @param g_options - `options`:
 * @param a_reserved_ranges - `reserved_range`: Range of reserved numeric values. Reserved numeric values may not be used by enum values in the same enum declaration. Reserved ranges may not overlap.
 * @param a_reserved_names - `reserved_name`: Reserved enum value names, which may not be reused. A given name may only be reserved once.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufEnumDescriptorProto = (s_name, a_values, atu8_options, a_reserved_ranges, a_reserved_names) => Protobuf() // ...
    .s(s_name) // string name = 1
    .B(a_values) // EnumValueDescriptorProto[] value = 2
    .b(atu8_options) // EnumOptions options = 3
    .B(a_reserved_ranges) // EnumReservedRange[] reserved_range = 4
    .S(a_reserved_names) // string[] reserved_name = 5
    .o;
/**
 * Encodes a `EnumValueDescriptorProto` protobuf message: Describes a value within an enum.
 * @param s_name - `name`:
 * @param n_number - `number`:
 * @param g_options - `options`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufEnumValueDescriptorProto = (s_name, n_number, atu8_options) => Protobuf() // ...
    .s(s_name) // string name = 1
    .v(n_number) // int32 number = 2
    .b(atu8_options) // EnumValueOptions options = 3
    .o;
/**
 * Encodes a `EnumValueOptions` protobuf message:
 * @param b_deprecated - `deprecated`: Is this enum value deprecated? Depending on the target platform, this can emit Deprecated annotations for the enum value, or it will be completely ignored; in the very least, this is a formalization for deprecating enum values.
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param b_debug_redact - `debug_redact`: Indicate that fields annotated with this enum value should not be printed out when using debug formats, e.g. when the field contains sensitive credentials.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufEnumValueOptions = (b_deprecated, atu8_features, b_debug_redact, a_uninterpreted_options) => Protobuf() // ...
    .v(b_deprecated) // boolean deprecated = 1
    .b(atu8_features) // FeatureSet features = 2
    .v(b_debug_redact) // boolean debug_redact = 3
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `EnumOptions` protobuf message:
 * @param b_allow_alias - `allow_alias`: Set this option to true to allow mapping different tag names to the same value.
 * @param b_deprecated - `deprecated`: Is this enum deprecated? Depending on the target platform, this can emit Deprecated annotations for the enum, or it will be completely ignored; in the very least, this is a formalization for deprecating enums.
 * @param b_deprecated_legacy_json_field_conflicts - `deprecated_legacy_json_field_conflicts`: Enable the legacy handling of JSON field name conflicts.  This lowercases and strips underscored from the fields before comparison in proto3 only. The new behavior takes `json_name` into account and applies to proto2 as well. TODO Remove this legacy behavior once downstream teams have had time to migrate.
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufEnumOptions = (b_allow_alias, b_deprecated, b_deprecated_legacy_json_field_conflicts, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .v(b_allow_alias, 2) // boolean allow_alias = 2
    .v(b_deprecated, 3) // boolean deprecated = 3
    .v(b_deprecated_legacy_json_field_conflicts, 6) // boolean deprecated_legacy_json_field_conflicts = 6
    .b(atu8_features, 7) // FeatureSet features = 7
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `EnumReservedRange` protobuf message:
 * @param n_start - `start`:
 * @param n_end - `end`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufEnumDescriptorProtoEnumReservedRange = (n_start, n_end) => Protobuf() // ...
    .v(n_start) // int32 start = 1
    .v(n_end) // int32 end = 2
    .o;
/**
 * Encodes a `ExtensionRange` protobuf message:
 * @param n_start - `start`:
 * @param n_end - `end`:
 * @param g_options - `options`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufDescriptorProtoExtensionRange = (n_start, n_end, atu8_options) => Protobuf() // ...
    .v(n_start) // int32 start = 1
    .v(n_end) // int32 end = 2
    .b(atu8_options) // ExtensionRangeOptions options = 3
    .o;
/**
 * Encodes a `ExtensionRangeOptions` protobuf message:
 * @param a_declarations - `declaration`: The parser stores options it doesn't recognize here. See above.
 * @param xc_verification - `verification`: For external users: DO NOT USE. We are in the process of open sourcing extension declaration and executing internal cleanups before it can be used externally.
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The verification state of the range. TODO: flip the default to DECLARATION once all empty ranges are marked as UNVERIFIED.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufExtensionRangeOptions = (a_declarations, xc_verification, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .B(a_declarations, 2) // Declaration[] declaration = 2
    .v(xc_verification, 3) // VerificationState verification = 3
    .b(atu8_features, 50) // FeatureSet features = 50
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `Declaration` protobuf message: The verification state of the extension range.
 * @param n_number - `number`: All the extensions of the range must be declared.
 * @param s_full_name - `full_name`:
 * @param s_type - `type`:
 * @param b_reserved - `reserved`:
 * @param b_repeated - `repeated`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufExtensionRangeOptionsDeclaration = (n_number, s_full_name, s_type, b_reserved, b_repeated) => Protobuf() // ...
    .v(n_number) // int32 number = 1
    .s(s_full_name) // string full_name = 2
    .s(s_type) // string type = 3
    .v(b_reserved, 5) // boolean reserved = 5
    .v(b_repeated, 6) // boolean repeated = 6
    .o;
/**
 * Encodes a `MessageOptions` protobuf message:
 * @param b_message_set_wire_format - `message_set_wire_format`: Set true to use the old proto1 MessageSet wire format for extensions. This is provided for backwards-compatibility with the MessageSet wire format.  You should not use this for any other reason:  It's less efficient, has fewer features, and is more complicated. The message must be defined exactly as follows: message Foo { option message_set_wire_format = true; extensions 4 to max; } Note that the message cannot have any defined fields; MessageSets only have extensions. All extensions of your type must be singular messages; e.g. they cannot be int32s, enums, or repeated messages. Because this is an option, the above two restrictions are not enforced by the protocol compiler.
 * @param b_no_standard_descriptor_accessor - `no_standard_descriptor_accessor`: Disables the generation of the standard "descriptor()" accessor, which can conflict with a field of the same name.  This is meant to make migration from proto1 easier; new code should avoid fields named "descriptor".
 * @param b_deprecated - `deprecated`: Is this message deprecated? Depending on the target platform, this can emit Deprecated annotations for the message, or it will be completely ignored; in the very least, this is a formalization for deprecating messages.
 * @param b_map_entry - `map_entry`: NOTE: Do not set the option in .proto files. Always use the maps syntax instead. The option should only be implicitly set by the proto compiler parser. Whether the message is an automatically generated map entry type for the maps field. For maps fields: map<KeyType, ValueType> map_field = 1; The parsed descriptor looks like: message MapFieldEntry { option map_entry = true; optional KeyType key = 1; optional ValueType value = 2; } repeated MapFieldEntry map_field = 1; Implementations may choose not to generate the map_entry=true message, but use a native map in the target language to hold the keys and values. The reflection APIs in such implementations still need to work as if the field is a repeated message field.
 * @param b_deprecated_legacy_json_field_conflicts - `deprecated_legacy_json_field_conflicts`: Enable the legacy handling of JSON field name conflicts.  This lowercases and strips underscored from the fields before comparison in proto3 only. The new behavior takes `json_name` into account and applies to proto2 as well. This should only be used as a temporary measure against broken builds due to the change in behavior for JSON field name conflicts. TODO This is legacy behavior we plan to remove once downstream teams have had time to migrate.
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufMessageOptions = (b_message_set_wire_format, b_no_standard_descriptor_accessor, b_deprecated, b_map_entry, b_deprecated_legacy_json_field_conflicts, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .v(b_message_set_wire_format) // boolean message_set_wire_format = 1
    .v(b_no_standard_descriptor_accessor) // boolean no_standard_descriptor_accessor = 2
    .v(b_deprecated) // boolean deprecated = 3
    .v(b_map_entry, 7) // boolean map_entry = 7
    .v(b_deprecated_legacy_json_field_conflicts, 11) // boolean deprecated_legacy_json_field_conflicts = 11
    .b(atu8_features, 12) // FeatureSet features = 12
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `OneofDescriptorProto` protobuf message: Describes a oneof.
 * @param s_name - `name`:
 * @param g_options - `options`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufOneofDescriptorProto = (s_name, atu8_options) => Protobuf() // ...
    .s(s_name) // string name = 1
    .b(atu8_options) // OneofOptions options = 2
    .o;
/**
 * Encodes a `OneofOptions` protobuf message:
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufOneofOptions = (atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .b(atu8_features) // FeatureSet features = 1
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `ReservedRange` protobuf message:
 * @param n_start - `start`:
 * @param n_end - `end`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufDescriptorProtoReservedRange = (n_start, n_end) => Protobuf() // ...
    .v(n_start) // int32 start = 1
    .v(n_end) // int32 end = 2
    .o;
/**
 * Encodes a `ServiceDescriptorProto` protobuf message: Describes a service.
 * @param s_name - `name`:
 * @param a_methods - `method`:
 * @param g_options - `options`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufServiceDescriptorProto = (s_name, a_methods, atu8_options) => Protobuf() // ...
    .s(s_name) // string name = 1
    .B(a_methods) // MethodDescriptorProto[] method = 2
    .b(atu8_options) // ServiceOptions options = 3
    .o;
/**
 * Encodes a `MethodDescriptorProto` protobuf message: Describes a method of a service.
 * @param s_name - `name`:
 * @param s_input_type - `input_type`: Input and output type names.  These are resolved in the same way as FieldDescriptorProto.type_name, but must refer to a message type.
 * @param s_output_type - `output_type`:
 * @param g_options - `options`:
 * @param b_client_streaming - `client_streaming`: Identifies if client streams multiple client messages
 * @param b_server_streaming - `server_streaming`: Identifies if server streams multiple server messages
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufMethodDescriptorProto = (s_name, s_input_type, s_output_type, atu8_options, b_client_streaming, b_server_streaming) => Protobuf() // ...
    .s(s_name) // string name = 1
    .s(s_input_type) // string input_type = 2
    .s(s_output_type) // string output_type = 3
    .b(atu8_options) // MethodOptions options = 4
    .v(b_client_streaming) // boolean client_streaming = 5
    .v(b_server_streaming) // boolean server_streaming = 6
    .o;
/**
 * Encodes a `MethodOptions` protobuf message:
 * @param b_deprecated - `deprecated`: Is this method deprecated? Depending on the target platform, this can emit Deprecated annotations for the method, or it will be completely ignored; in the very least, this is a formalization for deprecating methods.
 * @param xc_idempotency_level - `idempotency_level`:
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufMethodOptions = (b_deprecated, xc_idempotency_level, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .v(b_deprecated, 33) // boolean deprecated = 33
    .v(xc_idempotency_level, 34) // IdempotencyLevel idempotency_level = 34
    .b(atu8_features, 35) // FeatureSet features = 35
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `ServiceOptions` protobuf message:
 * @param b_deprecated - `deprecated`: Any features defined in the specific edition.
 * @param g_features - `features`: Is this service deprecated? Depending on the target platform, this can emit Deprecated annotations for the service, or it will be completely ignored; in the very least, this is a formalization for deprecating services.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufServiceOptions = (b_deprecated, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .v(b_deprecated, 33) // boolean deprecated = 33
    .b(atu8_features, 34) // FeatureSet features = 34
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `FileOptions` protobuf message:
 * @param s_java_package - `java_package`: Sets the Java package where classes generated from this .proto will be placed.  By default, the proto package is used, but this is often inappropriate because proto packages do not normally start with backwards domain names.
 * @param s_java_outer_classname - `java_outer_classname`: Controls the name of the wrapper Java class generated for the .proto file. That class will always contain the .proto file's getDescriptor() method as well as any top-level extensions defined in the .proto file. If java_multiple_files is disabled, then all the other classes from the .proto file will be nested inside the single wrapper outer class.
 * @param xc_optimize_for - `optimize_for`: If enabled, then the Java code generator will generate a separate .java file for each top-level message, enum, and service defined in the .proto file.  Thus, these types will *not* be nested inside the wrapper class named by java_outer_classname.  However, the wrapper class will still be generated to contain the file's getDescriptor() method as well as any top-level extensions defined in the file.
 * @param b_java_multiple_files - `java_multiple_files`: This option does nothing.
 * @param s_go_package - `go_package`: If set true, then the Java2 code generator will generate code that throws an exception whenever an attempt is made to assign a non-UTF-8 byte sequence to a string field. Message reflection will do the same. However, an extension field still accepts non-UTF-8 byte sequences. This option has no effect on when used with the lite runtime.
 * @param b_cc_generic_services - `cc_generic_services`:
 * @param b_java_generic_services - `java_generic_services`: Sets the Go package where structs generated from this .proto will be placed. If omitted, the Go package will be derived from the following: - The basename of the package import path, if provided. - Otherwise, the package statement in the .proto file, if present. - Otherwise, the basename of the .proto file, without extension.
 * @param b_py_generic_services - `py_generic_services`: Should generic services be generated in each language?  "Generic" services are not specific to any particular RPC system.  They are generated by the main code generators in each language (without additional plugins). Generic services were the only kind of service generation supported by early versions of google.protobuf. Generic services are now considered deprecated in favor of using plugins that generate code specific to your particular RPC system.  Therefore, these default to false.  Old code which depends on generic services should explicitly set them to true.
 * @param b_java_generate_equals_and_hash - `java_generate_equals_and_hash`:
 * @param b_deprecated - `deprecated`:
 * @param b_java_string_check_utf8 - `java_string_check_utf8`:
 * @param b_cc_enable_arenas - `cc_enable_arenas`: Is this file deprecated? Depending on the target platform, this can emit Deprecated annotations for everything in the file, or it will be completely ignored; in the very least, this is a formalization for deprecating files.
 * @param s_objc_class_prefix - `objc_class_prefix`: Enables the use of arenas for the proto messages in this file. This applies only to generated classes for C++.
 * @param s_csharp_namespace - `csharp_namespace`: Sets the objective c class prefix which is prepended to all objective c generated classes from this .proto. There is no default.
 * @param s_swift_prefix - `swift_prefix`: Namespace for generated classes; defaults to the package.
 * @param s_php_class_prefix - `php_class_prefix`: By default Swift generators will take the proto package and CamelCase it replacing '.' with underscore and use that to prefix the types/symbols defined. When this options is provided, they will use this value instead to prefix the types/symbols defined.
 * @param s_php_namespace - `php_namespace`: Sets the php class prefix which is prepended to all php generated classes from this .proto. Default is empty.
 * @param b_php_generic_services - `php_generic_services`: Use this option to change the namespace of php generated classes. Default is empty. When this option is empty, the package name will be used for determining the namespace.
 * @param s_php_metadata_namespace - `php_metadata_namespace`: Use this option to change the namespace of php generated metadata classes. Default is empty. When this option is empty, the proto file name will be used for determining the namespace.
 * @param s_ruby_package - `ruby_package`: Use this option to change the package of ruby generated classes. Default is empty. When this option is not set, the package name will be used for determining the ruby package.
 * @param g_features - `features`: Any features defined in the specific edition.
 * @param a_uninterpreted_options - `uninterpreted_option`: The parser stores options it doesn't recognize here. See the documentation for the "Options" section above.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufFileOptions = (s_java_package, s_java_outer_classname, xc_optimize_for, b_java_multiple_files, s_go_package, b_cc_generic_services, b_java_generic_services, b_py_generic_services, b_java_generate_equals_and_hash, b_deprecated, b_java_string_check_utf8, b_cc_enable_arenas, s_objc_class_prefix, s_csharp_namespace, s_swift_prefix, s_php_class_prefix, s_php_namespace, b_php_generic_services, s_php_metadata_namespace, s_ruby_package, atu8_features, a_uninterpreted_options) => Protobuf() // ...
    .s(s_java_package) // string java_package = 1
    .s(s_java_outer_classname, 8) // string java_outer_classname = 8
    .v(xc_optimize_for, 9) // OptimizeMode optimize_for = 9
    .v(b_java_multiple_files, 10) // boolean java_multiple_files = 10
    .s(s_go_package, 11) // string go_package = 11
    .v(b_cc_generic_services, 16) // boolean cc_generic_services = 16
    .v(b_java_generic_services, 17) // boolean java_generic_services = 17
    .v(b_py_generic_services, 18) // boolean py_generic_services = 18
    .v(b_java_generate_equals_and_hash, 20) // boolean java_generate_equals_and_hash = 20
    .v(b_deprecated, 23) // boolean deprecated = 23
    .v(b_java_string_check_utf8, 27) // boolean java_string_check_utf8 = 27
    .v(b_cc_enable_arenas, 31) // boolean cc_enable_arenas = 31
    .s(s_objc_class_prefix, 36) // string objc_class_prefix = 36
    .s(s_csharp_namespace, 37) // string csharp_namespace = 37
    .s(s_swift_prefix, 39) // string swift_prefix = 39
    .s(s_php_class_prefix, 40) // string php_class_prefix = 40
    .s(s_php_namespace, 41) // string php_namespace = 41
    .v(b_php_generic_services, 42) // boolean php_generic_services = 42
    .s(s_php_metadata_namespace, 44) // string php_metadata_namespace = 44
    .s(s_ruby_package, 45) // string ruby_package = 45
    .b(atu8_features, 50) // FeatureSet features = 50
    .B(a_uninterpreted_options, 999) // UninterpretedOption[] uninterpreted_option = 999
    .o;
/**
 * Encodes a `SourceCodeInfo` protobuf message: Encapsulates information about the original source file from which a FileDescriptorProto was generated.
 * @param a_locations - `location`: A Location identifies a piece of source code in a .proto file which corresponds to a particular definition.  This information is intended to be useful to IDEs, code indexers, documentation generators, and similar tools. For example, say we have a file like: message Foo { optional string foo = 1; } Let's look at just the field definition: optional string foo = 1; ^       ^^     ^^  ^  ^^^ a       bc     de  f  ghi We have the following locations: span   path               represents [a,i)  [ 4, 0, 2, 0 ]     The whole field definition. [a,b)  [ 4, 0, 2, 0, 4 ]  The label (optional). [c,d)  [ 4, 0, 2, 0, 5 ]  The type (string). [e,f)  [ 4, 0, 2, 0, 1 ]  The name (foo). [g,h)  [ 4, 0, 2, 0, 3 ]  The number (1). Notes: - A location may refer to a repeated field itself (i.e. not to any particular index within it).  This is used whenever a set of elements are logically enclosed in a single code segment.  For example, an entire extend block (possibly containing multiple extension definitions) will have an outer location whose path refers to the "extensions" repeated field without an index. - Multiple locations may have the same path.  This happens when a single logical declaration is spread out across multiple places.  The most obvious example is the "extend" block again -- there may be multiple extend blocks in the same scope, each of which will have the same path. - A location's span is not always a subset of its parent's span.  For example, the "extendee" of an extension declaration appears at the beginning of the "extend" block and is shared by all extensions within the block. - Just because a location's span is a subset of some other location's span does not mean that it is a descendant.  For example, a "group" defines both a type and a field in a single declaration.  Thus, the locations corresponding to the type and field and their components will overlap. - Code which tries to interpret locations should probably be designed to ignore those that it doesn't understand, as more types of locations could be recorded in the future.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufSourceCodeInfo = (a_locations) => Protobuf() // ...
    .B(a_locations) // Location[] location = 1
    .o;
/**
 * Encodes a `Location` protobuf message:
 * @param a_paths - `path`:
 * @param a_spans - `span`:
 * @param s_leading_comments - `leading_comments`:
 * @param s_trailing_comments - `trailing_comments`:
 * @param a_leading_detached_comments - `leading_detached_comments`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGoogleProtobufSourceCodeInfoLocation = (a_paths, a_spans, s_leading_comments, s_trailing_comments, a_leading_detached_comments) => Protobuf() // ...
    .V(a_paths) // int32[] path = 1
    .V(a_spans) // int32[] span = 2
    .s(s_leading_comments) // string leading_comments = 3
    .s(s_trailing_comments) // string trailing_comments = 4
    .S(a_leading_detached_comments, 6) // string[] leading_detached_comments = 6
    .o;
/**
 * Decodes a protobuf FileDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufFileDescriptorProto}
 */
export const decodeGoogleProtobufFileDescriptorProto = ((atu8_payload, [s_name, s_package, a_dependencys, a_message_type, a_enum_type, a_service, a_extension, a_options, a_source_code_info, a_public_dependencys, a_weak_dependencys, s_syntax, , xc_edition] = decode_protobuf(atu8_payload, [5, 5, 4, 0, 0, 0, 0, 1, 1, 0, 0, 5, 0, 1], [0, 0, 0, decodeGoogleProtobufDescriptorProto, decodeGoogleProtobufEnumDescriptorProto, decodeGoogleProtobufServiceDescriptorProto, decodeGoogleProtobufFieldDescriptorProto, decodeGoogleProtobufFileOptions, decodeGoogleProtobufSourceCodeInfo])) => [s_name, s_package, a_dependencys, a_message_type, a_enum_type, a_service, a_extension, a_options, a_source_code_info, a_public_dependencys, a_weak_dependencys, s_syntax, __UNDEFINED, xc_edition]);
/**
 * Decodes a protobuf DescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufDescriptorProto}
 */
export const decodeGoogleProtobufDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0, 0, 0, 0, 0, 1, 0, 0, 4], [0, decodeGoogleProtobufFieldDescriptorProto, decodeGoogleProtobufDescriptorProto, decodeGoogleProtobufEnumDescriptorProto, decodeGoogleProtobufDescriptorProtoExtensionRange, decodeGoogleProtobufFieldDescriptorProto, decodeGoogleProtobufMessageOptions, decodeGoogleProtobufOneofDescriptorProto, decodeGoogleProtobufDescriptorProtoReservedRange]);
/**
 * Decodes a protobuf FieldDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufFieldDescriptorProto}
 */
export const decodeGoogleProtobufFieldDescriptorProto = ((atu8_payload, [s_name, s_extendee, n_number, xc_label, xc_type, s_type_name, s_default_value, a_options, n_oneof_index, s_json_name, , , , , , , b_proto3_optional] = decode_protobuf(atu8_payload, [5, 5, 1, 1, 1, 5, 5, 1, 1, 5, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFieldOptions])) => [s_name, s_extendee, n_number, xc_label, xc_type, s_type_name, s_default_value, a_options, n_oneof_index, s_json_name, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_proto3_optional]);
/**
 * Decodes a protobuf FieldOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufFieldOptions}
 */
export const decodeGoogleProtobufFieldOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFieldOptionsEditionDefault, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [xc_ctype, b_packed, b_deprecated, , b_lazy, xc_jstype, , , , b_weak, , , , , b_unverified_lazy, b_debug_redact, xc_retention, , a_targets, a_edition_defaults, a_features] = a_decoded) => oda([xc_ctype, b_packed, b_deprecated, __UNDEFINED, b_lazy, xc_jstype, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_weak, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_unverified_lazy, b_debug_redact, xc_retention, __UNDEFINED, a_targets, a_edition_defaults, a_features], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf EditionDefault message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufFieldOptionsEditionDefault}
 */
export const decodeGoogleProtobufFieldOptionsEditionDefault = ((atu8_payload, [, s_value, xc_edition] = decode_protobuf(atu8_payload, [0, 5, 1])) => [__UNDEFINED, s_value, xc_edition]);
/**
 * Decodes a protobuf FeatureSet message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufFeatureSet}
 */
export const decodeGoogleProtobufFeatureSet = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1, 1, 1, 1]);
/**
 * Decodes a protobuf UninterpretedOption message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufUninterpretedOption}
 */
export const decodeGoogleProtobufUninterpretedOption = ((atu8_payload, [, a_name, s_identifier_value, sg_positive_int_value, sg_negative_int_value, x_double_value, atu8_string_value, s_aggregate_value] = decode_protobuf(atu8_payload, [0, 0, 5, 3, 3, 1, 1, 5], [0, decodeGoogleProtobufUninterpretedOptionNamePart])) => [__UNDEFINED, a_name, s_identifier_value, sg_positive_int_value, sg_negative_int_value, x_double_value, atu8_string_value, s_aggregate_value]);
/**
 * Decodes a protobuf NamePart message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufUninterpretedOptionNamePart}
 */
export const decodeGoogleProtobufUninterpretedOptionNamePart = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1]);
/**
 * Decodes a protobuf EnumDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufEnumDescriptorProto}
 */
export const decodeGoogleProtobufEnumDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0, 1, 0, 4], [0, decodeGoogleProtobufEnumValueDescriptorProto, decodeGoogleProtobufEnumOptions, decodeGoogleProtobufEnumDescriptorProtoEnumReservedRange]);
/**
 * Decodes a protobuf EnumValueDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufEnumValueDescriptorProto}
 */
export const decodeGoogleProtobufEnumValueDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 1], [0, 0, decodeGoogleProtobufEnumValueOptions]);
/**
 * Decodes a protobuf EnumValueOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufEnumValueOptions}
 */
export const decodeGoogleProtobufEnumValueOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [1, 1, 1], [0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [b_deprecated, a_features, b_debug_redact] = a_decoded) => oda([b_deprecated, a_features, b_debug_redact], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf EnumOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufEnumOptions}
 */
export const decodeGoogleProtobufEnumOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [0, 1, 1, 0, 0, 1, 1], [0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [, b_allow_alias, b_deprecated, , , b_deprecated_legacy_json_field_conflicts, a_features] = a_decoded) => oda([__UNDEFINED, b_allow_alias, b_deprecated, __UNDEFINED, __UNDEFINED, b_deprecated_legacy_json_field_conflicts, a_features], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf EnumReservedRange message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufEnumDescriptorProtoEnumReservedRange}
 */
export const decodeGoogleProtobufEnumDescriptorProtoEnumReservedRange = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Decodes a protobuf ExtensionRange message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufDescriptorProtoExtensionRange}
 */
export const decodeGoogleProtobufDescriptorProtoExtensionRange = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1], [0, 0, decodeGoogleProtobufExtensionRangeOptions]);
/**
 * Decodes a protobuf ExtensionRangeOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufExtensionRangeOptions}
 */
export const decodeGoogleProtobufExtensionRangeOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [0, 0, 1], [0, decodeGoogleProtobufExtensionRangeOptionsDeclaration, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 49: a_features, 998: a_uninterpreted_option } = a_decoded, [, a_declaration, xc_verification] = a_decoded) => oda([__UNDEFINED, a_declaration, xc_verification], {
    49: a_features,
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf Declaration message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufExtensionRangeOptionsDeclaration}
 */
export const decodeGoogleProtobufExtensionRangeOptionsDeclaration = ((atu8_payload, [n_number, s_full_name, s_type, , b_reserved, b_repeated] = decode_protobuf(atu8_payload, [1, 5, 5, 0, 1, 1])) => [n_number, s_full_name, s_type, __UNDEFINED, b_reserved, b_repeated]);
/**
 * Decodes a protobuf MessageOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufMessageOptions}
 */
export const decodeGoogleProtobufMessageOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [b_message_set_wire_format, b_no_standard_descriptor_accessor, b_deprecated, , , , b_map_entry, , , , b_deprecated_legacy_json_field_conflicts, a_features] = a_decoded) => oda([b_message_set_wire_format, b_no_standard_descriptor_accessor, b_deprecated, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_map_entry, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_deprecated_legacy_json_field_conflicts, a_features], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf OneofDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufOneofDescriptorProto}
 */
export const decodeGoogleProtobufOneofDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeGoogleProtobufOneofOptions]);
/**
 * Decodes a protobuf OneofOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufOneofOptions}
 */
export const decodeGoogleProtobufOneofOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [1], [decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [a_features] = a_decoded) => oda([a_features], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf ReservedRange message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufDescriptorProtoReservedRange}
 */
export const decodeGoogleProtobufDescriptorProtoReservedRange = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Decodes a protobuf ServiceDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufServiceDescriptorProto}
 */
export const decodeGoogleProtobufServiceDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0, 1], [0, decodeGoogleProtobufMethodDescriptorProto, decodeGoogleProtobufServiceOptions]);
/**
 * Decodes a protobuf MethodDescriptorProto message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufMethodDescriptorProto}
 */
export const decodeGoogleProtobufMethodDescriptorProto = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5, 1, 1, 1], [0, 0, 0, decodeGoogleProtobufMethodOptions]);
/**
 * Decodes a protobuf MethodOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufMethodOptions}
 */
export const decodeGoogleProtobufMethodOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, __UNDEFINED, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 32: b_deprecated, 33: xc_idempotency_level, 34: a_features, 998: a_uninterpreted_option } = a_decoded) => oda([], {
    32: b_deprecated,
    33: xc_idempotency_level,
    34: a_features,
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf ServiceOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufServiceOptions}
 */
export const decodeGoogleProtobufServiceOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, __UNDEFINED, [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 32: b_deprecated, 33: a_features, 998: a_uninterpreted_option } = a_decoded) => oda([], {
    32: b_deprecated,
    33: a_features,
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf FileOptions message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufFileOptions}
 */
export const decodeGoogleProtobufFileOptions = ((atu8_payload, a_decoded = decode_protobuf(atu8_payload, [5, 0, 0, 0, 0, 0, 0, 5, 1, 1, 5, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 5, 5, 0, 5, 5, 5, 1, 0, 5, 5, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufFeatureSet, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, decodeGoogleProtobufUninterpretedOption]), { 998: a_uninterpreted_option } = a_decoded, [s_java_package, , , , , , , s_java_outer_classname, xc_optimize_for, b_java_multiple_files, s_go_package, , , , , b_cc_generic_services, b_java_generic_services, b_py_generic_services, , b_java_generate_equals_and_hash, , , b_deprecated, , , , b_java_string_check_utf8, , , , b_cc_enable_arenas, , , , , s_objc_class_prefix, s_csharp_namespace, , s_swift_prefix, s_php_class_prefix, s_php_namespace, b_php_generic_services, , s_php_metadata_namespace, s_ruby_package, , , , , a_features] = a_decoded) => oda([s_java_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, s_java_outer_classname, xc_optimize_for, b_java_multiple_files, s_go_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_cc_generic_services, b_java_generic_services, b_py_generic_services, __UNDEFINED, b_java_generate_equals_and_hash, __UNDEFINED, __UNDEFINED, b_deprecated, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_java_string_check_utf8, __UNDEFINED, __UNDEFINED, __UNDEFINED, b_cc_enable_arenas, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, s_objc_class_prefix, s_csharp_namespace, __UNDEFINED, s_swift_prefix, s_php_class_prefix, s_php_namespace, b_php_generic_services, __UNDEFINED, s_php_metadata_namespace, s_ruby_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, a_features], {
    998: a_uninterpreted_option,
}));
/**
 * Decodes a protobuf SourceCodeInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGoogleProtobufSourceCodeInfo}
 */
export const decodeGoogleProtobufSourceCodeInfo = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeGoogleProtobufSourceCodeInfoLocation]);
/**
 * Decodes a protobuf Location message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedGoogleProtobufSourceCodeInfoLocation}
 */
export const decodeGoogleProtobufSourceCodeInfoLocation = ((atu8_payload, [a_paths, a_spans, s_leading_comments, s_trailing_comments, , a_leading_detached_comments] = decode_protobuf(atu8_payload, [0, 0, 5, 5, 0, 4])) => [a_paths, a_spans, s_leading_comments, s_trailing_comments, __UNDEFINED, a_leading_detached_comments]);
/**
 * Destructures the fields of a {@link GoogleProtobufFileDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: package -
 *   - 2: dependency - Names of files imported by this file.
 *   - 3: message_type - Indexes of the public imported files in the dependency list above.
 *   - 4: enum_type - Indexes of the weak imported files in the dependency list. For Google-internal migration only. Do not use.
 *   - 5: service - All top-level definitions in this file.
 *   - 6: extension -
 *   - 7: options -
 *   - 8: source_code_info -
 *   - 9: public_dependency -
 *   - 10: weak_dependency - This field contains optional information about the original source code. You may safely remove this entire field without harming runtime functionality of the descriptors -- the information is needed only by development tools.
 *   - 11: syntax - The syntax of the proto file. The supported values are "proto2", "proto3", and "editions". If `edition` is present, this value must be "editions".
 *   - 13: edition - The edition of the proto file.
 */
export const destructGoogleProtobufFileDescriptorProto = (g_struct) => [g_struct.name, g_struct.package, g_struct.dependency, g_struct.message_type, g_struct.enum_type, g_struct.service, g_struct.extension, g_struct.options, g_struct.source_code_info, g_struct.public_dependency, g_struct.weak_dependency, g_struct.syntax, __UNDEFINED, g_struct.edition];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FILE_DESCRIPTOR_PROTO = '/google.protobuf.FileDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: field -
 *   - 2: nested_type -
 *   - 3: enum_type -
 *   - 4: extension_range -
 *   - 5: extension -
 *   - 6: options -
 *   - 7: oneof_decl -
 *   - 8: reserved_range -
 *   - 9: reserved_name - Reserved field names, which may not be used by fields in the same message. A given name may only be reserved once.
 */
export const destructGoogleProtobufDescriptorProto = (g_struct) => [g_struct.name, g_struct.field, g_struct.nested_type, g_struct.enum_type, g_struct.extension_range, g_struct.extension, g_struct.options, g_struct.oneof_decl, g_struct.reserved_range, g_struct.reserved_name];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_DESCRIPTOR_PROTO = '/google.protobuf.DescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufFieldDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: extendee -
 *   - 2: number -
 *   - 3: label - If type_name is set, this need not be set.  If both this and type_name are set, this must be one of TYPE_ENUM, TYPE_MESSAGE or TYPE_GROUP.
 *   - 4: type - For message and enum types, this is the name of the type.  If the name starts with a '.', it is fully-qualified.  Otherwise, C++-like scoping rules are used to find the type (i.e. first the nested types within this message are searched, then within the parent, on up to the root namespace).
 *   - 5: type_name - For extensions, this is the name of the type being extended.  It is resolved in the same manner as type_name.
 *   - 6: default_value - For numeric types, contains the original text representation of the value. For booleans, "true" or "false". For strings, contains the default text contents (not escaped in any way). For bytes, contains the C escaped value.  All bytes >= 128 are escaped.
 *   - 7: options - If set, gives the index of a oneof in the containing type's oneof_decl list.  This field is a member of that oneof.
 *   - 8: oneof_index - JSON name of this field. The value is set by protocol compiler. If the user has set a "json_name" option on this field, that option's value will be used. Otherwise, it's deduced from the field's name by converting it to camelCase.
 *   - 9: json_name -
 *   - 16: proto3_optional - If true, this is a proto3 "optional". When a proto3 field is optional, it tracks presence regardless of field type. When proto3_optional is true, this field must be belong to a oneof to signal to old proto3 clients that presence is tracked for this field. This oneof is known as a "synthetic" oneof, and this field must be its sole member (each proto3 optional field gets its own synthetic oneof). Synthetic oneofs exist in the descriptor only, and do not generate any API. Synthetic oneofs must be ordered after all "real" oneofs. For message fields, proto3_optional doesn't create any semantic change, since non-repeated message fields always track presence. However it still indicates the semantic detail of whether the user wrote "optional" or not. This can be useful for round-tripping the .proto file. For consistency we give message fields a synthetic oneof also, even though it is not required to track presence. This is especially important because the parser can't tell if a field is a message or an enum, so it must always create a synthetic oneof. Proto2 optional fields do not set this flag, because they already indicate optional with `LABEL_OPTIONAL`.
 */
export const destructGoogleProtobufFieldDescriptorProto = (g_struct) => [g_struct.name, g_struct.extendee, g_struct.number, g_struct.label, g_struct.type, g_struct.type_name, g_struct.default_value, g_struct.options, g_struct.oneof_index, g_struct.json_name, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.proto3_optional];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO = '/google.protobuf.FieldDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufFieldOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: ctype - The ctype option instructs the C++ code generator to use a different representation of the field than it normally would.  See the specific options below.  This option is only implemented to support use of [ctype=CORD] and [ctype=STRING] (the default) on non-repeated fields of type "bytes" in the open source release -- sorry, we'll try to include other types in a future version!
 *   - 1: packed - The packed option can be enabled for repeated primitive fields to enable a more efficient representation on the wire. Rather than repeatedly writing the tag and type for each element, the entire array is encoded as a single length-delimited blob. In proto3, only explicit setting it to false will avoid using packed encoding.  This option is prohibited in Editions, but the `repeated_field_encoding` feature can be used to control the behavior.
 *   - 2: deprecated - The jstype option determines the JavaScript type used for values of the field.  The option is permitted only for 64 bit integral and fixed types (int64, uint64, sint64, fixed64, sfixed64).  A field with jstype JS_STRING is represented as JavaScript string, which avoids loss of precision that can happen when a large value is converted to a floating point JavaScript. Specifying JS_NUMBER for the jstype causes the generated JavaScript code to use the JavaScript "number" type.  The behavior of the default option JS_NORMAL is implementation dependent. This option is an enum to permit additional types to be added, e.g. goog.math.Integer.
 *   - 4: lazy - Should this field be parsed lazily?  Lazy applies only to message-type fields.  It means that when the outer message is initially parsed, the inner message's contents will not be parsed but instead stored in encoded form.  The inner message will actually be parsed when it is first accessed. This is only a hint.  Implementations are free to choose whether to use eager or lazy parsing regardless of the value of this option.  However, setting this option true suggests that the protocol author believes that using lazy parsing on this field is worth the additional bookkeeping overhead typically needed to implement it. This option does not affect the public interface of any generated code; all method signatures remain the same.  Furthermore, thread-safety of the interface is not affected by this option; const methods remain safe to call from multiple threads concurrently, while non-const methods continue to require exclusive access. Note that implementations may choose not to check required fields within a lazy sub-message.  That is, calling IsInitialized() on the outer message may return true even if the inner message has missing required fields. This is necessary because otherwise the inner message would have to be parsed in order to perform the check, defeating the purpose of lazy parsing.  An implementation which chooses not to check required fields must be consistent about it.  That is, for any particular sub-message, the implementation must either *always* check its required fields, or *never* check its required fields, regardless of whether or not the message has been parsed. As of May 2022, lazy verifies the contents of the byte stream during parsing.  An invalid byte stream will cause the overall parsing to fail.
 *   - 5: jstype - unverified_lazy does no correctness checks on the byte stream. This should only be used where lazy with verification is prohibitive for performance reasons.
 *   - 9: weak - Is this field deprecated? Depending on the target platform, this can emit Deprecated annotations for accessors, or it will be completely ignored; in the very least, this is a formalization for deprecating fields.
 *   - 14: unverified_lazy - For Google-internal migration only. Do not use.
 *   - 15: debug_redact - Indicate that the field value should not be printed out when using debug formats, e.g. when the field contains sensitive credentials.
 *   - 16: retention -
 *   - 18: targets -
 *   - 19: edition_defaults -
 *   - 20: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufFieldOptions = (g_struct) => oda([g_struct.ctype, g_struct.packed, g_struct.deprecated, __UNDEFINED, g_struct.lazy, g_struct.jstype, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.weak, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.unverified_lazy, g_struct.debug_redact, g_struct.retention, __UNDEFINED, g_struct.targets, g_struct.edition_defaults, g_struct.features], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FIELD_OPTIONS = '/google.protobuf.FieldOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufFieldOptionsEditionDefault} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: value - Default mode.
 *   - 2: edition - The option [ctype=CORD] may be applied to a non-repeated field of type "bytes". It indicates that in C++, the data should be stored in a Cord instead of a string.  For very large strings, this may reduce memory fragmentation. It may also allow better performance when parsing from a Cord, or when parsing with aliasing enabled, as the parsed Cord may then alias the original buffer.
 */
export const destructGoogleProtobufFieldOptionsEditionDefault = (g_struct) => [__UNDEFINED, g_struct.value, g_struct.edition];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FIELD_OPTIONS_EDITION_DEFAULT = '/google.protobuf.FieldOptions.EditionDefault';
/**
 * Destructures the fields of a {@link GoogleProtobufFeatureSet} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: field_presence -
 *   - 1: enum_type -
 *   - 2: repeated_field_encoding -
 *   - 3: utf8_validation -
 *   - 4: message_encoding -
 *   - 5: json_format -
 */
export const destructGoogleProtobufFeatureSet = (g_struct) => [g_struct.field_presence, g_struct.enum_type, g_struct.repeated_field_encoding, g_struct.utf8_validation, g_struct.message_encoding, g_struct.json_format];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FEATURE_SET = '/google.protobuf.FeatureSet';
/**
 * Destructures the fields of a {@link GoogleProtobufUninterpretedOption} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: name -
 *   - 2: identifier_value - The value of the uninterpreted option, in whatever type the tokenizer identified it as during parsing. Exactly one of these should be set.
 *   - 3: positive_int_value -
 *   - 4: negative_int_value -
 *   - 5: double_value -
 *   - 6: string_value -
 *   - 7: aggregate_value -
 */
export const destructGoogleProtobufUninterpretedOption = (g_struct) => [__UNDEFINED, g_struct.name, g_struct.identifier_value, g_struct.positive_int_value, g_struct.negative_int_value, g_struct.double_value, safe_base64_to_bytes(g_struct.string_value), g_struct.aggregate_value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_UNINTERPRETED_OPTION = '/google.protobuf.UninterpretedOption';
/**
 * Destructures the fields of a {@link GoogleProtobufUninterpretedOptionNamePart} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name_part -
 *   - 1: is_extension -
 */
export const destructGoogleProtobufUninterpretedOptionNamePart = (g_struct) => [g_struct.name_part, g_struct.is_extension];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_UNINTERPRETED_OPTION_NAME_PART = '/google.protobuf.UninterpretedOption.NamePart';
/**
 * Destructures the fields of a {@link GoogleProtobufEnumDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: value -
 *   - 2: options -
 *   - 3: reserved_range - Range of reserved numeric values. Reserved numeric values may not be used by enum values in the same enum declaration. Reserved ranges may not overlap.
 *   - 4: reserved_name - Reserved enum value names, which may not be reused. A given name may only be reserved once.
 */
export const destructGoogleProtobufEnumDescriptorProto = (g_struct) => [g_struct.name, g_struct.value, g_struct.options, g_struct.reserved_range, g_struct.reserved_name];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ENUM_DESCRIPTOR_PROTO = '/google.protobuf.EnumDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufEnumValueDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: number -
 *   - 2: options -
 */
export const destructGoogleProtobufEnumValueDescriptorProto = (g_struct) => [g_struct.name, g_struct.number, g_struct.options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ENUM_VALUE_DESCRIPTOR_PROTO = '/google.protobuf.EnumValueDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufEnumValueOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: deprecated - Is this enum value deprecated? Depending on the target platform, this can emit Deprecated annotations for the enum value, or it will be completely ignored; in the very least, this is a formalization for deprecating enum values.
 *   - 1: features - Any features defined in the specific edition.
 *   - 2: debug_redact - Indicate that fields annotated with this enum value should not be printed out when using debug formats, e.g. when the field contains sensitive credentials.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufEnumValueOptions = (g_struct) => oda([g_struct.deprecated, g_struct.features, g_struct.debug_redact], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ENUM_VALUE_OPTIONS = '/google.protobuf.EnumValueOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufEnumOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: allow_alias - Set this option to true to allow mapping different tag names to the same value.
 *   - 2: deprecated - Is this enum deprecated? Depending on the target platform, this can emit Deprecated annotations for the enum, or it will be completely ignored; in the very least, this is a formalization for deprecating enums.
 *   - 5: deprecated_legacy_json_field_conflicts - Enable the legacy handling of JSON field name conflicts.  This lowercases and strips underscored from the fields before comparison in proto3 only. The new behavior takes `json_name` into account and applies to proto2 as well. TODO Remove this legacy behavior once downstream teams have had time to migrate.
 *   - 6: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufEnumOptions = (g_struct) => oda([__UNDEFINED, g_struct.allow_alias, g_struct.deprecated, __UNDEFINED, __UNDEFINED, g_struct.deprecated_legacy_json_field_conflicts, g_struct.features], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ENUM_OPTIONS = '/google.protobuf.EnumOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufEnumDescriptorProtoEnumReservedRange} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: start -
 *   - 1: end -
 */
export const destructGoogleProtobufEnumDescriptorProtoEnumReservedRange = (g_struct) => [g_struct.start, g_struct.end];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ENUM_DESCRIPTOR_PROTO_ENUM_RESERVED_RANGE = '/google.protobuf.EnumDescriptorProto.EnumReservedRange';
/**
 * Destructures the fields of a {@link GoogleProtobufDescriptorProtoExtensionRange} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: start -
 *   - 1: end -
 *   - 2: options -
 */
export const destructGoogleProtobufDescriptorProtoExtensionRange = (g_struct) => [g_struct.start, g_struct.end, g_struct.options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_DESCRIPTOR_PROTO_EXTENSION_RANGE = '/google.protobuf.DescriptorProto.ExtensionRange';
/**
 * Destructures the fields of a {@link GoogleProtobufExtensionRangeOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: declaration - The parser stores options it doesn't recognize here. See above.
 *   - 2: verification - For external users: DO NOT USE. We are in the process of open sourcing extension declaration and executing internal cleanups before it can be used externally.
 *   - 49: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The verification state of the range. TODO: flip the default to DECLARATION once all empty ranges are marked as UNVERIFIED.
 */
export const destructGoogleProtobufExtensionRangeOptions = (g_struct) => oda([__UNDEFINED, g_struct.declaration, g_struct.verification], {
    50: g_struct.features,
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS = '/google.protobuf.ExtensionRangeOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufExtensionRangeOptionsDeclaration} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: number - All the extensions of the range must be declared.
 *   - 1: full_name -
 *   - 2: type -
 *   - 4: reserved -
 *   - 5: repeated -
 */
export const destructGoogleProtobufExtensionRangeOptionsDeclaration = (g_struct) => [g_struct.number, g_struct.full_name, g_struct.type, __UNDEFINED, g_struct.reserved, g_struct.repeated];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION = '/google.protobuf.ExtensionRangeOptions.Declaration';
/**
 * Destructures the fields of a {@link GoogleProtobufMessageOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: message_set_wire_format - Set true to use the old proto1 MessageSet wire format for extensions. This is provided for backwards-compatibility with the MessageSet wire format.  You should not use this for any other reason:  It's less efficient, has fewer features, and is more complicated. The message must be defined exactly as follows: message Foo { option message_set_wire_format = true; extensions 4 to max; } Note that the message cannot have any defined fields; MessageSets only have extensions. All extensions of your type must be singular messages; e.g. they cannot be int32s, enums, or repeated messages. Because this is an option, the above two restrictions are not enforced by the protocol compiler.
 *   - 1: no_standard_descriptor_accessor - Disables the generation of the standard "descriptor()" accessor, which can conflict with a field of the same name.  This is meant to make migration from proto1 easier; new code should avoid fields named "descriptor".
 *   - 2: deprecated - Is this message deprecated? Depending on the target platform, this can emit Deprecated annotations for the message, or it will be completely ignored; in the very least, this is a formalization for deprecating messages.
 *   - 6: map_entry - NOTE: Do not set the option in .proto files. Always use the maps syntax instead. The option should only be implicitly set by the proto compiler parser. Whether the message is an automatically generated map entry type for the maps field. For maps fields: map<KeyType, ValueType> map_field = 1; The parsed descriptor looks like: message MapFieldEntry { option map_entry = true; optional KeyType key = 1; optional ValueType value = 2; } repeated MapFieldEntry map_field = 1; Implementations may choose not to generate the map_entry=true message, but use a native map in the target language to hold the keys and values. The reflection APIs in such implementations still need to work as if the field is a repeated message field.
 *   - 10: deprecated_legacy_json_field_conflicts - Enable the legacy handling of JSON field name conflicts.  This lowercases and strips underscored from the fields before comparison in proto3 only. The new behavior takes `json_name` into account and applies to proto2 as well. This should only be used as a temporary measure against broken builds due to the change in behavior for JSON field name conflicts. TODO This is legacy behavior we plan to remove once downstream teams have had time to migrate.
 *   - 11: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufMessageOptions = (g_struct) => oda([g_struct.message_set_wire_format, g_struct.no_standard_descriptor_accessor, g_struct.deprecated, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.map_entry, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.deprecated_legacy_json_field_conflicts, g_struct.features], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_MESSAGE_OPTIONS = '/google.protobuf.MessageOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufOneofDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: options -
 */
export const destructGoogleProtobufOneofDescriptorProto = (g_struct) => [g_struct.name, g_struct.options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ONEOF_DESCRIPTOR_PROTO = '/google.protobuf.OneofDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufOneofOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufOneofOptions = (g_struct) => oda([g_struct.features], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_ONEOF_OPTIONS = '/google.protobuf.OneofOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufDescriptorProtoReservedRange} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: start -
 *   - 1: end -
 */
export const destructGoogleProtobufDescriptorProtoReservedRange = (g_struct) => [g_struct.start, g_struct.end];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_DESCRIPTOR_PROTO_RESERVED_RANGE = '/google.protobuf.DescriptorProto.ReservedRange';
/**
 * Destructures the fields of a {@link GoogleProtobufServiceDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: method -
 *   - 2: options -
 */
export const destructGoogleProtobufServiceDescriptorProto = (g_struct) => [g_struct.name, g_struct.method, g_struct.options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_SERVICE_DESCRIPTOR_PROTO = '/google.protobuf.ServiceDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufMethodDescriptorProto} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: input_type - Input and output type names.  These are resolved in the same way as FieldDescriptorProto.type_name, but must refer to a message type.
 *   - 2: output_type -
 *   - 3: options -
 *   - 4: client_streaming - Identifies if client streams multiple client messages
 *   - 5: server_streaming - Identifies if server streams multiple server messages
 */
export const destructGoogleProtobufMethodDescriptorProto = (g_struct) => [g_struct.name, g_struct.input_type, g_struct.output_type, g_struct.options, g_struct.client_streaming, g_struct.server_streaming];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_METHOD_DESCRIPTOR_PROTO = '/google.protobuf.MethodDescriptorProto';
/**
 * Destructures the fields of a {@link GoogleProtobufMethodOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 32: deprecated - Is this method deprecated? Depending on the target platform, this can emit Deprecated annotations for the method, or it will be completely ignored; in the very least, this is a formalization for deprecating methods.
 *   - 33: idempotency_level -
 *   - 34: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufMethodOptions = (g_struct) => oda([], {
    33: g_struct.deprecated,
    34: g_struct.idempotency_level,
    35: g_struct.features,
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_METHOD_OPTIONS = '/google.protobuf.MethodOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufServiceOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 32: deprecated - Any features defined in the specific edition.
 *   - 33: features - Is this service deprecated? Depending on the target platform, this can emit Deprecated annotations for the service, or it will be completely ignored; in the very least, this is a formalization for deprecating services.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See above.
 */
export const destructGoogleProtobufServiceOptions = (g_struct) => oda([], {
    33: g_struct.deprecated,
    34: g_struct.features,
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_SERVICE_OPTIONS = '/google.protobuf.ServiceOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufFileOptions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: java_package - Sets the Java package where classes generated from this .proto will be placed.  By default, the proto package is used, but this is often inappropriate because proto packages do not normally start with backwards domain names.
 *   - 7: java_outer_classname - Controls the name of the wrapper Java class generated for the .proto file. That class will always contain the .proto file's getDescriptor() method as well as any top-level extensions defined in the .proto file. If java_multiple_files is disabled, then all the other classes from the .proto file will be nested inside the single wrapper outer class.
 *   - 8: optimize_for - If enabled, then the Java code generator will generate a separate .java file for each top-level message, enum, and service defined in the .proto file.  Thus, these types will *not* be nested inside the wrapper class named by java_outer_classname.  However, the wrapper class will still be generated to contain the file's getDescriptor() method as well as any top-level extensions defined in the file.
 *   - 9: java_multiple_files - This option does nothing.
 *   - 10: go_package - If set true, then the Java2 code generator will generate code that throws an exception whenever an attempt is made to assign a non-UTF-8 byte sequence to a string field. Message reflection will do the same. However, an extension field still accepts non-UTF-8 byte sequences. This option has no effect on when used with the lite runtime.
 *   - 15: cc_generic_services -
 *   - 16: java_generic_services - Sets the Go package where structs generated from this .proto will be placed. If omitted, the Go package will be derived from the following: - The basename of the package import path, if provided. - Otherwise, the package statement in the .proto file, if present. - Otherwise, the basename of the .proto file, without extension.
 *   - 17: py_generic_services - Should generic services be generated in each language?  "Generic" services are not specific to any particular RPC system.  They are generated by the main code generators in each language (without additional plugins). Generic services were the only kind of service generation supported by early versions of google.protobuf. Generic services are now considered deprecated in favor of using plugins that generate code specific to your particular RPC system.  Therefore, these default to false.  Old code which depends on generic services should explicitly set them to true.
 *   - 19: java_generate_equals_and_hash -
 *   - 22: deprecated -
 *   - 26: java_string_check_utf8 -
 *   - 30: cc_enable_arenas - Is this file deprecated? Depending on the target platform, this can emit Deprecated annotations for everything in the file, or it will be completely ignored; in the very least, this is a formalization for deprecating files.
 *   - 35: objc_class_prefix - Enables the use of arenas for the proto messages in this file. This applies only to generated classes for C++.
 *   - 36: csharp_namespace - Sets the objective c class prefix which is prepended to all objective c generated classes from this .proto. There is no default.
 *   - 38: swift_prefix - Namespace for generated classes; defaults to the package.
 *   - 39: php_class_prefix - By default Swift generators will take the proto package and CamelCase it replacing '.' with underscore and use that to prefix the types/symbols defined. When this options is provided, they will use this value instead to prefix the types/symbols defined.
 *   - 40: php_namespace - Sets the php class prefix which is prepended to all php generated classes from this .proto. Default is empty.
 *   - 41: php_generic_services - Use this option to change the namespace of php generated classes. Default is empty. When this option is empty, the package name will be used for determining the namespace.
 *   - 43: php_metadata_namespace - Use this option to change the namespace of php generated metadata classes. Default is empty. When this option is empty, the proto file name will be used for determining the namespace.
 *   - 44: ruby_package - Use this option to change the package of ruby generated classes. Default is empty. When this option is not set, the package name will be used for determining the ruby package.
 *   - 49: features - Any features defined in the specific edition.
 *   - 998: uninterpreted_option - The parser stores options it doesn't recognize here. See the documentation for the "Options" section above.
 */
export const destructGoogleProtobufFileOptions = (g_struct) => oda([g_struct.java_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.java_outer_classname, g_struct.optimize_for, g_struct.java_multiple_files, g_struct.go_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.cc_generic_services, g_struct.java_generic_services, g_struct.py_generic_services, __UNDEFINED, g_struct.java_generate_equals_and_hash, __UNDEFINED, __UNDEFINED, g_struct.deprecated, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.java_string_check_utf8, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.cc_enable_arenas, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.objc_class_prefix, g_struct.csharp_namespace, __UNDEFINED, g_struct.swift_prefix, g_struct.php_class_prefix, g_struct.php_namespace, g_struct.php_generic_services, __UNDEFINED, g_struct.php_metadata_namespace, g_struct.ruby_package, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.features], {
    999: g_struct.uninterpreted_option,
});
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_FILE_OPTIONS = '/google.protobuf.FileOptions';
/**
 * Destructures the fields of a {@link GoogleProtobufSourceCodeInfo} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: location - A Location identifies a piece of source code in a .proto file which corresponds to a particular definition.  This information is intended to be useful to IDEs, code indexers, documentation generators, and similar tools. For example, say we have a file like: message Foo { optional string foo = 1; } Let's look at just the field definition: optional string foo = 1; ^       ^^     ^^  ^  ^^^ a       bc     de  f  ghi We have the following locations: span   path               represents [a,i)  [ 4, 0, 2, 0 ]     The whole field definition. [a,b)  [ 4, 0, 2, 0, 4 ]  The label (optional). [c,d)  [ 4, 0, 2, 0, 5 ]  The type (string). [e,f)  [ 4, 0, 2, 0, 1 ]  The name (foo). [g,h)  [ 4, 0, 2, 0, 3 ]  The number (1). Notes: - A location may refer to a repeated field itself (i.e. not to any particular index within it).  This is used whenever a set of elements are logically enclosed in a single code segment.  For example, an entire extend block (possibly containing multiple extension definitions) will have an outer location whose path refers to the "extensions" repeated field without an index. - Multiple locations may have the same path.  This happens when a single logical declaration is spread out across multiple places.  The most obvious example is the "extend" block again -- there may be multiple extend blocks in the same scope, each of which will have the same path. - A location's span is not always a subset of its parent's span.  For example, the "extendee" of an extension declaration appears at the beginning of the "extend" block and is shared by all extensions within the block. - Just because a location's span is a subset of some other location's span does not mean that it is a descendant.  For example, a "group" defines both a type and a field in a single declaration.  Thus, the locations corresponding to the type and field and their components will overlap. - Code which tries to interpret locations should probably be designed to ignore those that it doesn't understand, as more types of locations could be recorded in the future.
 */
export const destructGoogleProtobufSourceCodeInfo = (g_struct) => [g_struct.location];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_SOURCE_CODE_INFO = '/google.protobuf.SourceCodeInfo';
/**
 * Destructures the fields of a {@link GoogleProtobufSourceCodeInfoLocation} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: path -
 *   - 1: span -
 *   - 2: leading_comments -
 *   - 3: trailing_comments -
 *   - 5: leading_detached_comments -
 */
export const destructGoogleProtobufSourceCodeInfoLocation = (g_struct) => [g_struct.path, g_struct.span, g_struct.leading_comments, g_struct.trailing_comments, __UNDEFINED, g_struct.leading_detached_comments];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GOOGLE_PROTOBUF_SOURCE_CODE_INFO_LOCATION = '/google.protobuf.SourceCodeInfo.Location';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_OPTIONAL** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL = 1;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_OPTIONAL** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL = 'LABEL_OPTIONAL';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_REPEATED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED = 3;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_REPEATED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED = 'LABEL_REPEATED';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_REQUIRED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED = 2;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Label`.
 *
 * **LABEL_REQUIRED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoLabel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED = 'LABEL_REQUIRED';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldDescriptorProto.Label` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldDescriptorProtoLabel = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldDescriptorProto.Label` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldDescriptorProtoLabel = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_OPTIONAL,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REPEATED,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_LABEL_REQUIRED,
};
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_DOUBLE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE = 1;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_DOUBLE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE = 'TYPE_DOUBLE';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FLOAT** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT = 2;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FLOAT** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT = 'TYPE_FLOAT';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_INT64** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64 = 3;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_INT64** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64 = 'TYPE_INT64';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_UINT64** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64 = 4;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_UINT64** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64 = 'TYPE_UINT64';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_INT32** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32 = 5;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_INT32** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32 = 'TYPE_INT32';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FIXED64** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64 = 6;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FIXED64** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64 = 'TYPE_FIXED64';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FIXED32** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32 = 7;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_FIXED32** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32 = 'TYPE_FIXED32';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_BOOL** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL = 8;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_BOOL** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL = 'TYPE_BOOL';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_STRING** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING = 9;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_STRING** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING = 'TYPE_STRING';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_GROUP** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP = 10;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_GROUP** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP = 'TYPE_GROUP';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_MESSAGE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE = 11;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_MESSAGE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE = 'TYPE_MESSAGE';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_BYTES** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES = 12;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_BYTES** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES = 'TYPE_BYTES';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_UINT32** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32 = 13;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_UINT32** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32 = 'TYPE_UINT32';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_ENUM** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM = 14;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_ENUM** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM = 'TYPE_ENUM';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SFIXED32** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32 = 15;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SFIXED32** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32 = 'TYPE_SFIXED32';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SFIXED64** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64 = 16;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SFIXED64** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64 = 'TYPE_SFIXED64';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SINT32** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32 = 17;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SINT32** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32 = 'TYPE_SINT32';
/**
 * Protobuf enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SINT64** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64 = 18;
/**
 * JSON enum value for `google.protobuf.FieldDescriptorProto.Type`.
 *
 * **TYPE_SINT64** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldDescriptorProtoType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64 = 'TYPE_SINT64';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldDescriptorProto.Type` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldDescriptorProtoType = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64]: SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldDescriptorProto.Type` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldDescriptorProtoType = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_DOUBLE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FLOAT,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT64,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT64,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_INT32,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED64,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_FIXED32,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BOOL,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_STRING,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_GROUP,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_MESSAGE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_BYTES,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_UINT32,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_ENUM,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED32,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SFIXED64,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT32,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_DESCRIPTOR_PROTO_TYPE_SINT64,
};
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **STRING** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsCType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING = 0;
/**
 * JSON enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **STRING** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsCType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING = 'STRING';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **CORD** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsCType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD = 1;
/**
 * JSON enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **CORD** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsCType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD = 'CORD';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **STRING_PIECE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsCType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE = 2;
/**
 * JSON enum value for `google.protobuf.FieldOptions.CType`.
 *
 * **STRING_PIECE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsCType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE = 'STRING_PIECE';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldOptions.CType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldOptionsCType = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldOptions.CType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldOptionsCType = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_CORD,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_CTYPE_STRING_PIECE,
};
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_NORMAL** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsJSType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL = 0;
/**
 * JSON enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_NORMAL** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsJSType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL = 'JS_NORMAL';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_STRING** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsJSType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING = 1;
/**
 * JSON enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_STRING** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsJSType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING = 'JS_STRING';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_NUMBER** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsJSType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER = 2;
/**
 * JSON enum value for `google.protobuf.FieldOptions.JSType`.
 *
 * **JS_NUMBER** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsJSType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER = 'JS_NUMBER';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldOptions.JSType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldOptionsJSType = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldOptions.JSType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldOptionsJSType = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NORMAL,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_STRING,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_JSTYPE_JS_NUMBER,
};
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN = 'RETENTION_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_RUNTIME** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME = 1;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_RUNTIME** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME = 'RETENTION_RUNTIME';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_SOURCE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE = 2;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionRetention`.
 *
 * **RETENTION_SOURCE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionRetention}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE = 'RETENTION_SOURCE';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldOptions.OptionRetention` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldOptionsOptionRetention = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldOptions.OptionRetention` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldOptionsOptionRetention = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_RUNTIME,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_RETENTION_SOURCE,
};
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN = 'TARGET_TYPE_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_FILE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE = 1;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_FILE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE = 'TARGET_TYPE_FILE';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_EXTENSION_RANGE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE = 2;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_EXTENSION_RANGE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE = 'TARGET_TYPE_EXTENSION_RANGE';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_MESSAGE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE = 3;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_MESSAGE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE = 'TARGET_TYPE_MESSAGE';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_FIELD** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD = 4;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_FIELD** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD = 'TARGET_TYPE_FIELD';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ONEOF** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF = 5;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ONEOF** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF = 'TARGET_TYPE_ONEOF';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ENUM** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM = 6;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ENUM** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM = 'TARGET_TYPE_ENUM';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ENUM_ENTRY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY = 7;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_ENUM_ENTRY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY = 'TARGET_TYPE_ENUM_ENTRY';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_SERVICE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE = 8;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_SERVICE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE = 'TARGET_TYPE_SERVICE';
/**
 * Protobuf enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_METHOD** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD = 9;
/**
 * JSON enum value for `google.protobuf.FieldOptions.OptionTargetType`.
 *
 * **TARGET_TYPE_METHOD** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFieldOptionsOptionTargetType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD = 'TARGET_TYPE_METHOD';
/**
 * Maps a protobuf enum int value for `google.protobuf.FieldOptions.OptionTargetType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFieldOptionsOptionTargetType = {
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE,
    [XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD]: SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FieldOptions.OptionTargetType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFieldOptionsOptionTargetType = {
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FILE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_EXTENSION_RANGE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_MESSAGE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_FIELD,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ONEOF,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_ENUM_ENTRY,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_SERVICE,
    [SI_JSON_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD]: XC_PROTO_GOOGLE_PROTOBUF_FIELD_OPTIONS_TARGET_TYPE_METHOD,
};
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_UNKNOWN** - A placeholder for an unknown edition value.
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_UNKNOWN** - A placeholder for an unknown edition value.
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_UNKNOWN = 'EDITION_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_PROTO2** - Legacy syntax "editions".  These pre-date editions, but behave much like distinct editions.  These can't be used to specify the edition of proto files, but feature definitions must supply proto2/proto3 defaults for backwards compatibility.
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO2 = 998;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_PROTO2** - Legacy syntax "editions".  These pre-date editions, but behave much like distinct editions.  These can't be used to specify the edition of proto files, but feature definitions must supply proto2/proto3 defaults for backwards compatibility.
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO2 = 'EDITION_PROTO2';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_PROTO3** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO3 = 999;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_PROTO3** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO3 = 'EDITION_PROTO3';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_2023** - Editions that have been released.  The specific values are arbitrary and should not be depended on, but they will always be time-ordered for easy comparison.
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_2023 = 1000;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_2023** - Editions that have been released.  The specific values are arbitrary and should not be depended on, but they will always be time-ordered for easy comparison.
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_2023 = 'EDITION_2023';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_1_TEST_ONLY** - Placeholder editions for testing feature resolution.  These should not be used or relyed on outside of tests.
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY = 1;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_1_TEST_ONLY** - Placeholder editions for testing feature resolution.  These should not be used or relyed on outside of tests.
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY = 'EDITION_1_TEST_ONLY';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_2_TEST_ONLY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY = 2;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_2_TEST_ONLY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY = 'EDITION_2_TEST_ONLY';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99997_TEST_ONLY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY = 99997;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99997_TEST_ONLY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY = 'EDITION_99997_TEST_ONLY';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99998_TEST_ONLY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY = 99998;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99998_TEST_ONLY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY = 'EDITION_99998_TEST_ONLY';
/**
 * Protobuf enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99999_TEST_ONLY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufEdition}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY = 99999;
/**
 * JSON enum value for `google.protobuf.Edition`.
 *
 * **EDITION_99999_TEST_ONLY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufEdition}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY = 'EDITION_99999_TEST_ONLY';
/**
 * Maps a protobuf enum int value for `google.protobuf.Edition` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufEdition = {
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_EDITION_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO2]: SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO2,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO3]: SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO3,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_2023]: SI_JSON_GOOGLE_PROTOBUF_EDITION_2023,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY]: SI_JSON_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY]: SI_JSON_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY]: SI_JSON_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY]: SI_JSON_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY,
    [XC_PROTO_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY]: SI_JSON_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY,
};
/**
 * Maps a JSON enum string value for `google.protobuf.Edition` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufEdition = {
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO2]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO2,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_PROTO3]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_PROTO3,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_2023]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_2023,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_1_TEST_ONLY,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_2_TEST_ONLY,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_99997_TEST_ONLY,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_99998_TEST_ONLY,
    [SI_JSON_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY]: XC_PROTO_GOOGLE_PROTOBUF_EDITION_99999_TEST_ONLY,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **FIELD_PRESENCE_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **FIELD_PRESENCE_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN = 'FIELD_PRESENCE_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **EXPLICIT** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **EXPLICIT** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT = 'EXPLICIT';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **IMPLICIT** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **IMPLICIT** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT = 'IMPLICIT';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **LEGACY_REQUIRED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED = 3;
/**
 * JSON enum value for `google.protobuf.FeatureSet.FieldPresence`.
 *
 * **LEGACY_REQUIRED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetFieldPresence}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED = 'LEGACY_REQUIRED';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.FieldPresence` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetFieldPresence = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.FieldPresence` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetFieldPresence = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_FIELD_PRESENCE_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPLICIT,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_IMPLICIT,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_REQUIRED,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **ENUM_TYPE_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetEnumType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **ENUM_TYPE_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetEnumType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN = 'ENUM_TYPE_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **OPEN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetEnumType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_OPEN = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **OPEN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetEnumType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_OPEN = 'OPEN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **CLOSED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetEnumType}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.EnumType`.
 *
 * **CLOSED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetEnumType}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED = 'CLOSED';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.EnumType` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetEnumType = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_OPEN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_OPEN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.EnumType` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetEnumType = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ENUM_TYPE_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_OPEN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_OPEN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_CLOSED,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **REPEATED_FIELD_ENCODING_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **REPEATED_FIELD_ENCODING_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN = 'REPEATED_FIELD_ENCODING_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **PACKED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_PACKED = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **PACKED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_PACKED = 'PACKED';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **EXPANDED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.RepeatedFieldEncoding`.
 *
 * **EXPANDED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetRepeatedFieldEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED = 'EXPANDED';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.RepeatedFieldEncoding` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetRepeatedFieldEncoding = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_PACKED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_PACKED,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.RepeatedFieldEncoding` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetRepeatedFieldEncoding = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_REPEATED_FIELD_ENCODING_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_PACKED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_PACKED,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_EXPANDED,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **UTF8_VALIDATION_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **UTF8_VALIDATION_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN = 'UTF8_VALIDATION_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **NONE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **NONE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE = 'NONE';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **VERIFY** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.Utf8Validation`.
 *
 * **VERIFY** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetUtf8Validation}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY = 'VERIFY';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.Utf8Validation` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetUtf8Validation = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.Utf8Validation` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetUtf8Validation = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_UTF8_VALIDATION_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_NONE,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_UTF8VALIDATION_VERIFY,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **MESSAGE_ENCODING_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **MESSAGE_ENCODING_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN = 'MESSAGE_ENCODING_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **LENGTH_PREFIXED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **LENGTH_PREFIXED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED = 'LENGTH_PREFIXED';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **DELIMITED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.MessageEncoding`.
 *
 * **DELIMITED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetMessageEncoding}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED = 'DELIMITED';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.MessageEncoding` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetMessageEncoding = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.MessageEncoding` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetMessageEncoding = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_MESSAGE_ENCODING_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LENGTH_PREFIXED,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_DELIMITED,
};
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **JSON_FORMAT_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **JSON_FORMAT_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN = 'JSON_FORMAT_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **ALLOW** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW = 1;
/**
 * JSON enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **ALLOW** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW = 'ALLOW';
/**
 * Protobuf enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **LEGACY_BEST_EFFORT** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT = 2;
/**
 * JSON enum value for `google.protobuf.FeatureSet.JsonFormat`.
 *
 * **LEGACY_BEST_EFFORT** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFeatureSetJsonFormat}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT = 'LEGACY_BEST_EFFORT';
/**
 * Maps a protobuf enum int value for `google.protobuf.FeatureSet.JsonFormat` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFeatureSetJsonFormat = {
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW,
    [XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT]: SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FeatureSet.JsonFormat` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFeatureSetJsonFormat = {
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_JSON_FORMAT_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_ALLOW,
    [SI_JSON_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT]: XC_PROTO_GOOGLE_PROTOBUF_FEATURE_SET_LEGACY_BEST_EFFORT,
};
/**
 * Protobuf enum value for `google.protobuf.ExtensionRangeOptions.VerificationState`.
 *
 * **DECLARATION** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufExtensionRangeOptionsVerificationState}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION = 0;
/**
 * JSON enum value for `google.protobuf.ExtensionRangeOptions.VerificationState`.
 *
 * **DECLARATION** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufExtensionRangeOptionsVerificationState}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION = 'DECLARATION';
/**
 * Protobuf enum value for `google.protobuf.ExtensionRangeOptions.VerificationState`.
 *
 * **UNVERIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufExtensionRangeOptionsVerificationState}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED = 1;
/**
 * JSON enum value for `google.protobuf.ExtensionRangeOptions.VerificationState`.
 *
 * **UNVERIFIED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufExtensionRangeOptionsVerificationState}
 */
export const SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED = 'UNVERIFIED';
/**
 * Maps a protobuf enum int value for `google.protobuf.ExtensionRangeOptions.VerificationState` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufExtensionRangeOptionsVerificationState = {
    [XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION]: SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION,
    [XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED]: SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED,
};
/**
 * Maps a JSON enum string value for `google.protobuf.ExtensionRangeOptions.VerificationState` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufExtensionRangeOptionsVerificationState = {
    [SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION]: XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_DECLARATION,
    [SI_JSON_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED]: XC_PROTO_GOOGLE_PROTOBUF_EXTENSION_RANGE_OPTIONS_UNVERIFIED,
};
/**
 * Protobuf enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **IDEMPOTENCY_UNKNOWN** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN = 0;
/**
 * JSON enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **IDEMPOTENCY_UNKNOWN** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN = 'IDEMPOTENCY_UNKNOWN';
/**
 * Protobuf enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **NO_SIDE_EFFECTS** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS = 1;
/**
 * JSON enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **NO_SIDE_EFFECTS** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS = 'NO_SIDE_EFFECTS';
/**
 * Protobuf enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **IDEMPOTENT** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT = 2;
/**
 * JSON enum value for `google.protobuf.MethodOptions.IdempotencyLevel`.
 *
 * **IDEMPOTENT** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufMethodOptionsIdempotencyLevel}
 */
export const SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT = 'IDEMPOTENT';
/**
 * Maps a protobuf enum int value for `google.protobuf.MethodOptions.IdempotencyLevel` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufMethodOptionsIdempotencyLevel = {
    [XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN]: SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN,
    [XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS]: SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS,
    [XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT]: SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT,
};
/**
 * Maps a JSON enum string value for `google.protobuf.MethodOptions.IdempotencyLevel` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufMethodOptionsIdempotencyLevel = {
    [SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN]: XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENCY_UNKNOWN,
    [SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS]: XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_NO_SIDE_EFFECTS,
    [SI_JSON_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT]: XC_PROTO_GOOGLE_PROTOBUF_METHOD_OPTIONS_IDEMPOTENT,
};
/**
 * Protobuf enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **SPEED** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED = 1;
/**
 * JSON enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **SPEED** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED = 'SPEED';
/**
 * Protobuf enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **CODE_SIZE** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE = 2;
/**
 * JSON enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **CODE_SIZE** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE = 'CODE_SIZE';
/**
 * Protobuf enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **LITE_RUNTIME** -
 *
 * Belongs to enum type {@link ProtoEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME = 3;
/**
 * JSON enum value for `google.protobuf.FileOptions.OptimizeMode`.
 *
 * **LITE_RUNTIME** -
 *
 * Belongs to enum type {@link JsonEnumGoogleProtobufFileOptionsOptimizeMode}
 */
export const SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME = 'LITE_RUNTIME';
/**
 * Maps a protobuf enum int value for `google.protobuf.FileOptions.OptimizeMode` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumGoogleProtobufFileOptionsOptimizeMode = {
    [XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED]: SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED,
    [XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE]: SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE,
    [XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME]: SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME,
};
/**
 * Maps a JSON enum string value for `google.protobuf.FileOptions.OptimizeMode` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumGoogleProtobufFileOptionsOptimizeMode = {
    [SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED]: XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_SPEED,
    [SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE]: XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_CODE_SIZE,
    [SI_JSON_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME]: XC_PROTO_GOOGLE_PROTOBUF_FILE_OPTIONS_LITE_RUNTIME,
};
