// @solar-republic/cosmos-grpc/build/dist/proto/axelar/axelarnet/v1beta1/params.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `Params` protobuf message: represent the genesis parameters for the module
 * @param sg_route_timeout_window - `route_timeout_window`: IBC packet route timeout window
 * @param sg_transfer_limit - `transfer_limit`:
 * @param sg_end_blocker_limit - `end_blocker_limit`:
 * @param a_call_contracts_proposal_min_deposits - `call_contracts_proposal_min_deposits`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetParams = (sg_route_timeout_window, sg_transfer_limit, sg_end_blocker_limit, a_call_contracts_proposal_min_deposits) => Protobuf() // ...
    .g(sg_route_timeout_window) // uint64 route_timeout_window = 1
    .g(sg_transfer_limit, 3) // uint64 transfer_limit = 3
    .g(sg_end_blocker_limit, 4) // uint64 end_blocker_limit = 4
    .B(a_call_contracts_proposal_min_deposits, 5) // CallContractProposalMinDeposit[] call_contracts_proposal_min_deposits = 5
    .o;
/**
 * Encodes a `CallContractProposalMinDeposit` protobuf message:
 * @param s_chain - `chain`:
 * @param sa_contract - `contract_address`:
 * @param a_min_deposits - `min_deposits`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarAxelarnetCallContractProposalMinDeposit = (s_chain, sa_contract, a_min_deposits) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .s(sa_contract) // string contract_address = 2
    .B(coins(a_min_deposits)) // Coin[] min_deposits = 3
    .o;
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedAxelarAxelarnetParams}
 */
export const decodeAxelarAxelarnetParams = ((atu8_payload, [sg_route_timeout_window, , sg_transfer_limit, sg_end_blocker_limit, a_call_contracts_proposal_min_deposits] = decode_protobuf(atu8_payload, [3, 0, 3, 3, 0], [0, 0, 0, 0, decodeAxelarAxelarnetCallContractProposalMinDeposit])) => [sg_route_timeout_window, __UNDEFINED, sg_transfer_limit, sg_end_blocker_limit, a_call_contracts_proposal_min_deposits]);
/**
 * Decodes a protobuf CallContractProposalMinDeposit message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarAxelarnetCallContractProposalMinDeposit}
 */
export const decodeAxelarAxelarnetCallContractProposalMinDeposit = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decode_coin]);
/**
 * Destructures the fields of a {@link AxelarAxelarnetParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: route_timeout_window - IBC packet route timeout window
 *   - 2: transfer_limit -
 *   - 3: end_blocker_limit -
 *   - 4: call_contracts_proposal_min_deposits -
 */
export const destructAxelarAxelarnetParams = (g_struct) => [g_struct.route_timeout_window, __UNDEFINED, g_struct.transfer_limit, g_struct.end_blocker_limit, g_struct.call_contracts_proposal_min_deposits];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_PARAMS = '/axelar.axelarnet.v1beta1.Params';
/**
 * Destructures the fields of a {@link AxelarAxelarnetCallContractProposalMinDeposit} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 *   - 1: contract_address -
 *   - 2: min_deposits -
 */
export const destructAxelarAxelarnetCallContractProposalMinDeposit = (g_struct) => [g_struct.chain, g_struct.contract_address, g_struct.min_deposits];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_AXELARNET_CALL_CONTRACT_PROPOSAL_MIN_DEPOSIT = '/axelar.axelarnet.v1beta1.CallContractProposalMinDeposit';
