// @solar-republic/cosmos-grpc/build/dist/proto/axelar/multisig/exported/v1beta1/types.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedKeyState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED = 0;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedKeyState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED = 'Inactive';
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_ASSIGNED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedKeyState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED = 1;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_ASSIGNED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedKeyState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED = 'Assigned';
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_ACTIVE** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedKeyState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE = 2;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.KeyState`.
 *
 * **KEY_STATE_ACTIVE** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedKeyState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE = 'Active';
/**
 * Maps a protobuf enum int value for `axelar.multisig.exported.v1beta1.KeyState` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAxelarMultisigExportedKeyState = {
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED]: SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED,
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED]: SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED,
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE]: SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE,
};
/**
 * Maps a JSON enum string value for `axelar.multisig.exported.v1beta1.KeyState` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAxelarMultisigExportedKeyState = {
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_UNSPECIFIED,
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ASSIGNED,
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_KEY_STATE_ACTIVE,
};
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedMultisigState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED = 0;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedMultisigState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED = 'NonExistent';
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_PENDING** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedMultisigState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING = 1;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_PENDING** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedMultisigState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING = 'Pending';
/**
 * Protobuf enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_COMPLETED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarMultisigExportedMultisigState}
 */
export const XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED = 2;
/**
 * JSON enum value for `axelar.multisig.exported.v1beta1.MultisigState`.
 *
 * **MULTISIG_STATE_COMPLETED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarMultisigExportedMultisigState}
 */
export const SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED = 'Completed';
/**
 * Maps a protobuf enum int value for `axelar.multisig.exported.v1beta1.MultisigState` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAxelarMultisigExportedMultisigState = {
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED]: SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED,
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING]: SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING,
    [XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED]: SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED,
};
/**
 * Maps a JSON enum string value for `axelar.multisig.exported.v1beta1.MultisigState` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAxelarMultisigExportedMultisigState = {
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_UNSPECIFIED,
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_PENDING,
    [SI_JSON_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED]: XC_PROTO_AXELAR_MULTISIG_EXPORTED_MULTISIG_STATE_COMPLETED,
};
