// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/v1beta1/shared.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `MigrationRecords` protobuf message: contains all the links between balancer and concentrated pools
 * @param a_balancer_to_concentrated_pool_links - `balancer_to_concentrated_pool_links`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMigrationRecords = (a_balancer_to_concentrated_pool_links) => Protobuf() // ...
    .B(a_balancer_to_concentrated_pool_links) // BalancerToConcentratedPoolLink[] balancer_to_concentrated_pool_links = 1
    .o;
/**
 * Encodes a `BalancerToConcentratedPoolLink` protobuf message: defines a single link between a single balancer pool and a single concentrated liquidity pool. This link is used to allow a balancer pool to migrate to a single canonical full range concentrated liquidity pool position A balancer pool can be linked to a maximum of one cl pool, and a cl pool can be linked to a maximum of one balancer pool.
 * @param sg_balancer_pool_id - `balancer_pool_id`:
 * @param sg_cl_pool_id - `cl_pool_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammBalancerToConcentratedPoolLink = (sg_balancer_pool_id, sg_cl_pool_id) => Protobuf() // ...
    .g(sg_balancer_pool_id) // uint64 balancer_pool_id = 1
    .g(sg_cl_pool_id) // uint64 cl_pool_id = 2
    .o;
/**
 * Decodes a protobuf MigrationRecords message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMigrationRecords}
 */
export const decodeOsmosisGammMigrationRecords = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisGammBalancerToConcentratedPoolLink]);
/**
 * Decodes a protobuf BalancerToConcentratedPoolLink message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammBalancerToConcentratedPoolLink}
 */
export const decodeOsmosisGammBalancerToConcentratedPoolLink = (atu8_payload) => decode_protobuf(atu8_payload, [3, 3]);
/**
 * Destructures the fields of a {@link OsmosisGammBalancerToConcentratedPoolLink} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: balancer_pool_id -
 *   - 1: cl_pool_id -
 */
export const destructOsmosisGammBalancerToConcentratedPoolLink = (g_struct) => [g_struct.balancer_pool_id, g_struct.cl_pool_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_BALANCER_TO_CONCENTRATED_POOL_LINK = '/osmosis.gamm.v1beta1.BalancerToConcentratedPoolLink';
/**
 * Destructures the fields of a {@link OsmosisGammMigrationRecords} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: balancer_to_concentrated_pool_links -
 */
export const destructOsmosisGammMigrationRecords = (g_struct) => [g_struct.balancer_to_concentrated_pool_links];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MIGRATION_RECORDS = '/osmosis.gamm.v1beta1.MigrationRecords';
