// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/v1beta1/tx.js
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coin, coins } from '../../../../api/protobuf-writer.js';
import { decodeOsmosisPoolmanagerSwapAmountInRoute, decodeOsmosisPoolmanagerSwapAmountOutRoute } from '../../poolmanager/v1beta1/swap_route.js';
/**
 * Encodes a `MsgJoinPool` protobuf message: ===================== MsgJoinPool This is really MsgJoinPoolNoSwap
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param s_share_out_amount - `share_out_amount`:
 * @param a_token_in_maxs - `token_in_maxs`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinPool = (sa_sender, sg_pool_id, s_share_out_amount, a_token_in_maxs) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .s(s_share_out_amount) // string share_out_amount = 3
    .B(coins(a_token_in_maxs)) // Coin[] token_in_maxs = 4
    .o;
/**
 * Encodes a `MsgJoinPoolResponse` protobuf message:
 * @param s_share_out_amount - `share_out_amount`:
 * @param a_token_ins - `token_in`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinPoolResponse = (s_share_out_amount, a_token_ins) => Protobuf() // ...
    .s(s_share_out_amount) // string share_out_amount = 1
    .B(coins(a_token_ins)) // Coin[] token_in = 2
    .o;
/**
 * Encodes a `MsgExitPool` protobuf message: ===================== MsgExitPool
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param s_share_in_amount - `share_in_amount`:
 * @param a_token_out_mins - `token_out_mins`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitPool = (sa_sender, sg_pool_id, s_share_in_amount, a_token_out_mins) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .s(s_share_in_amount) // string share_in_amount = 3
    .B(coins(a_token_out_mins)) // Coin[] token_out_mins = 4
    .o;
/**
 * Encodes a `MsgExitPoolResponse` protobuf message:
 * @param a_token_outs - `token_out`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitPoolResponse = (a_token_outs) => Protobuf() // ...
    .B(coins(a_token_outs)) // Coin[] token_out = 1
    .o;
/**
 * Encodes a `MsgSwapExactAmountIn` protobuf message: ===================== MsgSwapExactAmountIn
 * @param sa_sender - `sender`:
 * @param a_routes - `routes`:
 * @param a_token_in - `token_in`:
 * @param s_token_out_min_amount - `token_out_min_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgSwapExactAmountIn = (sa_sender, a_routes, a_token_in, s_token_out_min_amount) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .B(a_routes) // SwapAmountInRoute[] routes = 2
    .b(coin(a_token_in)) // Coin token_in = 3
    .s(s_token_out_min_amount) // string token_out_min_amount = 4
    .o;
/**
 * Encodes a `MsgSwapExactAmountInResponse` protobuf message:
 * @param s_token_out_amount - `token_out_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgSwapExactAmountInResponse = (s_token_out_amount) => Protobuf() // ...
    .s(s_token_out_amount) // string token_out_amount = 1
    .o;
/**
 * Encodes a `MsgSwapExactAmountOut` protobuf message:
 * @param sa_sender - `sender`:
 * @param a_routes - `routes`:
 * @param s_token_in_max_amount - `token_in_max_amount`:
 * @param a_token_out - `token_out`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgSwapExactAmountOut = (sa_sender, a_routes, s_token_in_max_amount, a_token_out) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .B(a_routes) // SwapAmountOutRoute[] routes = 2
    .s(s_token_in_max_amount) // string token_in_max_amount = 3
    .b(coin(a_token_out)) // Coin token_out = 4
    .o;
/**
 * Encodes a `MsgSwapExactAmountOutResponse` protobuf message:
 * @param s_token_in_amount - `token_in_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgSwapExactAmountOutResponse = (s_token_in_amount) => Protobuf() // ...
    .s(s_token_in_amount) // string token_in_amount = 1
    .o;
/**
 * Encodes a `MsgJoinSwapExternAmountIn` protobuf message: ===================== MsgJoinSwapExternAmountIn TODO: Rename to MsgJoinSwapExactAmountIn
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param a_token_in - `token_in`:
 * @param s_share_out_min_amount - `share_out_min_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinSwapExternAmountIn = (sa_sender, sg_pool_id, a_token_in, s_share_out_min_amount) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .b(coin(a_token_in)) // Coin token_in = 3
    .s(s_share_out_min_amount) // string share_out_min_amount = 4
    .o;
/**
 * Encodes a `MsgJoinSwapExternAmountInResponse` protobuf message:
 * @param s_share_out_amount - `share_out_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinSwapExternAmountInResponse = (s_share_out_amount) => Protobuf() // ...
    .s(s_share_out_amount) // string share_out_amount = 1
    .o;
/**
 * Encodes a `MsgJoinSwapShareAmountOut` protobuf message: ===================== MsgJoinSwapShareAmountOut
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param s_token_in_denom - `token_in_denom`:
 * @param s_share_out_amount - `share_out_amount`:
 * @param s_token_in_max_amount - `token_in_max_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinSwapShareAmountOut = (sa_sender, sg_pool_id, s_token_in_denom, s_share_out_amount, s_token_in_max_amount) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .s(s_token_in_denom) // string token_in_denom = 3
    .s(s_share_out_amount) // string share_out_amount = 4
    .s(s_token_in_max_amount) // string token_in_max_amount = 5
    .o;
/**
 * Encodes a `MsgJoinSwapShareAmountOutResponse` protobuf message:
 * @param s_token_in_amount - `token_in_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgJoinSwapShareAmountOutResponse = (s_token_in_amount) => Protobuf() // ...
    .s(s_token_in_amount) // string token_in_amount = 1
    .o;
/**
 * Encodes a `MsgExitSwapShareAmountIn` protobuf message: ===================== MsgExitSwapShareAmountIn
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param s_token_out_denom - `token_out_denom`:
 * @param s_share_in_amount - `share_in_amount`:
 * @param s_token_out_min_amount - `token_out_min_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitSwapShareAmountIn = (sa_sender, sg_pool_id, s_token_out_denom, s_share_in_amount, s_token_out_min_amount) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .s(s_token_out_denom) // string token_out_denom = 3
    .s(s_share_in_amount) // string share_in_amount = 4
    .s(s_token_out_min_amount) // string token_out_min_amount = 5
    .o;
/**
 * Encodes a `MsgExitSwapShareAmountInResponse` protobuf message:
 * @param s_token_out_amount - `token_out_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitSwapShareAmountInResponse = (s_token_out_amount) => Protobuf() // ...
    .s(s_token_out_amount) // string token_out_amount = 1
    .o;
/**
 * Encodes a `MsgExitSwapExternAmountOut` protobuf message: ===================== MsgExitSwapExternAmountOut
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param a_token_out - `token_out`:
 * @param s_share_in_max_amount - `share_in_max_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitSwapExternAmountOut = (sa_sender, sg_pool_id, a_token_out, s_share_in_max_amount) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .b(coin(a_token_out)) // Coin token_out = 3
    .s(s_share_in_max_amount) // string share_in_max_amount = 4
    .o;
/**
 * Encodes a `MsgExitSwapExternAmountOutResponse` protobuf message:
 * @param s_share_in_amount - `share_in_amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammMsgExitSwapExternAmountOutResponse = (s_share_in_amount) => Protobuf() // ...
    .s(s_share_in_amount) // string share_in_amount = 1
    .o;
/**
 * Decodes a protobuf JoinPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinPool}
 */
export const decodeOsmosisGammMsgJoinPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 5, 0], [0, 0, 0, decode_coin]);
/**
 * Decodes a protobuf JoinPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinPoolResponse}
 */
export const decodeOsmosisGammMsgJoinPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0], [0, decode_coin]);
/**
 * Decodes a protobuf ExitPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitPool}
 */
export const decodeOsmosisGammMsgExitPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 5, 0], [0, 0, 0, decode_coin]);
/**
 * Decodes a protobuf ExitPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitPoolResponse}
 */
export const decodeOsmosisGammMsgExitPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decode_coin]);
/**
 * Decodes a protobuf SwapExactAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgSwapExactAmountIn}
 */
export const decodeOsmosisGammMsgSwapExactAmountIn = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0, 1, 5], [0, decodeOsmosisPoolmanagerSwapAmountInRoute, decode_coin]);
/**
 * Decodes a protobuf SwapExactAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgSwapExactAmountInResponse}
 */
export const decodeOsmosisGammMsgSwapExactAmountInResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf SwapExactAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgSwapExactAmountOut}
 */
export const decodeOsmosisGammMsgSwapExactAmountOut = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0, 5, 1], [0, decodeOsmosisPoolmanagerSwapAmountOutRoute, 0, decode_coin]);
/**
 * Decodes a protobuf SwapExactAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgSwapExactAmountOutResponse}
 */
export const decodeOsmosisGammMsgSwapExactAmountOutResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf JoinSwapExternAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinSwapExternAmountIn}
 */
export const decodeOsmosisGammMsgJoinSwapExternAmountIn = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 1, 5], [0, 0, decode_coin]);
/**
 * Decodes a protobuf JoinSwapExternAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinSwapExternAmountInResponse}
 */
export const decodeOsmosisGammMsgJoinSwapExternAmountInResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf JoinSwapShareAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinSwapShareAmountOut}
 */
export const decodeOsmosisGammMsgJoinSwapShareAmountOut = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 5, 5, 5]);
/**
 * Decodes a protobuf JoinSwapShareAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgJoinSwapShareAmountOutResponse}
 */
export const decodeOsmosisGammMsgJoinSwapShareAmountOutResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ExitSwapShareAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitSwapShareAmountIn}
 */
export const decodeOsmosisGammMsgExitSwapShareAmountIn = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 5, 5, 5]);
/**
 * Decodes a protobuf ExitSwapShareAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitSwapShareAmountInResponse}
 */
export const decodeOsmosisGammMsgExitSwapShareAmountInResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ExitSwapExternAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitSwapExternAmountOut}
 */
export const decodeOsmosisGammMsgExitSwapExternAmountOut = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 1, 5], [0, 0, decode_coin]);
/**
 * Decodes a protobuf ExitSwapExternAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammMsgExitSwapExternAmountOutResponse}
 */
export const decodeOsmosisGammMsgExitSwapExternAmountOutResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: share_out_amount -
 *   - 3: token_in_maxs -
 */
export const destructOsmosisGammMsgJoinPool = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.share_out_amount, g_struct.token_in_maxs];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_POOL = '/osmosis.gamm.v1beta1.MsgJoinPool';
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: share_out_amount -
 *   - 1: token_in -
 */
export const destructOsmosisGammMsgJoinPoolResponse = (g_struct) => [g_struct.share_out_amount, g_struct.token_in];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_POOL_RESPONSE = '/osmosis.gamm.v1beta1.MsgJoinPoolResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: share_in_amount -
 *   - 3: token_out_mins -
 */
export const destructOsmosisGammMsgExitPool = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.share_in_amount, g_struct.token_out_mins];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_POOL = '/osmosis.gamm.v1beta1.MsgExitPool';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_out -
 */
export const destructOsmosisGammMsgExitPoolResponse = (g_struct) => [g_struct.token_out];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_POOL_RESPONSE = '/osmosis.gamm.v1beta1.MsgExitPoolResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgSwapExactAmountIn} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: routes -
 *   - 2: token_in -
 *   - 3: token_out_min_amount -
 */
export const destructOsmosisGammMsgSwapExactAmountIn = (g_struct) => [g_struct.sender, g_struct.routes, g_struct.token_in, g_struct.token_out_min_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_SWAP_EXACT_AMOUNT_IN = '/osmosis.gamm.v1beta1.MsgSwapExactAmountIn';
/**
 * Destructures the fields of a {@link OsmosisGammMsgSwapExactAmountInResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_out_amount -
 */
export const destructOsmosisGammMsgSwapExactAmountInResponse = (g_struct) => [g_struct.token_out_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_SWAP_EXACT_AMOUNT_IN_RESPONSE = '/osmosis.gamm.v1beta1.MsgSwapExactAmountInResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgSwapExactAmountOut} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: routes -
 *   - 2: token_in_max_amount -
 *   - 3: token_out -
 */
export const destructOsmosisGammMsgSwapExactAmountOut = (g_struct) => [g_struct.sender, g_struct.routes, g_struct.token_in_max_amount, g_struct.token_out];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_SWAP_EXACT_AMOUNT_OUT = '/osmosis.gamm.v1beta1.MsgSwapExactAmountOut';
/**
 * Destructures the fields of a {@link OsmosisGammMsgSwapExactAmountOutResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_in_amount -
 */
export const destructOsmosisGammMsgSwapExactAmountOutResponse = (g_struct) => [g_struct.token_in_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_SWAP_EXACT_AMOUNT_OUT_RESPONSE = '/osmosis.gamm.v1beta1.MsgSwapExactAmountOutResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinSwapExternAmountIn} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: token_in -
 *   - 3: share_out_min_amount -
 */
export const destructOsmosisGammMsgJoinSwapExternAmountIn = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.token_in, g_struct.share_out_min_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_SWAP_EXTERN_AMOUNT_IN = '/osmosis.gamm.v1beta1.MsgJoinSwapExternAmountIn';
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinSwapExternAmountInResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: share_out_amount -
 */
export const destructOsmosisGammMsgJoinSwapExternAmountInResponse = (g_struct) => [g_struct.share_out_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_SWAP_EXTERN_AMOUNT_IN_RESPONSE = '/osmosis.gamm.v1beta1.MsgJoinSwapExternAmountInResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinSwapShareAmountOut} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: token_in_denom -
 *   - 3: share_out_amount -
 *   - 4: token_in_max_amount -
 */
export const destructOsmosisGammMsgJoinSwapShareAmountOut = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.token_in_denom, g_struct.share_out_amount, g_struct.token_in_max_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_SWAP_SHARE_AMOUNT_OUT = '/osmosis.gamm.v1beta1.MsgJoinSwapShareAmountOut';
/**
 * Destructures the fields of a {@link OsmosisGammMsgJoinSwapShareAmountOutResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_in_amount -
 */
export const destructOsmosisGammMsgJoinSwapShareAmountOutResponse = (g_struct) => [g_struct.token_in_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_JOIN_SWAP_SHARE_AMOUNT_OUT_RESPONSE = '/osmosis.gamm.v1beta1.MsgJoinSwapShareAmountOutResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitSwapShareAmountIn} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: token_out_denom -
 *   - 3: share_in_amount -
 *   - 4: token_out_min_amount -
 */
export const destructOsmosisGammMsgExitSwapShareAmountIn = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.token_out_denom, g_struct.share_in_amount, g_struct.token_out_min_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_SWAP_SHARE_AMOUNT_IN = '/osmosis.gamm.v1beta1.MsgExitSwapShareAmountIn';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitSwapShareAmountInResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_out_amount -
 */
export const destructOsmosisGammMsgExitSwapShareAmountInResponse = (g_struct) => [g_struct.token_out_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_SWAP_SHARE_AMOUNT_IN_RESPONSE = '/osmosis.gamm.v1beta1.MsgExitSwapShareAmountInResponse';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitSwapExternAmountOut} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: token_out -
 *   - 3: share_in_max_amount -
 */
export const destructOsmosisGammMsgExitSwapExternAmountOut = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.token_out, g_struct.share_in_max_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_SWAP_EXTERN_AMOUNT_OUT = '/osmosis.gamm.v1beta1.MsgExitSwapExternAmountOut';
/**
 * Destructures the fields of a {@link OsmosisGammMsgExitSwapExternAmountOutResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: share_in_amount -
 */
export const destructOsmosisGammMsgExitSwapExternAmountOutResponse = (g_struct) => [g_struct.share_in_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_MSG_EXIT_SWAP_EXTERN_AMOUNT_OUT_RESPONSE = '/osmosis.gamm.v1beta1.MsgExitSwapExternAmountOutResponse';
