// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/pool-incentives/v1beta1/query.js
import { parse_duration } from '../../../../api/json.js';
import { decode_protobuf, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, map, temporal } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeOsmosisIncentivesGauge } from '../../incentives/gauge.js';
import { decodeOsmosisPoolincentivesDistrInfo, decodeOsmosisPoolincentivesParams } from './incentives.js';
const SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY = '/osmosis/pool-incentives/v1beta1/';
/**
 * Takes the pool id and returns the matching gauge ids and durations
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sg_pool_id -
 * @returns a tuple where:
 *   - 0: gauge_ids_with_duration - the 'gauge_ids_with_duration' response property
 */
export const queryOsmosisPoolincentivesGaugeIds = restful_grpc(sg_pool_id => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'gauge-ids/' + sg_pool_id]);
/**
 * Returns the pool's matching gauge ids and weights.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: distr_info - the 'distr_info' response property
 */
export const queryOsmosisPoolincentivesDistrInfo = restful_grpc(() => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'distr_info']);
/**
 * Returns pool incentives params.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - the 'params' response property
 */
export const queryOsmosisPoolincentivesParams = restful_grpc(() => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'params']);
/**
 * Returns lock durations for pools.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: lockable_durations - the 'lockable_durations' response property
 */
export const queryOsmosisPoolincentivesLockableDurations = restful_grpc(() => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'lockable_durations']);
/**
 * Returns currently incentivized pools
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: incentivized_pools - the 'incentivized_pools' response property
 */
export const queryOsmosisPoolincentivesIncentivizedPools = restful_grpc(() => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'incentivized_pools']);
/**
 * Returns external incentive gauges.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: data - the 'data' response property
 */
export const queryOsmosisPoolincentivesExternalIncentiveGauges = restful_grpc(() => [SR_LCD_OSMOSIS_POOLINCENTIVES_QUERY + 'external_incentive_gauges']);
/**
 * Encodes a `QueryGaugeIdsRequest` protobuf message:
 * @param sg_pool_id - `pool_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryGaugeIdsRequest = (sg_pool_id) => Protobuf() // ...
    .g(sg_pool_id) // uint64 pool_id = 1
    .o;
/**
 * Encodes a `QueryGaugeIdsResponse` protobuf message:
 * @param a_gauge_ids_with_durations - `gauge_ids_with_duration`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryGaugeIdsResponse = (a_gauge_ids_with_durations) => Protobuf() // ...
    .B(a_gauge_ids_with_durations) // GaugeIdWithDuration[] gauge_ids_with_duration = 1
    .o;
/**
 * Encodes a `QueryDistrInfoRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryDistrInfoRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryDistrInfoResponse` protobuf message:
 * @param g_distr_info - `distr_info`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryDistrInfoResponse = (atu8_distr_info) => Protobuf() // ...
    .b(atu8_distr_info) // DistrInfo distr_info = 1
    .o;
/**
 * Encodes a `QueryParamsRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryParamsResponse` protobuf message:
 * @param g_params - `params`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Encodes a `QueryLockableDurationsRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryLockableDurationsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryLockableDurationsResponse` protobuf message:
 * @param a_lockable_durations - `lockable_durations`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryLockableDurationsResponse = (a_lockable_durations) => Protobuf() // ...
    .B(map(a_lockable_durations, temporal)) // Duration[] lockable_durations = 1
    .o;
/**
 * Encodes a `QueryIncentivizedPoolsRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryIncentivizedPoolsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `IncentivizedPool` protobuf message:
 * @param sg_pool_id - `pool_id`:
 * @param xt_lockable_duration - `lockable_duration`:
 * @param sg_gauge_id - `gauge_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesIncentivizedPool = (sg_pool_id, xt_lockable_duration, sg_gauge_id) => Protobuf() // ...
    .g(sg_pool_id) // uint64 pool_id = 1
    .b(temporal(xt_lockable_duration)) // Duration lockable_duration = 2
    .g(sg_gauge_id) // uint64 gauge_id = 3
    .o;
/**
 * Encodes a `QueryIncentivizedPoolsResponse` protobuf message:
 * @param a_incentivized_pools - `incentivized_pools`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryIncentivizedPoolsResponse = (a_incentivized_pools) => Protobuf() // ...
    .B(a_incentivized_pools) // IncentivizedPool[] incentivized_pools = 1
    .o;
/**
 * Encodes a `QueryExternalIncentiveGaugesRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryExternalIncentiveGaugesRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryExternalIncentiveGaugesResponse` protobuf message:
 * @param a_datas - `data`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryExternalIncentiveGaugesResponse = (a_datas) => Protobuf() // ...
    .B(a_datas) // Gauge[] data = 1
    .o;
/**
 * Encodes a `GaugeIdWithDuration` protobuf message:
 * @param sg_gauge_id - `gauge_id`:
 * @param xt_duration - `duration`:
 * @param s_gauge_incentive_percentage - `gauge_incentive_percentage`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration = (sg_gauge_id, xt_duration, s_gauge_incentive_percentage) => Protobuf() // ...
    .g(sg_gauge_id) // uint64 gauge_id = 1
    .b(temporal(xt_duration)) // Duration duration = 2
    .s(s_gauge_incentive_percentage) // string gauge_incentive_percentage = 3
    .o;
/**
 * Decodes a protobuf QueryGaugeIdsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryGaugeIdsRequest}
 */
export const decodeOsmosisPoolincentivesQueryGaugeIdsRequest = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Decodes a protobuf QueryGaugeIds message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryGaugeIdsResponse}
 */
export const decodeOsmosisPoolincentivesQueryGaugeIdsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration]);
/**
 * Decodes a protobuf QueryDistrInfoRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryDistrInfoRequest}
 */
export const decodeOsmosisPoolincentivesQueryDistrInfoRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryDistrInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryDistrInfoResponse}
 */
export const decodeOsmosisPoolincentivesQueryDistrInfoResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisPoolincentivesDistrInfo]);
/**
 * Decodes a protobuf QueryParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryParamsRequest}
 */
export const decodeOsmosisPoolincentivesQueryParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryParamsResponse}
 */
export const decodeOsmosisPoolincentivesQueryParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisPoolincentivesParams]);
/**
 * Decodes a protobuf QueryLockableDurationsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryLockableDurationsRequest}
 */
export const decodeOsmosisPoolincentivesQueryLockableDurationsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryLockableDurations message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryLockableDurationsResponse}
 */
export const decodeOsmosisPoolincentivesQueryLockableDurationsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [[3, 1]], [decode_temporal]);
/**
 * Decodes a protobuf QueryIncentivizedPoolsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryIncentivizedPoolsRequest}
 */
export const decodeOsmosisPoolincentivesQueryIncentivizedPoolsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf IncentivizedPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesIncentivizedPool}
 */
export const decodeOsmosisPoolincentivesIncentivizedPool = (atu8_payload) => decode_protobuf(atu8_payload, [3, [3, 1], 3], [0, decode_temporal]);
/**
 * Decodes a protobuf QueryIncentivizedPools message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryIncentivizedPoolsResponse}
 */
export const decodeOsmosisPoolincentivesQueryIncentivizedPoolsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisPoolincentivesIncentivizedPool]);
/**
 * Decodes a protobuf QueryExternalIncentiveGaugesRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryExternalIncentiveGaugesRequest}
 */
export const decodeOsmosisPoolincentivesQueryExternalIncentiveGaugesRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryExternalIncentiveGauges message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryExternalIncentiveGaugesResponse}
 */
export const decodeOsmosisPoolincentivesQueryExternalIncentiveGaugesResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisIncentivesGauge]);
/**
 * Decodes a protobuf GaugeIdWithDuration message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration}
 */
export const decodeOsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration = (atu8_payload) => decode_protobuf(atu8_payload, [3, [3, 1], 5], [0, decode_temporal]);
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryGaugeIdsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_id -
 */
export const destructOsmosisPoolincentivesQueryGaugeIdsRequest = (g_struct) => [g_struct.pool_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_GAUGE_IDS_REQUEST = '/osmosis.poolincentives.v1beta1.QueryGaugeIdsRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryGaugeIdsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: gauge_ids_with_duration -
 */
export const destructOsmosisPoolincentivesQueryGaugeIdsResponse = (g_struct) => [g_struct.gauge_ids_with_duration];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_GAUGE_IDS_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryGaugeIdsResponse';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: gauge_id -
 *   - 1: duration -
 *   - 2: gauge_incentive_percentage -
 */
export const destructOsmosisPoolincentivesQueryGaugeIdsResponseGaugeIdWithDuration = (g_struct) => [g_struct.gauge_id, parse_duration(g_struct.duration), g_struct.gauge_incentive_percentage];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_GAUGE_IDS_RESPONSE_GAUGE_ID_WITH_DURATION = '/osmosis.poolincentives.v1beta1.QueryGaugeIdsResponse.GaugeIdWithDuration';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryDistrInfoRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisPoolincentivesQueryDistrInfoRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_DISTR_INFO_REQUEST = '/osmosis.poolincentives.v1beta1.QueryDistrInfoRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryDistrInfoResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: distr_info -
 */
export const destructOsmosisPoolincentivesQueryDistrInfoResponse = (g_struct) => [g_struct.distr_info];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_DISTR_INFO_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryDistrInfoResponse';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisPoolincentivesQueryParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_PARAMS_REQUEST = '/osmosis.poolincentives.v1beta1.QueryParamsRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params -
 */
export const destructOsmosisPoolincentivesQueryParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_PARAMS_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryParamsResponse';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryLockableDurationsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisPoolincentivesQueryLockableDurationsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_LOCKABLE_DURATIONS_REQUEST = '/osmosis.poolincentives.v1beta1.QueryLockableDurationsRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryLockableDurationsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: lockable_durations -
 */
export const destructOsmosisPoolincentivesQueryLockableDurationsResponse = (g_struct) => [map(g_struct.lockable_durations, parse_duration)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_LOCKABLE_DURATIONS_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryLockableDurationsResponse';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryIncentivizedPoolsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisPoolincentivesQueryIncentivizedPoolsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_INCENTIVIZED_POOLS_REQUEST = '/osmosis.poolincentives.v1beta1.QueryIncentivizedPoolsRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesIncentivizedPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_id -
 *   - 1: lockable_duration -
 *   - 2: gauge_id -
 */
export const destructOsmosisPoolincentivesIncentivizedPool = (g_struct) => [g_struct.pool_id, parse_duration(g_struct.lockable_duration), g_struct.gauge_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_INCENTIVIZED_POOL = '/osmosis.poolincentives.v1beta1.IncentivizedPool';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryIncentivizedPoolsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: incentivized_pools -
 */
export const destructOsmosisPoolincentivesQueryIncentivizedPoolsResponse = (g_struct) => [g_struct.incentivized_pools];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_INCENTIVIZED_POOLS_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryIncentivizedPoolsResponse';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryExternalIncentiveGaugesRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisPoolincentivesQueryExternalIncentiveGaugesRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_EXTERNAL_INCENTIVE_GAUGES_REQUEST = '/osmosis.poolincentives.v1beta1.QueryExternalIncentiveGaugesRequest';
/**
 * Destructures the fields of a {@link OsmosisPoolincentivesQueryExternalIncentiveGaugesResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: data -
 */
export const destructOsmosisPoolincentivesQueryExternalIncentiveGaugesResponse = (g_struct) => [g_struct.data];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_POOLINCENTIVES_QUERY_EXTERNAL_INCENTIVE_GAUGES_RESPONSE = '/osmosis.poolincentives.v1beta1.QueryExternalIncentiveGaugesResponse';
