// @solar-republic/cosmos-grpc/build/dist/proto/juno/cwhooks/v1/genesis.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `GenesisState` protobuf message: - initial state of module
 * @param g_params - `params`: Params of this module
 * @param a_staking_contract_addresses - `staking_contract_addresses`:
 * @param a_gov_contract_addresses - `gov_contract_addresses`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksGenesisState = (atu8_params, a_staking_contract_addresses, a_gov_contract_addresses) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .S(a_staking_contract_addresses) // string[] staking_contract_addresses = 2
    .S(a_gov_contract_addresses) // string[] gov_contract_addresses = 3
    .o;
/**
 * Encodes a `Params` protobuf message: defines the set of module parameters.
 * @param sg_contract_gas_limit - `contract_gas_limit`: the contract call gas limit
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksParams = (sg_contract_gas_limit) => Protobuf() // ...
    .g(sg_contract_gas_limit) // uint64 contract_gas_limit = 1
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksGenesisState}
 */
export const decodeJunoCwhooksGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1, 4, 4], [decodeJunoCwhooksParams]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksParams}
 */
export const decodeJunoCwhooksParams = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Destructures the fields of a {@link JunoCwhooksParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_gas_limit - the contract call gas limit
 */
export const destructJunoCwhooksParams = (g_struct) => [g_struct.contract_gas_limit];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_PARAMS = '/juno.cwhooks.v1.Params';
/**
 * Destructures the fields of a {@link JunoCwhooksGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - Params of this module
 *   - 1: staking_contract_addresses -
 *   - 2: gov_contract_addresses -
 */
export const destructJunoCwhooksGenesisState = (g_struct) => [g_struct.params, g_struct.staking_contract_addresses, g_struct.gov_contract_addresses];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_GENESIS_STATE = '/juno.cwhooks.v1.GenesisState';
