// @solar-republic/cosmos-grpc/build/dist/proto/juno/cwhooks/v1/tx.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeJunoCwhooksParams } from './genesis.js';
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type. Since: cosmos-sdk 0.47
 * @param s_authority - `authority`: the address of the governance account.
 * @param g_params - `params`: defines the x/clock parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUpdateParams = (s_authority, atu8_params) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgRegisterStaking` protobuf message:
 * @param sa_contract - `contract_address`:
 * @param sa_register - `register_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgRegisterStaking = (sa_contract, sa_register) => Protobuf() // ...
    .s(sa_contract) // string contract_address = 1
    .s(sa_register) // string register_address = 2
    .o;
/**
 * Encodes a `MsgRegisterStakingResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgRegisterStakingResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgRegisterGovernance` protobuf message:
 * @param sa_contract - `contract_address`:
 * @param sa_register - `register_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgRegisterGovernance = (sa_contract, sa_register) => Protobuf() // ...
    .s(sa_contract) // string contract_address = 1
    .s(sa_register) // string register_address = 2
    .o;
/**
 * Encodes a `MsgRegisterGovernanceResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgRegisterGovernanceResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUnregisterGovernance` protobuf message:
 * @param sa_contract - `contract_address`:
 * @param sa_register - `register_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUnregisterGovernance = (sa_contract, sa_register) => Protobuf() // ...
    .s(sa_contract) // string contract_address = 1
    .s(sa_register) // string register_address = 2
    .o;
/**
 * Encodes a `MsgUnregisterGovernanceResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUnregisterGovernanceResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUnregisterStaking` protobuf message:
 * @param sa_contract - `contract_address`:
 * @param sa_register - `register_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUnregisterStaking = (sa_contract, sa_register) => Protobuf() // ...
    .s(sa_contract) // string contract_address = 1
    .s(sa_register) // string register_address = 2
    .o;
/**
 * Encodes a `MsgUnregisterStakingResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksMsgUnregisterStakingResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUpdateParams}
 */
export const decodeJunoCwhooksMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeJunoCwhooksParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUpdateParamsResponse}
 */
export const decodeJunoCwhooksMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf RegisterStaking message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgRegisterStaking}
 */
export const decodeJunoCwhooksMsgRegisterStaking = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf RegisterStaking message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgRegisterStakingResponse}
 */
export const decodeJunoCwhooksMsgRegisterStakingResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf RegisterGovernance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgRegisterGovernance}
 */
export const decodeJunoCwhooksMsgRegisterGovernance = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf RegisterGovernance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgRegisterGovernanceResponse}
 */
export const decodeJunoCwhooksMsgRegisterGovernanceResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UnregisterGovernance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUnregisterGovernance}
 */
export const decodeJunoCwhooksMsgUnregisterGovernance = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf UnregisterGovernance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUnregisterGovernanceResponse}
 */
export const decodeJunoCwhooksMsgUnregisterGovernanceResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UnregisterStaking message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUnregisterStaking}
 */
export const decodeJunoCwhooksMsgUnregisterStaking = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf UnregisterStaking message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksMsgUnregisterStakingResponse}
 */
export const decodeJunoCwhooksMsgUnregisterStakingResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address of the governance account.
 *   - 1: params - defines the x/clock parameters to update. NOTE: All parameters must be supplied.
 */
export const destructJunoCwhooksMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UPDATE_PARAMS = '/juno.cwhooks.v1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UPDATE_PARAMS_RESPONSE = '/juno.cwhooks.v1.MsgUpdateParamsResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgRegisterStaking} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_address -
 *   - 1: register_address -
 */
export const destructJunoCwhooksMsgRegisterStaking = (g_struct) => [g_struct.contract_address, g_struct.register_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_REGISTER_STAKING = '/juno.cwhooks.v1.MsgRegisterStaking';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgRegisterStakingResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksMsgRegisterStakingResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_REGISTER_STAKING_RESPONSE = '/juno.cwhooks.v1.MsgRegisterStakingResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgRegisterGovernance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_address -
 *   - 1: register_address -
 */
export const destructJunoCwhooksMsgRegisterGovernance = (g_struct) => [g_struct.contract_address, g_struct.register_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_REGISTER_GOVERNANCE = '/juno.cwhooks.v1.MsgRegisterGovernance';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgRegisterGovernanceResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksMsgRegisterGovernanceResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_REGISTER_GOVERNANCE_RESPONSE = '/juno.cwhooks.v1.MsgRegisterGovernanceResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUnregisterGovernance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_address -
 *   - 1: register_address -
 */
export const destructJunoCwhooksMsgUnregisterGovernance = (g_struct) => [g_struct.contract_address, g_struct.register_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UNREGISTER_GOVERNANCE = '/juno.cwhooks.v1.MsgUnregisterGovernance';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUnregisterGovernanceResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksMsgUnregisterGovernanceResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UNREGISTER_GOVERNANCE_RESPONSE = '/juno.cwhooks.v1.MsgUnregisterGovernanceResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUnregisterStaking} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_address -
 *   - 1: register_address -
 */
export const destructJunoCwhooksMsgUnregisterStaking = (g_struct) => [g_struct.contract_address, g_struct.register_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UNREGISTER_STAKING = '/juno.cwhooks.v1.MsgUnregisterStaking';
/**
 * Destructures the fields of a {@link JunoCwhooksMsgUnregisterStakingResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksMsgUnregisterStakingResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_MSG_UNREGISTER_STAKING_RESPONSE = '/juno.cwhooks.v1.MsgUnregisterStakingResponse';
