// @solar-republic/cosmos-grpc/build/dist/proto/juno/drip/v1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeJunoDripParams } from './genesis.js';
const SR_LCD_JUNO_DRIP_QUERY = '/juno/drip/v1/params';
/**
 * Retrieves the Drip module params
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - the returned parameter from the module
 */
export const queryJunoDripParams = restful_grpc(() => [SR_LCD_JUNO_DRIP_QUERY]);
/**
 * Encodes a `QueryParamsRequest` protobuf message: the request type for the Query/Params RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoDripQueryParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryParamsResponse` protobuf message: the response type for the Query/Params RPC method.
 * @param g_params - `params`: the returned parameter from the module
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoDripQueryParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Decodes a protobuf QueryParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoDripQueryParamsRequest}
 */
export const decodeJunoDripQueryParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoDripQueryParamsResponse}
 */
export const decodeJunoDripQueryParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeJunoDripParams]);
/**
 * Destructures the fields of a {@link JunoDripQueryParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoDripQueryParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_DRIP_QUERY_PARAMS_REQUEST = '/juno.drip.v1.QueryParamsRequest';
/**
 * Destructures the fields of a {@link JunoDripQueryParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - the returned parameter from the module
 */
export const destructJunoDripQueryParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_DRIP_QUERY_PARAMS_RESPONSE = '/juno.drip.v1.QueryParamsResponse';
