// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/staking/v1beta1/tx.js
import { parse_timestamp } from '../../../../api/json.js';
import { decode_protobuf, decode_coin, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, temporal, coin } from '../../../../api/protobuf-writer.js';
import { decodeCosmosStakingDescription, decodeCosmosStakingCommissionRates, decodeCosmosStakingParams } from './staking.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
/**
 * Encodes a `MsgCreateValidator` protobuf message: defines a SDK message for creating a new validator.
 * @param g_description - `description`:
 * @param g_commission - `commission`:
 * @param s_min_self_delegation - `min_self_delegation`:
 * @param sa_delegator - `delegator_address`: Deprecated: Use of Delegator Address in MsgCreateValidator is deprecated. The validator address bytes and delegator address bytes refer to the same account while creating validator (defer only in bech32 notation).
 * @param sa_validator - `validator_address`:
 * @param atu8_pubkey - `pubkey`:
 * @param a_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgCreateValidator = (atu8_description, atu8_commission, s_min_self_delegation, sa_delegator, sa_validator, atu8_pubkey, a_value) => Protobuf() // ...
    .b(atu8_description) // Description description = 1
    .b(atu8_commission) // CommissionRates commission = 2
    .s(s_min_self_delegation) // string min_self_delegation = 3
    .s(sa_delegator) // string delegator_address = 4
    .s(sa_validator) // string validator_address = 5
    .b(atu8_pubkey) // Any pubkey = 6
    .b(coin(a_value)) // Coin value = 7
    .o;
/**
 * Encodes a `MsgCreateValidatorResponse` protobuf message: defines the Msg/CreateValidator response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgCreateValidatorResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgEditValidator` protobuf message: defines a SDK message for editing an existing validator.
 * @param g_description - `description`:
 * @param sa_validator - `validator_address`:
 * @param s_commission_rate - `commission_rate`: We pass a reference to the new commission rate and min self delegation as it's not mandatory to update. If not updated, the deserialized rate will be zero with no way to distinguish if an update was intended. REF: #2373
 * @param s_min_self_delegation - `min_self_delegation`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgEditValidator = (atu8_description, sa_validator, s_commission_rate, s_min_self_delegation) => Protobuf() // ...
    .b(atu8_description) // Description description = 1
    .s(sa_validator) // string validator_address = 2
    .s(s_commission_rate) // string commission_rate = 3
    .s(s_min_self_delegation) // string min_self_delegation = 4
    .o;
/**
 * Encodes a `MsgEditValidatorResponse` protobuf message: defines the Msg/EditValidator response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgEditValidatorResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgDelegate` protobuf message: defines a SDK message for performing a delegation of coins from a delegator to a validator.
 * @param sa_delegator - `delegator_address`:
 * @param sa_validator - `validator_address`:
 * @param a_amount - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgDelegate = (sa_delegator, sa_validator, a_amount) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_validator) // string validator_address = 2
    .b(coin(a_amount)) // Coin amount = 3
    .o;
/**
 * Encodes a `MsgDelegateResponse` protobuf message: defines the Msg/Delegate response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgDelegateResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgBeginRedelegate` protobuf message: defines a SDK message for performing a redelegation of coins from a delegator and source validator to a destination validator.
 * @param sa_delegator - `delegator_address`:
 * @param sa_validator_src - `validator_src_address`:
 * @param sa_validator_dst - `validator_dst_address`:
 * @param a_amount - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgBeginRedelegate = (sa_delegator, sa_validator_src, sa_validator_dst, a_amount) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_validator_src) // string validator_src_address = 2
    .s(sa_validator_dst) // string validator_dst_address = 3
    .b(coin(a_amount)) // Coin amount = 4
    .o;
/**
 * Encodes a `MsgBeginRedelegateResponse` protobuf message: defines the Msg/BeginRedelegate response type.
 * @param xt_completion_time - `completion_time`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgBeginRedelegateResponse = (xt_completion_time) => Protobuf() // ...
    .b(temporal(xt_completion_time)) // Timestamp completion_time = 1
    .o;
/**
 * Encodes a `MsgUndelegate` protobuf message: defines a SDK message for performing an undelegation from a delegate and a validator.
 * @param sa_delegator - `delegator_address`:
 * @param sa_validator - `validator_address`:
 * @param a_amount - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgUndelegate = (sa_delegator, sa_validator, a_amount) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_validator) // string validator_address = 2
    .b(coin(a_amount)) // Coin amount = 3
    .o;
/**
 * Encodes a `MsgUndelegateResponse` protobuf message: defines the Msg/Undelegate response type.
 * @param xt_completion_time - `completion_time`:
 * @param a_amount - `amount`: returns the amount of undelegated coins Since: cosmos-sdk 0.50
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgUndelegateResponse = (xt_completion_time, a_amount) => Protobuf() // ...
    .b(temporal(xt_completion_time)) // Timestamp completion_time = 1
    .b(coin(a_amount)) // Coin amount = 2
    .o;
/**
 * Encodes a `MsgCancelUnbondingDelegation` protobuf message: defines the SDK message for performing a cancel unbonding delegation for delegator Since: cosmos-sdk 0.46
 * @param sa_delegator - `delegator_address`:
 * @param sa_validator - `validator_address`:
 * @param a_amount - `amount`: always less than or equal to unbonding delegation entry balance
 * @param sg_creation_height - `creation_height`: the height which the unbonding took place.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgCancelUnbondingDelegation = (sa_delegator, sa_validator, a_amount, sg_creation_height) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_validator) // string validator_address = 2
    .b(coin(a_amount)) // Coin amount = 3
    .g(sg_creation_height) // int32 creation_height = 4
    .o;
/**
 * Encodes a `MsgCancelUnbondingDelegationResponse` protobuf message: Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgCancelUnbondingDelegationResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type. Since: cosmos-sdk 0.47
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param g_params - `params`: defines the x/staking parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgUpdateParams = (s_authority, atu8_params) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosStakingMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf CreateValidator message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgCreateValidator}
 */
export const decodeCosmosStakingMsgCreateValidator = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 5, 5, 5, 1, 1], [decodeCosmosStakingDescription, decodeCosmosStakingCommissionRates, 0, 0, 0, decodeGoogleProtobufAny, decode_coin]);
/**
 * Decodes a protobuf CreateValidator message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgCreateValidatorResponse}
 */
export const decodeCosmosStakingMsgCreateValidatorResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf EditValidator message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgEditValidator}
 */
export const decodeCosmosStakingMsgEditValidator = (atu8_payload) => decode_protobuf(atu8_payload, [1, 5, 5, 5], [decodeCosmosStakingDescription]);
/**
 * Decodes a protobuf EditValidator message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgEditValidatorResponse}
 */
export const decodeCosmosStakingMsgEditValidatorResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf Delegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgDelegate}
 */
export const decodeCosmosStakingMsgDelegate = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 1], [0, 0, decode_coin]);
/**
 * Decodes a protobuf Delegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgDelegateResponse}
 */
export const decodeCosmosStakingMsgDelegateResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf BeginRedelegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgBeginRedelegate}
 */
export const decodeCosmosStakingMsgBeginRedelegate = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5, 1], [0, 0, 0, decode_coin]);
/**
 * Decodes a protobuf BeginRedelegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgBeginRedelegateResponse}
 */
export const decodeCosmosStakingMsgBeginRedelegateResponse = (atu8_payload) => decode_protobuf(atu8_payload, [[3, 1]], [decode_temporal]);
/**
 * Decodes a protobuf Undelegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgUndelegate}
 */
export const decodeCosmosStakingMsgUndelegate = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 1], [0, 0, decode_coin]);
/**
 * Decodes a protobuf Undelegate message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgUndelegateResponse}
 */
export const decodeCosmosStakingMsgUndelegateResponse = (atu8_payload) => decode_protobuf(atu8_payload, [[3, 1], 1], [decode_temporal, decode_coin]);
/**
 * Decodes a protobuf CancelUnbondingDelegation message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgCancelUnbondingDelegation}
 */
export const decodeCosmosStakingMsgCancelUnbondingDelegation = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 1, 3], [0, 0, decode_coin]);
/**
 * Decodes a protobuf CancelUnbondingDelegation message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgCancelUnbondingDelegationResponse}
 */
export const decodeCosmosStakingMsgCancelUnbondingDelegationResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgUpdateParams}
 */
export const decodeCosmosStakingMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosStakingParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosStakingMsgUpdateParamsResponse}
 */
export const decodeCosmosStakingMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosStakingMsgCreateValidator} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: description -
 *   - 1: commission -
 *   - 2: min_self_delegation -
 *   - 3: delegator_address - Deprecated: Use of Delegator Address in MsgCreateValidator is deprecated. The validator address bytes and delegator address bytes refer to the same account while creating validator (defer only in bech32 notation).
 *   - 4: validator_address -
 *   - 5: pubkey -
 *   - 6: value -
 */
export const destructCosmosStakingMsgCreateValidator = (g_struct) => [g_struct.description, g_struct.commission, g_struct.min_self_delegation, g_struct.delegator_address, g_struct.validator_address, g_struct.pubkey, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_CREATE_VALIDATOR = '/cosmos.staking.v1beta1.MsgCreateValidator';
/**
 * Destructures the fields of a {@link CosmosStakingMsgCreateValidatorResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosStakingMsgCreateValidatorResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_CREATE_VALIDATOR_RESPONSE = '/cosmos.staking.v1beta1.MsgCreateValidatorResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgEditValidator} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: description -
 *   - 1: validator_address -
 *   - 2: commission_rate - We pass a reference to the new commission rate and min self delegation as it's not mandatory to update. If not updated, the deserialized rate will be zero with no way to distinguish if an update was intended. REF: #2373
 *   - 3: min_self_delegation -
 */
export const destructCosmosStakingMsgEditValidator = (g_struct) => [g_struct.description, g_struct.validator_address, g_struct.commission_rate, g_struct.min_self_delegation];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_EDIT_VALIDATOR = '/cosmos.staking.v1beta1.MsgEditValidator';
/**
 * Destructures the fields of a {@link CosmosStakingMsgEditValidatorResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosStakingMsgEditValidatorResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_EDIT_VALIDATOR_RESPONSE = '/cosmos.staking.v1beta1.MsgEditValidatorResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgDelegate} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: validator_address -
 *   - 2: amount -
 */
export const destructCosmosStakingMsgDelegate = (g_struct) => [g_struct.delegator_address, g_struct.validator_address, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_DELEGATE = '/cosmos.staking.v1beta1.MsgDelegate';
/**
 * Destructures the fields of a {@link CosmosStakingMsgDelegateResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosStakingMsgDelegateResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_DELEGATE_RESPONSE = '/cosmos.staking.v1beta1.MsgDelegateResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgBeginRedelegate} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: validator_src_address -
 *   - 2: validator_dst_address -
 *   - 3: amount -
 */
export const destructCosmosStakingMsgBeginRedelegate = (g_struct) => [g_struct.delegator_address, g_struct.validator_src_address, g_struct.validator_dst_address, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_BEGIN_REDELEGATE = '/cosmos.staking.v1beta1.MsgBeginRedelegate';
/**
 * Destructures the fields of a {@link CosmosStakingMsgBeginRedelegateResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: completion_time -
 */
export const destructCosmosStakingMsgBeginRedelegateResponse = (g_struct) => [parse_timestamp(g_struct.completion_time)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_BEGIN_REDELEGATE_RESPONSE = '/cosmos.staking.v1beta1.MsgBeginRedelegateResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgUndelegate} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: validator_address -
 *   - 2: amount -
 */
export const destructCosmosStakingMsgUndelegate = (g_struct) => [g_struct.delegator_address, g_struct.validator_address, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_UNDELEGATE = '/cosmos.staking.v1beta1.MsgUndelegate';
/**
 * Destructures the fields of a {@link CosmosStakingMsgUndelegateResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: completion_time -
 *   - 1: amount - returns the amount of undelegated coins Since: cosmos-sdk 0.50
 */
export const destructCosmosStakingMsgUndelegateResponse = (g_struct) => [parse_timestamp(g_struct.completion_time), g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_UNDELEGATE_RESPONSE = '/cosmos.staking.v1beta1.MsgUndelegateResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgCancelUnbondingDelegation} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: validator_address -
 *   - 2: amount - always less than or equal to unbonding delegation entry balance
 *   - 3: creation_height - the height which the unbonding took place.
 */
export const destructCosmosStakingMsgCancelUnbondingDelegation = (g_struct) => [g_struct.delegator_address, g_struct.validator_address, g_struct.amount, g_struct.creation_height];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_CANCEL_UNBONDING_DELEGATION = '/cosmos.staking.v1beta1.MsgCancelUnbondingDelegation';
/**
 * Destructures the fields of a {@link CosmosStakingMsgCancelUnbondingDelegationResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosStakingMsgCancelUnbondingDelegationResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_CANCEL_UNBONDING_DELEGATION_RESPONSE = '/cosmos.staking.v1beta1.MsgCancelUnbondingDelegationResponse';
/**
 * Destructures the fields of a {@link CosmosStakingMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: params - defines the x/staking parameters to update. NOTE: All parameters must be supplied.
 */
export const destructCosmosStakingMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_UPDATE_PARAMS = '/cosmos.staking.v1beta1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link CosmosStakingMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosStakingMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_STAKING_MSG_UPDATE_PARAMS_RESPONSE = '/cosmos.staking.v1beta1.MsgUpdateParamsResponse';
