// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/pool-models/stableswap/tx.js
import { decode_protobuf, decode_coin } from '../../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../../api/protobuf-writer.js';
import { decodeOsmosisGammPoolmodelsStableswapPoolParams } from './stableswap_pool.js';
/**
 * Encodes a `MsgCreateStableswapPool` protobuf message: ===================== MsgCreatePool
 * @param sa_sender - `sender`:
 * @param g_pool_params - `pool_params`:
 * @param a_initial_pool_liquiditys - `initial_pool_liquidity`:
 * @param a_scaling_factors - `scaling_factors`:
 * @param s_future_pool_governor - `future_pool_governor`:
 * @param s_scaling_factor_controller - `scaling_factor_controller`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapMsgCreateStableswapPool = (sa_sender, atu8_pool_params, a_initial_pool_liquiditys, a_scaling_factors, s_future_pool_governor, s_scaling_factor_controller) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .b(atu8_pool_params) // PoolParams pool_params = 2
    .B(coins(a_initial_pool_liquiditys)) // Coin[] initial_pool_liquidity = 3
    .G(a_scaling_factors) // uint64[] scaling_factors = 4
    .s(s_future_pool_governor) // string future_pool_governor = 5
    .s(s_scaling_factor_controller) // string scaling_factor_controller = 6
    .o;
/**
 * Encodes a `MsgCreateStableswapPoolResponse` protobuf message: Returns a poolID with custom poolName.
 * @param sg_pool_id - `pool_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapMsgCreateStableswapPoolResponse = (sg_pool_id) => Protobuf() // ...
    .g(sg_pool_id) // uint64 pool_id = 1
    .o;
/**
 * Encodes a `MsgStableSwapAdjustScalingFactors` protobuf message: Sender must be the pool's scaling_factor_governor in order for the tx to succeed. Adjusts stableswap scaling factors.
 * @param sa_sender - `sender`:
 * @param sg_pool_id - `pool_id`:
 * @param a_scaling_factors - `scaling_factors`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactors = (sa_sender, sg_pool_id, a_scaling_factors) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .g(sg_pool_id) // uint64 pool_id = 2
    .G(a_scaling_factors) // uint64[] scaling_factors = 3
    .o;
/**
 * Encodes a `MsgStableSwapAdjustScalingFactorsResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactorsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf CreateStableswapPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapMsgCreateStableswapPool}
 */
export const decodeOsmosisGammPoolmodelsStableswapMsgCreateStableswapPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 0, 2, 5, 5], [0, decodeOsmosisGammPoolmodelsStableswapPoolParams, decode_coin]);
/**
 * Decodes a protobuf CreateStableswapPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapMsgCreateStableswapPoolResponse}
 */
export const decodeOsmosisGammPoolmodelsStableswapMsgCreateStableswapPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Decodes a protobuf StableSwapAdjustScalingFactors message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactors}
 */
export const decodeOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactors = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 2]);
/**
 * Decodes a protobuf StableSwapAdjustScalingFactors message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactorsResponse}
 */
export const decodeOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactorsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapMsgCreateStableswapPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_params -
 *   - 2: initial_pool_liquidity -
 *   - 3: scaling_factors -
 *   - 4: future_pool_governor -
 *   - 5: scaling_factor_controller -
 */
export const destructOsmosisGammPoolmodelsStableswapMsgCreateStableswapPool = (g_struct) => [g_struct.sender, g_struct.pool_params, g_struct.initial_pool_liquidity, g_struct.scaling_factors, g_struct.future_pool_governor, g_struct.scaling_factor_controller];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_MSG_CREATE_STABLESWAP_POOL = '/osmosis.gamm.poolmodels.stableswap.v1beta1.MsgCreateStableswapPool';
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapMsgCreateStableswapPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_id -
 */
export const destructOsmosisGammPoolmodelsStableswapMsgCreateStableswapPoolResponse = (g_struct) => [g_struct.pool_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_MSG_CREATE_STABLESWAP_POOL_RESPONSE = '/osmosis.gamm.poolmodels.stableswap.v1beta1.MsgCreateStableswapPoolResponse';
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactors} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: pool_id -
 *   - 2: scaling_factors -
 */
export const destructOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactors = (g_struct) => [g_struct.sender, g_struct.pool_id, g_struct.scaling_factors];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_MSG_STABLE_SWAP_ADJUST_SCALING_FACTORS = '/osmosis.gamm.poolmodels.stableswap.v1beta1.MsgStableSwapAdjustScalingFactors';
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactorsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisGammPoolmodelsStableswapMsgStableSwapAdjustScalingFactorsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_MSG_STABLE_SWAP_ADJUST_SCALING_FACTORS_RESPONSE = '/osmosis.gamm.poolmodels.stableswap.v1beta1.MsgStableSwapAdjustScalingFactorsResponse';
