// @solar-republic/cosmos-grpc/build/dist/proto/juno/cwhooks/v1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeJunoCwhooksParams } from './genesis.js';
const SR_LCD_JUNO_CWHOOKS_QUERY = '/juno/cwhooks/v1/';
/**
 *
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - the 'params' response property
 */
export const queryJunoCwhooksParams = restful_grpc(() => [SR_LCD_JUNO_CWHOOKS_QUERY + 'params']);
/**
 *
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: contracts - the 'contracts' response property
 */
export const queryJunoCwhooksStakingContracts = restful_grpc(() => [SR_LCD_JUNO_CWHOOKS_QUERY + 'staking_contracts']);
/**
 *
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: contracts - the 'contracts' response property
 */
export const queryJunoCwhooksGovernanceContracts = restful_grpc(() => [SR_LCD_JUNO_CWHOOKS_QUERY + 'governance_contracts']);
/**
 * Encodes a `QueryParamsRequest` protobuf message: QueryParams is the request type to get all module params.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryParamsResponse` protobuf message: QueryClockContractsResponse is the response type for the Query/ClockContracts RPC method.
 * @param g_params - `params`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Encodes a `QueryStakingContractsRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryStakingContractsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryStakingContractsResponse` protobuf message:
 * @param a_contracts - `contracts`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryStakingContractsResponse = (a_contracts) => Protobuf() // ...
    .S(a_contracts) // string[] contracts = 1
    .o;
/**
 * Encodes a `QueryGovernanceContractsRequest` protobuf message: QueryStakingContractsRequest
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryGovernanceContractsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryGovernanceContractsResponse` protobuf message:
 * @param a_contracts - `contracts`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoCwhooksQueryGovernanceContractsResponse = (a_contracts) => Protobuf() // ...
    .S(a_contracts) // string[] contracts = 1
    .o;
/**
 * Decodes a protobuf QueryParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryParamsRequest}
 */
export const decodeJunoCwhooksQueryParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryParamsResponse}
 */
export const decodeJunoCwhooksQueryParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeJunoCwhooksParams]);
/**
 * Decodes a protobuf QueryStakingContractsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryStakingContractsRequest}
 */
export const decodeJunoCwhooksQueryStakingContractsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryStakingContracts message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryStakingContractsResponse}
 */
export const decodeJunoCwhooksQueryStakingContractsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [4]);
/**
 * Decodes a protobuf QueryGovernanceContractsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryGovernanceContractsRequest}
 */
export const decodeJunoCwhooksQueryGovernanceContractsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryGovernanceContracts message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoCwhooksQueryGovernanceContractsResponse}
 */
export const decodeJunoCwhooksQueryGovernanceContractsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [4]);
/**
 * Destructures the fields of a {@link JunoCwhooksQueryParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksQueryParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_PARAMS_REQUEST = '/juno.cwhooks.v1.QueryParamsRequest';
/**
 * Destructures the fields of a {@link JunoCwhooksQueryParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params -
 */
export const destructJunoCwhooksQueryParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_PARAMS_RESPONSE = '/juno.cwhooks.v1.QueryParamsResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksQueryStakingContractsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksQueryStakingContractsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_STAKING_CONTRACTS_REQUEST = '/juno.cwhooks.v1.QueryStakingContractsRequest';
/**
 * Destructures the fields of a {@link JunoCwhooksQueryStakingContractsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contracts -
 */
export const destructJunoCwhooksQueryStakingContractsResponse = (g_struct) => [g_struct.contracts];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_STAKING_CONTRACTS_RESPONSE = '/juno.cwhooks.v1.QueryStakingContractsResponse';
/**
 * Destructures the fields of a {@link JunoCwhooksQueryGovernanceContractsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoCwhooksQueryGovernanceContractsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_GOVERNANCE_CONTRACTS_REQUEST = '/juno.cwhooks.v1.QueryGovernanceContractsRequest';
/**
 * Destructures the fields of a {@link JunoCwhooksQueryGovernanceContractsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contracts -
 */
export const destructJunoCwhooksQueryGovernanceContractsResponse = (g_struct) => [g_struct.contracts];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CWHOOKS_QUERY_GOVERNANCE_CONTRACTS_RESPONSE = '/juno.cwhooks.v1.QueryGovernanceContractsResponse';
