// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/downtime-detector/v1beta1/downtime_duration.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_30S** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S = 0;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_30S** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S = 'DURATION_30S';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M = 1;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M = 'DURATION_1M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M = 2;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M = 'DURATION_2M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_3M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M = 3;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_3M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M = 'DURATION_3M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_4M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M = 4;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_4M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M = 'DURATION_4M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_5M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M = 5;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_5M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M = 'DURATION_5M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_10M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M = 6;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_10M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M = 'DURATION_10M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_20M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M = 7;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_20M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M = 'DURATION_20M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_30M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M = 8;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_30M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M = 'DURATION_30M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_40M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M = 9;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_40M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M = 'DURATION_40M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_50M** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M = 10;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_50M** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M = 'DURATION_50M';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H = 11;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H = 'DURATION_1H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1_5H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H = 12;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_1_5H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H = 'DURATION_1_5H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H = 13;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H = 'DURATION_2H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2_5H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H = 14;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_2_5H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H = 'DURATION_2_5H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_3H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H = 15;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_3H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H = 'DURATION_3H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_4H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H = 16;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_4H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H = 'DURATION_4H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_5H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H = 17;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_5H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H = 'DURATION_5H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_6H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H = 18;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_6H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H = 'DURATION_6H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_9H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H = 19;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_9H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H = 'DURATION_9H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_12H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H = 20;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_12H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H = 'DURATION_12H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_18H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H = 21;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_18H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H = 'DURATION_18H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_24H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H = 22;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_24H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H = 'DURATION_24H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_36H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H = 23;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_36H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H = 'DURATION_36H';
/**
 * Protobuf enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_48H** -
 *
 * Belongs to enum type {@link ProtoEnumOsmosisDowntimedetectorDowntime}
 */
export const XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H = 24;
/**
 * JSON enum value for `osmosis.downtimedetector.v1beta1.Downtime`.
 *
 * **DURATION_48H** -
 *
 * Belongs to enum type {@link JsonEnumOsmosisDowntimedetectorDowntime}
 */
export const SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H = 'DURATION_48H';
/**
 * Maps a protobuf enum int value for `osmosis.downtimedetector.v1beta1.Downtime` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumOsmosisDowntimedetectorDowntime = {
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H,
    [XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H]: SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H,
};
/**
 * Maps a JSON enum string value for `osmosis.downtimedetector.v1beta1.Downtime` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumOsmosisDowntimedetectorDowntime = {
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30S,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_10M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_20M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_30M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_40M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_50M,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_1_5H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_2_5H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_3H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_4H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_5H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_6H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_9H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_12H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_18H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_24H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_36H,
    [SI_JSON_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H]: XC_PROTO_OSMOSIS_DOWNTIMEDETECTOR_DURATION_48H,
};
