// @solar-republic/cosmos-grpc/build/dist/proto/akash/deployment/v1beta2/deployment.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
/**
 * Encodes a `DeploymentID` protobuf message: stores owner and sequence number
 * @param sa_owner - `owner`:
 * @param sg_dseq - `dseq`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentDeploymentID = (sa_owner, sg_dseq) => Protobuf() // ...
    .s(sa_owner) // string owner = 1
    .g(sg_dseq) // uint64 dseq = 2
    .o;
/**
 * Encodes a `Deployment` protobuf message: stores deploymentID, state and version details
 * @param g_deployment_id - `deployment_id`:
 * @param xc_state - `state`:
 * @param atu8_version - `version`:
 * @param sg_created_at - `created_at`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentDeployment = (atu8_deployment_id, xc_state, atu8_version, sg_created_at) => Protobuf() // ...
    .b(atu8_deployment_id) // DeploymentID deployment_id = 1
    .v(xc_state) // State state = 2
    .b(atu8_version) // bytes version = 3
    .g(sg_created_at) // int32 created_at = 4
    .o;
/**
 * Encodes a `DeploymentFilters` protobuf message: defines filters used to filter deployments
 * @param sa_owner - `owner`:
 * @param sg_dseq - `dseq`:
 * @param s_state - `state`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentDeploymentFilters = (sa_owner, sg_dseq, s_state) => Protobuf() // ...
    .s(sa_owner) // string owner = 1
    .g(sg_dseq) // uint64 dseq = 2
    .s(s_state) // string state = 3
    .o;
/**
 * Decodes a protobuf DeploymentID message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentDeploymentID}
 */
export const decodeAkashDeploymentDeploymentID = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3]);
/**
 * Decodes a protobuf Deployment message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentDeployment}
 */
export const decodeAkashDeploymentDeployment = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1, 3], [decodeAkashDeploymentDeploymentID]);
/**
 * Decodes a protobuf DeploymentFilters message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentDeploymentFilters}
 */
export const decodeAkashDeploymentDeploymentFilters = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 5]);
/**
 * Destructures the fields of a {@link AkashDeploymentDeploymentID} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: owner -
 *   - 1: dseq -
 */
export const destructAkashDeploymentDeploymentID = (g_struct) => [g_struct.owner, g_struct.dseq];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_DEPLOYMENT_I_D = '/akash.deployment.v1beta2.DeploymentID';
/**
 * Destructures the fields of a {@link AkashDeploymentDeployment} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: deployment_id -
 *   - 1: state -
 *   - 2: version -
 *   - 3: created_at -
 */
export const destructAkashDeploymentDeployment = (g_struct) => [g_struct.deployment_id, g_struct.state, safe_base64_to_bytes(g_struct.version), g_struct.created_at];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_DEPLOYMENT = '/akash.deployment.v1beta2.Deployment';
/**
 * Destructures the fields of a {@link AkashDeploymentDeploymentFilters} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: owner -
 *   - 1: dseq -
 *   - 2: state -
 */
export const destructAkashDeploymentDeploymentFilters = (g_struct) => [g_struct.owner, g_struct.dseq, g_struct.state];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_DEPLOYMENT_FILTERS = '/akash.deployment.v1beta2.DeploymentFilters';
/**
 * Protobuf enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **invalid** -
 *
 * Belongs to enum type {@link ProtoEnumAkashDeploymentDeploymentState}
 */
export const XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_invalid = 0;
/**
 * JSON enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **invalid** -
 *
 * Belongs to enum type {@link JsonEnumAkashDeploymentDeploymentState}
 */
export const SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_invalid = 'DeploymentStateInvalid';
/**
 * Protobuf enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **active** -
 *
 * Belongs to enum type {@link ProtoEnumAkashDeploymentDeploymentState}
 */
export const XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_active = 1;
/**
 * JSON enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **active** -
 *
 * Belongs to enum type {@link JsonEnumAkashDeploymentDeploymentState}
 */
export const SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_active = 'DeploymentActive';
/**
 * Protobuf enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **closed** -
 *
 * Belongs to enum type {@link ProtoEnumAkashDeploymentDeploymentState}
 */
export const XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_closed = 2;
/**
 * JSON enum value for `akash.deployment.v1beta2.Deployment.State`.
 *
 * **closed** -
 *
 * Belongs to enum type {@link JsonEnumAkashDeploymentDeploymentState}
 */
export const SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_closed = 'DeploymentClosed';
/**
 * Maps a protobuf enum int value for `akash.deployment.v1beta2.Deployment.State` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAkashDeploymentDeploymentState = {
    [XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_invalid]: SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_invalid,
    [XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_active]: SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_active,
    [XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_closed]: SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_closed,
};
/**
 * Maps a JSON enum string value for `akash.deployment.v1beta2.Deployment.State` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAkashDeploymentDeploymentState = {
    [SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_invalid]: XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_invalid,
    [SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_active]: XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_active,
    [SI_JSON_AKASH_DEPLOYMENT_DEPLOYMENT_closed]: XC_PROTO_AKASH_DEPLOYMENT_DEPLOYMENT_closed,
};
