// @solar-republic/cosmos-grpc/build/dist/proto/akash/deployment/v1beta2/groupmsg.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeAkashDeploymentGroupID } from './groupid.js';
/**
 * Encodes a `MsgCloseGroup` protobuf message: defines SDK message to close a single Group within a Deployment.
 * @param g_id - `id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgCloseGroup = (atu8_id) => Protobuf() // ...
    .b(atu8_id) // GroupID id = 1
    .o;
/**
 * Encodes a `MsgCloseGroupResponse` protobuf message: defines the Msg/CloseGroup response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgCloseGroupResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgPauseGroup` protobuf message: defines SDK message to close a single Group within a Deployment.
 * @param g_id - `id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgPauseGroup = (atu8_id) => Protobuf() // ...
    .b(atu8_id) // GroupID id = 1
    .o;
/**
 * Encodes a `MsgPauseGroupResponse` protobuf message: defines the Msg/PauseGroup response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgPauseGroupResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgStartGroup` protobuf message: defines SDK message to close a single Group within a Deployment.
 * @param g_id - `id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgStartGroup = (atu8_id) => Protobuf() // ...
    .b(atu8_id) // GroupID id = 1
    .o;
/**
 * Encodes a `MsgStartGroupResponse` protobuf message: defines the Msg/StartGroup response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashDeploymentMsgStartGroupResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf CloseGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgCloseGroup}
 */
export const decodeAkashDeploymentMsgCloseGroup = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAkashDeploymentGroupID]);
/**
 * Decodes a protobuf CloseGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgCloseGroupResponse}
 */
export const decodeAkashDeploymentMsgCloseGroupResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf PauseGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgPauseGroup}
 */
export const decodeAkashDeploymentMsgPauseGroup = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAkashDeploymentGroupID]);
/**
 * Decodes a protobuf PauseGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgPauseGroupResponse}
 */
export const decodeAkashDeploymentMsgPauseGroupResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf StartGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgStartGroup}
 */
export const decodeAkashDeploymentMsgStartGroup = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAkashDeploymentGroupID]);
/**
 * Decodes a protobuf StartGroup message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashDeploymentMsgStartGroupResponse}
 */
export const decodeAkashDeploymentMsgStartGroupResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link AkashDeploymentMsgCloseGroup} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: id -
 */
export const destructAkashDeploymentMsgCloseGroup = (g_struct) => [g_struct.id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_CLOSE_GROUP = '/akash.deployment.v1beta2.MsgCloseGroup';
/**
 * Destructures the fields of a {@link AkashDeploymentMsgCloseGroupResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAkashDeploymentMsgCloseGroupResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_CLOSE_GROUP_RESPONSE = '/akash.deployment.v1beta2.MsgCloseGroupResponse';
/**
 * Destructures the fields of a {@link AkashDeploymentMsgPauseGroup} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: id -
 */
export const destructAkashDeploymentMsgPauseGroup = (g_struct) => [g_struct.id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_PAUSE_GROUP = '/akash.deployment.v1beta2.MsgPauseGroup';
/**
 * Destructures the fields of a {@link AkashDeploymentMsgPauseGroupResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAkashDeploymentMsgPauseGroupResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_PAUSE_GROUP_RESPONSE = '/akash.deployment.v1beta2.MsgPauseGroupResponse';
/**
 * Destructures the fields of a {@link AkashDeploymentMsgStartGroup} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: id -
 */
export const destructAkashDeploymentMsgStartGroup = (g_struct) => [g_struct.id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_START_GROUP = '/akash.deployment.v1beta2.MsgStartGroup';
/**
 * Destructures the fields of a {@link AkashDeploymentMsgStartGroupResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAkashDeploymentMsgStartGroupResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_DEPLOYMENT_MSG_START_GROUP_RESPONSE = '/akash.deployment.v1beta2.MsgStartGroupResponse';
