// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/gov/v1beta1/gov.js
import { parse_duration, parse_timestamp } from '../../../../api/json.js';
import { decode_protobuf, decode_coin, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, temporal, any, coins } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
/**
 * Encodes a `TextProposal` protobuf message wrapped in the `Any` container: defines a standard text proposal whose changes need to be manually updated in case of approval.
 * @param s_title - `title`: of the proposal.
 * @param s_description - `description`: associated with the proposal.
 * @returns a strongly subtyped Uint8Array representing an `Any` protobuf message
 */
export const anyTextProposal = (s_title, s_description) => any('/cosmos.gov.v1beta1.TextProposal', Protobuf() // ...
    .s(s_title) // string title = 1
    .s(s_description) // string description = 2
    .o);
/**
 * Encodes a `WeightedVoteOption` protobuf message: defines a unit of vote for vote split. Since: cosmos-sdk 0.43
 * @param xc_option - `option`: defines the valid vote options, it must not contain duplicate vote options.
 * @param s_weight - `weight`: the vote weight associated with the vote option.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovWeightedVoteOption = (xc_option, s_weight) => Protobuf() // ...
    .v(xc_option) // VoteOption option = 1
    .s(s_weight) // string weight = 2
    .o;
/**
 * Encodes a `TextProposal` protobuf message: defines a standard text proposal whose changes need to be manually updated in case of approval.
 * @param s_title - `title`: of the proposal.
 * @param s_description - `description`: associated with the proposal.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovTextProposal = (s_title, s_description) => Protobuf() // ...
    .s(s_title) // string title = 1
    .s(s_description) // string description = 2
    .o;
/**
 * Encodes a `Deposit` protobuf message: defines an amount deposited by an account address to an active proposal.
 * @param sg_proposal_id - `proposal_id`: defines the unique id of the proposal.
 * @param sa_depositor - `depositor`: defines the deposit addresses from the proposals.
 * @param a_amounts - `amount`: to be deposited by depositor.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovDeposit = (sg_proposal_id, sa_depositor, a_amounts) => Protobuf() // ...
    .g(sg_proposal_id) // uint64 proposal_id = 1
    .s(sa_depositor) // string depositor = 2
    .B(coins(a_amounts)) // Coin[] amount = 3
    .o;
/**
 * Encodes a `Proposal` protobuf message: defines the core field members of a governance proposal.
 * @param sg_proposal_id - `proposal_id`: defines the unique id of the proposal.
 * @param atu8_content - `content`: the proposal's content.
 * @param xc_status - `status`: defines the proposal status.
 * @param g_final_tally_result - `final_tally_result`: the final tally result of the proposal. When querying a proposal via gRPC, this field is not populated until the proposal's voting period has ended.
 * @param xt_submit_time - `submit_time`: the time of proposal submission.
 * @param xt_deposit_end_time - `deposit_end_time`: the end time for deposition.
 * @param a_total_deposits - `total_deposit`: the total deposit on the proposal.
 * @param xt_voting_start_time - `voting_start_time`: the starting time to vote on a proposal.
 * @param xt_voting_end_time - `voting_end_time`: the end time of voting on a proposal.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovProposal = (sg_proposal_id, atu8_content, xc_status, atu8_final_tally_result, xt_submit_time, xt_deposit_end_time, a_total_deposits, xt_voting_start_time, xt_voting_end_time) => Protobuf() // ...
    .g(sg_proposal_id) // uint64 proposal_id = 1
    .b(atu8_content) // Any content = 2
    .v(xc_status) // ProposalStatus status = 3
    .b(atu8_final_tally_result) // TallyResult final_tally_result = 4
    .b(temporal(xt_submit_time)) // Timestamp submit_time = 5
    .b(temporal(xt_deposit_end_time)) // Timestamp deposit_end_time = 6
    .B(coins(a_total_deposits)) // Coin[] total_deposit = 7
    .b(temporal(xt_voting_start_time)) // Timestamp voting_start_time = 8
    .b(temporal(xt_voting_end_time)) // Timestamp voting_end_time = 9
    .o;
/**
 * Encodes a `TallyResult` protobuf message: defines a standard tally for a governance proposal.
 * @param s_yes - `yes`: the number of yes votes on a proposal.
 * @param s_abstain - `abstain`: the number of abstain votes on a proposal.
 * @param s_no - `no`: the number of no votes on a proposal.
 * @param s_no_with_veto - `no_with_veto`: the number of no with veto votes on a proposal.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovTallyResult = (s_yes, s_abstain, s_no, s_no_with_veto) => Protobuf() // ...
    .s(s_yes) // string yes = 1
    .s(s_abstain) // string abstain = 2
    .s(s_no) // string no = 3
    .s(s_no_with_veto) // string no_with_veto = 4
    .o;
/**
 * Encodes a `Vote` protobuf message: defines a vote on a governance proposal. A Vote consists of a proposal ID, the voter, and the vote option.
 * @param sg_proposal_id - `proposal_id`: defines the unique id of the proposal.
 * @param sa_voter - `voter`: the voter address of the proposal.
 * @param xc_option - `option`: Deprecated: Prefer to use `options` instead. This field is set in queries if and only if `len(options) == 1` and that option has weight 1. In all other cases, this field will default to VOTE_OPTION_UNSPECIFIED.
 * @param a_options - `options`: the weighted vote options. Since: cosmos-sdk 0.43
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovVote = (sg_proposal_id, sa_voter, xc_option, a_options) => Protobuf() // ...
    .g(sg_proposal_id) // uint64 proposal_id = 1
    .s(sa_voter) // string voter = 2
    .v(xc_option) // VoteOption option = 3
    .B(a_options) // WeightedVoteOption[] options = 4
    .o;
/**
 * Encodes a `DepositParams` protobuf message: defines the params for deposits on governance proposals.
 * @param a_min_deposits - `min_deposit`: Minimum deposit for a proposal to enter voting period.
 * @param xt_max_deposit_period - `max_deposit_period`: Maximum period for Atom holders to deposit on a proposal. Initial value: 2 months.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovDepositParams = (a_min_deposits, xt_max_deposit_period) => Protobuf() // ...
    .B(coins(a_min_deposits)) // Coin[] min_deposit = 1
    .b(temporal(xt_max_deposit_period)) // Duration max_deposit_period = 2
    .o;
/**
 * Encodes a `VotingParams` protobuf message: defines the params for voting on governance proposals.
 * @param xt_voting_period - `voting_period`: Duration of the voting period.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovVotingParams = (xt_voting_period) => Protobuf() // ...
    .b(temporal(xt_voting_period)) // Duration voting_period = 1
    .o;
/**
 * Encodes a `TallyParams` protobuf message: defines the params for tallying votes on governance proposals.
 * @param atu8_quorum - `quorum`: Minimum percentage of total stake needed to vote for a result to be considered valid.
 * @param atu8_threshold - `threshold`: Minimum proportion of Yes votes for proposal to pass. Default value: 0.5.
 * @param atu8_veto_threshold - `veto_threshold`: Minimum value of Veto votes to Total votes ratio for proposal to be vetoed. Default value: 1/3.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosGovTallyParams = (atu8_quorum, atu8_threshold, atu8_veto_threshold) => Protobuf() // ...
    .b(atu8_quorum) // bytes quorum = 1
    .b(atu8_threshold) // bytes threshold = 2
    .b(atu8_veto_threshold) // bytes veto_threshold = 3
    .o;
/**
 * Decodes a protobuf WeightedVoteOption message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovWeightedVoteOption}
 */
export const decodeCosmosGovWeightedVoteOption = (atu8_payload) => decode_protobuf(atu8_payload, [1, 5]);
/**
 * Decodes a protobuf TextProposal message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovTextProposal}
 */
export const decodeCosmosGovTextProposal = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf Deposit message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovDeposit}
 */
export const decodeCosmosGovDeposit = (atu8_payload) => decode_protobuf(atu8_payload, [3, 5, 0], [0, 0, decode_coin]);
/**
 * Decodes a protobuf Proposal message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovProposal}
 */
export const decodeCosmosGovProposal = (atu8_payload) => decode_protobuf(atu8_payload, [3, 1, 1, 1, [3, 1], [3, 1], 0, [3, 1], [3, 1]], [0, decodeGoogleProtobufAny, 0, decodeCosmosGovTallyResult, decode_temporal, decode_temporal, decode_coin, decode_temporal, decode_temporal]);
/**
 * Decodes a protobuf TallyResult message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovTallyResult}
 */
export const decodeCosmosGovTallyResult = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5, 5]);
/**
 * Decodes a protobuf Vote message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovVote}
 */
export const decodeCosmosGovVote = (atu8_payload) => decode_protobuf(atu8_payload, [3, 5, 1, 0], [0, 0, 0, decodeCosmosGovWeightedVoteOption]);
/**
 * Decodes a protobuf DepositParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovDepositParams}
 */
export const decodeCosmosGovDepositParams = (atu8_payload) => decode_protobuf(atu8_payload, [0, [3, 1]], [decode_coin, decode_temporal]);
/**
 * Decodes a protobuf VotingParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovVotingParams}
 */
export const decodeCosmosGovVotingParams = (atu8_payload) => decode_protobuf(atu8_payload, [[3, 1]], [decode_temporal]);
/**
 * Decodes a protobuf TallyParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosGovTallyParams}
 */
export const decodeCosmosGovTallyParams = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1]);
/**
 * Destructures the fields of a {@link CosmosGovWeightedVoteOption} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: option - defines the valid vote options, it must not contain duplicate vote options.
 *   - 1: weight - the vote weight associated with the vote option.
 */
export const destructCosmosGovWeightedVoteOption = (g_struct) => [g_struct.option, g_struct.weight];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_WEIGHTED_VOTE_OPTION = '/cosmos.gov.v1beta1.WeightedVoteOption';
/**
 * Destructures the fields of a {@link CosmosGovTextProposal} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: title - of the proposal.
 *   - 1: description - associated with the proposal.
 */
export const destructCosmosGovTextProposal = (g_struct) => [g_struct.title, g_struct.description];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_TEXT_PROPOSAL = '/cosmos.gov.v1beta1.TextProposal';
/**
 * Destructures the fields of a {@link CosmosGovDeposit} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: proposal_id - defines the unique id of the proposal.
 *   - 1: depositor - defines the deposit addresses from the proposals.
 *   - 2: amount - to be deposited by depositor.
 */
export const destructCosmosGovDeposit = (g_struct) => [g_struct.proposal_id, g_struct.depositor, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_DEPOSIT = '/cosmos.gov.v1beta1.Deposit';
/**
 * Destructures the fields of a {@link CosmosGovProposal} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: proposal_id - defines the unique id of the proposal.
 *   - 1: content - the proposal's content.
 *   - 2: status - defines the proposal status.
 *   - 3: final_tally_result - the final tally result of the proposal. When querying a proposal via gRPC, this field is not populated until the proposal's voting period has ended.
 *   - 4: submit_time - the time of proposal submission.
 *   - 5: deposit_end_time - the end time for deposition.
 *   - 6: total_deposit - the total deposit on the proposal.
 *   - 7: voting_start_time - the starting time to vote on a proposal.
 *   - 8: voting_end_time - the end time of voting on a proposal.
 */
export const destructCosmosGovProposal = (g_struct) => [g_struct.proposal_id, g_struct.content, g_struct.status, g_struct.final_tally_result, parse_timestamp(g_struct.submit_time), parse_timestamp(g_struct.deposit_end_time), g_struct.total_deposit, parse_timestamp(g_struct.voting_start_time), parse_timestamp(g_struct.voting_end_time)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_PROPOSAL = '/cosmos.gov.v1beta1.Proposal';
/**
 * Destructures the fields of a {@link CosmosGovTallyResult} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: yes - the number of yes votes on a proposal.
 *   - 1: abstain - the number of abstain votes on a proposal.
 *   - 2: no - the number of no votes on a proposal.
 *   - 3: no_with_veto - the number of no with veto votes on a proposal.
 */
export const destructCosmosGovTallyResult = (g_struct) => [g_struct.yes, g_struct.abstain, g_struct.no, g_struct.no_with_veto];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_TALLY_RESULT = '/cosmos.gov.v1beta1.TallyResult';
/**
 * Destructures the fields of a {@link CosmosGovVote} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: proposal_id - defines the unique id of the proposal.
 *   - 1: voter - the voter address of the proposal.
 *   - 2: option - Deprecated: Prefer to use `options` instead. This field is set in queries if and only if `len(options) == 1` and that option has weight 1. In all other cases, this field will default to VOTE_OPTION_UNSPECIFIED.
 *   - 3: options - the weighted vote options. Since: cosmos-sdk 0.43
 */
export const destructCosmosGovVote = (g_struct) => [g_struct.proposal_id, g_struct.voter, g_struct.option, g_struct.options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_VOTE = '/cosmos.gov.v1beta1.Vote';
/**
 * Destructures the fields of a {@link CosmosGovDepositParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: min_deposit - Minimum deposit for a proposal to enter voting period.
 *   - 1: max_deposit_period - Maximum period for Atom holders to deposit on a proposal. Initial value: 2 months.
 */
export const destructCosmosGovDepositParams = (g_struct) => [g_struct.min_deposit, parse_duration(g_struct.max_deposit_period)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_DEPOSIT_PARAMS = '/cosmos.gov.v1beta1.DepositParams';
/**
 * Destructures the fields of a {@link CosmosGovVotingParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: voting_period - Duration of the voting period.
 */
export const destructCosmosGovVotingParams = (g_struct) => [parse_duration(g_struct.voting_period)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_VOTING_PARAMS = '/cosmos.gov.v1beta1.VotingParams';
/**
 * Destructures the fields of a {@link CosmosGovTallyParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: quorum - Minimum percentage of total stake needed to vote for a result to be considered valid.
 *   - 1: threshold - Minimum proportion of Yes votes for proposal to pass. Default value: 0.5.
 *   - 2: veto_threshold - Minimum value of Veto votes to Total votes ratio for proposal to be vetoed. Default value: 1/3.
 */
export const destructCosmosGovTallyParams = (g_struct) => [safe_base64_to_bytes(g_struct.quorum), safe_base64_to_bytes(g_struct.threshold), safe_base64_to_bytes(g_struct.veto_threshold)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_GOV_TALLY_PARAMS = '/cosmos.gov.v1beta1.TallyParams';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_UNSPECIFIED** - defines a no-op vote option.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovVoteOption}
 */
export const XC_PROTO_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED = 0;
/**
 * JSON enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_UNSPECIFIED** - defines a no-op vote option.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovVoteOption}
 */
export const SI_JSON_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED = 'OptionEmpty';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_YES** - defines a yes vote option.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovVoteOption}
 */
export const XC_PROTO_COSMOS_GOV_VOTE_OPTION_YES = 1;
/**
 * JSON enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_YES** - defines a yes vote option.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovVoteOption}
 */
export const SI_JSON_COSMOS_GOV_VOTE_OPTION_YES = 'OptionYes';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_ABSTAIN** - defines an abstain vote option.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovVoteOption}
 */
export const XC_PROTO_COSMOS_GOV_VOTE_OPTION_ABSTAIN = 2;
/**
 * JSON enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_ABSTAIN** - defines an abstain vote option.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovVoteOption}
 */
export const SI_JSON_COSMOS_GOV_VOTE_OPTION_ABSTAIN = 'OptionAbstain';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_NO** - defines a no vote option.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovVoteOption}
 */
export const XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO = 3;
/**
 * JSON enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_NO** - defines a no vote option.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovVoteOption}
 */
export const SI_JSON_COSMOS_GOV_VOTE_OPTION_NO = 'OptionNo';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_NO_WITH_VETO** - defines a no with veto vote option.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovVoteOption}
 */
export const XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO = 4;
/**
 * JSON enum value for `cosmos.gov.v1beta1.VoteOption`.
 *
 * **VOTE_OPTION_NO_WITH_VETO** - defines a no with veto vote option.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovVoteOption}
 */
export const SI_JSON_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO = 'OptionNoWithVeto';
/**
 * Maps a protobuf enum int value for `cosmos.gov.v1beta1.VoteOption` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumCosmosGovVoteOption = {
    [XC_PROTO_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED]: SI_JSON_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED,
    [XC_PROTO_COSMOS_GOV_VOTE_OPTION_YES]: SI_JSON_COSMOS_GOV_VOTE_OPTION_YES,
    [XC_PROTO_COSMOS_GOV_VOTE_OPTION_ABSTAIN]: SI_JSON_COSMOS_GOV_VOTE_OPTION_ABSTAIN,
    [XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO]: SI_JSON_COSMOS_GOV_VOTE_OPTION_NO,
    [XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO]: SI_JSON_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO,
};
/**
 * Maps a JSON enum string value for `cosmos.gov.v1beta1.VoteOption` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumCosmosGovVoteOption = {
    [SI_JSON_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED]: XC_PROTO_COSMOS_GOV_VOTE_OPTION_UNSPECIFIED,
    [SI_JSON_COSMOS_GOV_VOTE_OPTION_YES]: XC_PROTO_COSMOS_GOV_VOTE_OPTION_YES,
    [SI_JSON_COSMOS_GOV_VOTE_OPTION_ABSTAIN]: XC_PROTO_COSMOS_GOV_VOTE_OPTION_ABSTAIN,
    [SI_JSON_COSMOS_GOV_VOTE_OPTION_NO]: XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO,
    [SI_JSON_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO]: XC_PROTO_COSMOS_GOV_VOTE_OPTION_NO_WITH_VETO,
};
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_UNSPECIFIED** - defines the default proposal status.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED = 0;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_UNSPECIFIED** - defines the default proposal status.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED = 'StatusNil';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_DEPOSIT_PERIOD** - defines a proposal status during the deposit period.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD = 1;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_DEPOSIT_PERIOD** - defines a proposal status during the deposit period.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD = 'StatusDepositPeriod';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_VOTING_PERIOD** - defines a proposal status during the voting period.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD = 2;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_VOTING_PERIOD** - defines a proposal status during the voting period.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD = 'StatusVotingPeriod';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_PASSED** - defines a proposal status of a proposal that has passed.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_PASSED = 3;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_PASSED** - defines a proposal status of a proposal that has passed.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_PASSED = 'StatusPassed';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_REJECTED** - defines a proposal status of a proposal that has been rejected.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_REJECTED = 4;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_REJECTED** - defines a proposal status of a proposal that has been rejected.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_REJECTED = 'StatusRejected';
/**
 * Protobuf enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_FAILED** - defines a proposal status of a proposal that has failed.
 *
 * Belongs to enum type {@link ProtoEnumCosmosGovProposalStatus}
 */
export const XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_FAILED = 5;
/**
 * JSON enum value for `cosmos.gov.v1beta1.ProposalStatus`.
 *
 * **PROPOSAL_STATUS_FAILED** - defines a proposal status of a proposal that has failed.
 *
 * Belongs to enum type {@link JsonEnumCosmosGovProposalStatus}
 */
export const SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_FAILED = 'StatusFailed';
/**
 * Maps a protobuf enum int value for `cosmos.gov.v1beta1.ProposalStatus` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumCosmosGovProposalStatus = {
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED,
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD,
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD,
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_PASSED]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_PASSED,
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_REJECTED]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_REJECTED,
    [XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_FAILED]: SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_FAILED,
};
/**
 * Maps a JSON enum string value for `cosmos.gov.v1beta1.ProposalStatus` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumCosmosGovProposalStatus = {
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_UNSPECIFIED,
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_DEPOSIT_PERIOD,
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_VOTING_PERIOD,
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_PASSED]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_PASSED,
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_REJECTED]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_REJECTED,
    [SI_JSON_COSMOS_GOV_PROPOSAL_STATUS_FAILED]: XC_PROTO_COSMOS_GOV_PROPOSAL_STATUS_FAILED,
};
