// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/bank/v1beta1/genesis.js
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../api/protobuf-writer.js';
import { decodeCosmosBankParams, decodeCosmosBankMetadata, decodeCosmosBankSendEnabled } from './bank.js';
/**
 * Encodes a `GenesisState` protobuf message: defines the bank module's genesis state.
 * @param g_params - `params`: defines all the parameters of the module.
 * @param a_balances - `balances`: an array containing the balances of all the accounts.
 * @param a_supplys - `supply`: represents the total supply. If it is left empty, then supply will be calculated based on the provided balances. Otherwise, it will be used to validate that the sum of the balances equals this amount.
 * @param a_denom_metadatas - `denom_metadata`: defines the metadata of the different coins.
 * @param a_send_enableds - `send_enabled`: defines the denoms where send is enabled or disabled. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBankGenesisState = (atu8_params, a_balances, a_supplys, a_denom_metadatas, a_send_enableds) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .B(a_balances) // Balance[] balances = 2
    .B(coins(a_supplys)) // Coin[] supply = 3
    .B(a_denom_metadatas) // Metadata[] denom_metadata = 4
    .B(a_send_enableds) // SendEnabled[] send_enabled = 5
    .o;
/**
 * Encodes a `Balance` protobuf message: defines an account address and balance pair used in the bank module's genesis state.
 * @param sa_address - `address`: the address of the balance holder.
 * @param a_coins - `coins`: defines the different coins this balance holds.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBankBalance = (sa_address, a_coins) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .B(coins(a_coins)) // Coin[] coins = 2
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBankGenesisState}
 */
export const decodeCosmosBankGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0, 0, 0, 0], [decodeCosmosBankParams, decodeCosmosBankBalance, decode_coin, decodeCosmosBankMetadata, decodeCosmosBankSendEnabled]);
/**
 * Decodes a protobuf Balance message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBankBalance}
 */
export const decodeCosmosBankBalance = (atu8_payload) => decode_protobuf(atu8_payload, [5, 0], [0, decode_coin]);
/**
 * Destructures the fields of a {@link CosmosBankGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - defines all the parameters of the module.
 *   - 1: balances - an array containing the balances of all the accounts.
 *   - 2: supply - represents the total supply. If it is left empty, then supply will be calculated based on the provided balances. Otherwise, it will be used to validate that the sum of the balances equals this amount.
 *   - 3: denom_metadata - defines the metadata of the different coins.
 *   - 4: send_enabled - defines the denoms where send is enabled or disabled. Since: cosmos-sdk 0.47
 */
export const destructCosmosBankGenesisState = (g_struct) => [g_struct.params, g_struct.balances, g_struct.supply, g_struct.denom_metadata, g_struct.send_enabled];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BANK_GENESIS_STATE = '/cosmos.bank.v1beta1.GenesisState';
/**
 * Destructures the fields of a {@link CosmosBankBalance} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address - the address of the balance holder.
 *   - 1: coins - defines the different coins this balance holds.
 */
export const destructCosmosBankBalance = (g_struct) => [g_struct.address, g_struct.coins];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BANK_BALANCE = '/cosmos.bank.v1beta1.Balance';
