// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/pool-models/stableswap/stableswap_pool.js
import { decode_protobuf, decode_coin } from '../../../../../api/protobuf-reader.js';
import { Protobuf, any, coin, coins } from '../../../../../api/protobuf-writer.js';
/**
 * Encodes a `Pool` protobuf message wrapped in the `Any` container: the stableswap Pool struct
 * @param sa_address - `address`:
 * @param sg_id - `id`:
 * @param g_pool_params - `pool_params`:
 * @param s_future_pool_governor - `future_pool_governor`: This string specifies who will govern the pool in the future. Valid forms of this are: {token name},{duration} {duration} where {token name} if specified is the token which determines the governor, and if not specified is the LP token for this pool.duration is a time specified as 0w,1w,2w, etc. which specifies how long the token would need to be locked up to count in governance. 0w means no lockup.
 * @param a_total_shares - `total_shares`: sum of all LP shares
 * @param a_pool_liquiditys - `pool_liquidity`: assets in the pool
 * @param a_scaling_factors - `scaling_factors`: for calculation amognst assets with different precisions
 * @param s_scaling_factor_controller - `scaling_factor_controller`: the address can adjust pool scaling factors
 * @returns a strongly subtyped Uint8Array representing an `Any` protobuf message
 */
export const anyPool = (sa_address, sg_id, atu8_pool_params, s_future_pool_governor, a_total_shares, a_pool_liquiditys, a_scaling_factors, s_scaling_factor_controller) => any('/osmosis.gamm.poolmodels.stableswap.v1beta1.Pool', Protobuf() // ...
    .s(sa_address) // string address = 1
    .g(sg_id) // uint64 id = 2
    .b(atu8_pool_params) // PoolParams pool_params = 3
    .s(s_future_pool_governor) // string future_pool_governor = 4
    .b(coin(a_total_shares)) // Coin total_shares = 5
    .B(coins(a_pool_liquiditys)) // Coin[] pool_liquidity = 6
    .G(a_scaling_factors) // uint64[] scaling_factors = 7
    .s(s_scaling_factor_controller) // string scaling_factor_controller = 8
    .o);
/**
 * Encodes a `PoolParams` protobuf message: defined the parameters that will be managed by the pool governance in the future. This params are not managed by the chain governance. Instead they will be managed by the token holders of the pool. The pool's token holders are specified in future_pool_governor.
 * @param s_swap_fee - `swap_fee`:
 * @param s_exit_fee - `exit_fee`: N.B.: exit fee is disabled during pool creation in x/poolmanager. While old pools can maintain a non-zero fee. No new pool can be created with non-zero fee anymore
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapPoolParams = (s_swap_fee, s_exit_fee) => Protobuf() // ...
    .s(s_swap_fee) // string swap_fee = 1
    .s(s_exit_fee) // string exit_fee = 2
    .o;
/**
 * Encodes a `Pool` protobuf message: the stableswap Pool struct
 * @param sa_address - `address`:
 * @param sg_id - `id`:
 * @param g_pool_params - `pool_params`:
 * @param s_future_pool_governor - `future_pool_governor`: This string specifies who will govern the pool in the future. Valid forms of this are: {token name},{duration} {duration} where {token name} if specified is the token which determines the governor, and if not specified is the LP token for this pool.duration is a time specified as 0w,1w,2w, etc. which specifies how long the token would need to be locked up to count in governance. 0w means no lockup.
 * @param a_total_shares - `total_shares`: sum of all LP shares
 * @param a_pool_liquiditys - `pool_liquidity`: assets in the pool
 * @param a_scaling_factors - `scaling_factors`: for calculation amognst assets with different precisions
 * @param s_scaling_factor_controller - `scaling_factor_controller`: the address can adjust pool scaling factors
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammPoolmodelsStableswapPool = (sa_address, sg_id, atu8_pool_params, s_future_pool_governor, a_total_shares, a_pool_liquiditys, a_scaling_factors, s_scaling_factor_controller) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .g(sg_id) // uint64 id = 2
    .b(atu8_pool_params) // PoolParams pool_params = 3
    .s(s_future_pool_governor) // string future_pool_governor = 4
    .b(coin(a_total_shares)) // Coin total_shares = 5
    .B(coins(a_pool_liquiditys)) // Coin[] pool_liquidity = 6
    .G(a_scaling_factors) // uint64[] scaling_factors = 7
    .s(s_scaling_factor_controller) // string scaling_factor_controller = 8
    .o;
/**
 * Decodes a protobuf PoolParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapPoolParams}
 */
export const decodeOsmosisGammPoolmodelsStableswapPoolParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf Pool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammPoolmodelsStableswapPool}
 */
export const decodeOsmosisGammPoolmodelsStableswapPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 3, 1, 5, 1, 0, 2, 5], [0, 0, decodeOsmosisGammPoolmodelsStableswapPoolParams, 0, decode_coin, decode_coin]);
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapPoolParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: swap_fee -
 *   - 1: exit_fee - N.B.: exit fee is disabled during pool creation in x/poolmanager. While old pools can maintain a non-zero fee. No new pool can be created with non-zero fee anymore
 */
export const destructOsmosisGammPoolmodelsStableswapPoolParams = (g_struct) => [g_struct.swap_fee, g_struct.exit_fee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_POOL_PARAMS = '/osmosis.gamm.poolmodels.stableswap.v1beta1.PoolParams';
/**
 * Destructures the fields of a {@link OsmosisGammPoolmodelsStableswapPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address -
 *   - 1: id -
 *   - 2: pool_params -
 *   - 3: future_pool_governor - This string specifies who will govern the pool in the future. Valid forms of this are: {token name},{duration} {duration} where {token name} if specified is the token which determines the governor, and if not specified is the LP token for this pool.duration is a time specified as 0w,1w,2w, etc. which specifies how long the token would need to be locked up to count in governance. 0w means no lockup.
 *   - 4: total_shares - sum of all LP shares
 *   - 5: pool_liquidity - assets in the pool
 *   - 6: scaling_factors - for calculation amognst assets with different precisions
 *   - 7: scaling_factor_controller - the address can adjust pool scaling factors
 */
export const destructOsmosisGammPoolmodelsStableswapPool = (g_struct) => [g_struct.address, g_struct.id, g_struct.pool_params, g_struct.future_pool_governor, g_struct.total_shares, g_struct.pool_liquidity, g_struct.scaling_factors, g_struct.scaling_factor_controller];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_POOLMODELS_STABLESWAP_POOL = '/osmosis.gamm.poolmodels.stableswap.v1beta1.Pool';
