// @solar-republic/cosmos-grpc/build/dist/proto/axelar/nexus/v1beta1/query.js
import { parse_duration } from '../../../../api/json.js';
import { decode_protobuf, decode_coin, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, map, temporal, coin } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
import { decodeAxelarNexusExportedCrossChainTransfer, decodeAxelarNexusExportedFeeInfo, decodeAxelarNexusExportedGeneralMessage } from '../exported/v1beta1/types.js';
import { decodeAxelarNexusParams } from './params.js';
import { decodeAxelarNexusChainState } from './types.js';
import { decodeCosmosBaseQueryPageRequest, decodeCosmosBaseQueryPageResponse } from '../../../cosmos/base/query/v1beta1/pagination.js';
/**
 * Encodes a `ChainMaintainersRequest` protobuf message: represents a message that queries the chain maintainers for the specified chain
 * @param s_chain - `chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainMaintainersRequest = (s_chain) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .o;
/**
 * Encodes a `ChainMaintainersResponse` protobuf message:
 * @param a_maintainers - `maintainers`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainMaintainersResponse = (a_maintainers) => Protobuf() // ...
    .B(a_maintainers) // bytes[] maintainers = 1
    .o;
/**
 * Encodes a `LatestDepositAddressRequest` protobuf message: represents a message that queries a deposit address by recipient address
 * @param sa_recipient_addr - `recipient_addr`:
 * @param s_recipient_chain - `recipient_chain`:
 * @param s_deposit_chain - `deposit_chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusLatestDepositAddressRequest = (sa_recipient_addr, s_recipient_chain, s_deposit_chain) => Protobuf() // ...
    .s(sa_recipient_addr) // string recipient_addr = 1
    .s(s_recipient_chain) // string recipient_chain = 2
    .s(s_deposit_chain) // string deposit_chain = 3
    .o;
/**
 * Encodes a `LatestDepositAddressResponse` protobuf message:
 * @param sa_deposit_addr - `deposit_addr`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusLatestDepositAddressResponse = (sa_deposit_addr) => Protobuf() // ...
    .s(sa_deposit_addr) // string deposit_addr = 1
    .o;
/**
 * Encodes a `TransfersForChainRequest` protobuf message: represents a message that queries the transfers for the specified chain
 * @param s_chain - `chain`:
 * @param xc_state - `state`:
 * @param g_pagination - `pagination`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransfersForChainRequest = (s_chain, xc_state, atu8_pagination) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .v(xc_state) // TransferState state = 2
    .b(atu8_pagination) // PageRequest pagination = 3
    .o;
/**
 * Encodes a `TransfersForChainResponse` protobuf message:
 * @param a_transfers - `transfers`:
 * @param g_pagination - `pagination`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransfersForChainResponse = (a_transfers, atu8_pagination) => Protobuf() // ...
    .B(a_transfers) // CrossChainTransfer[] transfers = 1
    .b(atu8_pagination) // PageResponse pagination = 2
    .o;
/**
 * Encodes a `FeeInfoRequest` protobuf message: represents a message that queries the transfer fees associated to an asset on a chain
 * @param s_chain - `chain`:
 * @param s_asset - `asset`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusFeeInfoRequest = (s_chain, s_asset) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .s(s_asset) // string asset = 2
    .o;
/**
 * Encodes a `FeeInfoResponse` protobuf message:
 * @param g_fee_info - `fee_info`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusFeeInfoResponse = (atu8_fee_info) => Protobuf() // ...
    .b(atu8_fee_info) // FeeInfo fee_info = 1
    .o;
/**
 * Encodes a `TransferFeeRequest` protobuf message: represents a message that queries the fees charged by the network for a cross-chain transfer
 * @param s_source_chain - `source_chain`:
 * @param s_destination_chain - `destination_chain`:
 * @param s_amount - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransferFeeRequest = (s_source_chain, s_destination_chain, s_amount) => Protobuf() // ...
    .s(s_source_chain) // string source_chain = 1
    .s(s_destination_chain) // string destination_chain = 2
    .s(s_amount) // string amount = 3
    .o;
/**
 * Encodes a `TransferFeeResponse` protobuf message:
 * @param a_fee - `fee`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransferFeeResponse = (a_fee) => Protobuf() // ...
    .b(coin(a_fee)) // Coin fee = 1
    .o;
/**
 * Encodes a `ChainsRequest` protobuf message: represents a message that queries the chains registered on the network
 * @param xc_status - `status`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainsRequest = (xc_status) => Protobuf() // ...
    .v(xc_status) // ChainStatus status = 1
    .o;
/**
 * Encodes a `ChainsResponse` protobuf message:
 * @param a_chains - `chains`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainsResponse = (a_chains) => Protobuf() // ...
    .S(a_chains) // string[] chains = 1
    .o;
/**
 * Encodes a `AssetsRequest` protobuf message: represents a message that queries the registered assets of a chain
 * @param s_chain - `chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusAssetsRequest = (s_chain) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .o;
/**
 * Encodes a `AssetsResponse` protobuf message:
 * @param a_assets - `assets`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusAssetsResponse = (a_assets) => Protobuf() // ...
    .S(a_assets) // string[] assets = 1
    .o;
/**
 * Encodes a `ChainStateRequest` protobuf message: represents a message that queries the state of a chain registered on the network
 * @param s_chain - `chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainStateRequest = (s_chain) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .o;
/**
 * Encodes a `ChainStateResponse` protobuf message:
 * @param g_state - `state`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainStateResponse = (atu8_state) => Protobuf() // ...
    .b(atu8_state) // ChainState state = 1
    .o;
/**
 * Encodes a `ChainsByAssetRequest` protobuf message: represents a message that queries the chains that support an asset on the network
 * @param s_asset - `asset`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainsByAssetRequest = (s_asset) => Protobuf() // ...
    .s(s_asset) // string asset = 1
    .o;
/**
 * Encodes a `ChainsByAssetResponse` protobuf message:
 * @param a_chains - `chains`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusChainsByAssetResponse = (a_chains) => Protobuf() // ...
    .S(a_chains) // string[] chains = 1
    .o;
/**
 * Encodes a `RecipientAddressRequest` protobuf message: represents a message that queries the registered recipient address for a given deposit address
 * @param sa_deposit_addr - `deposit_addr`:
 * @param s_deposit_chain - `deposit_chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusRecipientAddressRequest = (sa_deposit_addr, s_deposit_chain) => Protobuf() // ...
    .s(sa_deposit_addr) // string deposit_addr = 1
    .s(s_deposit_chain) // string deposit_chain = 2
    .o;
/**
 * Encodes a `RecipientAddressResponse` protobuf message:
 * @param sa_recipient_addr - `recipient_addr`:
 * @param s_recipient_chain - `recipient_chain`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusRecipientAddressResponse = (sa_recipient_addr, s_recipient_chain) => Protobuf() // ...
    .s(sa_recipient_addr) // string recipient_addr = 1
    .s(s_recipient_chain) // string recipient_chain = 2
    .o;
/**
 * Encodes a `TransferRateLimitRequest` protobuf message: represents a message that queries the registered transfer rate limit and current transfer amounts for a given chain and asset
 * @param s_chain - `chain`:
 * @param s_asset - `asset`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransferRateLimitRequest = (s_chain, s_asset) => Protobuf() // ...
    .s(s_chain) // string chain = 1
    .s(s_asset) // string asset = 2
    .o;
/**
 * Encodes a `TransferRateLimitResponse` protobuf message:
 * @param g_transfer_rate_limit - `transfer_rate_limit`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransferRateLimitResponse = (atu8_transfer_rate_limit) => Protobuf() // ...
    .b(atu8_transfer_rate_limit) // TransferRateLimit transfer_rate_limit = 1
    .o;
/**
 * Encodes a `TransferRateLimit` protobuf message:
 * @param atu8_limit - `limit`:
 * @param xt_window - `window`:
 * @param atu8_incoming - `incoming`:
 * @param atu8_outgoing - `outgoing`:
 * @param xt_time_left - `time_left`: indicates the time left in the rate limit window
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusTransferRateLimit = (atu8_limit, xt_window, atu8_incoming, atu8_outgoing, xt_time_left) => Protobuf() // ...
    .b(atu8_limit) // bytes limit = 1
    .b(temporal(xt_window)) // Duration window = 2
    .b(atu8_incoming) // bytes incoming = 3
    .b(atu8_outgoing) // bytes outgoing = 4
    .b(temporal(xt_time_left)) // Duration time_left = 5
    .o;
/**
 * Encodes a `MessageRequest` protobuf message:
 * @param s_id - `id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusMessageRequest = (s_id) => Protobuf() // ...
    .s(s_id) // string id = 1
    .o;
/**
 * Encodes a `MessageResponse` protobuf message:
 * @param g_message - `message`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusMessageResponse = (atu8_message) => Protobuf() // ...
    .b(atu8_message) // GeneralMessage message = 1
    .o;
/**
 * Encodes a `ParamsRequest` protobuf message: represents a message that queries the params
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `ParamsResponse` protobuf message:
 * @param g_params - `params`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarNexusParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Decodes a protobuf ChainMaintainersRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainMaintainersRequest}
 */
export const decodeAxelarNexusChainMaintainersRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ChainMaintainers message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainMaintainersResponse}
 */
export const decodeAxelarNexusChainMaintainersResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0]);
/**
 * Decodes a protobuf LatestDepositAddressRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusLatestDepositAddressRequest}
 */
export const decodeAxelarNexusLatestDepositAddressRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5]);
/**
 * Decodes a protobuf LatestDepositAddress message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusLatestDepositAddressResponse}
 */
export const decodeAxelarNexusLatestDepositAddressResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf TransfersForChainRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransfersForChainRequest}
 */
export const decodeAxelarNexusTransfersForChainRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 1], [0, 0, decodeCosmosBaseQueryPageRequest]);
/**
 * Decodes a protobuf TransfersForChain message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransfersForChainResponse}
 */
export const decodeAxelarNexusTransfersForChainResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0, 1], [decodeAxelarNexusExportedCrossChainTransfer, decodeCosmosBaseQueryPageResponse]);
/**
 * Decodes a protobuf FeeInfoRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusFeeInfoRequest}
 */
export const decodeAxelarNexusFeeInfoRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf FeeInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusFeeInfoResponse}
 */
export const decodeAxelarNexusFeeInfoResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarNexusExportedFeeInfo]);
/**
 * Decodes a protobuf TransferFeeRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransferFeeRequest}
 */
export const decodeAxelarNexusTransferFeeRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5]);
/**
 * Decodes a protobuf TransferFee message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransferFeeResponse}
 */
export const decodeAxelarNexusTransferFeeResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decode_coin]);
/**
 * Decodes a protobuf ChainsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainsRequest}
 */
export const decodeAxelarNexusChainsRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Decodes a protobuf Chains message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainsResponse}
 */
export const decodeAxelarNexusChainsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [4]);
/**
 * Decodes a protobuf AssetsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusAssetsRequest}
 */
export const decodeAxelarNexusAssetsRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf Assets message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusAssetsResponse}
 */
export const decodeAxelarNexusAssetsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [4]);
/**
 * Decodes a protobuf ChainStateRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainStateRequest}
 */
export const decodeAxelarNexusChainStateRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ChainState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainStateResponse}
 */
export const decodeAxelarNexusChainStateResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarNexusChainState]);
/**
 * Decodes a protobuf ChainsByAssetRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainsByAssetRequest}
 */
export const decodeAxelarNexusChainsByAssetRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ChainsByAsset message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusChainsByAssetResponse}
 */
export const decodeAxelarNexusChainsByAssetResponse = (atu8_payload) => decode_protobuf(atu8_payload, [4]);
/**
 * Decodes a protobuf RecipientAddressRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusRecipientAddressRequest}
 */
export const decodeAxelarNexusRecipientAddressRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf RecipientAddress message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusRecipientAddressResponse}
 */
export const decodeAxelarNexusRecipientAddressResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf TransferRateLimitRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransferRateLimitRequest}
 */
export const decodeAxelarNexusTransferRateLimitRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf TransferRateLimit message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransferRateLimitResponse}
 */
export const decodeAxelarNexusTransferRateLimitResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarNexusTransferRateLimit]);
/**
 * Decodes a protobuf TransferRateLimit message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusTransferRateLimit}
 */
export const decodeAxelarNexusTransferRateLimit = (atu8_payload) => decode_protobuf(atu8_payload, [1, [3, 1], 1, 1, [3, 1]], [0, decode_temporal, 0, 0, decode_temporal]);
/**
 * Decodes a protobuf MessageRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusMessageRequest}
 */
export const decodeAxelarNexusMessageRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf Message message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusMessageResponse}
 */
export const decodeAxelarNexusMessageResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarNexusExportedGeneralMessage]);
/**
 * Decodes a protobuf ParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusParamsRequest}
 */
export const decodeAxelarNexusParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarNexusParamsResponse}
 */
export const decodeAxelarNexusParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeAxelarNexusParams]);
/**
 * Destructures the fields of a {@link AxelarNexusChainMaintainersRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 */
export const destructAxelarNexusChainMaintainersRequest = (g_struct) => [g_struct.chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAIN_MAINTAINERS_REQUEST = '/axelar.nexus.v1beta1.ChainMaintainersRequest';
/**
 * Destructures the fields of a {@link AxelarNexusChainMaintainersResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: maintainers -
 */
export const destructAxelarNexusChainMaintainersResponse = (g_struct) => [map(g_struct.maintainers, safe_base64_to_bytes)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAIN_MAINTAINERS_RESPONSE = '/axelar.nexus.v1beta1.ChainMaintainersResponse';
/**
 * Destructures the fields of a {@link AxelarNexusLatestDepositAddressRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: recipient_addr -
 *   - 1: recipient_chain -
 *   - 2: deposit_chain -
 */
export const destructAxelarNexusLatestDepositAddressRequest = (g_struct) => [g_struct.recipient_addr, g_struct.recipient_chain, g_struct.deposit_chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_LATEST_DEPOSIT_ADDRESS_REQUEST = '/axelar.nexus.v1beta1.LatestDepositAddressRequest';
/**
 * Destructures the fields of a {@link AxelarNexusLatestDepositAddressResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: deposit_addr -
 */
export const destructAxelarNexusLatestDepositAddressResponse = (g_struct) => [g_struct.deposit_addr];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_LATEST_DEPOSIT_ADDRESS_RESPONSE = '/axelar.nexus.v1beta1.LatestDepositAddressResponse';
/**
 * Destructures the fields of a {@link AxelarNexusTransfersForChainRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 *   - 1: state -
 *   - 2: pagination -
 */
export const destructAxelarNexusTransfersForChainRequest = (g_struct) => [g_struct.chain, g_struct.state, g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFERS_FOR_CHAIN_REQUEST = '/axelar.nexus.v1beta1.TransfersForChainRequest';
/**
 * Destructures the fields of a {@link AxelarNexusTransfersForChainResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: transfers -
 *   - 1: pagination -
 */
export const destructAxelarNexusTransfersForChainResponse = (g_struct) => [g_struct.transfers, g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFERS_FOR_CHAIN_RESPONSE = '/axelar.nexus.v1beta1.TransfersForChainResponse';
/**
 * Destructures the fields of a {@link AxelarNexusFeeInfoRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 *   - 1: asset -
 */
export const destructAxelarNexusFeeInfoRequest = (g_struct) => [g_struct.chain, g_struct.asset];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_FEE_INFO_REQUEST = '/axelar.nexus.v1beta1.FeeInfoRequest';
/**
 * Destructures the fields of a {@link AxelarNexusFeeInfoResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: fee_info -
 */
export const destructAxelarNexusFeeInfoResponse = (g_struct) => [g_struct.fee_info];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_FEE_INFO_RESPONSE = '/axelar.nexus.v1beta1.FeeInfoResponse';
/**
 * Destructures the fields of a {@link AxelarNexusTransferFeeRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: source_chain -
 *   - 1: destination_chain -
 *   - 2: amount -
 */
export const destructAxelarNexusTransferFeeRequest = (g_struct) => [g_struct.source_chain, g_struct.destination_chain, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFER_FEE_REQUEST = '/axelar.nexus.v1beta1.TransferFeeRequest';
/**
 * Destructures the fields of a {@link AxelarNexusTransferFeeResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: fee -
 */
export const destructAxelarNexusTransferFeeResponse = (g_struct) => [g_struct.fee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFER_FEE_RESPONSE = '/axelar.nexus.v1beta1.TransferFeeResponse';
/**
 * Destructures the fields of a {@link AxelarNexusChainsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: status -
 */
export const destructAxelarNexusChainsRequest = (g_struct) => [g_struct.status];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAINS_REQUEST = '/axelar.nexus.v1beta1.ChainsRequest';
/**
 * Destructures the fields of a {@link AxelarNexusChainsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chains -
 */
export const destructAxelarNexusChainsResponse = (g_struct) => [g_struct.chains];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAINS_RESPONSE = '/axelar.nexus.v1beta1.ChainsResponse';
/**
 * Destructures the fields of a {@link AxelarNexusAssetsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 */
export const destructAxelarNexusAssetsRequest = (g_struct) => [g_struct.chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_ASSETS_REQUEST = '/axelar.nexus.v1beta1.AssetsRequest';
/**
 * Destructures the fields of a {@link AxelarNexusAssetsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: assets -
 */
export const destructAxelarNexusAssetsResponse = (g_struct) => [g_struct.assets];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_ASSETS_RESPONSE = '/axelar.nexus.v1beta1.AssetsResponse';
/**
 * Destructures the fields of a {@link AxelarNexusChainStateRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 */
export const destructAxelarNexusChainStateRequest = (g_struct) => [g_struct.chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAIN_STATE_REQUEST = '/axelar.nexus.v1beta1.ChainStateRequest';
/**
 * Destructures the fields of a {@link AxelarNexusChainStateResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: state -
 */
export const destructAxelarNexusChainStateResponse = (g_struct) => [g_struct.state];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAIN_STATE_RESPONSE = '/axelar.nexus.v1beta1.ChainStateResponse';
/**
 * Destructures the fields of a {@link AxelarNexusChainsByAssetRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: asset -
 */
export const destructAxelarNexusChainsByAssetRequest = (g_struct) => [g_struct.asset];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAINS_BY_ASSET_REQUEST = '/axelar.nexus.v1beta1.ChainsByAssetRequest';
/**
 * Destructures the fields of a {@link AxelarNexusChainsByAssetResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chains -
 */
export const destructAxelarNexusChainsByAssetResponse = (g_struct) => [g_struct.chains];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_CHAINS_BY_ASSET_RESPONSE = '/axelar.nexus.v1beta1.ChainsByAssetResponse';
/**
 * Destructures the fields of a {@link AxelarNexusRecipientAddressRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: deposit_addr -
 *   - 1: deposit_chain -
 */
export const destructAxelarNexusRecipientAddressRequest = (g_struct) => [g_struct.deposit_addr, g_struct.deposit_chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_RECIPIENT_ADDRESS_REQUEST = '/axelar.nexus.v1beta1.RecipientAddressRequest';
/**
 * Destructures the fields of a {@link AxelarNexusRecipientAddressResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: recipient_addr -
 *   - 1: recipient_chain -
 */
export const destructAxelarNexusRecipientAddressResponse = (g_struct) => [g_struct.recipient_addr, g_struct.recipient_chain];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_RECIPIENT_ADDRESS_RESPONSE = '/axelar.nexus.v1beta1.RecipientAddressResponse';
/**
 * Destructures the fields of a {@link AxelarNexusTransferRateLimitRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: chain -
 *   - 1: asset -
 */
export const destructAxelarNexusTransferRateLimitRequest = (g_struct) => [g_struct.chain, g_struct.asset];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFER_RATE_LIMIT_REQUEST = '/axelar.nexus.v1beta1.TransferRateLimitRequest';
/**
 * Destructures the fields of a {@link AxelarNexusTransferRateLimitResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: transfer_rate_limit -
 */
export const destructAxelarNexusTransferRateLimitResponse = (g_struct) => [g_struct.transfer_rate_limit];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFER_RATE_LIMIT_RESPONSE = '/axelar.nexus.v1beta1.TransferRateLimitResponse';
/**
 * Destructures the fields of a {@link AxelarNexusTransferRateLimit} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: limit -
 *   - 1: window -
 *   - 2: incoming -
 *   - 3: outgoing -
 *   - 4: time_left - indicates the time left in the rate limit window
 */
export const destructAxelarNexusTransferRateLimit = (g_struct) => [safe_base64_to_bytes(g_struct.limit), parse_duration(g_struct.window), safe_base64_to_bytes(g_struct.incoming), safe_base64_to_bytes(g_struct.outgoing), parse_duration(g_struct.time_left)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_TRANSFER_RATE_LIMIT = '/axelar.nexus.v1beta1.TransferRateLimit';
/**
 * Destructures the fields of a {@link AxelarNexusMessageRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: id -
 */
export const destructAxelarNexusMessageRequest = (g_struct) => [g_struct.id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_MESSAGE_REQUEST = '/axelar.nexus.v1beta1.MessageRequest';
/**
 * Destructures the fields of a {@link AxelarNexusMessageResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: message -
 */
export const destructAxelarNexusMessageResponse = (g_struct) => [g_struct.message];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_MESSAGE_RESPONSE = '/axelar.nexus.v1beta1.MessageResponse';
/**
 * Destructures the fields of a {@link AxelarNexusParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAxelarNexusParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_PARAMS_REQUEST = '/axelar.nexus.v1beta1.ParamsRequest';
/**
 * Destructures the fields of a {@link AxelarNexusParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params -
 */
export const destructAxelarNexusParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_NEXUS_PARAMS_RESPONSE = '/axelar.nexus.v1beta1.ParamsResponse';
/**
 * Protobuf enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_UNSPECIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarNexusChainStatus}
 */
export const XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED = 0;
/**
 * JSON enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_UNSPECIFIED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarNexusChainStatus}
 */
export const SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED = 'Unspecified';
/**
 * Protobuf enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_ACTIVATED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarNexusChainStatus}
 */
export const XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED = 1;
/**
 * JSON enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_ACTIVATED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarNexusChainStatus}
 */
export const SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED = 'Activated';
/**
 * Protobuf enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_DEACTIVATED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarNexusChainStatus}
 */
export const XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED = 2;
/**
 * JSON enum value for `axelar.nexus.v1beta1.ChainStatus`.
 *
 * **CHAIN_STATUS_DEACTIVATED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarNexusChainStatus}
 */
export const SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED = 'Deactivated';
/**
 * Maps a protobuf enum int value for `axelar.nexus.v1beta1.ChainStatus` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAxelarNexusChainStatus = {
    [XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED]: SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED,
    [XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED]: SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED,
    [XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED]: SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED,
};
/**
 * Maps a JSON enum string value for `axelar.nexus.v1beta1.ChainStatus` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAxelarNexusChainStatus = {
    [SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED]: XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_UNSPECIFIED,
    [SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED]: XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_ACTIVATED,
    [SI_JSON_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED]: XC_PROTO_AXELAR_NEXUS_CHAIN_STATUS_DEACTIVATED,
};
