// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/concentrated-liquidity/gov.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../api/protobuf-reader.js';
import { Protobuf } from '../../../api/protobuf-writer.js';
/**
 * Encodes a `CreateConcentratedLiquidityPoolsProposal` protobuf message: a gov Content type for creating concentrated liquidity pools. If a CreateConcentratedLiquidityPoolsProposal passes, the pools are created via pool manager module account.
 * @param s_title - `title`:
 * @param s_description - `description`:
 * @param a_pool_records - `pool_records`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisConcentratedliquidityCreateConcentratedLiquidityPoolsProposal = (s_title, s_description, a_pool_records) => Protobuf() // ...
    .s(s_title) // string title = 1
    .s(s_description) // string description = 2
    .B(a_pool_records) // PoolRecord[] pool_records = 3
    .o;
/**
 * Encodes a `TickSpacingDecreaseProposal` protobuf message: a gov Content type for proposing a tick spacing decrease for a pool. The proposal will fail if one of the pools do not exist, or if the new tick spacing is not less than the current tick spacing.
 * @param s_title - `title`:
 * @param s_description - `description`:
 * @param a_pool_id_to_tick_spacing_records - `pool_id_to_tick_spacing_records`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisConcentratedliquidityTickSpacingDecreaseProposal = (s_title, s_description, a_pool_id_to_tick_spacing_records) => Protobuf() // ...
    .s(s_title) // string title = 1
    .s(s_description) // string description = 2
    .B(a_pool_id_to_tick_spacing_records) // PoolIdToTickSpacingRecord[] pool_id_to_tick_spacing_records = 3
    .o;
/**
 * Encodes a `PoolIdToTickSpacingRecord` protobuf message: a struct that contains a pool id to new tick spacing pair.
 * @param sg_pool_id - `pool_id`:
 * @param sg_new_tick_spacing - `new_tick_spacing`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisConcentratedliquidityPoolIdToTickSpacingRecord = (sg_pool_id, sg_new_tick_spacing) => Protobuf() // ...
    .g(sg_pool_id) // uint64 pool_id = 1
    .g(sg_new_tick_spacing) // uint64 new_tick_spacing = 2
    .o;
/**
 * Encodes a `PoolRecord` protobuf message:
 * @param s_denom0 - `denom0`:
 * @param s_denom1 - `denom1`:
 * @param sg_tick_spacing - `tick_spacing`:
 * @param s_spread_factor - `spread_factor`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisConcentratedliquidityPoolRecord = (s_denom0, s_denom1, sg_tick_spacing, s_spread_factor) => Protobuf() // ...
    .s(s_denom0) // string denom0 = 1
    .s(s_denom1) // string denom1 = 2
    .g(sg_tick_spacing) // uint64 tick_spacing = 3
    .s(s_spread_factor, 5) // string spread_factor = 5
    .o;
/**
 * Decodes a protobuf CreateConcentratedLiquidityPoolsProposal message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisConcentratedliquidityCreateConcentratedLiquidityPoolsProposal}
 */
export const decodeOsmosisConcentratedliquidityCreateConcentratedLiquidityPoolsProposal = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decodeOsmosisConcentratedliquidityPoolRecord]);
/**
 * Decodes a protobuf TickSpacingDecreaseProposal message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisConcentratedliquidityTickSpacingDecreaseProposal}
 */
export const decodeOsmosisConcentratedliquidityTickSpacingDecreaseProposal = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decodeOsmosisConcentratedliquidityPoolIdToTickSpacingRecord]);
/**
 * Decodes a protobuf PoolIdToTickSpacingRecord message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisConcentratedliquidityPoolIdToTickSpacingRecord}
 */
export const decodeOsmosisConcentratedliquidityPoolIdToTickSpacingRecord = (atu8_payload) => decode_protobuf(atu8_payload, [3, 3]);
/**
 * Decodes a protobuf PoolRecord message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedOsmosisConcentratedliquidityPoolRecord}
 */
export const decodeOsmosisConcentratedliquidityPoolRecord = ((atu8_payload, [s_denom0, s_denom1, sg_tick_spacing, , s_spread_factor] = decode_protobuf(atu8_payload, [5, 5, 3, 0, 5])) => [s_denom0, s_denom1, sg_tick_spacing, __UNDEFINED, s_spread_factor]);
/**
 * Destructures the fields of a {@link OsmosisConcentratedliquidityCreateConcentratedLiquidityPoolsProposal} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: title -
 *   - 1: description -
 *   - 2: pool_records -
 */
export const destructOsmosisConcentratedliquidityCreateConcentratedLiquidityPoolsProposal = (g_struct) => [g_struct.title, g_struct.description, g_struct.pool_records];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_CONCENTRATEDLIQUIDITY_CREATE_CONCENTRATED_LIQUIDITY_POOLS_PROPOSAL = '/osmosis.concentratedliquidity.v1beta1.CreateConcentratedLiquidityPoolsProposal';
/**
 * Destructures the fields of a {@link OsmosisConcentratedliquidityPoolRecord} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: denom0 -
 *   - 1: denom1 -
 *   - 2: tick_spacing -
 *   - 4: spread_factor -
 */
export const destructOsmosisConcentratedliquidityPoolRecord = (g_struct) => [g_struct.denom0, g_struct.denom1, g_struct.tick_spacing, __UNDEFINED, g_struct.spread_factor];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_CONCENTRATEDLIQUIDITY_POOL_RECORD = '/osmosis.concentratedliquidity.v1beta1.PoolRecord';
/**
 * Destructures the fields of a {@link OsmosisConcentratedliquidityTickSpacingDecreaseProposal} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: title -
 *   - 1: description -
 *   - 2: pool_id_to_tick_spacing_records -
 */
export const destructOsmosisConcentratedliquidityTickSpacingDecreaseProposal = (g_struct) => [g_struct.title, g_struct.description, g_struct.pool_id_to_tick_spacing_records];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_CONCENTRATEDLIQUIDITY_TICK_SPACING_DECREASE_PROPOSAL = '/osmosis.concentratedliquidity.v1beta1.TickSpacingDecreaseProposal';
/**
 * Destructures the fields of a {@link OsmosisConcentratedliquidityPoolIdToTickSpacingRecord} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_id -
 *   - 1: new_tick_spacing -
 */
export const destructOsmosisConcentratedliquidityPoolIdToTickSpacingRecord = (g_struct) => [g_struct.pool_id, g_struct.new_tick_spacing];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_CONCENTRATEDLIQUIDITY_POOL_ID_TO_TICK_SPACING_RECORD = '/osmosis.concentratedliquidity.v1beta1.PoolIdToTickSpacingRecord';
