// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/authz/v1beta1/event.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `EventGrant` protobuf message: emitted on Msg/Grant
 * @param s_msg_type_url - `msg_type_url`: Msg type URL for which an autorization is granted
 * @param sa_granter - `granter`: Granter account address
 * @param sa_grantee - `grantee`: Grantee account address
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAuthzEventGrant = (s_msg_type_url, sa_granter, sa_grantee) => Protobuf() // ...
    .s(s_msg_type_url, 2) // string msg_type_url = 2
    .s(sa_granter, 3) // string granter = 3
    .s(sa_grantee, 4) // string grantee = 4
    .o;
/**
 * Encodes a `EventRevoke` protobuf message: emitted on Msg/Revoke
 * @param s_msg_type_url - `msg_type_url`: Msg type URL for which an autorization is revoked
 * @param sa_granter - `granter`: Granter account address
 * @param sa_grantee - `grantee`: Grantee account address
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAuthzEventRevoke = (s_msg_type_url, sa_granter, sa_grantee) => Protobuf() // ...
    .s(s_msg_type_url, 2) // string msg_type_url = 2
    .s(sa_granter, 3) // string granter = 3
    .s(sa_grantee, 4) // string grantee = 4
    .o;
/**
 * Decodes a protobuf EventGrant message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedCosmosAuthzEventGrant}
 */
export const decodeCosmosAuthzEventGrant = ((atu8_payload, [, s_msg_type_url, sa_granter, sa_grantee] = decode_protobuf(atu8_payload, [0, 5, 5, 5])) => [__UNDEFINED, s_msg_type_url, sa_granter, sa_grantee]);
/**
 * Decodes a protobuf EventRevoke message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedCosmosAuthzEventRevoke}
 */
export const decodeCosmosAuthzEventRevoke = ((atu8_payload, [, s_msg_type_url, sa_granter, sa_grantee] = decode_protobuf(atu8_payload, [0, 5, 5, 5])) => [__UNDEFINED, s_msg_type_url, sa_granter, sa_grantee]);
/**
 * Destructures the fields of a {@link CosmosAuthzEventGrant} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: msg_type_url - Msg type URL for which an autorization is granted
 *   - 2: granter - Granter account address
 *   - 3: grantee - Grantee account address
 */
export const destructCosmosAuthzEventGrant = (g_struct) => [__UNDEFINED, g_struct.msg_type_url, g_struct.granter, g_struct.grantee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_AUTHZ_EVENT_GRANT = '/cosmos.authz.v1beta1.EventGrant';
/**
 * Destructures the fields of a {@link CosmosAuthzEventRevoke} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: msg_type_url - Msg type URL for which an autorization is revoked
 *   - 2: granter - Granter account address
 *   - 3: grantee - Grantee account address
 */
export const destructCosmosAuthzEventRevoke = (g_struct) => [__UNDEFINED, g_struct.msg_type_url, g_struct.granter, g_struct.grantee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_AUTHZ_EVENT_REVOKE = '/cosmos.authz.v1beta1.EventRevoke';
