// @solar-republic/cosmos-grpc/build/dist/proto/juno/feepay/v1/genesis.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeJunoFeepayFeePayContract } from './feepay.js';
/**
 * Encodes a `GenesisState` protobuf message: defines the module's genesis state.
 * @param g_params - `params`: the feepay module parameters
 * @param a_fee_pay_contracts - `fee_pay_contracts`: the feepay module contracts
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoFeepayGenesisState = (atu8_params, a_fee_pay_contracts) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .B(a_fee_pay_contracts) // FeePayContract[] fee_pay_contracts = 2
    .o;
/**
 * Encodes a `Params` protobuf message: defines the feepay module params
 * @param b_enable_feepay - `enable_feepay`: defines a parameter to enable the feepay module
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoFeepayParams = (b_enable_feepay) => Protobuf() // ...
    .v(b_enable_feepay) // boolean enable_feepay = 1
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoFeepayGenesisState}
 */
export const decodeJunoFeepayGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1, 0], [decodeJunoFeepayParams, decodeJunoFeepayFeePayContract]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoFeepayParams}
 */
export const decodeJunoFeepayParams = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Destructures the fields of a {@link JunoFeepayParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: enable_feepay - defines a parameter to enable the feepay module
 */
export const destructJunoFeepayParams = (g_struct) => [g_struct.enable_feepay];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_FEEPAY_PARAMS = '/juno.feepay.v1.Params';
/**
 * Destructures the fields of a {@link JunoFeepayGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - the feepay module parameters
 *   - 1: fee_pay_contracts - the feepay module contracts
 */
export const destructJunoFeepayGenesisState = (g_struct) => [g_struct.params, g_struct.fee_pay_contracts];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_FEEPAY_GENESIS_STATE = '/juno.feepay.v1.GenesisState';
