// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/distribution/v1beta1/tx.js
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../api/protobuf-writer.js';
import { decodeCosmosDistributionParams } from './distribution.js';
/**
 * Encodes a `MsgSetWithdrawAddress` protobuf message: sets the withdraw address for a delegator (or validator self-delegation).
 * @param sa_delegator - `delegator_address`:
 * @param sa_withdraw - `withdraw_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgSetWithdrawAddress = (sa_delegator, sa_withdraw) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_withdraw) // string withdraw_address = 2
    .o;
/**
 * Encodes a `MsgSetWithdrawAddressResponse` protobuf message: defines the Msg/SetWithdrawAddress response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgSetWithdrawAddressResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgWithdrawDelegatorReward` protobuf message: represents delegation withdrawal to a delegator from a single validator.
 * @param sa_delegator - `delegator_address`:
 * @param sa_validator - `validator_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgWithdrawDelegatorReward = (sa_delegator, sa_validator) => Protobuf() // ...
    .s(sa_delegator) // string delegator_address = 1
    .s(sa_validator) // string validator_address = 2
    .o;
/**
 * Encodes a `MsgWithdrawDelegatorRewardResponse` protobuf message: defines the Msg/WithdrawDelegatorReward response type.
 * @param a_amounts - `amount`: Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgWithdrawDelegatorRewardResponse = (a_amounts) => Protobuf() // ...
    .B(coins(a_amounts)) // Coin[] amount = 1
    .o;
/**
 * Encodes a `MsgWithdrawValidatorCommission` protobuf message: withdraws the full commission to the validator address.
 * @param sa_validator - `validator_address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgWithdrawValidatorCommission = (sa_validator) => Protobuf() // ...
    .s(sa_validator) // string validator_address = 1
    .o;
/**
 * Encodes a `MsgWithdrawValidatorCommissionResponse` protobuf message: defines the Msg/WithdrawValidatorCommission response type.
 * @param a_amounts - `amount`: Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgWithdrawValidatorCommissionResponse = (a_amounts) => Protobuf() // ...
    .B(coins(a_amounts)) // Coin[] amount = 1
    .o;
/**
 * Encodes a `MsgFundCommunityPool` protobuf message: allows an account to directly fund the community pool. Deprecated: Use x/protocolpool module's MsgFundCommunityPool instead. Since: cosmos-sdk 0.50
 * @param a_amounts - `amount`:
 * @param sa_depositor - `depositor`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgFundCommunityPool = (a_amounts, sa_depositor) => Protobuf() // ...
    .B(coins(a_amounts)) // Coin[] amount = 1
    .s(sa_depositor) // string depositor = 2
    .o;
/**
 * Encodes a `MsgFundCommunityPoolResponse` protobuf message: defines the Msg/FundCommunityPool response type. Deprecated Since: cosmos-sdk 0.50
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgFundCommunityPoolResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type. Since: cosmos-sdk 0.47
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param g_params - `params`: defines the x/distribution parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgUpdateParams = (s_authority, atu8_params) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgCommunityPoolSpend` protobuf message: Deprecated: Use x/protocolpool module's MsgCommunityPoolSpend instead Since: cosmos-sdk 0.50
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param sa_recipient - `recipient`:
 * @param a_amounts - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgCommunityPoolSpend = (s_authority, sa_recipient, a_amounts) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .s(sa_recipient) // string recipient = 2
    .B(coins(a_amounts)) // Coin[] amount = 3
    .o;
/**
 * Encodes a `MsgCommunityPoolSpendResponse` protobuf message: defines the response to executing a MsgCommunityPoolSpend message. Deprecated Since: cosmos-sdk 0.50
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgCommunityPoolSpendResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgDepositValidatorRewardsPool` protobuf message: DepositValidatorRewardsPool defines the request structure to provide additional rewards to delegators from a specific validator. Since: cosmos-sdk 0.50
 * @param sa_depositor - `depositor`:
 * @param sa_validator - `validator_address`:
 * @param a_amounts - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgDepositValidatorRewardsPool = (sa_depositor, sa_validator, a_amounts) => Protobuf() // ...
    .s(sa_depositor) // string depositor = 1
    .s(sa_validator) // string validator_address = 2
    .B(coins(a_amounts)) // Coin[] amount = 3
    .o;
/**
 * Encodes a `MsgDepositValidatorRewardsPoolResponse` protobuf message: defines the response to executing a MsgDepositValidatorRewardsPool message. Since: cosmos-sdk 0.50
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosDistributionMsgDepositValidatorRewardsPoolResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf SetWithdrawAddress message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgSetWithdrawAddress}
 */
export const decodeCosmosDistributionMsgSetWithdrawAddress = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf SetWithdrawAddress message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgSetWithdrawAddressResponse}
 */
export const decodeCosmosDistributionMsgSetWithdrawAddressResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf WithdrawDelegatorReward message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgWithdrawDelegatorReward}
 */
export const decodeCosmosDistributionMsgWithdrawDelegatorReward = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf WithdrawDelegatorReward message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgWithdrawDelegatorRewardResponse}
 */
export const decodeCosmosDistributionMsgWithdrawDelegatorRewardResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decode_coin]);
/**
 * Decodes a protobuf WithdrawValidatorCommission message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgWithdrawValidatorCommission}
 */
export const decodeCosmosDistributionMsgWithdrawValidatorCommission = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf WithdrawValidatorCommission message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgWithdrawValidatorCommissionResponse}
 */
export const decodeCosmosDistributionMsgWithdrawValidatorCommissionResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decode_coin]);
/**
 * Decodes a protobuf FundCommunityPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgFundCommunityPool}
 */
export const decodeCosmosDistributionMsgFundCommunityPool = (atu8_payload) => decode_protobuf(atu8_payload, [0, 5], [decode_coin]);
/**
 * Decodes a protobuf FundCommunityPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgFundCommunityPoolResponse}
 */
export const decodeCosmosDistributionMsgFundCommunityPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgUpdateParams}
 */
export const decodeCosmosDistributionMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosDistributionParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgUpdateParamsResponse}
 */
export const decodeCosmosDistributionMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf CommunityPoolSpend message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgCommunityPoolSpend}
 */
export const decodeCosmosDistributionMsgCommunityPoolSpend = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decode_coin]);
/**
 * Decodes a protobuf CommunityPoolSpend message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgCommunityPoolSpendResponse}
 */
export const decodeCosmosDistributionMsgCommunityPoolSpendResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf DepositValidatorRewardsPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgDepositValidatorRewardsPool}
 */
export const decodeCosmosDistributionMsgDepositValidatorRewardsPool = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decode_coin]);
/**
 * Decodes a protobuf DepositValidatorRewardsPool message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosDistributionMsgDepositValidatorRewardsPoolResponse}
 */
export const decodeCosmosDistributionMsgDepositValidatorRewardsPoolResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosDistributionMsgSetWithdrawAddress} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: withdraw_address -
 */
export const destructCosmosDistributionMsgSetWithdrawAddress = (g_struct) => [g_struct.delegator_address, g_struct.withdraw_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_SET_WITHDRAW_ADDRESS = '/cosmos.distribution.v1beta1.MsgSetWithdrawAddress';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgSetWithdrawAddressResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosDistributionMsgSetWithdrawAddressResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_SET_WITHDRAW_ADDRESS_RESPONSE = '/cosmos.distribution.v1beta1.MsgSetWithdrawAddressResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgWithdrawDelegatorReward} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: delegator_address -
 *   - 1: validator_address -
 */
export const destructCosmosDistributionMsgWithdrawDelegatorReward = (g_struct) => [g_struct.delegator_address, g_struct.validator_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_WITHDRAW_DELEGATOR_REWARD = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgWithdrawDelegatorRewardResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: amount - Since: cosmos-sdk 0.46
 */
export const destructCosmosDistributionMsgWithdrawDelegatorRewardResponse = (g_struct) => [g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_WITHDRAW_DELEGATOR_REWARD_RESPONSE = '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorRewardResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgWithdrawValidatorCommission} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: validator_address -
 */
export const destructCosmosDistributionMsgWithdrawValidatorCommission = (g_struct) => [g_struct.validator_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_WITHDRAW_VALIDATOR_COMMISSION = '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgWithdrawValidatorCommissionResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: amount - Since: cosmos-sdk 0.46
 */
export const destructCosmosDistributionMsgWithdrawValidatorCommissionResponse = (g_struct) => [g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_WITHDRAW_VALIDATOR_COMMISSION_RESPONSE = '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommissionResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgFundCommunityPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: amount -
 *   - 1: depositor -
 */
export const destructCosmosDistributionMsgFundCommunityPool = (g_struct) => [g_struct.amount, g_struct.depositor];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_FUND_COMMUNITY_POOL = '/cosmos.distribution.v1beta1.MsgFundCommunityPool';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgFundCommunityPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosDistributionMsgFundCommunityPoolResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_FUND_COMMUNITY_POOL_RESPONSE = '/cosmos.distribution.v1beta1.MsgFundCommunityPoolResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: params - defines the x/distribution parameters to update. NOTE: All parameters must be supplied.
 */
export const destructCosmosDistributionMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_UPDATE_PARAMS = '/cosmos.distribution.v1beta1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosDistributionMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_UPDATE_PARAMS_RESPONSE = '/cosmos.distribution.v1beta1.MsgUpdateParamsResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgCommunityPoolSpend} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: recipient -
 *   - 2: amount -
 */
export const destructCosmosDistributionMsgCommunityPoolSpend = (g_struct) => [g_struct.authority, g_struct.recipient, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_COMMUNITY_POOL_SPEND = '/cosmos.distribution.v1beta1.MsgCommunityPoolSpend';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgCommunityPoolSpendResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosDistributionMsgCommunityPoolSpendResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_COMMUNITY_POOL_SPEND_RESPONSE = '/cosmos.distribution.v1beta1.MsgCommunityPoolSpendResponse';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgDepositValidatorRewardsPool} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: depositor -
 *   - 1: validator_address -
 *   - 2: amount -
 */
export const destructCosmosDistributionMsgDepositValidatorRewardsPool = (g_struct) => [g_struct.depositor, g_struct.validator_address, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_DEPOSIT_VALIDATOR_REWARDS_POOL = '/cosmos.distribution.v1beta1.MsgDepositValidatorRewardsPool';
/**
 * Destructures the fields of a {@link CosmosDistributionMsgDepositValidatorRewardsPoolResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosDistributionMsgDepositValidatorRewardsPoolResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_DISTRIBUTION_MSG_DEPOSIT_VALIDATOR_REWARDS_POOL_RESPONSE = '/cosmos.distribution.v1beta1.MsgDepositValidatorRewardsPoolResponse';
