// @solar-republic/cosmos-grpc/build/dist/proto/akash/base/v1beta1/endpoint.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `Endpoint` protobuf message: describes a publicly accessible IP service
 * @param xc_kind - `kind`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAkashBaseEndpoint = (xc_kind) => Protobuf() // ...
    .v(xc_kind) // Kind kind = 1
    .o;
/**
 * Decodes a protobuf Endpoint message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAkashBaseEndpoint}
 */
export const decodeAkashBaseEndpoint = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Destructures the fields of a {@link AkashBaseEndpoint} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: kind -
 */
export const destructAkashBaseEndpoint = (g_struct) => [g_struct.kind];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AKASH_BASE_ENDPOINT = '/akash.base.v1beta1.Endpoint';
/**
 * Protobuf enum value for `akash.base.v1beta1.Endpoint.Kind`.
 *
 * **SHARED_HTTP** -
 *
 * Belongs to enum type {@link ProtoEnumAkashBaseEndpointKind}
 */
export const XC_PROTO_AKASH_BASE_ENDPOINT_SHARED_HTTP = 0;
/**
 * JSON enum value for `akash.base.v1beta1.Endpoint.Kind`.
 *
 * **SHARED_HTTP** -
 *
 * Belongs to enum type {@link JsonEnumAkashBaseEndpointKind}
 */
export const SI_JSON_AKASH_BASE_ENDPOINT_SHARED_HTTP = 'SHARED_HTTP';
/**
 * Protobuf enum value for `akash.base.v1beta1.Endpoint.Kind`.
 *
 * **RANDOM_PORT** -
 *
 * Belongs to enum type {@link ProtoEnumAkashBaseEndpointKind}
 */
export const XC_PROTO_AKASH_BASE_ENDPOINT_RANDOM_PORT = 1;
/**
 * JSON enum value for `akash.base.v1beta1.Endpoint.Kind`.
 *
 * **RANDOM_PORT** -
 *
 * Belongs to enum type {@link JsonEnumAkashBaseEndpointKind}
 */
export const SI_JSON_AKASH_BASE_ENDPOINT_RANDOM_PORT = 'RANDOM_PORT';
/**
 * Maps a protobuf enum int value for `akash.base.v1beta1.Endpoint.Kind` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAkashBaseEndpointKind = {
    [XC_PROTO_AKASH_BASE_ENDPOINT_SHARED_HTTP]: SI_JSON_AKASH_BASE_ENDPOINT_SHARED_HTTP,
    [XC_PROTO_AKASH_BASE_ENDPOINT_RANDOM_PORT]: SI_JSON_AKASH_BASE_ENDPOINT_RANDOM_PORT,
};
/**
 * Maps a JSON enum string value for `akash.base.v1beta1.Endpoint.Kind` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAkashBaseEndpointKind = {
    [SI_JSON_AKASH_BASE_ENDPOINT_SHARED_HTTP]: XC_PROTO_AKASH_BASE_ENDPOINT_SHARED_HTTP,
    [SI_JSON_AKASH_BASE_ENDPOINT_RANDOM_PORT]: XC_PROTO_AKASH_BASE_ENDPOINT_RANDOM_PORT,
};
