// @solar-republic/cosmos-grpc/build/dist/proto/axelar/snapshot/exported/v1beta1/types.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { parse_timestamp } from '../../../../../api/json.js';
import { decode_protobuf, decode_temporal } from '../../../../../api/protobuf-reader.js';
import { Protobuf, temporal } from '../../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../../api/util.js';
/**
 * Encodes a `Participant` protobuf message:
 * @param atu8_address - `address`:
 * @param atu8_weight - `weight`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotExportedParticipant = (atu8_address, atu8_weight) => Protobuf() // ...
    .b(atu8_address) // bytes address = 1
    .b(atu8_weight) // bytes weight = 2
    .o;
/**
 * Encodes a `Snapshot` protobuf message:
 * @param xt_timestamp - `timestamp`:
 * @param sg_height - `height`:
 * @param a_participants - `participants`:
 * @param atu8_bonded_weight - `bonded_weight`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotExportedSnapshot = (xt_timestamp, sg_height, a_participants, atu8_bonded_weight) => Protobuf() // ...
    .b(temporal(xt_timestamp), 2) // Timestamp timestamp = 2
    .g(sg_height, 3) // int32 height = 3
    .B(a_participants, 8) // ParticipantsEntry[] participants = 8
    .b(atu8_bonded_weight, 9) // bytes bonded_weight = 9
    .o;
/**
 * Encodes a `ParticipantsEntry` protobuf message:
 * @param s_key - `key`:
 * @param g_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotExportedSnapshotParticipantsEntry = (s_key, atu8_value) => Protobuf() // ...
    .s(s_key) // string key = 1
    .b(atu8_value) // Participant value = 2
    .o;
/**
 * Decodes a protobuf Participant message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotExportedParticipant}
 */
export const decodeAxelarSnapshotExportedParticipant = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Decodes a protobuf Snapshot message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedAxelarSnapshotExportedSnapshot}
 */
export const decodeAxelarSnapshotExportedSnapshot = ((atu8_payload, [, a_timestamp, sg_height, , , , , a_participants, atu8_bonded_weight] = decode_protobuf(atu8_payload, [0, [3, 1], 3, 0, 0, 0, 0, 0, 1], [0, decode_temporal, 0, 0, 0, 0, 0, decodeAxelarSnapshotExportedSnapshotParticipantsEntry])) => [__UNDEFINED, a_timestamp, sg_height, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, a_participants, atu8_bonded_weight]);
/**
 * Decodes a protobuf ParticipantsEntry message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotExportedSnapshotParticipantsEntry}
 */
export const decodeAxelarSnapshotExportedSnapshotParticipantsEntry = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeAxelarSnapshotExportedParticipant]);
/**
 * Destructures the fields of a {@link AxelarSnapshotExportedSnapshot} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 1: timestamp -
 *   - 2: height -
 *   - 7: participants -
 *   - 8: bonded_weight -
 */
export const destructAxelarSnapshotExportedSnapshot = (g_struct) => [__UNDEFINED, parse_timestamp(g_struct.timestamp), g_struct.height, __UNDEFINED, __UNDEFINED, __UNDEFINED, __UNDEFINED, g_struct.participants, safe_base64_to_bytes(g_struct.bonded_weight)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_EXPORTED_SNAPSHOT = '/axelar.snapshot.exported.v1beta1.Snapshot';
/**
 * Destructures the fields of a {@link AxelarSnapshotExportedSnapshotParticipantsEntry} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 */
export const destructAxelarSnapshotExportedSnapshotParticipantsEntry = (g_struct) => [g_struct.key, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_EXPORTED_SNAPSHOT_PARTICIPANTS_ENTRY = '/axelar.snapshot.exported.v1beta1.Snapshot.ParticipantsEntry';
/**
 * Destructures the fields of a {@link AxelarSnapshotExportedParticipant} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address -
 *   - 1: weight -
 */
export const destructAxelarSnapshotExportedParticipant = (g_struct) => [safe_base64_to_bytes(g_struct.address), safe_base64_to_bytes(g_struct.weight)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_EXPORTED_PARTICIPANT = '/axelar.snapshot.exported.v1beta1.Participant';
