// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/app/v1alpha1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeCosmosAppConfig } from './config.js';
/**
 * Encodes a `QueryConfigRequest` protobuf message: the Query/Config request type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAppQueryConfigRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryConfigResponse` protobuf message: QueryConfigRequest is the Query/Config response type.
 * @param g_config - `config`: the current app config.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAppQueryConfigResponse = (atu8_config) => Protobuf() // ...
    .b(atu8_config) // Config config = 1
    .o;
/**
 * Decodes a protobuf QueryConfigRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAppQueryConfigRequest}
 */
export const decodeCosmosAppQueryConfigRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryConfig message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAppQueryConfigResponse}
 */
export const decodeCosmosAppQueryConfigResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeCosmosAppConfig]);
/**
 * Destructures the fields of a {@link CosmosAppQueryConfigRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosAppQueryConfigRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_APP_QUERY_CONFIG_REQUEST = '/cosmos.app.v1alpha1.QueryConfigRequest';
/**
 * Destructures the fields of a {@link CosmosAppQueryConfigResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: config - the current app config.
 */
export const destructCosmosAppQueryConfigResponse = (g_struct) => [g_struct.config];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_APP_QUERY_CONFIG_RESPONSE = '/cosmos.app.v1alpha1.QueryConfigResponse';
