// @solar-republic/cosmos-grpc/build/dist/proto/secret/emergencybutton/v1beta1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeSecretEmergencybuttonParams } from './params.js';
const SR_LCD_SECRET_EMERGENCYBUTTON_QUERY = '/emergencybutton/v1beta1/params';
/**
 * Defines a gRPC query method that returns the emergencybutton module's parameters.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const querySecretEmergencybuttonParams = restful_grpc(() => [SR_LCD_SECRET_EMERGENCYBUTTON_QUERY]);
/**
 * Encodes a `ParamsRequest` protobuf message: the request type for the Query/Params RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeSecretEmergencybuttonParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `ParamsResponse` protobuf message: the response type for the Query/Params RPC method.
 * @param g_params - `params`: defines the parameters of the module.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeSecretEmergencybuttonParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Decodes a protobuf ParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedSecretEmergencybuttonParamsRequest}
 */
export const decodeSecretEmergencybuttonParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedSecretEmergencybuttonParamsResponse}
 */
export const decodeSecretEmergencybuttonParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeSecretEmergencybuttonParams]);
/**
 * Destructures the fields of a {@link SecretEmergencybuttonParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructSecretEmergencybuttonParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_SECRET_EMERGENCYBUTTON_PARAMS_REQUEST = '/secret.emergencybutton.v1beta1.ParamsRequest';
/**
 * Destructures the fields of a {@link SecretEmergencybuttonParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const destructSecretEmergencybuttonParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_SECRET_EMERGENCYBUTTON_PARAMS_RESPONSE = '/secret.emergencybutton.v1beta1.ParamsResponse';
