// @solar-republic/cosmos-grpc/build/dist/proto/secret/emergencybutton/v1beta1/tx.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `MsgToggleIbcSwitch` protobuf message: represents a message to toggle the emergencybutton status by the defined pauser.
 * @param sa_sender - `sender`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeSecretEmergencybuttonMsgToggleIbcSwitch = (sa_sender) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .o;
/**
 * Encodes a `MsgToggleIbcSwitchResponse` protobuf message: defines the response type for the toggle.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeSecretEmergencybuttonMsgToggleIbcSwitchResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf ToggleIbcSwitch message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedSecretEmergencybuttonMsgToggleIbcSwitch}
 */
export const decodeSecretEmergencybuttonMsgToggleIbcSwitch = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf ToggleIbcSwitch message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedSecretEmergencybuttonMsgToggleIbcSwitchResponse}
 */
export const decodeSecretEmergencybuttonMsgToggleIbcSwitchResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link SecretEmergencybuttonMsgToggleIbcSwitch} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 */
export const destructSecretEmergencybuttonMsgToggleIbcSwitch = (g_struct) => [g_struct.sender];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_SECRET_EMERGENCYBUTTON_MSG_TOGGLE_IBC_SWITCH = '/secret.emergencybutton.v1beta1.MsgToggleIbcSwitch';
/**
 * Destructures the fields of a {@link SecretEmergencybuttonMsgToggleIbcSwitchResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructSecretEmergencybuttonMsgToggleIbcSwitchResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_SECRET_EMERGENCYBUTTON_MSG_TOGGLE_IBC_SWITCH_RESPONSE = '/secret.emergencybutton.v1beta1.MsgToggleIbcSwitchResponse';
