// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/epochs/query.js
import { decode_protobuf } from '../../../api/protobuf-reader.js';
import { Protobuf } from '../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../api/transport.js';
import { decodeOsmosisEpochsEpochInfo } from './genesis.js';
const SR_LCD_OSMOSIS_EPOCHS_QUERY = '/osmosis/epochs/v1beta1/';
/**
 * Provide running epochInfos
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: epochs - the 'epochs' response property
 */
export const queryOsmosisEpochsEpochInfos = restful_grpc(() => [SR_LCD_OSMOSIS_EPOCHS_QUERY + 'epochs']);
/**
 * Provide current epoch of specified identifier
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param s_identifier -
 * @returns a tuple where:
 *   - 0: current_epoch - the 'current_epoch' response property
 */
export const queryOsmosisEpochsCurrentEpoch = restful_grpc(s_identifier => [SR_LCD_OSMOSIS_EPOCHS_QUERY + 'current_epoch', {
        identifier: s_identifier,
    }]);
/**
 * Encodes a `QueryEpochsInfoRequest` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisEpochsQueryEpochsInfoRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryEpochsInfoResponse` protobuf message:
 * @param a_epochs - `epochs`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisEpochsQueryEpochsInfoResponse = (a_epochs) => Protobuf() // ...
    .B(a_epochs) // EpochInfo[] epochs = 1
    .o;
/**
 * Encodes a `QueryCurrentEpochRequest` protobuf message:
 * @param s_identifier - `identifier`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisEpochsQueryCurrentEpochRequest = (s_identifier) => Protobuf() // ...
    .s(s_identifier) // string identifier = 1
    .o;
/**
 * Encodes a `QueryCurrentEpochResponse` protobuf message:
 * @param sg_current_epoch - `current_epoch`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisEpochsQueryCurrentEpochResponse = (sg_current_epoch) => Protobuf() // ...
    .g(sg_current_epoch) // int32 current_epoch = 1
    .o;
/**
 * Decodes a protobuf QueryEpochsInfoRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisEpochsQueryEpochsInfoRequest}
 */
export const decodeOsmosisEpochsQueryEpochsInfoRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryEpochsInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisEpochsQueryEpochsInfoResponse}
 */
export const decodeOsmosisEpochsQueryEpochsInfoResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisEpochsEpochInfo]);
/**
 * Decodes a protobuf QueryCurrentEpochRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisEpochsQueryCurrentEpochRequest}
 */
export const decodeOsmosisEpochsQueryCurrentEpochRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf QueryCurrentEpoch message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisEpochsQueryCurrentEpochResponse}
 */
export const decodeOsmosisEpochsQueryCurrentEpochResponse = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Destructures the fields of a {@link OsmosisEpochsQueryEpochsInfoRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisEpochsQueryEpochsInfoRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_EPOCHS_QUERY_EPOCHS_INFO_REQUEST = '/osmosis.epochs.v1beta1.QueryEpochsInfoRequest';
/**
 * Destructures the fields of a {@link OsmosisEpochsQueryEpochsInfoResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: epochs -
 */
export const destructOsmosisEpochsQueryEpochsInfoResponse = (g_struct) => [g_struct.epochs];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_EPOCHS_QUERY_EPOCHS_INFO_RESPONSE = '/osmosis.epochs.v1beta1.QueryEpochsInfoResponse';
/**
 * Destructures the fields of a {@link OsmosisEpochsQueryCurrentEpochRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: identifier -
 */
export const destructOsmosisEpochsQueryCurrentEpochRequest = (g_struct) => [g_struct.identifier];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_EPOCHS_QUERY_CURRENT_EPOCH_REQUEST = '/osmosis.epochs.v1beta1.QueryCurrentEpochRequest';
/**
 * Destructures the fields of a {@link OsmosisEpochsQueryCurrentEpochResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: current_epoch -
 */
export const destructOsmosisEpochsQueryCurrentEpochResponse = (g_struct) => [g_struct.current_epoch];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_EPOCHS_QUERY_CURRENT_EPOCH_RESPONSE = '/osmosis.epochs.v1beta1.QueryCurrentEpochResponse';
