// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/valset-pref/v1beta1/query.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeOsmosisValsetprefValidatorPreference } from './state.js';
const SR_LCD_OSMOSIS_VALSETPREF_QUERY = '/osmosis/valset-pref/v1beta1/';
/**
 * Returns the list of ValidatorPreferences for the user.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sa_address - user account address
 * @returns a tuple where:
 *   - 0: preferences - the 'preferences' response property
 */
export const queryOsmosisValsetprefUserValidatorPreferences = restful_grpc(sa_address => [SR_LCD_OSMOSIS_VALSETPREF_QUERY + sa_address]);
/**
 * Encodes a `UserValidatorPreferencesRequest` protobuf message: Request type for UserValidatorPreferences.
 * @param sa_address - `address`: user account address
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisValsetprefUserValidatorPreferencesRequest = (sa_address) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .o;
/**
 * Encodes a `UserValidatorPreferencesResponse` protobuf message: Response type the QueryUserValidatorPreferences query request
 * @param a_preferences - `preferences`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisValsetprefUserValidatorPreferencesResponse = (a_preferences) => Protobuf() // ...
    .B(a_preferences) // ValidatorPreference[] preferences = 1
    .o;
/**
 * Decodes a protobuf UserValidatorPreferencesRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisValsetprefUserValidatorPreferencesRequest}
 */
export const decodeOsmosisValsetprefUserValidatorPreferencesRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf UserValidatorPreferences message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisValsetprefUserValidatorPreferencesResponse}
 */
export const decodeOsmosisValsetprefUserValidatorPreferencesResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeOsmosisValsetprefValidatorPreference]);
/**
 * Destructures the fields of a {@link OsmosisValsetprefUserValidatorPreferencesRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address - user account address
 */
export const destructOsmosisValsetprefUserValidatorPreferencesRequest = (g_struct) => [g_struct.address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_VALSETPREF_USER_VALIDATOR_PREFERENCES_REQUEST = '/osmosis.valsetpref.v1beta1.UserValidatorPreferencesRequest';
/**
 * Destructures the fields of a {@link OsmosisValsetprefUserValidatorPreferencesResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: preferences -
 */
export const destructOsmosisValsetprefUserValidatorPreferencesResponse = (g_struct) => [g_struct.preferences];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_VALSETPREF_USER_VALIDATOR_PREFERENCES_RESPONSE = '/osmosis.valsetpref.v1beta1.UserValidatorPreferencesResponse';
