// @solar-republic/cosmos-grpc/build/dist/proto/axelar/permission/exported/v1beta1/types.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
/**
 * Protobuf enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarPermissionExportedRole}
 */
export const XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED = 0;
/**
 * JSON enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarPermissionExportedRole}
 */
export const SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED = 'ROLE_UNSPECIFIED';
/**
 * Protobuf enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_UNRESTRICTED** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarPermissionExportedRole}
 */
export const XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED = 1;
/**
 * JSON enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_UNRESTRICTED** -
 *
 * Belongs to enum type {@link JsonEnumAxelarPermissionExportedRole}
 */
export const SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED = 'ROLE_UNRESTRICTED';
/**
 * Protobuf enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_CHAIN_MANAGEMENT** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarPermissionExportedRole}
 */
export const XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT = 2;
/**
 * JSON enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_CHAIN_MANAGEMENT** -
 *
 * Belongs to enum type {@link JsonEnumAxelarPermissionExportedRole}
 */
export const SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT = 'ROLE_CHAIN_MANAGEMENT';
/**
 * Protobuf enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_ACCESS_CONTROL** -
 *
 * Belongs to enum type {@link ProtoEnumAxelarPermissionExportedRole}
 */
export const XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL = 3;
/**
 * JSON enum value for `axelar.permission.exported.v1beta1.Role`.
 *
 * **ROLE_ACCESS_CONTROL** -
 *
 * Belongs to enum type {@link JsonEnumAxelarPermissionExportedRole}
 */
export const SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL = 'ROLE_ACCESS_CONTROL';
/**
 * Maps a protobuf enum int value for `axelar.permission.exported.v1beta1.Role` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumAxelarPermissionExportedRole = {
    [XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED]: SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED,
    [XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED]: SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED,
    [XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT]: SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT,
    [XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL]: SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL,
};
/**
 * Maps a JSON enum string value for `axelar.permission.exported.v1beta1.Role` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumAxelarPermissionExportedRole = {
    [SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED]: XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNSPECIFIED,
    [SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED]: XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_UNRESTRICTED,
    [SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT]: XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_CHAIN_MANAGEMENT,
    [SI_JSON_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL]: XC_PROTO_AXELAR_PERMISSION_EXPORTED_ROLE_ACCESS_CONTROL,
};
