// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/cosmwasmpool/v1beta1/model/module_sudo_msg.js
import { decode_protobuf, decode_coin } from '../../../../../api/protobuf-reader.js';
import { Protobuf, coin } from '../../../../../api/protobuf-writer.js';
/**
 * Encodes a `SwapExactAmountIn` protobuf message: ===================== SwapExactAmountIn
 * @param sa_sender - `sender`:
 * @param a_token_in - `token_in`: the token to be sent to the pool.
 * @param s_token_out_denom - `token_out_denom`: the token denom to be received from the pool.
 * @param s_token_out_min_amount - `token_out_min_amount`: the minimum amount of token_out to be received from the pool.
 * @param s_swap_fee - `swap_fee`: the swap fee for this swap estimate.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountIn = (sa_sender, a_token_in, s_token_out_denom, s_token_out_min_amount, s_swap_fee) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .b(coin(a_token_in)) // Coin token_in = 2
    .s(s_token_out_denom) // string token_out_denom = 3
    .s(s_token_out_min_amount) // string token_out_min_amount = 4
    .s(s_swap_fee) // string swap_fee = 5
    .o;
/**
 * Encodes a `SwapExactAmountInSudoMsg` protobuf message:
 * @param g_swap_exact_amount_in - `swap_exact_amount_in`: the structure containing all the request information for this message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountInSudoMsg = (atu8_swap_exact_amount_in) => Protobuf() // ...
    .b(atu8_swap_exact_amount_in) // SwapExactAmountIn swap_exact_amount_in = 1
    .o;
/**
 * Encodes a `SwapExactAmountInSudoMsgResponse` protobuf message:
 * @param s_token_out_amount - `token_out_amount`: the token out computed from this swap estimate call.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountInSudoMsgResponse = (s_token_out_amount) => Protobuf() // ...
    .s(s_token_out_amount) // string token_out_amount = 1
    .o;
/**
 * Encodes a `SwapExactAmountOut` protobuf message: ===================== SwapExactAmountOut
 * @param sa_sender - `sender`:
 * @param a_token_out - `token_out`: the token to be sent out of the pool.
 * @param s_token_in_denom - `token_in_denom`: the token denom to be sent too the pool.
 * @param s_token_in_max_amount - `token_in_max_amount`: the maximum amount of token_in to be sent to the pool.
 * @param s_swap_fee - `swap_fee`: the swap fee for this swap estimate.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountOut = (sa_sender, a_token_out, s_token_in_denom, s_token_in_max_amount, s_swap_fee) => Protobuf() // ...
    .s(sa_sender) // string sender = 1
    .b(coin(a_token_out)) // Coin token_out = 2
    .s(s_token_in_denom) // string token_in_denom = 3
    .s(s_token_in_max_amount) // string token_in_max_amount = 4
    .s(s_swap_fee) // string swap_fee = 5
    .o;
/**
 * Encodes a `SwapExactAmountOutSudoMsg` protobuf message:
 * @param g_swap_exact_amount_out - `swap_exact_amount_out`: the structure containing all the request information for this message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountOutSudoMsg = (atu8_swap_exact_amount_out) => Protobuf() // ...
    .b(atu8_swap_exact_amount_out) // SwapExactAmountOut swap_exact_amount_out = 1
    .o;
/**
 * Encodes a `SwapExactAmountOutSudoMsgResponse` protobuf message:
 * @param s_token_in_amount - `token_in_amount`: the token in computed from this swap estimate call.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolSwapExactAmountOutSudoMsgResponse = (s_token_in_amount) => Protobuf() // ...
    .s(s_token_in_amount) // string token_in_amount = 1
    .o;
/**
 * Decodes a protobuf SwapExactAmountIn message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountIn}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountIn = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 5, 5, 5], [0, decode_coin]);
/**
 * Decodes a protobuf SwapExactAmountInSudoMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountInSudoMsg}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountInSudoMsg = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisCosmwasmpoolSwapExactAmountIn]);
/**
 * Decodes a protobuf SwapExactAmountInSudoMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountInSudoMsgResponse}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountInSudoMsgResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf SwapExactAmountOut message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountOut}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountOut = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 5, 5, 5], [0, decode_coin]);
/**
 * Decodes a protobuf SwapExactAmountOutSudoMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountOutSudoMsg}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountOutSudoMsg = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisCosmwasmpoolSwapExactAmountOut]);
/**
 * Decodes a protobuf SwapExactAmountOutSudoMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolSwapExactAmountOutSudoMsgResponse}
 */
export const decodeOsmosisCosmwasmpoolSwapExactAmountOutSudoMsgResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountIn} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: token_in - the token to be sent to the pool.
 *   - 2: token_out_denom - the token denom to be received from the pool.
 *   - 3: token_out_min_amount - the minimum amount of token_out to be received from the pool.
 *   - 4: swap_fee - the swap fee for this swap estimate.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountIn = (g_struct) => [g_struct.sender, g_struct.token_in, g_struct.token_out_denom, g_struct.token_out_min_amount, g_struct.swap_fee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_IN = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountIn';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountInSudoMsg} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: swap_exact_amount_in - the structure containing all the request information for this message.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountInSudoMsg = (g_struct) => [g_struct.swap_exact_amount_in];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_IN_SUDO_MSG = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountInSudoMsg';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountInSudoMsgResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_out_amount - the token out computed from this swap estimate call.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountInSudoMsgResponse = (g_struct) => [g_struct.token_out_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_IN_SUDO_MSG_RESPONSE = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountInSudoMsgResponse';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountOut} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: token_out - the token to be sent out of the pool.
 *   - 2: token_in_denom - the token denom to be sent too the pool.
 *   - 3: token_in_max_amount - the maximum amount of token_in to be sent to the pool.
 *   - 4: swap_fee - the swap fee for this swap estimate.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountOut = (g_struct) => [g_struct.sender, g_struct.token_out, g_struct.token_in_denom, g_struct.token_in_max_amount, g_struct.swap_fee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_OUT = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountOut';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountOutSudoMsg} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: swap_exact_amount_out - the structure containing all the request information for this message.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountOutSudoMsg = (g_struct) => [g_struct.swap_exact_amount_out];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_OUT_SUDO_MSG = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountOutSudoMsg';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolSwapExactAmountOutSudoMsgResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: token_in_amount - the token in computed from this swap estimate call.
 */
export const destructOsmosisCosmwasmpoolSwapExactAmountOutSudoMsgResponse = (g_struct) => [g_struct.token_in_amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_SWAP_EXACT_AMOUNT_OUT_SUDO_MSG_RESPONSE = '/osmosis.cosmwasmpool.v1beta1.SwapExactAmountOutSudoMsgResponse';
