// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/cosmwasmpool/v1beta1/model/transmuter_msgs.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../api/protobuf-writer.js';
/**
 * Encodes a `EmptyRequest` protobuf message: ===================== JoinPoolExecuteMsg
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolEmptyRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `JoinPoolExecuteMsgRequest` protobuf message:
 * @param g_join_pool - `join_pool`: the structure containing all request fields of the join pool execute message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolJoinPoolExecuteMsgRequest = (atu8_join_pool) => Protobuf() // ...
    .b(atu8_join_pool) // EmptyRequest join_pool = 1
    .o;
/**
 * Encodes a `JoinPoolExecuteMsgResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolJoinPoolExecuteMsgResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `ExitPoolExecuteMsgRequest` protobuf message: ===================== ExitPoolExecuteMsg
 * @param g_exit_pool - `exit_pool`: the structure containing all request fields of the exit pool execute message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolExitPoolExecuteMsgRequest = (atu8_exit_pool) => Protobuf() // ...
    .b(atu8_exit_pool) // EmptyRequest exit_pool = 1
    .o;
/**
 * Encodes a `ExitPoolExecuteMsgResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisCosmwasmpoolExitPoolExecuteMsgResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf EmptyRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolEmptyRequest}
 */
export const decodeOsmosisCosmwasmpoolEmptyRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf JoinPoolExecuteMsgRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolJoinPoolExecuteMsgRequest}
 */
export const decodeOsmosisCosmwasmpoolJoinPoolExecuteMsgRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisCosmwasmpoolEmptyRequest]);
/**
 * Decodes a protobuf JoinPoolExecuteMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolJoinPoolExecuteMsgResponse}
 */
export const decodeOsmosisCosmwasmpoolJoinPoolExecuteMsgResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf ExitPoolExecuteMsgRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolExitPoolExecuteMsgRequest}
 */
export const decodeOsmosisCosmwasmpoolExitPoolExecuteMsgRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeOsmosisCosmwasmpoolEmptyRequest]);
/**
 * Decodes a protobuf ExitPoolExecuteMsg message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisCosmwasmpoolExitPoolExecuteMsgResponse}
 */
export const decodeOsmosisCosmwasmpoolExitPoolExecuteMsgResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolEmptyRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisCosmwasmpoolEmptyRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_EMPTY_REQUEST = '/osmosis.cosmwasmpool.v1beta1.EmptyRequest';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolJoinPoolExecuteMsgRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: join_pool - the structure containing all request fields of the join pool execute message.
 */
export const destructOsmosisCosmwasmpoolJoinPoolExecuteMsgRequest = (g_struct) => [g_struct.join_pool];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_JOIN_POOL_EXECUTE_MSG_REQUEST = '/osmosis.cosmwasmpool.v1beta1.JoinPoolExecuteMsgRequest';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolJoinPoolExecuteMsgResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisCosmwasmpoolJoinPoolExecuteMsgResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_JOIN_POOL_EXECUTE_MSG_RESPONSE = '/osmosis.cosmwasmpool.v1beta1.JoinPoolExecuteMsgResponse';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolExitPoolExecuteMsgRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: exit_pool - the structure containing all request fields of the exit pool execute message.
 */
export const destructOsmosisCosmwasmpoolExitPoolExecuteMsgRequest = (g_struct) => [g_struct.exit_pool];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_EXIT_POOL_EXECUTE_MSG_REQUEST = '/osmosis.cosmwasmpool.v1beta1.ExitPoolExecuteMsgRequest';
/**
 * Destructures the fields of a {@link OsmosisCosmwasmpoolExitPoolExecuteMsgResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructOsmosisCosmwasmpoolExitPoolExecuteMsgResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_COSMWASMPOOL_EXIT_POOL_EXECUTE_MSG_RESPONSE = '/osmosis.cosmwasmpool.v1beta1.ExitPoolExecuteMsgResponse';
