// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/consensus/v1/tx.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeTendermintTypesBlockParams, decodeTendermintTypesEvidenceParams, decodeTendermintTypesValidatorParams, decodeTendermintTypesABCIParams } from '../../../tendermint/types/params.js';
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type.
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param g_block - `block`: params defines the x/consensus parameters to update. VersionsParams is not included in this Msg because it is tracked separarately in x/upgrade. NOTE: All parameters must be supplied.
 * @param g_evidence - `evidence`:
 * @param g_validator - `validator`:
 * @param g_abci - `abci`: Since: cosmos-sdk 0.50
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosConsensusMsgUpdateParams = (s_authority, atu8_block, atu8_evidence, atu8_validator, atu8_abci) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_block) // BlockParams block = 2
    .b(atu8_evidence) // EvidenceParams evidence = 3
    .b(atu8_validator) // ValidatorParams validator = 4
    .b(atu8_abci) // ABCIParams abci = 5
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosConsensusMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosConsensusMsgUpdateParams}
 */
export const decodeCosmosConsensusMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 1, 1, 1], [0, decodeTendermintTypesBlockParams, decodeTendermintTypesEvidenceParams, decodeTendermintTypesValidatorParams, decodeTendermintTypesABCIParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosConsensusMsgUpdateParamsResponse}
 */
export const decodeCosmosConsensusMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosConsensusMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: block - params defines the x/consensus parameters to update. VersionsParams is not included in this Msg because it is tracked separarately in x/upgrade. NOTE: All parameters must be supplied.
 *   - 2: evidence -
 *   - 3: validator -
 *   - 4: abci - Since: cosmos-sdk 0.50
 */
export const destructCosmosConsensusMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.block, g_struct.evidence, g_struct.validator, g_struct.abci];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_CONSENSUS_MSG_UPDATE_PARAMS = '/cosmos.consensus.v1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link CosmosConsensusMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosConsensusMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_CONSENSUS_MSG_UPDATE_PARAMS_RESPONSE = '/cosmos.consensus.v1.MsgUpdateParamsResponse';
