// @solar-republic/cosmos-grpc/build/dist/proto/axelar/snapshot/v1beta1/tx.js
import { bech32_decode } from '@solar-republic/crypto';
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { safe_base64_to_bytes } from '../../../../api/util.js';
/**
 * Encodes a `RegisterProxyRequest` protobuf message:
 * @param sa_sender - `sender`:
 * @param atu8_proxy_addr - `proxy_addr`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotRegisterProxyRequest = (sa_sender, atu8_proxy_addr) => Protobuf() // ...
    .b(bech32_decode(sa_sender)) // bytes sender = 1
    .b(atu8_proxy_addr) // bytes proxy_addr = 2
    .o;
/**
 * Encodes a `RegisterProxyResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotRegisterProxyResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `DeactivateProxyRequest` protobuf message:
 * @param sa_sender - `sender`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotDeactivateProxyRequest = (sa_sender) => Protobuf() // ...
    .b(bech32_decode(sa_sender)) // bytes sender = 1
    .o;
/**
 * Encodes a `DeactivateProxyResponse` protobuf message:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeAxelarSnapshotDeactivateProxyResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf RegisterProxyRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotRegisterProxyRequest}
 */
export const decodeAxelarSnapshotRegisterProxyRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1]);
/**
 * Decodes a protobuf RegisterProxy message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotRegisterProxyResponse}
 */
export const decodeAxelarSnapshotRegisterProxyResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf DeactivateProxyRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotDeactivateProxyRequest}
 */
export const decodeAxelarSnapshotDeactivateProxyRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Decodes a protobuf DeactivateProxy message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedAxelarSnapshotDeactivateProxyResponse}
 */
export const decodeAxelarSnapshotDeactivateProxyResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link AxelarSnapshotRegisterProxyRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 *   - 1: proxy_addr -
 */
export const destructAxelarSnapshotRegisterProxyRequest = (g_struct) => [g_struct.sender, safe_base64_to_bytes(g_struct.proxy_addr)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_REGISTER_PROXY_REQUEST = '/axelar.snapshot.v1beta1.RegisterProxyRequest';
/**
 * Destructures the fields of a {@link AxelarSnapshotRegisterProxyResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAxelarSnapshotRegisterProxyResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_REGISTER_PROXY_RESPONSE = '/axelar.snapshot.v1beta1.RegisterProxyResponse';
/**
 * Destructures the fields of a {@link AxelarSnapshotDeactivateProxyRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender -
 */
export const destructAxelarSnapshotDeactivateProxyRequest = (g_struct) => [g_struct.sender];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_DEACTIVATE_PROXY_REQUEST = '/axelar.snapshot.v1beta1.DeactivateProxyRequest';
/**
 * Destructures the fields of a {@link AxelarSnapshotDeactivateProxyResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructAxelarSnapshotDeactivateProxyResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_AXELAR_SNAPSHOT_DEACTIVATE_PROXY_RESPONSE = '/axelar.snapshot.v1beta1.DeactivateProxyResponse';
