// @solar-republic/cosmos-grpc/build/dist/proto/ibc/applications/interchain_accounts/controller/v1/query.js
import { decode_protobuf } from '../../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../../../api/transport.js';
import { decodeIbcApplicationsInterchainAccountsControllerParams } from './controller.js';
const SR_LCD_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY = '/ibc/apps/interchain_accounts/controller/v1/';
/**
 * Returns the interchain account address for a given owner address on a given connection
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sa_owner -
 * @param si_connection -
 * @returns a tuple where:
 *   - 0: address - the 'address' response property
 */
export const queryIbcApplicationsInterchainAccountsControllerInterchainAccount = restful_grpc((sa_owner, si_connection) => [SR_LCD_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY + 'owners/' + sa_owner + '/connections/' + si_connection]);
/**
 * Queries all parameters of the ICA controller submodule.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const queryIbcApplicationsInterchainAccountsControllerParams = restful_grpc(() => [SR_LCD_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY + 'params']);
/**
 * Encodes a `QueryInterchainAccountRequest` protobuf message: the request type for the Query/InterchainAccount RPC method.
 * @param sa_owner - `owner`:
 * @param si_connection - `connection_id`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsControllerQueryInterchainAccountRequest = (sa_owner, si_connection) => Protobuf() // ...
    .s(sa_owner) // string owner = 1
    .s(si_connection) // string connection_id = 2
    .o;
/**
 * Encodes a `QueryInterchainAccountResponse` protobuf message: the response type for the Query/InterchainAccount RPC method.
 * @param sa_address - `address`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsControllerQueryInterchainAccountResponse = (sa_address) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .o;
/**
 * Encodes a `QueryParamsRequest` protobuf message: the request type for the Query/Params RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsControllerQueryParamsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `QueryParamsResponse` protobuf message: the response type for the Query/Params RPC method.
 * @param g_params - `params`: defines the parameters of the module.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeIbcApplicationsInterchainAccountsControllerQueryParamsResponse = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Decodes a protobuf QueryInterchainAccountRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsControllerQueryInterchainAccountRequest}
 */
export const decodeIbcApplicationsInterchainAccountsControllerQueryInterchainAccountRequest = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf QueryInterchainAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsControllerQueryInterchainAccountResponse}
 */
export const decodeIbcApplicationsInterchainAccountsControllerQueryInterchainAccountResponse = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf QueryParamsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsControllerQueryParamsRequest}
 */
export const decodeIbcApplicationsInterchainAccountsControllerQueryParamsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf QueryParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedIbcApplicationsInterchainAccountsControllerQueryParamsResponse}
 */
export const decodeIbcApplicationsInterchainAccountsControllerQueryParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeIbcApplicationsInterchainAccountsControllerParams]);
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsControllerQueryInterchainAccountRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: owner -
 *   - 1: connection_id -
 */
export const destructIbcApplicationsInterchainAccountsControllerQueryInterchainAccountRequest = (g_struct) => [g_struct.owner, g_struct.connection_id];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY_INTERCHAIN_ACCOUNT_REQUEST = '/ibc.applications.interchain_accounts.controller.v1.QueryInterchainAccountRequest';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsControllerQueryInterchainAccountResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address -
 */
export const destructIbcApplicationsInterchainAccountsControllerQueryInterchainAccountResponse = (g_struct) => [g_struct.address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY_INTERCHAIN_ACCOUNT_RESPONSE = '/ibc.applications.interchain_accounts.controller.v1.QueryInterchainAccountResponse';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsControllerQueryParamsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructIbcApplicationsInterchainAccountsControllerQueryParamsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY_PARAMS_REQUEST = '/ibc.applications.interchain_accounts.controller.v1.QueryParamsRequest';
/**
 * Destructures the fields of a {@link IbcApplicationsInterchainAccountsControllerQueryParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - defines the parameters of the module.
 */
export const destructIbcApplicationsInterchainAccountsControllerQueryParamsResponse = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_IBC_APPLICATIONS_INTERCHAIN_ACCOUNTS_CONTROLLER_QUERY_PARAMS_RESPONSE = '/ibc.applications.interchain_accounts.controller.v1.QueryParamsResponse';
