// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/autocli/v1/query.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeCosmosAutocliModuleOptions } from './options.js';
/**
 * Encodes a `AppOptionsRequest` protobuf message: the RemoteInfoService/AppOptions request type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAutocliAppOptionsRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `AppOptionsResponse` protobuf message: the RemoteInfoService/AppOptions response type.
 * @param a_module_options - `module_options`: a map of module name to autocli module options.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAutocliAppOptionsResponse = (a_module_options) => Protobuf() // ...
    .B(a_module_options) // ModuleOptionsEntry[] module_options = 1
    .o;
/**
 * Encodes a `ModuleOptionsEntry` protobuf message:
 * @param s_key - `key`:
 * @param g_value - `value`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAutocliAppOptionsResponseModuleOptionsEntry = (s_key, atu8_value) => Protobuf() // ...
    .s(s_key) // string key = 1
    .b(atu8_value) // ModuleOptions value = 2
    .o;
/**
 * Decodes a protobuf AppOptionsRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAutocliAppOptionsRequest}
 */
export const decodeCosmosAutocliAppOptionsRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf AppOptions message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAutocliAppOptionsResponse}
 */
export const decodeCosmosAutocliAppOptionsResponse = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeCosmosAutocliAppOptionsResponseModuleOptionsEntry]);
/**
 * Decodes a protobuf ModuleOptionsEntry message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAutocliAppOptionsResponseModuleOptionsEntry}
 */
export const decodeCosmosAutocliAppOptionsResponseModuleOptionsEntry = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosAutocliModuleOptions]);
/**
 * Destructures the fields of a {@link CosmosAutocliAppOptionsRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosAutocliAppOptionsRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_AUTOCLI_APP_OPTIONS_REQUEST = '/cosmos.autocli.v1.AppOptionsRequest';
/**
 * Destructures the fields of a {@link CosmosAutocliAppOptionsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: module_options - a map of module name to autocli module options.
 */
export const destructCosmosAutocliAppOptionsResponse = (g_struct) => [g_struct.module_options];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_AUTOCLI_APP_OPTIONS_RESPONSE = '/cosmos.autocli.v1.AppOptionsResponse';
/**
 * Destructures the fields of a {@link CosmosAutocliAppOptionsResponseModuleOptionsEntry} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: key -
 *   - 1: value -
 */
export const destructCosmosAutocliAppOptionsResponseModuleOptionsEntry = (g_struct) => [g_struct.key, g_struct.value];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_AUTOCLI_APP_OPTIONS_RESPONSE_MODULE_OPTIONS_ENTRY = '/cosmos.autocli.v1.AppOptionsResponse.ModuleOptionsEntry';
