// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/vesting/v1beta1/tx.js
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../api/protobuf-writer.js';
import { decodeCosmosVestingPeriod } from './vesting.js';
/**
 * Encodes a `MsgCreateVestingAccount` protobuf message: defines a message that enables creating a vesting account.
 * @param sa_from - `from_address`:
 * @param sa_to - `to_address`:
 * @param a_amounts - `amount`:
 * @param sg_end_time - `end_time`: end of vesting as unix time (in seconds).
 * @param b_delayed - `delayed`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreateVestingAccount = (sa_from, sa_to, a_amounts, sg_end_time, b_delayed) => Protobuf() // ...
    .s(sa_from) // string from_address = 1
    .s(sa_to) // string to_address = 2
    .B(coins(a_amounts)) // Coin[] amount = 3
    .g(sg_end_time) // int32 end_time = 4
    .v(b_delayed) // boolean delayed = 5
    .o;
/**
 * Encodes a `MsgCreateVestingAccountResponse` protobuf message: defines the Msg/CreateVestingAccount response type.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreateVestingAccountResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgCreatePermanentLockedAccount` protobuf message: defines a message that enables creating a permanent locked account. Since: cosmos-sdk 0.46
 * @param sa_from - `from_address`:
 * @param sa_to - `to_address`:
 * @param a_amounts - `amount`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreatePermanentLockedAccount = (sa_from, sa_to, a_amounts) => Protobuf() // ...
    .s(sa_from) // string from_address = 1
    .s(sa_to) // string to_address = 2
    .B(coins(a_amounts)) // Coin[] amount = 3
    .o;
/**
 * Encodes a `MsgCreatePermanentLockedAccountResponse` protobuf message: defines the Msg/CreatePermanentLockedAccount response type. Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreatePermanentLockedAccountResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgCreatePeriodicVestingAccount` protobuf message: MsgCreateVestingAccount defines a message that enables creating a vesting account. Since: cosmos-sdk 0.46
 * @param sa_from - `from_address`:
 * @param sa_to - `to_address`:
 * @param sg_start_time - `start_time`: start of vesting as unix time (in seconds).
 * @param a_vesting_periods - `vesting_periods`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreatePeriodicVestingAccount = (sa_from, sa_to, sg_start_time, a_vesting_periods) => Protobuf() // ...
    .s(sa_from) // string from_address = 1
    .s(sa_to) // string to_address = 2
    .g(sg_start_time) // int32 start_time = 3
    .B(a_vesting_periods) // Period[] vesting_periods = 4
    .o;
/**
 * Encodes a `MsgCreatePeriodicVestingAccountResponse` protobuf message: MsgCreateVestingAccountResponse defines the Msg/CreatePeriodicVestingAccount response type. Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosVestingMsgCreatePeriodicVestingAccountResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf CreateVestingAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreateVestingAccount}
 */
export const decodeCosmosVestingMsgCreateVestingAccount = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0, 3, 1], [0, 0, decode_coin]);
/**
 * Decodes a protobuf CreateVestingAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreateVestingAccountResponse}
 */
export const decodeCosmosVestingMsgCreateVestingAccountResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf CreatePermanentLockedAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreatePermanentLockedAccount}
 */
export const decodeCosmosVestingMsgCreatePermanentLockedAccount = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 0], [0, 0, decode_coin]);
/**
 * Decodes a protobuf CreatePermanentLockedAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreatePermanentLockedAccountResponse}
 */
export const decodeCosmosVestingMsgCreatePermanentLockedAccountResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf CreatePeriodicVestingAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreatePeriodicVestingAccount}
 */
export const decodeCosmosVestingMsgCreatePeriodicVestingAccount = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 3, 0], [0, 0, 0, decodeCosmosVestingPeriod]);
/**
 * Decodes a protobuf CreatePeriodicVestingAccount message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosVestingMsgCreatePeriodicVestingAccountResponse}
 */
export const decodeCosmosVestingMsgCreatePeriodicVestingAccountResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreateVestingAccount} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: from_address -
 *   - 1: to_address -
 *   - 2: amount -
 *   - 3: end_time - end of vesting as unix time (in seconds).
 *   - 4: delayed -
 */
export const destructCosmosVestingMsgCreateVestingAccount = (g_struct) => [g_struct.from_address, g_struct.to_address, g_struct.amount, g_struct.end_time, g_struct.delayed];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_VESTING_ACCOUNT = '/cosmos.vesting.v1beta1.MsgCreateVestingAccount';
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreateVestingAccountResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosVestingMsgCreateVestingAccountResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_VESTING_ACCOUNT_RESPONSE = '/cosmos.vesting.v1beta1.MsgCreateVestingAccountResponse';
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreatePermanentLockedAccount} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: from_address -
 *   - 1: to_address -
 *   - 2: amount -
 */
export const destructCosmosVestingMsgCreatePermanentLockedAccount = (g_struct) => [g_struct.from_address, g_struct.to_address, g_struct.amount];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_PERMANENT_LOCKED_ACCOUNT = '/cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccount';
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreatePermanentLockedAccountResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosVestingMsgCreatePermanentLockedAccountResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_PERMANENT_LOCKED_ACCOUNT_RESPONSE = '/cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccountResponse';
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreatePeriodicVestingAccount} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: from_address -
 *   - 1: to_address -
 *   - 2: start_time - start of vesting as unix time (in seconds).
 *   - 3: vesting_periods -
 */
export const destructCosmosVestingMsgCreatePeriodicVestingAccount = (g_struct) => [g_struct.from_address, g_struct.to_address, g_struct.start_time, g_struct.vesting_periods];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_PERIODIC_VESTING_ACCOUNT = '/cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccount';
/**
 * Destructures the fields of a {@link CosmosVestingMsgCreatePeriodicVestingAccountResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosVestingMsgCreatePeriodicVestingAccountResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_VESTING_MSG_CREATE_PERIODIC_VESTING_ACCOUNT_RESPONSE = '/cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccountResponse';
