// @solar-republic/cosmos-grpc/build/dist/proto/juno/clock/v1/tx.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../api/transport.js';
import { decodeJunoClockParams } from './genesis.js';
const SR_LCD_JUNO_CLOCK_MSG = '/juno/clock/v1/tx/';
/**
 * Defines the endpoint for registering a new clock contract.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sa_sender - The address of the sender.
 * @param sa_contract - The address of the contract to register.
 * @returns an empty tuple
 */
export const submitJunoClockRegisterClockContract = restful_grpc((sa_sender, sa_contract) => [SR_LCD_JUNO_CLOCK_MSG + 'register', {
        sender_address: sa_sender,
        contract_address: sa_contract,
    }], 1);
/**
 * Defines the endpoint for unregistering a clock contract.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sa_sender - The address of the sender.
 * @param sa_contract - The address of the contract to unregister.
 * @returns an empty tuple
 */
export const submitJunoClockUnregisterClockContract = restful_grpc((sa_sender, sa_contract) => [SR_LCD_JUNO_CLOCK_MSG + 'unregister', {
        sender_address: sa_sender,
        contract_address: sa_contract,
    }], 1);
/**
 * Defines the endpoint for unjailing a clock contract.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sa_sender - The address of the sender.
 * @param sa_contract - The address of the contract to unjail.
 * @returns an empty tuple
 */
export const submitJunoClockUnjailClockContract = restful_grpc((sa_sender, sa_contract) => [SR_LCD_JUNO_CLOCK_MSG + 'unjail', {
        sender_address: sa_sender,
        contract_address: sa_contract,
    }], 1);
/**
 * Encodes a `MsgRegisterClockContract` protobuf message: the Msg/RegisterClockContract request type.
 * @param sa_sender - `sender_address`: The address of the sender.
 * @param sa_contract - `contract_address`: The address of the contract to register.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgRegisterClockContract = (sa_sender, sa_contract) => Protobuf() // ...
    .s(sa_sender) // string sender_address = 1
    .s(sa_contract) // string contract_address = 2
    .o;
/**
 * Encodes a `MsgRegisterClockContractResponse` protobuf message: defines the response structure for executing a MsgRegisterClockContract message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgRegisterClockContractResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUnregisterClockContract` protobuf message: the Msg/UnregisterClockContract request type.
 * @param sa_sender - `sender_address`: The address of the sender.
 * @param sa_contract - `contract_address`: The address of the contract to unregister.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUnregisterClockContract = (sa_sender, sa_contract) => Protobuf() // ...
    .s(sa_sender) // string sender_address = 1
    .s(sa_contract) // string contract_address = 2
    .o;
/**
 * Encodes a `MsgUnregisterClockContractResponse` protobuf message: defines the response structure for executing a MsgUnregisterClockContract message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUnregisterClockContractResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUnjailClockContract` protobuf message: the Msg/UnjailClockContract request type.
 * @param sa_sender - `sender_address`: The address of the sender.
 * @param sa_contract - `contract_address`: The address of the contract to unjail.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUnjailClockContract = (sa_sender, sa_contract) => Protobuf() // ...
    .s(sa_sender) // string sender_address = 1
    .s(sa_contract) // string contract_address = 2
    .o;
/**
 * Encodes a `MsgUnjailClockContractResponse` protobuf message: defines the response structure for executing a MsgUnjailClockContract message.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUnjailClockContractResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type. Since: cosmos-sdk 0.47
 * @param s_authority - `authority`: the address of the governance account.
 * @param g_params - `params`: defines the x/clock parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUpdateParams = (s_authority, atu8_params) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf RegisterClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgRegisterClockContract}
 */
export const decodeJunoClockMsgRegisterClockContract = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf RegisterClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgRegisterClockContractResponse}
 */
export const decodeJunoClockMsgRegisterClockContractResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UnregisterClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUnregisterClockContract}
 */
export const decodeJunoClockMsgUnregisterClockContract = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf UnregisterClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUnregisterClockContractResponse}
 */
export const decodeJunoClockMsgUnregisterClockContractResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UnjailClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUnjailClockContract}
 */
export const decodeJunoClockMsgUnjailClockContract = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Decodes a protobuf UnjailClockContract message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUnjailClockContractResponse}
 */
export const decodeJunoClockMsgUnjailClockContractResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUpdateParams}
 */
export const decodeJunoClockMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeJunoClockParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockMsgUpdateParamsResponse}
 */
export const decodeJunoClockMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link JunoClockMsgRegisterClockContract} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender_address - The address of the sender.
 *   - 1: contract_address - The address of the contract to register.
 */
export const destructJunoClockMsgRegisterClockContract = (g_struct) => [g_struct.sender_address, g_struct.contract_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_REGISTER_CLOCK_CONTRACT = '/juno.clock.v1.MsgRegisterClockContract';
/**
 * Destructures the fields of a {@link JunoClockMsgRegisterClockContractResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoClockMsgRegisterClockContractResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_REGISTER_CLOCK_CONTRACT_RESPONSE = '/juno.clock.v1.MsgRegisterClockContractResponse';
/**
 * Destructures the fields of a {@link JunoClockMsgUnregisterClockContract} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender_address - The address of the sender.
 *   - 1: contract_address - The address of the contract to unregister.
 */
export const destructJunoClockMsgUnregisterClockContract = (g_struct) => [g_struct.sender_address, g_struct.contract_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UNREGISTER_CLOCK_CONTRACT = '/juno.clock.v1.MsgUnregisterClockContract';
/**
 * Destructures the fields of a {@link JunoClockMsgUnregisterClockContractResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoClockMsgUnregisterClockContractResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UNREGISTER_CLOCK_CONTRACT_RESPONSE = '/juno.clock.v1.MsgUnregisterClockContractResponse';
/**
 * Destructures the fields of a {@link JunoClockMsgUnjailClockContract} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: sender_address - The address of the sender.
 *   - 1: contract_address - The address of the contract to unjail.
 */
export const destructJunoClockMsgUnjailClockContract = (g_struct) => [g_struct.sender_address, g_struct.contract_address];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UNJAIL_CLOCK_CONTRACT = '/juno.clock.v1.MsgUnjailClockContract';
/**
 * Destructures the fields of a {@link JunoClockMsgUnjailClockContractResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoClockMsgUnjailClockContractResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UNJAIL_CLOCK_CONTRACT_RESPONSE = '/juno.clock.v1.MsgUnjailClockContractResponse';
/**
 * Destructures the fields of a {@link JunoClockMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address of the governance account.
 *   - 1: params - defines the x/clock parameters to update. NOTE: All parameters must be supplied.
 */
export const destructJunoClockMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UPDATE_PARAMS = '/juno.clock.v1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link JunoClockMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructJunoClockMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_MSG_UPDATE_PARAMS_RESPONSE = '/juno.clock.v1.MsgUpdateParamsResponse';
