// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/base/tendermint/v1beta1/query.js
import { __UNDEFINED } from '@blake.regalia/belt';
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../api/protobuf-writer.js';
import { restful_grpc } from '../../../../../api/transport.js';
import { safe_bytes_to_base64, safe_base64_to_bytes } from '../../../../../api/util.js';
import { decodeCosmosBaseQueryPageRequest, decodeCosmosBaseQueryPageResponse } from '../../query/v1beta1/pagination.js';
import { decodeCosmosBaseTendermintBlock } from './types.js';
import { decodeGoogleProtobufAny } from '../../../../google/protobuf/any.js';
import { decodeTendermintP2pDefaultNodeInfo } from '../../../../tendermint/p2p/types.js';
import { decodeTendermintTypesBlock } from '../../../../tendermint/types/block.js';
import { decodeTendermintTypesBlockID } from '../../../../tendermint/types/types.js';
const SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE = '/cosmos/base/tendermint/v1beta1/';
/**
 * Queries the current node info.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: default_node_info - the 'default_node_info' response property
 *   - 1: application_version - the 'default_node_info' response property
 */
export const queryCosmosBaseTendermintGetNodeInfo = restful_grpc(() => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'node_info']);
/**
 * Queries node syncing.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: syncing - the 'syncing' response property
 */
export const queryCosmosBaseTendermintGetSyncing = restful_grpc(() => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'syncing']);
/**
 * Returns the latest block.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @returns a tuple where:
 *   - 0: block_id - the 'block_id' response property
 *   - 1: block - the 'block_id' response property
 *   - 2: sdk_block - the 'block_id' response property
 */
export const queryCosmosBaseTendermintGetLatestBlock = restful_grpc(() => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'blocks/latest']);
/**
 * Queries block for given height.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sg_height -
 * @returns a tuple where:
 *   - 0: block_id - the 'block_id' response property
 *   - 1: block - the 'block_id' response property
 *   - 2: sdk_block - the 'block_id' response property
 */
export const queryCosmosBaseTendermintGetBlockByHeight = restful_grpc(sg_height => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'blocks/' + sg_height]);
/**
 * Queries latest validator-set.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param g_pagination - defines an pagination for the request.
 * @returns a tuple where:
 *   - 0: block_height - the 'block_height' response property
 *   - 1: validators - the 'block_height' response property
 *   - 2: pagination - the 'block_height' response property
 */
export const queryCosmosBaseTendermintGetLatestValidatorSet = restful_grpc(g_pagination => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'validatorsets/latest', {
        pagination: g_pagination,
    }]);
/**
 * Queries validator-set at a given height.
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param sg_height -
 * @param g_pagination - defines an pagination for the request.
 * @returns a tuple where:
 *   - 0: block_height - the 'block_height' response property
 *   - 1: validators - the 'block_height' response property
 *   - 2: pagination - the 'block_height' response property
 */
export const queryCosmosBaseTendermintGetValidatorSetByHeight = restful_grpc((sg_height, g_pagination) => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'validatorsets/' + sg_height, {
        pagination: g_pagination,
    }]);
/**
 * Defines a query handler that supports ABCI queries directly to the application, bypassing Tendermint completely. The ABCI query must contain a valid and supported path, including app, custom, p2p, and store. Since: cosmos-sdk 0.46
 * @param z_req - URL origin of LCD endpoint or {@link RequestInit}
 * @param atu8_data -
 * @param s_path -
 * @param sg_height -
 * @param b_prove -
 * @returns a tuple where:
 *   - 0: code - the 'code' response property
 *   - 1: log - the 'code' response property
 *   - 2: info - the 'code' response property
 *   - 3: index - the 'code' response property
 *   - 4: key - the 'code' response property
 *   - 5: value - the 'code' response property
 *   - 6: proof_ops - the 'code' response property
 *   - 7: height - the 'code' response property
 *   - 8: codespace - the 'code' response property
 */
export const queryCosmosBaseTendermintABCIQuery = restful_grpc((atu8_data, s_path, sg_height, b_prove) => [SR_LCD_COSMOS_BASE_TENDERMINT_SERVICE + 'abci_query', {
        data: safe_bytes_to_base64(atu8_data),
        path: s_path,
        height: sg_height,
        prove: b_prove,
    }]);
/**
 * Encodes a `GetValidatorSetByHeightRequest` protobuf message: the request type for the Query/GetValidatorSetByHeight RPC method.
 * @param sg_height - `height`:
 * @param g_pagination - `pagination`: defines an pagination for the request.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetValidatorSetByHeightRequest = (sg_height, atu8_pagination) => Protobuf() // ...
    .g(sg_height) // int32 height = 1
    .b(atu8_pagination) // PageRequest pagination = 2
    .o;
/**
 * Encodes a `GetValidatorSetByHeightResponse` protobuf message: the response type for the Query/GetValidatorSetByHeight RPC method.
 * @param sg_block_height - `block_height`:
 * @param a_validators - `validators`:
 * @param g_pagination - `pagination`: defines an pagination for the response.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetValidatorSetByHeightResponse = (sg_block_height, a_validators, atu8_pagination) => Protobuf() // ...
    .g(sg_block_height) // int32 block_height = 1
    .B(a_validators) // Validator[] validators = 2
    .b(atu8_pagination) // PageResponse pagination = 3
    .o;
/**
 * Encodes a `GetLatestValidatorSetRequest` protobuf message: the request type for the Query/GetValidatorSetByHeight RPC method.
 * @param g_pagination - `pagination`: defines an pagination for the request.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetLatestValidatorSetRequest = (atu8_pagination) => Protobuf() // ...
    .b(atu8_pagination) // PageRequest pagination = 1
    .o;
/**
 * Encodes a `GetLatestValidatorSetResponse` protobuf message: the response type for the Query/GetValidatorSetByHeight RPC method.
 * @param sg_block_height - `block_height`:
 * @param a_validators - `validators`:
 * @param g_pagination - `pagination`: defines an pagination for the response.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetLatestValidatorSetResponse = (sg_block_height, a_validators, atu8_pagination) => Protobuf() // ...
    .g(sg_block_height) // int32 block_height = 1
    .B(a_validators) // Validator[] validators = 2
    .b(atu8_pagination) // PageResponse pagination = 3
    .o;
/**
 * Encodes a `Validator` protobuf message: the type for the validator-set.
 * @param sa_address - `address`:
 * @param atu8_pub_key - `pub_key`:
 * @param sg_voting_power - `voting_power`:
 * @param sg_proposer_priority - `proposer_priority`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintValidator = (sa_address, atu8_pub_key, sg_voting_power, sg_proposer_priority) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .b(atu8_pub_key) // Any pub_key = 2
    .g(sg_voting_power) // int32 voting_power = 3
    .g(sg_proposer_priority) // int32 proposer_priority = 4
    .o;
/**
 * Encodes a `GetBlockByHeightRequest` protobuf message: the request type for the Query/GetBlockByHeight RPC method.
 * @param sg_height - `height`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetBlockByHeightRequest = (sg_height) => Protobuf() // ...
    .g(sg_height) // int32 height = 1
    .o;
/**
 * Encodes a `GetBlockByHeightResponse` protobuf message: the response type for the Query/GetBlockByHeight RPC method.
 * @param g_block_id - `block_id`:
 * @param g_block - `block`: Deprecated: please use `sdk_block` instead
 * @param g_sdk_block - `sdk_block`: Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetBlockByHeightResponse = (atu8_block_id, atu8_block, atu8_sdk_block) => Protobuf() // ...
    .b(atu8_block_id) // BlockID block_id = 1
    .b(atu8_block) // Block block = 2
    .b(atu8_sdk_block) // Block sdk_block = 3
    .o;
/**
 * Encodes a `GetLatestBlockRequest` protobuf message: the request type for the Query/GetLatestBlock RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetLatestBlockRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `GetLatestBlockResponse` protobuf message: the response type for the Query/GetLatestBlock RPC method.
 * @param g_block_id - `block_id`:
 * @param g_block - `block`: Deprecated: please use `sdk_block` instead
 * @param g_sdk_block - `sdk_block`: Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetLatestBlockResponse = (atu8_block_id, atu8_block, atu8_sdk_block) => Protobuf() // ...
    .b(atu8_block_id) // BlockID block_id = 1
    .b(atu8_block) // Block block = 2
    .b(atu8_sdk_block) // Block sdk_block = 3
    .o;
/**
 * Encodes a `GetSyncingRequest` protobuf message: the request type for the Query/GetSyncing RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetSyncingRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `GetSyncingResponse` protobuf message: the response type for the Query/GetSyncing RPC method.
 * @param b_syncing - `syncing`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetSyncingResponse = (b_syncing) => Protobuf() // ...
    .v(b_syncing) // boolean syncing = 1
    .o;
/**
 * Encodes a `GetNodeInfoRequest` protobuf message: the request type for the Query/GetNodeInfo RPC method.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetNodeInfoRequest = () => Protobuf() // ...
    .o;
/**
 * Encodes a `GetNodeInfoResponse` protobuf message: the response type for the Query/GetNodeInfo RPC method.
 * @param g_default_node_info - `default_node_info`:
 * @param g_application_version - `application_version`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintGetNodeInfoResponse = (atu8_default_node_info, atu8_application_version) => Protobuf() // ...
    .b(atu8_default_node_info) // DefaultNodeInfo default_node_info = 1
    .b(atu8_application_version) // VersionInfo application_version = 2
    .o;
/**
 * Encodes a `VersionInfo` protobuf message: the type for the GetNodeInfoResponse message.
 * @param s_name - `name`:
 * @param s_app_name - `app_name`:
 * @param s_version - `version`:
 * @param s_git_commit - `git_commit`:
 * @param s_build_tags - `build_tags`:
 * @param s_go_version - `go_version`:
 * @param a_build_deps - `build_deps`:
 * @param s_cosmos_sdk_version - `cosmos_sdk_version`: Since: cosmos-sdk 0.43
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintVersionInfo = (s_name, s_app_name, s_version, s_git_commit, s_build_tags, s_go_version, a_build_deps, s_cosmos_sdk_version) => Protobuf() // ...
    .s(s_name) // string name = 1
    .s(s_app_name) // string app_name = 2
    .s(s_version) // string version = 3
    .s(s_git_commit) // string git_commit = 4
    .s(s_build_tags) // string build_tags = 5
    .s(s_go_version) // string go_version = 6
    .B(a_build_deps) // Module[] build_deps = 7
    .s(s_cosmos_sdk_version) // string cosmos_sdk_version = 8
    .o;
/**
 * Encodes a `Module` protobuf message: the type for VersionInfo
 * @param s_path - `path`: module path
 * @param s_version - `version`: module version
 * @param s_sum - `sum`: checksum
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintModule = (s_path, s_version, s_sum) => Protobuf() // ...
    .s(s_path) // string path = 1
    .s(s_version) // string version = 2
    .s(s_sum) // string sum = 3
    .o;
/**
 * Encodes a `ABCIQueryRequest` protobuf message: defines the request structure for the ABCIQuery gRPC query.
 * @param atu8_data - `data`:
 * @param s_path - `path`:
 * @param sg_height - `height`:
 * @param b_prove - `prove`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintABCIQueryRequest = (atu8_data, s_path, sg_height, b_prove) => Protobuf() // ...
    .b(atu8_data) // bytes data = 1
    .s(s_path) // string path = 2
    .g(sg_height) // int32 height = 3
    .v(b_prove) // boolean prove = 4
    .o;
/**
 * Encodes a `ABCIQueryResponse` protobuf message: defines the response structure for the ABCIQuery gRPC query. Note: This type is a duplicate of the ResponseQuery proto type defined in Tendermint.
 * @param n_code - `code`:
 * @param s_log - `log`:
 * @param s_info - `info`:
 * @param sg_index - `index`:
 * @param atu8_key - `key`:
 * @param atu8_value - `value`:
 * @param g_proof_ops - `proof_ops`:
 * @param sg_height - `height`:
 * @param s_codespace - `codespace`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintABCIQueryResponse = (n_code, s_log, s_info, sg_index, atu8_key, atu8_value, atu8_proof_ops, sg_height, s_codespace) => Protobuf() // ...
    .v(n_code) // uint32 code = 1
    .s(s_log, 3) // string log = 3
    .s(s_info, 4) // string info = 4
    .g(sg_index, 5) // int32 index = 5
    .b(atu8_key, 6) // bytes key = 6
    .b(atu8_value, 7) // bytes value = 7
    .b(atu8_proof_ops, 8) // ProofOps proof_ops = 8
    .g(sg_height, 9) // int32 height = 9
    .s(s_codespace, 10) // string codespace = 10
    .o;
/**
 * Encodes a `ProofOp` protobuf message: defines an operation used for calculating Merkle root. The data could be arbitrary format, providing necessary data for example neighbouring node hash. Note: This type is a duplicate of the ProofOp proto type defined in Tendermint.
 * @param s_type - `type`:
 * @param atu8_key - `key`:
 * @param atu8_data - `data`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintProofOp = (s_type, atu8_key, atu8_data) => Protobuf() // ...
    .s(s_type) // string type = 1
    .b(atu8_key) // bytes key = 2
    .b(atu8_data) // bytes data = 3
    .o;
/**
 * Encodes a `ProofOps` protobuf message: Merkle proof defined by the list of ProofOps. Note: This type is a duplicate of the ProofOps proto type defined in Tendermint.
 * @param a_ops - `ops`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosBaseTendermintProofOps = (a_ops) => Protobuf() // ...
    .B(a_ops) // ProofOp[] ops = 1
    .o;
/**
 * Decodes a protobuf GetValidatorSetByHeightRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetValidatorSetByHeightRequest}
 */
export const decodeCosmosBaseTendermintGetValidatorSetByHeightRequest = (atu8_payload) => decode_protobuf(atu8_payload, [3, 1], [0, decodeCosmosBaseQueryPageRequest]);
/**
 * Decodes a protobuf GetValidatorSetByHeight message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetValidatorSetByHeightResponse}
 */
export const decodeCosmosBaseTendermintGetValidatorSetByHeightResponse = (atu8_payload) => decode_protobuf(atu8_payload, [3, 0, 1], [0, decodeCosmosBaseTendermintValidator, decodeCosmosBaseQueryPageResponse]);
/**
 * Decodes a protobuf GetLatestValidatorSetRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetLatestValidatorSetRequest}
 */
export const decodeCosmosBaseTendermintGetLatestValidatorSetRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeCosmosBaseQueryPageRequest]);
/**
 * Decodes a protobuf GetLatestValidatorSet message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetLatestValidatorSetResponse}
 */
export const decodeCosmosBaseTendermintGetLatestValidatorSetResponse = (atu8_payload) => decode_protobuf(atu8_payload, [3, 0, 1], [0, decodeCosmosBaseTendermintValidator, decodeCosmosBaseQueryPageResponse]);
/**
 * Decodes a protobuf Validator message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintValidator}
 */
export const decodeCosmosBaseTendermintValidator = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 3, 3], [0, decodeGoogleProtobufAny]);
/**
 * Decodes a protobuf GetBlockByHeightRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetBlockByHeightRequest}
 */
export const decodeCosmosBaseTendermintGetBlockByHeightRequest = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Decodes a protobuf GetBlockByHeight message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetBlockByHeightResponse}
 */
export const decodeCosmosBaseTendermintGetBlockByHeightResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1], [decodeTendermintTypesBlockID, decodeTendermintTypesBlock, decodeCosmosBaseTendermintBlock]);
/**
 * Decodes a protobuf GetLatestBlockRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetLatestBlockRequest}
 */
export const decodeCosmosBaseTendermintGetLatestBlockRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf GetLatestBlock message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetLatestBlockResponse}
 */
export const decodeCosmosBaseTendermintGetLatestBlockResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1, 1], [decodeTendermintTypesBlockID, decodeTendermintTypesBlock, decodeCosmosBaseTendermintBlock]);
/**
 * Decodes a protobuf GetSyncingRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetSyncingRequest}
 */
export const decodeCosmosBaseTendermintGetSyncingRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf GetSyncing message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetSyncingResponse}
 */
export const decodeCosmosBaseTendermintGetSyncingResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1]);
/**
 * Decodes a protobuf GetNodeInfoRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetNodeInfoRequest}
 */
export const decodeCosmosBaseTendermintGetNodeInfoRequest = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf GetNodeInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintGetNodeInfoResponse}
 */
export const decodeCosmosBaseTendermintGetNodeInfoResponse = (atu8_payload) => decode_protobuf(atu8_payload, [1, 1], [decodeTendermintP2pDefaultNodeInfo, decodeCosmosBaseTendermintVersionInfo]);
/**
 * Decodes a protobuf VersionInfo message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintVersionInfo}
 */
export const decodeCosmosBaseTendermintVersionInfo = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5, 5, 5, 5, 0, 5], [0, 0, 0, 0, 0, 0, decodeCosmosBaseTendermintModule]);
/**
 * Decodes a protobuf Module message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintModule}
 */
export const decodeCosmosBaseTendermintModule = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5, 5]);
/**
 * Decodes a protobuf ABCIQueryRequest message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintABCIQueryRequest}
 */
export const decodeCosmosBaseTendermintABCIQueryRequest = (atu8_payload) => decode_protobuf(atu8_payload, [1, 5, 3, 1]);
/**
 * Decodes a protobuf ABCIQuery message
 * @param atu8_payload - raw bytes to decode
 * @param RESERVED - a second argument is explicitly forbidden. make sure not to pass this function by reference to some callback argument
 * @returns a {@link DecodedCosmosBaseTendermintABCIQueryResponse}
 */
export const decodeCosmosBaseTendermintABCIQueryResponse = ((atu8_payload, [n_code, , s_log, s_info, sg_index, atu8_key, atu8_value, a_proof_ops, sg_height, s_codespace] = decode_protobuf(atu8_payload, [1, 0, 5, 5, 3, 1, 1, 1, 3, 5], [0, 0, 0, 0, 0, 0, 0, decodeCosmosBaseTendermintProofOps])) => [n_code, __UNDEFINED, s_log, s_info, sg_index, atu8_key, atu8_value, a_proof_ops, sg_height, s_codespace]);
/**
 * Decodes a protobuf ProofOp message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintProofOp}
 */
export const decodeCosmosBaseTendermintProofOp = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1, 1]);
/**
 * Decodes a protobuf ProofOps message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosBaseTendermintProofOps}
 */
export const decodeCosmosBaseTendermintProofOps = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decodeCosmosBaseTendermintProofOp]);
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetValidatorSetByHeightRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: height -
 *   - 1: pagination - defines an pagination for the request.
 */
export const destructCosmosBaseTendermintGetValidatorSetByHeightRequest = (g_struct) => [g_struct.height, g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_VALIDATOR_SET_BY_HEIGHT_REQUEST = '/cosmos.base.tendermint.v1beta1.GetValidatorSetByHeightRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetValidatorSetByHeightResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: block_height -
 *   - 1: validators -
 *   - 2: pagination - defines an pagination for the response.
 */
export const destructCosmosBaseTendermintGetValidatorSetByHeightResponse = (g_struct) => [g_struct.block_height, g_struct.validators, g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_VALIDATOR_SET_BY_HEIGHT_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetValidatorSetByHeightResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintValidator} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address -
 *   - 1: pub_key -
 *   - 2: voting_power -
 *   - 3: proposer_priority -
 */
export const destructCosmosBaseTendermintValidator = (g_struct) => [g_struct.address, g_struct.pub_key, g_struct.voting_power, g_struct.proposer_priority];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_VALIDATOR = '/cosmos.base.tendermint.v1beta1.Validator';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetLatestValidatorSetRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pagination - defines an pagination for the request.
 */
export const destructCosmosBaseTendermintGetLatestValidatorSetRequest = (g_struct) => [g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_LATEST_VALIDATOR_SET_REQUEST = '/cosmos.base.tendermint.v1beta1.GetLatestValidatorSetRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetLatestValidatorSetResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: block_height -
 *   - 1: validators -
 *   - 2: pagination - defines an pagination for the response.
 */
export const destructCosmosBaseTendermintGetLatestValidatorSetResponse = (g_struct) => [g_struct.block_height, g_struct.validators, g_struct.pagination];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_LATEST_VALIDATOR_SET_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetLatestValidatorSetResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetBlockByHeightRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: height -
 */
export const destructCosmosBaseTendermintGetBlockByHeightRequest = (g_struct) => [g_struct.height];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_BLOCK_BY_HEIGHT_REQUEST = '/cosmos.base.tendermint.v1beta1.GetBlockByHeightRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetBlockByHeightResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: block_id -
 *   - 1: block - Deprecated: please use `sdk_block` instead
 *   - 2: sdk_block - Since: cosmos-sdk 0.47
 */
export const destructCosmosBaseTendermintGetBlockByHeightResponse = (g_struct) => [g_struct.block_id, g_struct.block, g_struct.sdk_block];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_BLOCK_BY_HEIGHT_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetBlockByHeightResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetLatestBlockRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosBaseTendermintGetLatestBlockRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_LATEST_BLOCK_REQUEST = '/cosmos.base.tendermint.v1beta1.GetLatestBlockRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetLatestBlockResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: block_id -
 *   - 1: block - Deprecated: please use `sdk_block` instead
 *   - 2: sdk_block - Since: cosmos-sdk 0.47
 */
export const destructCosmosBaseTendermintGetLatestBlockResponse = (g_struct) => [g_struct.block_id, g_struct.block, g_struct.sdk_block];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_LATEST_BLOCK_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetLatestBlockResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetSyncingRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosBaseTendermintGetSyncingRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_SYNCING_REQUEST = '/cosmos.base.tendermint.v1beta1.GetSyncingRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetSyncingResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: syncing -
 */
export const destructCosmosBaseTendermintGetSyncingResponse = (g_struct) => [g_struct.syncing];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_SYNCING_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetSyncingResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetNodeInfoRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosBaseTendermintGetNodeInfoRequest = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_NODE_INFO_REQUEST = '/cosmos.base.tendermint.v1beta1.GetNodeInfoRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintGetNodeInfoResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: default_node_info -
 *   - 1: application_version -
 */
export const destructCosmosBaseTendermintGetNodeInfoResponse = (g_struct) => [g_struct.default_node_info, g_struct.application_version];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_GET_NODE_INFO_RESPONSE = '/cosmos.base.tendermint.v1beta1.GetNodeInfoResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintVersionInfo} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: name -
 *   - 1: app_name -
 *   - 2: version -
 *   - 3: git_commit -
 *   - 4: build_tags -
 *   - 5: go_version -
 *   - 6: build_deps -
 *   - 7: cosmos_sdk_version - Since: cosmos-sdk 0.43
 */
export const destructCosmosBaseTendermintVersionInfo = (g_struct) => [g_struct.name, g_struct.app_name, g_struct.version, g_struct.git_commit, g_struct.build_tags, g_struct.go_version, g_struct.build_deps, g_struct.cosmos_sdk_version];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_VERSION_INFO = '/cosmos.base.tendermint.v1beta1.VersionInfo';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintModule} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: path - module path
 *   - 1: version - module version
 *   - 2: sum - checksum
 */
export const destructCosmosBaseTendermintModule = (g_struct) => [g_struct.path, g_struct.version, g_struct.sum];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_MODULE = '/cosmos.base.tendermint.v1beta1.Module';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintABCIQueryRequest} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: data -
 *   - 1: path -
 *   - 2: height -
 *   - 3: prove -
 */
export const destructCosmosBaseTendermintABCIQueryRequest = (g_struct) => [safe_base64_to_bytes(g_struct.data), g_struct.path, g_struct.height, g_struct.prove];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_A_B_C_I_QUERY_REQUEST = '/cosmos.base.tendermint.v1beta1.ABCIQueryRequest';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintABCIQueryResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: code -
 *   - 2: log -
 *   - 3: info -
 *   - 4: index -
 *   - 5: key -
 *   - 6: value -
 *   - 7: proof_ops -
 *   - 8: height -
 *   - 9: codespace -
 */
export const destructCosmosBaseTendermintABCIQueryResponse = (g_struct) => [g_struct.code, __UNDEFINED, g_struct.log, g_struct.info, g_struct.index, safe_base64_to_bytes(g_struct.key), safe_base64_to_bytes(g_struct.value), g_struct.proof_ops, g_struct.height, g_struct.codespace];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_A_B_C_I_QUERY_RESPONSE = '/cosmos.base.tendermint.v1beta1.ABCIQueryResponse';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintProofOps} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: ops -
 */
export const destructCosmosBaseTendermintProofOps = (g_struct) => [g_struct.ops];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_PROOF_OPS = '/cosmos.base.tendermint.v1beta1.ProofOps';
/**
 * Destructures the fields of a {@link CosmosBaseTendermintProofOp} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: type -
 *   - 1: key -
 *   - 2: data -
 */
export const destructCosmosBaseTendermintProofOp = (g_struct) => [g_struct.type, safe_base64_to_bytes(g_struct.key), safe_base64_to_bytes(g_struct.data)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_BASE_TENDERMINT_PROOF_OP = '/cosmos.base.tendermint.v1beta1.ProofOp';
