// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/upgrade/v1beta1/tx.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeCosmosUpgradePlan } from './upgrade.js';
/**
 * Encodes a `MsgSoftwareUpgrade` protobuf message: the Msg/SoftwareUpgrade request type. Since: cosmos-sdk 0.46
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param g_plan - `plan`: the upgrade plan.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosUpgradeMsgSoftwareUpgrade = (s_authority, atu8_plan) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_plan) // Plan plan = 2
    .o;
/**
 * Encodes a `MsgSoftwareUpgradeResponse` protobuf message: the Msg/SoftwareUpgrade response type. Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosUpgradeMsgSoftwareUpgradeResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgCancelUpgrade` protobuf message: the Msg/CancelUpgrade request type. Since: cosmos-sdk 0.46
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosUpgradeMsgCancelUpgrade = (s_authority) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .o;
/**
 * Encodes a `MsgCancelUpgradeResponse` protobuf message: the Msg/CancelUpgrade response type. Since: cosmos-sdk 0.46
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosUpgradeMsgCancelUpgradeResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf SoftwareUpgrade message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosUpgradeMsgSoftwareUpgrade}
 */
export const decodeCosmosUpgradeMsgSoftwareUpgrade = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosUpgradePlan]);
/**
 * Decodes a protobuf SoftwareUpgrade message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosUpgradeMsgSoftwareUpgradeResponse}
 */
export const decodeCosmosUpgradeMsgSoftwareUpgradeResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf CancelUpgrade message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosUpgradeMsgCancelUpgrade}
 */
export const decodeCosmosUpgradeMsgCancelUpgrade = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf CancelUpgrade message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosUpgradeMsgCancelUpgradeResponse}
 */
export const decodeCosmosUpgradeMsgCancelUpgradeResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosUpgradeMsgSoftwareUpgrade} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: plan - the upgrade plan.
 */
export const destructCosmosUpgradeMsgSoftwareUpgrade = (g_struct) => [g_struct.authority, g_struct.plan];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_UPGRADE_MSG_SOFTWARE_UPGRADE = '/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade';
/**
 * Destructures the fields of a {@link CosmosUpgradeMsgSoftwareUpgradeResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosUpgradeMsgSoftwareUpgradeResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_UPGRADE_MSG_SOFTWARE_UPGRADE_RESPONSE = '/cosmos.upgrade.v1beta1.MsgSoftwareUpgradeResponse';
/**
 * Destructures the fields of a {@link CosmosUpgradeMsgCancelUpgrade} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 */
export const destructCosmosUpgradeMsgCancelUpgrade = (g_struct) => [g_struct.authority];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_UPGRADE_MSG_CANCEL_UPGRADE = '/cosmos.upgrade.v1beta1.MsgCancelUpgrade';
/**
 * Destructures the fields of a {@link CosmosUpgradeMsgCancelUpgradeResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosUpgradeMsgCancelUpgradeResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_UPGRADE_MSG_CANCEL_UPGRADE_RESPONSE = '/cosmos.upgrade.v1beta1.MsgCancelUpgradeResponse';
