// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/circuit/v1/types.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `Permissions` protobuf message: the permissions that an account has to trip or reset the circuit breaker.
 * @param xc_level - `level`: the level of permissions granted to this account.
 * @param a_limit_type_urls - `limit_type_urls`: used with LEVEL_SOME_MSGS to limit the lists of Msg type URLs that the account can trip. It is an error to use limit_type_urls with a level other than LEVEL_SOME_MSGS.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosCircuitPermissions = (xc_level, a_limit_type_urls) => Protobuf() // ...
    .v(xc_level) // Level level = 1
    .S(a_limit_type_urls) // string[] limit_type_urls = 2
    .o;
/**
 * Encodes a `GenesisAccountPermissions` protobuf message: the account permissions for the circuit breaker in genesis
 * @param sa_address - `address`:
 * @param g_permissions - `permissions`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosCircuitGenesisAccountPermissions = (sa_address, atu8_permissions) => Protobuf() // ...
    .s(sa_address) // string address = 1
    .b(atu8_permissions) // Permissions permissions = 2
    .o;
/**
 * Encodes a `GenesisState` protobuf message: the state that must be provided at genesis.
 * @param a_account_permissions - `account_permissions`:
 * @param a_disabled_type_urls - `disabled_type_urls`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosCircuitGenesisState = (a_account_permissions, a_disabled_type_urls) => Protobuf() // ...
    .B(a_account_permissions) // GenesisAccountPermissions[] account_permissions = 1
    .S(a_disabled_type_urls) // string[] disabled_type_urls = 2
    .o;
/**
 * Decodes a protobuf Permissions message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosCircuitPermissions}
 */
export const decodeCosmosCircuitPermissions = (atu8_payload) => decode_protobuf(atu8_payload, [1, 4]);
/**
 * Decodes a protobuf GenesisAccountPermissions message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosCircuitGenesisAccountPermissions}
 */
export const decodeCosmosCircuitGenesisAccountPermissions = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosCircuitPermissions]);
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosCircuitGenesisState}
 */
export const decodeCosmosCircuitGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [0, 4], [decodeCosmosCircuitGenesisAccountPermissions]);
/**
 * Destructures the fields of a {@link CosmosCircuitPermissions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: level - the level of permissions granted to this account.
 *   - 1: limit_type_urls - used with LEVEL_SOME_MSGS to limit the lists of Msg type URLs that the account can trip. It is an error to use limit_type_urls with a level other than LEVEL_SOME_MSGS.
 */
export const destructCosmosCircuitPermissions = (g_struct) => [g_struct.level, g_struct.limit_type_urls];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_CIRCUIT_PERMISSIONS = '/cosmos.circuit.v1.Permissions';
/**
 * Destructures the fields of a {@link CosmosCircuitGenesisAccountPermissions} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: address -
 *   - 1: permissions -
 */
export const destructCosmosCircuitGenesisAccountPermissions = (g_struct) => [g_struct.address, g_struct.permissions];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_CIRCUIT_GENESIS_ACCOUNT_PERMISSIONS = '/cosmos.circuit.v1.GenesisAccountPermissions';
/**
 * Destructures the fields of a {@link CosmosCircuitGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: account_permissions -
 *   - 1: disabled_type_urls -
 */
export const destructCosmosCircuitGenesisState = (g_struct) => [g_struct.account_permissions, g_struct.disabled_type_urls];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_CIRCUIT_GENESIS_STATE = '/cosmos.circuit.v1.GenesisState';
/**
 * Protobuf enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_NONE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link ProtoEnumCosmosCircuitPermissionsLevel}
 */
export const XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED = 0;
/**
 * JSON enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_NONE_UNSPECIFIED** -
 *
 * Belongs to enum type {@link JsonEnumCosmosCircuitPermissionsLevel}
 */
export const SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED = 'LEVEL_NONE_UNSPECIFIED';
/**
 * Protobuf enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_SOME_MSGS** -
 *
 * Belongs to enum type {@link ProtoEnumCosmosCircuitPermissionsLevel}
 */
export const XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS = 1;
/**
 * JSON enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_SOME_MSGS** -
 *
 * Belongs to enum type {@link JsonEnumCosmosCircuitPermissionsLevel}
 */
export const SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS = 'LEVEL_SOME_MSGS';
/**
 * Protobuf enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_ALL_MSGS** -
 *
 * Belongs to enum type {@link ProtoEnumCosmosCircuitPermissionsLevel}
 */
export const XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS = 2;
/**
 * JSON enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_ALL_MSGS** -
 *
 * Belongs to enum type {@link JsonEnumCosmosCircuitPermissionsLevel}
 */
export const SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS = 'LEVEL_ALL_MSGS';
/**
 * Protobuf enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_SUPER_ADMIN** -
 *
 * Belongs to enum type {@link ProtoEnumCosmosCircuitPermissionsLevel}
 */
export const XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN = 3;
/**
 * JSON enum value for `cosmos.circuit.v1.Permissions.Level`.
 *
 * **LEVEL_SUPER_ADMIN** -
 *
 * Belongs to enum type {@link JsonEnumCosmosCircuitPermissionsLevel}
 */
export const SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN = 'LEVEL_SUPER_ADMIN';
/**
 * Maps a protobuf enum int value for `cosmos.circuit.v1.Permissions.Level` to is JSON equivalent enum string value
 */
export const ProtoToJsonEnumCosmosCircuitPermissionsLevel = {
    [XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED]: SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED,
    [XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS]: SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS,
    [XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS]: SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS,
    [XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN]: SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN,
};
/**
 * Maps a JSON enum string value for `cosmos.circuit.v1.Permissions.Level` to is protobuf equivalent enum int value
 */
export const JsonToProtoEnumCosmosCircuitPermissionsLevel = {
    [SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED]: XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_NONE_UNSPECIFIED,
    [SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS]: XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SOME_MSGS,
    [SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS]: XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_ALL_MSGS,
    [SI_JSON_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN]: XC_PROTO_COSMOS_CIRCUIT_PERMISSIONS_LEVEL_SUPER_ADMIN,
};
