// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/twap/v1beta1/genesis.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { parse_duration } from '../../../../api/json.js';
import { decode_protobuf, decode_temporal } from '../../../../api/protobuf-reader.js';
import { Protobuf, temporal } from '../../../../api/protobuf-writer.js';
import { decodeOsmosisTwapTwapRecord } from './twap_record.js';
/**
 * Encodes a `Params` protobuf message: holds parameters for the twap module
 * @param s_prune_epoch_identifier - `prune_epoch_identifier`:
 * @param xt_record_history_keep_period - `record_history_keep_period`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisTwapParams = (s_prune_epoch_identifier, xt_record_history_keep_period) => Protobuf() // ...
    .s(s_prune_epoch_identifier) // string prune_epoch_identifier = 1
    .b(temporal(xt_record_history_keep_period)) // Duration record_history_keep_period = 2
    .o;
/**
 * Encodes a `GenesisState` protobuf message: defines the twap module's genesis state.
 * @param a_twaps - `twaps`: the collection of all twap records.
 * @param g_params - `params`: the container of twap parameters.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisTwapGenesisState = (a_twaps, atu8_params) => Protobuf() // ...
    .B(a_twaps) // TwapRecord[] twaps = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisTwapParams}
 */
export const decodeOsmosisTwapParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, [3, 1]], [0, decode_temporal]);
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisTwapGenesisState}
 */
export const decodeOsmosisTwapGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [0, 1], [decodeOsmosisTwapTwapRecord, decodeOsmosisTwapParams]);
/**
 * Destructures the fields of a {@link OsmosisTwapParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: prune_epoch_identifier -
 *   - 1: record_history_keep_period -
 */
export const destructOsmosisTwapParams = (g_struct) => [g_struct.prune_epoch_identifier, parse_duration(g_struct.record_history_keep_period)];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_TWAP_PARAMS = '/osmosis.twap.v1beta1.Params';
/**
 * Destructures the fields of a {@link OsmosisTwapGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: twaps - the collection of all twap records.
 *   - 1: params - the container of twap parameters.
 */
export const destructOsmosisTwapGenesisState = (g_struct) => [g_struct.twaps, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_TWAP_GENESIS_STATE = '/osmosis.twap.v1beta1.GenesisState';
