// @solar-republic/cosmos-grpc/build/dist/proto/gaia/globalfee/v1beta1/genesis.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeCosmosBaseDecCoin } from '../../../cosmos/base/v1beta1/coin.js';
/**
 * Encodes a `GenesisState` protobuf message: - initial state of module
 * @param g_params - `params`: Params of this module
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGaiaGlobalfeeGenesisState = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Encodes a `Params` protobuf message: defines the set of module parameters.
 * @param a_minimum_gas_prices - `minimum_gas_prices`: stores the minimum gas price(s) for all TX on the chain. When multiple coins are defined then they are accepted alternatively. The list must be sorted by denoms asc. No duplicate denoms or zero amount values allowed. For more information see https://docs.cosmos.network/main/modules/auth#concepts
 * @param a_bypass_min_fee_msg_types - `bypass_min_fee_msg_types`: defines a list of message type urls that are free of fee charge.
 * @param sg_max_total_bypass_min_fee_msg_gas_usage - `max_total_bypass_min_fee_msg_gas_usage`: defines the total maximum gas usage allowed for a transaction containing only messages of types in bypass_min_fee_msg_types to bypass fee charge.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeGaiaGlobalfeeParams = (a_minimum_gas_prices, a_bypass_min_fee_msg_types, sg_max_total_bypass_min_fee_msg_gas_usage) => Protobuf() // ...
    .B(a_minimum_gas_prices) // DecCoin[] minimum_gas_prices = 1
    .S(a_bypass_min_fee_msg_types) // string[] bypass_min_fee_msg_types = 2
    .g(sg_max_total_bypass_min_fee_msg_gas_usage) // uint64 max_total_bypass_min_fee_msg_gas_usage = 3
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGaiaGlobalfeeGenesisState}
 */
export const decodeGaiaGlobalfeeGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeGaiaGlobalfeeParams]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedGaiaGlobalfeeParams}
 */
export const decodeGaiaGlobalfeeParams = (atu8_payload) => decode_protobuf(atu8_payload, [0, 4, 3], [decodeCosmosBaseDecCoin]);
/**
 * Destructures the fields of a {@link GaiaGlobalfeeParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: minimum_gas_prices - stores the minimum gas price(s) for all TX on the chain. When multiple coins are defined then they are accepted alternatively. The list must be sorted by denoms asc. No duplicate denoms or zero amount values allowed. For more information see https://docs.cosmos.network/main/modules/auth#concepts
 *   - 1: bypass_min_fee_msg_types - defines a list of message type urls that are free of fee charge.
 *   - 2: max_total_bypass_min_fee_msg_gas_usage - defines the total maximum gas usage allowed for a transaction containing only messages of types in bypass_min_fee_msg_types to bypass fee charge.
 */
export const destructGaiaGlobalfeeParams = (g_struct) => [g_struct.minimum_gas_prices, g_struct.bypass_min_fee_msg_types, g_struct.max_total_bypass_min_fee_msg_gas_usage];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GAIA_GLOBALFEE_PARAMS = '/gaia.globalfee.v1beta1.Params';
/**
 * Destructures the fields of a {@link GaiaGlobalfeeGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - Params of this module
 */
export const destructGaiaGlobalfeeGenesisState = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_GAIA_GLOBALFEE_GENESIS_STATE = '/gaia.globalfee.v1beta1.GenesisState';
