// @solar-republic/cosmos-grpc/build/dist/proto/osmosis/gamm/v1beta1/genesis.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf, decode_coin } from '../../../../api/protobuf-reader.js';
import { Protobuf, coins } from '../../../../api/protobuf-writer.js';
import { decodeGoogleProtobufAny } from '../../../google/protobuf/any.js';
import { decodeOsmosisGammMigrationRecords } from './shared.js';
/**
 * Encodes a `Params` protobuf message: holds parameters for the incentives module
 * @param a_pool_creation_fees - `pool_creation_fee`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammParams = (a_pool_creation_fees) => Protobuf() // ...
    .B(coins(a_pool_creation_fees)) // Coin[] pool_creation_fee = 1
    .o;
/**
 * Encodes a `GenesisState` protobuf message: defines the gamm module's genesis state.
 * @param a_pools - `pools`:
 * @param sg_next_pool_number - `next_pool_number`: will be renamed to next_pool_id in an upcoming version
 * @param g_params - `params`:
 * @param g_migration_records - `migration_records`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeOsmosisGammGenesisState = (a_pools, sg_next_pool_number, atu8_params, atu8_migration_records) => Protobuf() // ...
    .B(a_pools) // Any[] pools = 1
    .g(sg_next_pool_number) // uint64 next_pool_number = 2
    .b(atu8_params) // Params params = 3
    .b(atu8_migration_records) // MigrationRecords migration_records = 4
    .o;
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammParams}
 */
export const decodeOsmosisGammParams = (atu8_payload) => decode_protobuf(atu8_payload, [0], [decode_coin]);
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedOsmosisGammGenesisState}
 */
export const decodeOsmosisGammGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [0, 3, 1, 1], [decodeGoogleProtobufAny, 0, decodeOsmosisGammParams, decodeOsmosisGammMigrationRecords]);
/**
 * Destructures the fields of a {@link OsmosisGammParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pool_creation_fee -
 */
export const destructOsmosisGammParams = (g_struct) => [g_struct.pool_creation_fee];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_PARAMS = '/osmosis.gamm.v1beta1.Params';
/**
 * Destructures the fields of a {@link OsmosisGammGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: pools -
 *   - 1: next_pool_number - will be renamed to next_pool_id in an upcoming version
 *   - 2: params -
 *   - 3: migration_records -
 */
export const destructOsmosisGammGenesisState = (g_struct) => [g_struct.pools, g_struct.next_pool_number, g_struct.params, g_struct.migration_records];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_OSMOSIS_GAMM_GENESIS_STATE = '/osmosis.gamm.v1beta1.GenesisState';
