// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/slashing/v1beta1/tx.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
import { decodeCosmosSlashingParams } from './slashing.js';
/**
 * Encodes a `MsgUnjail` protobuf message: defines the Msg/Unjail request type
 * @param sa_validator - `validator_addr`:
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosSlashingMsgUnjail = (sa_validator) => Protobuf() // ...
    .s(sa_validator) // string validator_addr = 1
    .o;
/**
 * Encodes a `MsgUnjailResponse` protobuf message: defines the Msg/Unjail response type
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosSlashingMsgUnjailResponse = () => Protobuf() // ...
    .o;
/**
 * Encodes a `MsgUpdateParams` protobuf message: the Msg/UpdateParams request type. Since: cosmos-sdk 0.47
 * @param s_authority - `authority`: the address that controls the module (defaults to x/gov unless overwritten).
 * @param g_params - `params`: defines the x/slashing parameters to update. NOTE: All parameters must be supplied.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosSlashingMsgUpdateParams = (s_authority, atu8_params) => Protobuf() // ...
    .s(s_authority) // string authority = 1
    .b(atu8_params) // Params params = 2
    .o;
/**
 * Encodes a `MsgUpdateParamsResponse` protobuf message: defines the response structure for executing a MsgUpdateParams message. Since: cosmos-sdk 0.47
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosSlashingMsgUpdateParamsResponse = () => Protobuf() // ...
    .o;
/**
 * Decodes a protobuf Unjail message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosSlashingMsgUnjail}
 */
export const decodeCosmosSlashingMsgUnjail = (atu8_payload) => decode_protobuf(atu8_payload, [5]);
/**
 * Decodes a protobuf Unjail message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosSlashingMsgUnjailResponse}
 */
export const decodeCosmosSlashingMsgUnjailResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosSlashingMsgUpdateParams}
 */
export const decodeCosmosSlashingMsgUpdateParams = (atu8_payload) => decode_protobuf(atu8_payload, [5, 1], [0, decodeCosmosSlashingParams]);
/**
 * Decodes a protobuf UpdateParams message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosSlashingMsgUpdateParamsResponse}
 */
export const decodeCosmosSlashingMsgUpdateParamsResponse = (atu8_payload) => decode_protobuf(atu8_payload);
/**
 * Destructures the fields of a {@link CosmosSlashingMsgUnjail} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: validator_addr -
 */
export const destructCosmosSlashingMsgUnjail = (g_struct) => [g_struct.validator_addr];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_SLASHING_MSG_UNJAIL = '/cosmos.slashing.v1beta1.MsgUnjail';
/**
 * Destructures the fields of a {@link CosmosSlashingMsgUnjailResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosSlashingMsgUnjailResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_SLASHING_MSG_UNJAIL_RESPONSE = '/cosmos.slashing.v1beta1.MsgUnjailResponse';
/**
 * Destructures the fields of a {@link CosmosSlashingMsgUpdateParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: authority - the address that controls the module (defaults to x/gov unless overwritten).
 *   - 1: params - defines the x/slashing parameters to update. NOTE: All parameters must be supplied.
 */
export const destructCosmosSlashingMsgUpdateParams = (g_struct) => [g_struct.authority, g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_SLASHING_MSG_UPDATE_PARAMS = '/cosmos.slashing.v1beta1.MsgUpdateParams';
/**
 * Destructures the fields of a {@link CosmosSlashingMsgUpdateParamsResponse} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 */
export const destructCosmosSlashingMsgUpdateParamsResponse = (g_struct) => [];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_SLASHING_MSG_UPDATE_PARAMS_RESPONSE = '/cosmos.slashing.v1beta1.MsgUpdateParamsResponse';
