// @solar-republic/cosmos-grpc/build/dist/proto/juno/clock/v1/genesis.js
import { decode_protobuf } from '../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../api/protobuf-writer.js';
/**
 * Encodes a `GenesisState` protobuf message: - initial state of module
 * @param g_params - `params`: Params of this module
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockGenesisState = (atu8_params) => Protobuf() // ...
    .b(atu8_params) // Params params = 1
    .o;
/**
 * Encodes a `Params` protobuf message: defines the set of module parameters.
 * @param sg_contract_gas_limit - `contract_gas_limit`: defines the maximum amount of gas that can be used by a contract.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeJunoClockParams = (sg_contract_gas_limit) => Protobuf() // ...
    .g(sg_contract_gas_limit) // uint64 contract_gas_limit = 1
    .o;
/**
 * Decodes a protobuf GenesisState message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockGenesisState}
 */
export const decodeJunoClockGenesisState = (atu8_payload) => decode_protobuf(atu8_payload, [1], [decodeJunoClockParams]);
/**
 * Decodes a protobuf Params message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedJunoClockParams}
 */
export const decodeJunoClockParams = (atu8_payload) => decode_protobuf(atu8_payload, [3]);
/**
 * Destructures the fields of a {@link JunoClockParams} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: contract_gas_limit - defines the maximum amount of gas that can be used by a contract.
 */
export const destructJunoClockParams = (g_struct) => [g_struct.contract_gas_limit];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_PARAMS = '/juno.clock.v1.Params';
/**
 * Destructures the fields of a {@link JunoClockGenesisState} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: params - Params of this module
 */
export const destructJunoClockGenesisState = (g_struct) => [g_struct.params];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_JUNO_CLOCK_GENESIS_STATE = '/juno.clock.v1.GenesisState';
