// @solar-republic/cosmos-grpc/build/dist/proto/cosmos/app/runtime/v1alpha1/module.js
/*
* ================================
*     GENERATED FILE WARNING
* Do not edit this file manually.
* ================================
*/
import { decode_protobuf } from '../../../../../api/protobuf-reader.js';
import { Protobuf } from '../../../../../api/protobuf-writer.js';
/**
 * Encodes a `Module` protobuf message: the config object for the runtime module.
 * @param s_app_name - `app_name`: the name of the app.
 * @param a_begin_blockers - `begin_blockers`: specifies the module names of begin blockers to call in the order in which they should be called. If this is left empty no begin blocker will be registered.
 * @param a_end_blockers - `end_blockers`: specifies the module names of the end blockers to call in the order in which they should be called. If this is left empty no end blocker will be registered.
 * @param a_init_genesis - `init_genesis`: specifies the module names of init genesis functions to call in the order in which they should be called. If this is left empty no init genesis function will be registered.
 * @param a_export_genesis - `export_genesis`: specifies the order in which to export module genesis data. If this is left empty, the init_genesis order will be used for export genesis if it is specified.
 * @param a_override_store_keys - `override_store_keys`: an optional list of overrides for the module store keys to be used in keeper construction.
 * @param a_order_migrations - `order_migrations`: defines the order in which module migrations are performed. If this is left empty, it uses the default migration order. https://pkg.go.dev/github.com/cosmos/cosmos-sdk@v0.47.0-alpha2/types/module#DefaultMigrationsOrder
 * @param a_precommiters - `precommiters`: specifies the module names of the precommiters to call in the order in which they should be called. If this is left empty no precommit function will be registered.
 * @param a_prepare_check_staters - `prepare_check_staters`: specifies the module names of the prepare_check_staters to call in the order in which they should be called. If this is left empty no preparecheckstate function will be registered.
 * @param a_pre_blockers - `pre_blockers`: specifies the module names of pre blockers to call in the order in which they should be called. If this is left empty no pre blocker will be registered.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAppRuntimeModule = (s_app_name, a_begin_blockers, a_end_blockers, a_init_genesis, a_export_genesis, a_override_store_keys, a_order_migrations, a_precommiters, a_prepare_check_staters, a_pre_blockers) => Protobuf() // ...
    .s(s_app_name) // string app_name = 1
    .S(a_begin_blockers) // string[] begin_blockers = 2
    .S(a_end_blockers) // string[] end_blockers = 3
    .S(a_init_genesis) // string[] init_genesis = 4
    .S(a_export_genesis) // string[] export_genesis = 5
    .B(a_override_store_keys) // StoreKeyConfig[] override_store_keys = 6
    .S(a_order_migrations) // string[] order_migrations = 7
    .S(a_precommiters) // string[] precommiters = 8
    .S(a_prepare_check_staters) // string[] prepare_check_staters = 9
    .S(a_pre_blockers) // string[] pre_blockers = 10
    .o;
/**
 * Encodes a `StoreKeyConfig` protobuf message: may be supplied to override the default module store key, which is the module name.
 * @param s_module_name - `module_name`: name of the module to override the store key of
 * @param s_kv_store_key - `kv_store_key`: the kv store key to use instead of the module name.
 * @returns a strongly subtyped Uint8Array protobuf message
 */
export const encodeCosmosAppRuntimeStoreKeyConfig = (s_module_name, s_kv_store_key) => Protobuf() // ...
    .s(s_module_name) // string module_name = 1
    .s(s_kv_store_key) // string kv_store_key = 2
    .o;
/**
 * Decodes a protobuf Module message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAppRuntimeModule}
 */
export const decodeCosmosAppRuntimeModule = (atu8_payload) => decode_protobuf(atu8_payload, [5, 4, 4, 4, 4, 0, 4, 4, 4, 4], [0, 0, 0, 0, 0, decodeCosmosAppRuntimeStoreKeyConfig]);
/**
 * Decodes a protobuf StoreKeyConfig message
 * @param atu8_payload - raw bytes to decode
 * @returns a {@link DecodedCosmosAppRuntimeStoreKeyConfig}
 */
export const decodeCosmosAppRuntimeStoreKeyConfig = (atu8_payload) => decode_protobuf(atu8_payload, [5, 5]);
/**
 * Destructures the fields of a {@link CosmosAppRuntimeModule} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: app_name - the name of the app.
 *   - 1: begin_blockers - specifies the module names of begin blockers to call in the order in which they should be called. If this is left empty no begin blocker will be registered.
 *   - 2: end_blockers - specifies the module names of the end blockers to call in the order in which they should be called. If this is left empty no end blocker will be registered.
 *   - 3: init_genesis - specifies the module names of init genesis functions to call in the order in which they should be called. If this is left empty no init genesis function will be registered.
 *   - 4: export_genesis - specifies the order in which to export module genesis data. If this is left empty, the init_genesis order will be used for export genesis if it is specified.
 *   - 5: override_store_keys - an optional list of overrides for the module store keys to be used in keeper construction.
 *   - 6: order_migrations - defines the order in which module migrations are performed. If this is left empty, it uses the default migration order. https://pkg.go.dev/github.com/cosmos/cosmos-sdk@v0.47.0-alpha2/types/module#DefaultMigrationsOrder
 *   - 7: precommiters - specifies the module names of the precommiters to call in the order in which they should be called. If this is left empty no precommit function will be registered.
 *   - 8: prepare_check_staters - specifies the module names of the prepare_check_staters to call in the order in which they should be called. If this is left empty no preparecheckstate function will be registered.
 *   - 9: pre_blockers - specifies the module names of pre blockers to call in the order in which they should be called. If this is left empty no pre blocker will be registered.
 */
export const destructCosmosAppRuntimeModule = (g_struct) => [g_struct.app_name, g_struct.begin_blockers, g_struct.end_blockers, g_struct.init_genesis, g_struct.export_genesis, g_struct.override_store_keys, g_struct.order_migrations, g_struct.precommiters, g_struct.prepare_check_staters, g_struct.pre_blockers];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_APP_RUNTIME_MODULE = '/cosmos.app.runtime.v1alpha1.Module';
/**
 * Destructures the fields of a {@link CosmosAppRuntimeStoreKeyConfig} JSON message into a tuple of parsed ES equivalents
 * @param g_struct - the JSON message
 * @returns a tuple where:
 *   - 0: module_name - name of the module to override the store key of
 *   - 1: kv_store_key - the kv store key to use instead of the module name.
 */
export const destructCosmosAppRuntimeStoreKeyConfig = (g_struct) => [g_struct.module_name, g_struct.kv_store_key];
/**
 * Message type string constant
 */
export const SI_MESSAGE_TYPE_COSMOS_APP_RUNTIME_STORE_KEY_CONFIG = '/cosmos.app.runtime.v1alpha1.StoreKeyConfig';
